{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from typing import List\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "from joblib import Parallel, delayed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from malid import config\n",
    "from malid.datamodels import GeneLocus, healthy_label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Instructions for adding more\n",
    "\n",
    "**Merging cell type subsets from the same sample:** `specimen_label` should be consistent across entries with different `replicate_label`'s.\n",
    "\n",
    "**If we have external cohorts that are BCR+TCR:**\n",
    "\n",
    "We should have one row per locus per replicate-of-a-specimen. Often specimens are single-replicate, so we can think of this as: one low per locus per specimen.\n",
    "\n",
    "Set a `specimen_label_by_locus` column that is the globally-unique specimen label tailored to a particular locus, e.g. `$SPECIMENLABEL-IGH` or `$SPECIMENLABEL-TRB` format. (TODO: update this to be replicate_label_by_locus?)\n",
    "\n",
    "And set a `specimen_label` column that is equivalent across different-loci rows for that specimen.\n",
    "\n",
    "The row's `gene_locus` column should be set to the locus of that row (must be the name of a valid `GeneLocus` enum value), and the `fname` column should be set to the path to the file containing the data for that locus.\n",
    "\n",
    "_NOTE: The above is not so relevant for Adaptive -- we see below that this data is almost entirely TCR, and we will just ignore the BCR._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adaptive Covid TCR specimens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant_label</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>locus</th>\n",
       "      <th>study</th>\n",
       "      <th>species</th>\n",
       "      <th>sample_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ImmuneCode-190921</td>\n",
       "      <td>860011232_TCRB</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>COVID-19-HUniv12Oct - Hospitalized</td>\n",
       "      <td>53</td>\n",
       "      <td>F</td>\n",
       "      <td>Hispanic/Latino</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>Human</td>\n",
       "      <td>860011232_TCRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ImmuneCode-026</td>\n",
       "      <td>INCOV026-AC-3_TCRB</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>COVID-19-ISB</td>\n",
       "      <td>33</td>\n",
       "      <td>M</td>\n",
       "      <td>Hispanic/Latino</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>Human</td>\n",
       "      <td>INCOV026-AC-3_TCRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ImmuneCode-321977</td>\n",
       "      <td>860011116_TCRB</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>COVID-19-HUniv12Oct - Hospitalized</td>\n",
       "      <td>52</td>\n",
       "      <td>F</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>Human</td>\n",
       "      <td>860011116_TCRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ImmuneCode-087</td>\n",
       "      <td>INCOV087-BL-3_TCRB</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>COVID-19-ISB</td>\n",
       "      <td>56</td>\n",
       "      <td>M</td>\n",
       "      <td>Hispanic/Latino</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>Human</td>\n",
       "      <td>INCOV087-BL-3_TCRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ImmuneCode-0000051</td>\n",
       "      <td>BS-EQ-0014-T2-replacement_TCRB</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>COVID-19-NIH/NIAID - Hospitalized</td>\n",
       "      <td>55</td>\n",
       "      <td>M</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>Human</td>\n",
       "      <td>BS-EQ-0014-T2-replacement_TCRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>ImmuneCode-304752</td>\n",
       "      <td>860011117_TCRB</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>COVID-19-HUniv12Oct - Hospitalized - ICU</td>\n",
       "      <td>60</td>\n",
       "      <td>M</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>Human</td>\n",
       "      <td>860011117_TCRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>ImmuneCode-775827</td>\n",
       "      <td>860011106_TCRB</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>COVID-19-HUniv12Oct - Hospitalized</td>\n",
       "      <td>57</td>\n",
       "      <td>F</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>Human</td>\n",
       "      <td>860011106_TCRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>ImmuneCode-0000446</td>\n",
       "      <td>BS-GIGI_10-replacement_TCRB</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>COVID-19-NIH/NIAID - Hospitalized</td>\n",
       "      <td>75</td>\n",
       "      <td>M</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>Human</td>\n",
       "      <td>BS-GIGI_10-replacement_TCRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>ImmuneCode-0000160</td>\n",
       "      <td>BS-EQ-25-T1_BS-GIGI-71-replacement_TCRB</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>COVID-19-NIH/NIAID - Hospitalized - ICU</td>\n",
       "      <td>72</td>\n",
       "      <td>M</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>Human</td>\n",
       "      <td>BS-EQ-25-T1_BS-GIGI-71-replacement_TCRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>ImmuneCode-506345</td>\n",
       "      <td>860011240_TCRB</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>COVID-19-HUniv12Oct - Hospitalized</td>\n",
       "      <td>78</td>\n",
       "      <td>F</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>Human</td>\n",
       "      <td>860011240_TCRB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     participant_label                           specimen_label  disease  \\\n",
       "0    ImmuneCode-190921                           860011232_TCRB  Covid19   \n",
       "1       ImmuneCode-026                       INCOV026-AC-3_TCRB  Covid19   \n",
       "2    ImmuneCode-321977                           860011116_TCRB  Covid19   \n",
       "3       ImmuneCode-087                       INCOV087-BL-3_TCRB  Covid19   \n",
       "4   ImmuneCode-0000051           BS-EQ-0014-T2-replacement_TCRB  Covid19   \n",
       "..                 ...                                      ...      ...   \n",
       "83   ImmuneCode-304752                           860011117_TCRB  Covid19   \n",
       "84   ImmuneCode-775827                           860011106_TCRB  Covid19   \n",
       "85  ImmuneCode-0000446              BS-GIGI_10-replacement_TCRB  Covid19   \n",
       "86  ImmuneCode-0000160  BS-EQ-25-T1_BS-GIGI-71-replacement_TCRB  Covid19   \n",
       "87   ImmuneCode-506345                           860011240_TCRB  Covid19   \n",
       "\n",
       "                             disease_subtype  age sex        ethnicity  \\\n",
       "0         COVID-19-HUniv12Oct - Hospitalized   53   F  Hispanic/Latino   \n",
       "1                               COVID-19-ISB   33   M  Hispanic/Latino   \n",
       "2         COVID-19-HUniv12Oct - Hospitalized   52   F        Caucasian   \n",
       "3                               COVID-19-ISB   56   M  Hispanic/Latino   \n",
       "4          COVID-19-NIH/NIAID - Hospitalized   55   M        Caucasian   \n",
       "..                                       ...  ...  ..              ...   \n",
       "83  COVID-19-HUniv12Oct - Hospitalized - ICU   60   M        Caucasian   \n",
       "84        COVID-19-HUniv12Oct - Hospitalized   57   F        Caucasian   \n",
       "85         COVID-19-NIH/NIAID - Hospitalized   75   M        Caucasian   \n",
       "86   COVID-19-NIH/NIAID - Hospitalized - ICU   72   M        Caucasian   \n",
       "87        COVID-19-HUniv12Oct - Hospitalized   78   F        Caucasian   \n",
       "\n",
       "   sequencing_type locus       study species  \\\n",
       "0             gDNA  TCRB  immunecode   Human   \n",
       "1             gDNA  TCRB  immunecode   Human   \n",
       "2             gDNA  TCRB  immunecode   Human   \n",
       "3             gDNA  TCRB  immunecode   Human   \n",
       "4             gDNA  TCRB  immunecode   Human   \n",
       "..             ...   ...         ...     ...   \n",
       "83            gDNA  TCRB  immunecode   Human   \n",
       "84            gDNA  TCRB  immunecode   Human   \n",
       "85            gDNA  TCRB  immunecode   Human   \n",
       "86            gDNA  TCRB  immunecode   Human   \n",
       "87            gDNA  TCRB  immunecode   Human   \n",
       "\n",
       "                                sample_name  \n",
       "0                            860011232_TCRB  \n",
       "1                        INCOV026-AC-3_TCRB  \n",
       "2                            860011116_TCRB  \n",
       "3                        INCOV087-BL-3_TCRB  \n",
       "4            BS-EQ-0014-T2-replacement_TCRB  \n",
       "..                                      ...  \n",
       "83                           860011117_TCRB  \n",
       "84                           860011106_TCRB  \n",
       "85              BS-GIGI_10-replacement_TCRB  \n",
       "86  BS-EQ-25-T1_BS-GIGI-71-replacement_TCRB  \n",
       "87                           860011240_TCRB  \n",
       "\n",
       "[88 rows x 12 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This metadata was created in covid_tcr_immunecode_metadata.ipynb. We filtered to peak disease samples.\n",
    "adaptive_tcr_covid_specimens = (\n",
    "    pd.read_csv(\n",
    "        config.paths.metadata_dir\n",
    "        / \"adaptive\"\n",
    "        / \"generated.immunecode_covid_tcr.specimens.tsv\",\n",
    "        sep=\"\\t\",\n",
    "    )\n",
    "    .assign(\n",
    "        study=\"immunecode\",\n",
    "        species=\"Human\",\n",
    "    )\n",
    "    .rename(columns={\"ethnicity_condensed\": \"ethnicity\"})\n",
    ")\n",
    "adaptive_tcr_covid_specimens[\"sample_name\"] = adaptive_tcr_covid_specimens[\n",
    "    \"specimen_label\"\n",
    "]\n",
    "\n",
    "adaptive_tcr_covid_specimens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# All others\n",
    "\n",
    "First a general survey:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>study</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIP15860</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP14363</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*02, HLA-A*11, HLA-B*35, Human...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIP14178</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*01, HLA-A*03, HLA-B*08, HLA-B...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIP13944</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*02, HLA-A*03, HLA-B*18, HLA-B...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP13911</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>22 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1645</th>\n",
       "      <td>015V09001205_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.539726027 years_rel_to_art, 41.89589041 age...</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1646</th>\n",
       "      <td>015V09003735_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>0.073972603 years_rel_to_art, 41.89589041 age_...</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1647</th>\n",
       "      <td>015V10006273_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>1.2 years_rel_to_art, 41.89589041 age_at_min_e...</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1648</th>\n",
       "      <td>015V12002996_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>2.950819672 years_rel_to_art, 41.89589041 age_...</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1649</th>\n",
       "      <td>015V14000020_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>4.224657534 years_rel_to_art, 41.89589041 age_...</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1650 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample_name locus  \\\n",
       "0              HIP15860  TCRB   \n",
       "1              HIP14363  TCRB   \n",
       "2              HIP14178  TCRB   \n",
       "3              HIP13944  TCRB   \n",
       "4              HIP13911  TCRB   \n",
       "...                 ...   ...   \n",
       "1645  015V09001205_CFAR  TCRB   \n",
       "1646  015V09003735_CFAR  TCRB   \n",
       "1647  015V10006273_CFAR  TCRB   \n",
       "1648  015V12002996_CFAR  TCRB   \n",
       "1649  015V14000020_CFAR  TCRB   \n",
       "\n",
       "                                            sample_tags                study  \\\n",
       "0     Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B...  emerson-2017-natgen   \n",
       "1     Cohort 01, HLA-A*02, HLA-A*11, HLA-B*35, Human...  emerson-2017-natgen   \n",
       "2     Cohort 01, HLA-A*01, HLA-A*03, HLA-B*08, HLA-B...  emerson-2017-natgen   \n",
       "3     Cohort 01, HLA-A*02, HLA-A*03, HLA-B*18, HLA-B...  emerson-2017-natgen   \n",
       "4     22 Years, Caucasian, Cohort 01, Cytomegaloviru...  emerson-2017-natgen   \n",
       "...                                                 ...                  ...   \n",
       "1645  -0.539726027 years_rel_to_art, 41.89589041 age...   towlerton-2022-hiv   \n",
       "1646  0.073972603 years_rel_to_art, 41.89589041 age_...   towlerton-2022-hiv   \n",
       "1647  1.2 years_rel_to_art, 41.89589041 age_at_min_e...   towlerton-2022-hiv   \n",
       "1648  2.950819672 years_rel_to_art, 41.89589041 age_...   towlerton-2022-hiv   \n",
       "1649  4.224657534 years_rel_to_art, 41.89589041 age_...   towlerton-2022-hiv   \n",
       "\n",
       "     counting_method       primer_set species  \n",
       "0                NaN              NaN     NaN  \n",
       "1                NaN              NaN     NaN  \n",
       "2                NaN              NaN     NaN  \n",
       "3                NaN              NaN     NaN  \n",
       "4                NaN              NaN     NaN  \n",
       "...              ...              ...     ...  \n",
       "1645              v3  Human-TCRB-PD1x   Human  \n",
       "1646              v3  Human-TCRB-PD1x   Human  \n",
       "1647              v3  Human-TCRB-PD1x   Human  \n",
       "1648              v3  Human-TCRB-PD1x   Human  \n",
       "1649              v3  Human-TCRB-PD1x   Human  \n",
       "\n",
       "[1650 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive = []\n",
    "\n",
    "cols_desired = [\n",
    "    \"sample_name\",\n",
    "    \"locus\",\n",
    "    \"sample_tags\",\n",
    "    \"counting_method\",\n",
    "    \"primer_set\",\n",
    "    \"species\",\n",
    "    \"study\",\n",
    "]\n",
    "\n",
    "\n",
    "def _load_metadata_df(study, filter_cols=True):\n",
    "    df = pd.read_csv(\n",
    "        config.paths.metadata_dir / \"adaptive\" / f\"{study}.tsv\", sep=\"\\t\"\n",
    "    ).assign(study=study)\n",
    "    if filter_cols:\n",
    "        # Filter down to any matching cols\n",
    "        return df[[c for c in cols_desired if c in df.columns]]\n",
    "    return df\n",
    "\n",
    "\n",
    "for study in [\n",
    "    \"emerson-2017-natgen\",\n",
    "    \"mitchell-2022-jcii\",\n",
    "    \"mustjoki-2017-natcomms\",\n",
    "    \"ramesh-2015-ci\",\n",
    "    \"TCRBv4-control\",\n",
    "    \"towlerton-2022-hiv\",\n",
    "]:\n",
    "    dfs_adaptive.append(_load_metadata_df(study))\n",
    "dfs_adaptive = pd.concat(dfs_adaptive, axis=0).reset_index(drop=True)\n",
    "dfs_adaptive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Human    864\n",
       "Name: species, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive[\"species\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "locus  primer_set        counting_method\n",
       "TCRB   Human-TCRB-PD1x   v2                  36\n",
       "                         v3                 250\n",
       "       Human-TCRB-PD4bx  v4                 354\n",
       "       Human-TCRB-PD4x   v4                 163\n",
       "       Human-TCRB-none   v1                  61\n",
       "dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive.groupby([\"locus\", \"primer_set\", \"counting_method\"]).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>study</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cohort 01</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HLA-A*01</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HLA-A*68</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HLA-B*14</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HLA-B*57</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1644</th>\n",
       "      <td>-0.873972603 years_rel_to_art</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1645</th>\n",
       "      <td>-0.539726027 years_rel_to_art</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1646</th>\n",
       "      <td>0.073972603 years_rel_to_art</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1647</th>\n",
       "      <td>1.2 years_rel_to_art</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1649</th>\n",
       "      <td>4.224657534 years_rel_to_art</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1790 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        sample_tags                study\n",
       "0                         Cohort 01  emerson-2017-natgen\n",
       "0                          HLA-A*01  emerson-2017-natgen\n",
       "0                          HLA-A*68  emerson-2017-natgen\n",
       "0                          HLA-B*14  emerson-2017-natgen\n",
       "0                          HLA-B*57  emerson-2017-natgen\n",
       "...                             ...                  ...\n",
       "1644  -0.873972603 years_rel_to_art   towlerton-2022-hiv\n",
       "1645  -0.539726027 years_rel_to_art   towlerton-2022-hiv\n",
       "1646   0.073972603 years_rel_to_art   towlerton-2022-hiv\n",
       "1647           1.2 years_rel_to_art   towlerton-2022-hiv\n",
       "1649   4.224657534 years_rel_to_art   towlerton-2022-hiv\n",
       "\n",
       "[1790 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_tags = pd.merge(\n",
    "    # explode the sample tags\n",
    "    dfs_adaptive[\"sample_tags\"].str.split(\",\").explode(),\n",
    "    # link back to study\n",
    "    dfs_adaptive[\"study\"],\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    "    how=\"left\",\n",
    "    validate=\"m:1\",\n",
    ").dropna(subset=\"sample_tags\")\n",
    "\n",
    "sample_tags[\"sample_tags\"] = sample_tags[\"sample_tags\"].str.strip()\n",
    "sample_tags = sample_tags.drop_duplicates()\n",
    "sample_tags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "{\"message\": \"/var/folders/19/sqcl8gz12zg22_sgdy5g554r0000gp/T/ipykernel_36682/1693548041.py:1: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.\\n  for k, v in sample_tags.groupby(\\\"study\\\")[\\\"sample_tags\\\"].unique().iteritems():\\n\", \"time\": \"2024-04-01T14:13:14.648323\"}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCRBv4-control\n",
      "67 Years, Blood, Cohort 3, Healthy, Hispanic, Inactive, Male, Non-smoker, 60 Years, Caucasian, Female, 40 Years, Smoker, 29 Years, Black or African American, 61 Years, Head and Neck Cancer, Never Smoked, PBMC, 38 Years, Unknown Smoking Status, 68 Years, Ex-smoker, Lung Cancer, 36 Years, 27 Years, 52 Years, 45 Years, 55 Years, 69 Years, Cigar Smoker, 66 Years, 32 Years, 43 Years, 56 Years, 30 Years, 59 Years, Asian or Pacific Islander, 54 Years, 84 Years, 20 Years, 24 Years, 64 Years, 46 Years, 72 Years, 80 Years, 73 Years, 75 Years, Colorectal Cancer, 50 Years, FFPE, Melanoma, 35 Years, 48 Years, Native American or Alaska Native, 31 Years, 57 Years, 65 Years, 63 Years, 33 Years, 25 Years, 23 Years, 37 Years, 34 Years, 21 Years, 51 Years, 49 Years, 26 Years, Smokeless Tobacco User, 58 Years, 47 Years, 62 Years, 74 Years, 78 Years, 53 Years, 42 Years, 81 Years\n",
      "\n",
      "emerson-2017-natgen\n",
      "Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B*57, Human, Inferred HLA-A*01, Inferred HLA-A*33, Inferred HLA-B*14, Inferred HLA-B*57, PBMC, Peripheral blood lymphocytes (PBL), T cells, Unknown Ethnicity, Unknown racial group, gDNA, HLA-A*02, HLA-A*11, HLA-B*35, Inferred HLA-A*02, Inferred HLA-B*35, HLA-A*03, HLA-B*08, HLA-B*40, Inferred HLA-A*03, Inferred HLA-B*08, HLA-B*18, HLA-B*44, Inferred HLA-A*29, Inferred HLA-B*18, Inferred HLA-B*44, 22 Years, Caucasian, Cytomegalovirus +, HLA-B*15, HLA-B*38, Inferred CMV +, Inferred HLA-A*68, Inferred HLA-B*38, Male, Non-Hispanic or Latino, 56 Years, HLA-A*24, HLA-B*07, Inferred HLA-A*24, 17 Years, Cytomegalovirus -, Female, HLA-A*31, HLA-B*60, HLA-B*70, Inferred CMV -, 20 Years, Asian or Pacific Islander, Cohort 02, 24 Years, 23 Years, Hispanic or Latino, 19 Years, 26 Years, 25 Years, 28 Years, 21 Years, 27 Years, 31 Years, 33 Years, Black or African American, 32 Years, 34 Years, 35 Years, 29 Years, 30 Years, Native American or Alaska Native, 50 Years, HLA-B*62, Inferred HLA-B*15, Inferred HLA-B*62, 64 Years, Inferred HLA-B*60, HLA-A*30, HLA-B*13, HLA-B*61, 40 Years, HLA-A*23, HLA-B*49, Inferred HLA-A*23, Inferred HLA-B*07, Inferred HLA-B*49, HLA-A*66, HLA-B*41, HLA-A*26, 54 Years, HLA-B*51, Inferred HLA-B*51, Inferred HLA-A*11, HLA-B*37, HLA-A*29, Inferred HLA-A*30, Inferred HLA-B*13, 37 Years, HLA-A*34, Inferred HLA-B*52, Inferred HLA-B*61, HLA-A*32, 42 Years, HLA-A*80, HLA-B*58, Inferred HLA-A*32, 46 Years, 41 Years, HLA-A*33, HLA-B*27, Inferred HLA-B*27, 48 Years, HLA-B*45, HLA-B*50, 39 Years, 11 Years, HLA-B*55, Inferred HLA-B*55, 53 Years, 44 Years, Inferred HLA-A*26, HLA-B*39, 15 Years, 38 Years, 66 Years, Inferred HLA-B*39, HLA-A*28, 55 Years, Inferred HLA-B*40, 58 Years, 62 Years, HLA-A*74, HLA-B*53, 59 Years, 18 Years, 51 Years, 36 Years, 13 Years, 47 Years, 71 Years, HLA-B*48, HLA-B*54, 61 Years, HLA-A*25, HLA-B*52, 14 Years, 12 Years, 43 Years, 49 Years, Inferred HLA-B*37, 74 Years, 45 Years, 52 Years, 57 Years, Inferred HLA-A*25, 65 Years, 05 Years, 07 Years, HLA-B*63, 10 Years, 08 Years, HLA-A*69, Inferred HLA-A*31, 60 Years, Inferred HLA-B*58, HLA-B*56, 02 Years, 03 Years, 09 Years, HLA-B*05, HLA-B*17, 70 Years, 67 Years, 63 Years, HLA-B*42, HLA-B*75, 06 Years, 01 Years, 16 Years, HLA-B*47, HLA-B*22, 72 Years, HLA-B*73, HLA-B*78, HLA-B*82, HLA-A*36\n",
      "\n",
      "mitchell-2022-jcii\n",
      "03 Years, 12.6575342465753 Years at diagnosis, 3 Years at visit, Case, Caucasian, Female, GAD65 0, HLA-A*2402, HLA-A*3002, HLA-B*0801, HLA-B*1501, HLA-C*0304, HLA-C*0701, HLA-DPA1*0103, HLA-DPB1*0201, HLA-DPB1*0401, HLA-DQA1*0301, HLA-DQB1*0302, HLA-DRB1*0401, IA-2 0, IAA 0.005, Non-Hispanic or Latino, Subject 016, Timepoint 1, ZnT8 0, 06 Years, 6 Years at visit, Control, Control 016, HLA-A*0201, HLA-A*6601, HLA-B*1801, HLA-B*4001, HLA-C*0501, HLA-DPA1*0201, HLA-DPB1*1301, HLA-DQA1*0102, HLA-DQA1*0501, HLA-DQB1*0201, HLA-DQB1*0604, HLA-DRB1*0301, HLA-DRB1*1302, Hispanic or Latino, IA-2 3, IAA 0.001, Timepoint 4, ZnT8 NR, 11.5095890410959 Years at diagnosis, GAD65 698, HLA-A*0101, HLA-B*4402, HLA-DPB1*0402, HLA-DQA1*0303, HLA-DQB1*0301, IAA 0.228, Male, Subject 020, Timepoint 3, ZnT8 0.072, 0 Years at visit, 09 Months, Control 006, HLA-A*0301, IAA 0.004, 08 Years, 8 Years at visit, HLA-B*0702, HLA-C*0702, Subject 010, Timepoint 2, 05 Years, 5 Years at visit, Control 013, IAA 0, 07 Years, 10.558904109589 Years at diagnosis, 7 Years at visit, GAD65 7, HLA-B*1518, HLA-C*0704, HLA-DPB1*0301, HLA-DRB1*0404, IAA 0.002, Subject 012, Control 012, HLA-A*0206, HLA-A*2902, HLA-B*4002, HLA-C*0305, HLA-DPB1*0202, HLA-DPB1*1401, HLA-DRB1*0407, Control 009, HLA-A*1101, HLA-B*3501, HLA-C*0401, HLA-DQA1*0101, HLA-DQB1*0501, HLA-DRB1*0103, Control 021, HLA-C*0303, Control 003, HLA-B*4403, HLA-C*1601, HLA-DQA1*0505, HLA-DRB1*1101, IAA 0.003, 01 Years, 1 Years at visit, Control 019, HLA-A*3201, HLA-DQA1*0104, HLA-DQB1*0503, HLA-DRB1*1454, 6.87397260273973 Years at diagnosis, GAD65 137, HLA-A*2601, HLA-B*3502, HLA-DQB1*0305, HLA-DRB1*0403, Subject 013, ZnT8 0.11, 18.0438356164384 Years at diagnosis, Subject 021, 10 Years, 10 Years at visit, 14.4328767123288 Years at diagnosis, GAD65 68, HLA-A*2501, HLA-C*1203, HLA-DPB1*2301, HLA-DQA1*0103, HLA-DQB1*0602, HLA-DQB1*0603, HLA-DRB1*1301, HLA-DRB1*1501, Subject 019, 12 Years, 12 Years at visit, 11.2082191780822 Years at diagnosis, GAD65 81, HLA-B*2705, HLA-C*0202, HLA-DPB1*1101, IAA 0.006, Subject 024, ZnT8 0.105, 19 Years, 19 Years at visit, Control 011, HLA-B*5101, GAD65 94, HLA-A*6801, HLA-C*1502, IAA 0.019, Subject 014, GAD65 4, IA-2 321, IAA 0.011, ZnT8 0.039, 02 Years, 2 Years at visit, Control 017, HLA-C*0409, HLA-DRB1*0101, 6.37808219178082 Years at diagnosis, GAD65 99, HLA-B*3801, HLA-DQA1*0401, HLA-DQB1*0402, HLA-DRB1*0402, HLA-DRB1*0801, IA-2 456, IAA 0.145, Subject 022, Control 018, HLA-B*1503, HLA-C*0210, 13 Years, 13 Years at visit, 13.0465753424658 Years at diagnosis, GAD65 1, IA-2 414, ZnT8 0.543, Control 020, Control 001, HLA-B*3503, HLA-DPB1*2001, IA-2 5, 09 Years, 9 Years at visit, GAD65 466, IA-2 406, IAA 0.038, Control 007, HLA-DPB1*0101, 19.8630136986301 Years at diagnosis, Subject 008, ZnT8 0.128, Control 015, HLA-B*5701, HLA-C*0602, HLA-DQA1*0201, HLA-DQB1*0303, HLA-DRB1*0701, 11 Years, 11 Years at visit, GAD65 72, IAA 0.014, IA-2 81, ZnT8 0.036, GAD65 11, IA-2 273, ZnT8 0.32, GAD65 17, IAA 0.01, Subject 018, ZnT8 0.016, Control 005, HLA-B*4102, HLA-C*1703, GAD65 36, IA-2 161, ZnT8 0.141, GAD65 312, HLA-B*1302, HLA-DPA1*0206, HLA-DPB1*0501, IA-2 230, Subject 004, ZnT8 0.546, GAD65 326, IAA 0.052, ZnT8 0.009, HLA-DQB1*0202, Subject 017, ZnT8 0.001, 8.4054794520548 Years at diagnosis, GAD65 5, IA-2 422, IAA 0.026, Subject 007, ZnT8 0.25, 14 Years, 14 Years at visit, 14.5561643835616 Years at diagnosis, GAD65 63, HLA-B*5601, HLA-C*0102, IAA 0.02, Subject 023, 9.56986301369863 Years at diagnosis, HLA-DPB1*0601, IA-2 593, IAA 0.069, Subject 011, 04 Years, 4 Years at visit, GAD65 80, IA-2 517, GAD65 255, IAA 0.045, IAA 0.008, ZnT8 0.012, Control 022, IA-2 2, 11.6821917808219 Years at diagnosis, HLA-A*3101, IAA 0.046, Subject 005, IA-2 387, ZnT8 0.553, 10 Months, IAA 0.007, Control 010, HLA-B*3901, IA-2 193, ZnT8 0.003, GAD65 60, GAD65 25, IA-2 28, IAA 0.024, HLA-B*5001, Subject 006, ZnT8 0.027, 22.2739726027397 Years at diagnosis, GAD65 8, HLA-C*1602, Subject 009, Control 004, HLA-DPB1*1701, Control 025, HLA-DQB1*0502, HLA-DRB1*1601, GAD65 32, IA-2 34, Subject 002, ZnT8 0.033, 12.958904109589 Years at diagnosis, HLA-B*3906, Subject 001, IAA 0.11, Control 002, HLA-DRB1*1201, IAA 0.009, IAA 0.229, 08 Months, Control 024, HLA-A*2301, IA-2 1, GAD65 58, ZnT8 0.031, GAD65 24, Control 023, HLA-A*3001, HLA-B*5801, HLA-C*0302, HLA-DQB1*0609, GAD65 481, Subject 003, GAD65 371, IAA 0.047, GAD65 10, IA-2 457, Subject 015, ZnT8 0.691, GAD65 200, GAD65 62, IAA 0.033, Control 014, HLA-B*1402, HLA-C*0802, HLA-DPB1*135, HLA-DRB1*0102, GAD65 9, GAD65 528, IAA 0.174, GAD65 2, IA-2 NR, GAD65 625, ZnT8 0.53, IAA 0.025, GAD65 771, IAA 0.013, ZnT8 0.053, Control 008, IAA 0.075, GAD65 3, IA-2 50, GAD65 83, IA-2 164, IA-2 212, ZnT8 0.109, GAD65 182, IA-2 115, ZnT8 0.313, GAD65 35, IA-2 60, ZnT8 0.017, 11.5260273972603 Years at diagnosis, ZnT8 0.643, GAD65 163, IA-2 328, ZnT8 0.58, GAD65 157, ZnT8 0.596, GAD65 74, GAD65 42, GAD65 280, IA-2 187, IAA 0.022, ZnT8 0.295, GAD65 201, ZnT8 0.006, GAD65 298, IA-2 12, 17 Years, 17 Years at visit, GAD65 13, GAD65 70, IAA 0.012, IA-2 459, GAD65 78, IA-2 148, ZnT8 0.002, GAD65 249, IAA 0.27, ZnT8 0.008, 17.7561643835616 Years at diagnosis, GAD65 223, IA-2 228, ZnT8 0.24, IA-2 6, GAD65 116, IA-2 42, Subject 025, ZnT8 0.122, 10.0493150684932 Years at diagnosis, HLA-DQA1*0302, HLA-DRB1*0901, IA-2 536, IAA 0.017, Subject 029, ZnT8 0.594, 22 Years, 22 Years at visit, 22.9452054794521 Years at diagnosis, GAD65 519, HLA-A*0205, Subject 027, ZnT8 0.028, GAD65 266, HLA-DPB1*1601, IAA 0.755, Subject 028, ZnT8 0.223, HLA-B*0706, HLA-C*1505, HLA-DPB1*12401, HLA-DRB1*0405, IA-2 145, Subject 026, ZnT8 0.005, IA-2 472, ZnT8 0.239, GAD65 308, IA-2 4, GAD65 NR, IA-2 21, 15 Years, 15 Years at visit, 15.0465753424658 Years at diagnosis, IA-2 533, IAA 0.397, ZnT8 0.062, GAD65 340, IA-2 8, 12-18 Years (Adolescence), 15 Years at diagnosis, GAD65 522, IA-2 140, IAA 0.051, T1D 006, T1D for 2 days, Type 1 Diabetes, Whole Blood, ZnT8 0.01, 14 Years at diagnosis, T1D 002, T1D for 15 days, 02-12 (Childhood), 9 Years at diagnosis, GAD65 452, Hispanic, IA-2 238, T1D 041, T1D for 1 days, ZnT8 0.266, 8 Years at diagnosis, GAD65 124, T1D 116, T1D for 3 days, 12 Years at diagnosis, GAD65 265, IA-2 290, T1D 058, ZnT8 0.138, 5 Years at diagnosis, IA-2 246, T1D 004, ZnT8 0.03, T1D 079, T1D for 17 days, 10 Years at diagnosis, IA-2 220, IAA 0.04, T1D 104, T1D for 0 days, ZnT8 0.28, 11 Years at diagnosis, GAD65 79, IA-2 146, PBMC, T1D 110, T1D for 4 days, GAD65 226, IA-2 236, T1D 021, ZnT8 0.484, GAD65 29, IA-2 24, T1D 007, ZnT8 0.383, GAD65 150, HLA-DRB1*0410, IA-2 172, T1D 119, ZnT8 0.043, GAD65 580, IA-2 64, IAA 0.059, T1D 029, ZnT8 0.089, GAD65 264, IA-2 315, IAA 0.035, T1D 075, Unknown Ethnicity, Unknown racial group, ZnT8 0.783, 4 Years at diagnosis, GAD65 38, IA-2 184, IAA 0.161, T1D 105, T1D for 12 days, IA-2 464, T1D 062, T1D for 5 days, IA-2 312, IAA 0.41, T1D 069, T1D for 13 days, ZnT8 0.398, 18 Years, 18 Years at diagnosis, 18-24 Years (Early Adulthood), GAD65 126, IA-2 262, T1D 131, ZnT8 0.403, 13 Years at diagnosis, IA-2 133, IAA 0.015, T1D 048, ZnT8 0.761, HLA-DRB1*1305, IA-2 254, T1D 049, ZnT8 0.479, GAD65 33, IA-2 460, T1D 133, ZnT8 0.625, 16 Years, 16 Years at diagnosis, Black or African American, T1D 113, T1D for 14 days, ZnT8 0.004, GAD65 114, IA-2 213, IAA 0.15, T1D 123, ZnT8 0.804, GAD65 16, IA-2 336, T1D 020, T1D for 7 days, IA-2 235, T1D 130, ZnT8 0.231, 7 Years at diagnosis, IA-2 256, T1D 093, T1D for 6 days, IA-2 331, IAA 0.021, T1D 074, ZnT8 0.312, Asian or Pacific Islander, HLA-DQA1*0503, HLA-DRB1*1402, IA-2 198, T1D 061, GAD65 184, IA-2 69, T1D 071, T1D for 8 days, GAD65 107, IA-2 283, IAA 0.049, T1D 008, ZnT8 0.694, GAD65 658, HLA-DQB1*0401, IA-2 324, IAA 0.104, T1D 096, IA-2 278, IAA 0.03, T1D 129, T1D for 16 days, ZnT8 0.581, GAD65 699, IA-2 317, IAA 0.09, T1D 100, ZnT8 0.26, GAD65 53, IA-2 144, IAA 0.016, T1D 099, ZnT8 0.097, GAD65 642, IA-2 267, IAA 0.233, T1D 039, ZnT8 0.84, GAD65 6, IA-2 371, IAA 0.029, T1D 117, ZnT8 0.14, 17 Years at diagnosis, GAD65 750, T1D 109, ZnT8 0.652, GAD65 12, T1D 031, ZnT8 0.011, GAD65 762, T1D 053, T1D 027, ZnT8 0.45, GAD65 416, IA-2 203, T1D 121, ZnT8 0.05, T1D 111, GAD65 39, IA-2 179, T1D 080, ZnT8 0.236, GAD65 540, IA-2 465, IAA 0.073, T1D 118, GAD65 130, T1D 047, ZnT8 0.184, GAD65 19, HLA-DRB1*1104, T1D 112, ZnT8 0.023, 21 Years, 21 Years at diagnosis, 24-45 Years (Middle Adulthood), GAD65 921, IA-2 368, T1D 040, ZnT8 0.095, IA-2 86, T1D 077, 29 Years, 29 Years at diagnosis, GAD65 475, IA-2 565, IAA 0.336, T1D 052, ZnT8 0.119, 6 Years at diagnosis, GAD65 129, HLA-DRB1*1103, IA-2 393, T1D 063, ZnT8 0.12, GAD65 113, IA-2 334, IAA 0.142, T1D 070, T1D for 10 days, ZnT8 0.448, T1D 083, T1D 082, IA-2 37, T1D 059, ZnT8 0.021, T1D 140, ZnT8 0.458, GAD65 243, IA-2 412, T1D 087, ZnT8 0.303, GAD65 178, IA-2 195, T1D 016, ZnT8 0.041, IA-2 93, T1D 128, ZnT8 0.629, T1D 017, ZnT8 0.081, IA-2 239, T1D 005, ZnT8 0.27, GAD65 30, IA-2 71, T1D 132, IA-2 112, T1D 011, ZnT8 0.063, GAD65 82, IA-2 482, T1D 134, ZnT8 0.686, GAD65 14, IAA 0.089, T1D 143, T1D for 37 days, ZnT8 0.495, GAD65 23, T1D 001, ZnT8 0.621, GAD65 212, IAA 0.07, T1D 035, GAD65 662, IA-2 20, T1D 086, GAD65 168, HLA-DRB1*1404, IA-2 381, T1D 139, T1D for 55 days, GAD65 218, IA-2 94, T1D 067, HLA-DRB1*0802, T1D 115, GAD65 238, T1D 135, IA-2 149, T1D 025, T1D 098, GAD65 587, HLA-DQB1*0319, HLA-DRB1*1102, IA-2 357, T1D 092, ZnT8 0.453, IA-2 274, T1D 056, ZnT8 0.234, GAD65 513, T1D 036, ZnT8 0.401, T1D 012, T1D 085, ZnT8 0.359, 3 Years at diagnosis, IA-2 79, T1D 076, GAD65 785, IA-2 378, T1D 065, ZnT8 0.71, GAD65 768, IA-2 143, IAA 0.298, T1D 101, ZnT8 0.536, GAD65 279, HLA-DRB1*1602, IA-2 271, IAA 0.096, T1D 142, GAD65 64, IA-2 335, T1D 032, GAD65 622, IA-2 44, T1D 022, ZnT8 0.029, IA-2 285, IAA 0.055, T1D 072, ZnT8 0.157, IA-2 35, T1D 126, IA-2 15, T1D 136, ZnT8 0.047, 27 Years, 27 Years at diagnosis, GAD65 623, T1D 009, ZnT8 0.77, GAD65 613, IA-2 348, IAA 0.031, T1D 141, ZnT8 0.208, GAD65 270, T1D 013, ZnT8 0.038, GAD65 122, T1D 051, ZnT8 0.044, IA-2 118, T1D 088, ZnT8 0.329, GAD65 229, IA-2 147, T1D 106, GAD65 402, IA-2 313, T1D 114, ZnT8 0.007, 19 Years at diagnosis, T1D 055, T1D 138, ZnT8 0.228, GAD65 18, T1D 046, IA-2 385, T1D 019, ZnT8 0.351, GAD65 120, IA-2 245, IAA 0.043, T1D 097, IA-2 555, IAA 0.176, T1D 091, ZnT8 0.244, GAD65 770, IA-2 388, T1D 038, GAD65 98, T1D 068, GAD65 635, T1D 066, T1D for 50 days, ZnT8 0.159, GAD65 889, IA-2 359, T1D 090, GAD65, IA-2, IAA, T1D 034, ZnT8, GAD65 44, IA-2 248, T1D 042, ZnT8 0.429, GAD65 67, IA-2 207, T1D 127, ZnT8 0.061, IA-2 358, T1D 045, ZnT8 0.145, GAD65 109, IA-2 251, T1D 030, ZnT8 0.16, T1D 084, GAD65 26, T1D 033, IA-2 322, T1D 073, ZnT8 0.768, IA-2 319, T1D 054, ZnT8 0.175, GAD65 853, HLA-DQA1*0402, IA-2 409, T1D 026, IA-2 178, T1D 024, GAD65 27, IA-2 237, T1D 089, ZnT8 0.073, GAD65 69, IA-2 677, T1D 120, ZnT8 0.151, GAD65 303, IA-2 294, T1D 108, GAD65 446, HLA-DRB1*1303, IA-2 242, IAA 0.095, T1D 060, ZnT8 0.216, GAD65 28, IA-2 296, T1D 122, GAD65 31, IA-2 428, T1D 028, IAA 0.037, T1D 018, IA-2 261, T1D 125, GAD65 648, IAA 0.206, T1D 107, GAD65 847, IA-2 377, T1D 124, GAD65 703, HLA-DRB1*1001, IA-2 196, T1D 003, IA-2 53, T1D 050, ZnT8 0.202, IA-2 36, T1D 095, GAD65 54, IA-2 266, T1D 078, GAD65 51, T1D 064, GAD65 616, HLA-DQB1*0601, HLA-DRB1*1502, IA-2 500, T1D 103, T1D for 34 days, ZnT8 0.059, GAD65 757, IA-2 342, T1D 137, ZnT8 0.772, T1D 010, ZnT8 0.262, GAD65 185, HLA-DQB1*0304, HLA-DRB1*0408, IA-2 122, T1D 037, ZnT8 0.125, IA-2 384, T1D 014, ZnT8 0.805, GAD65 1002, T1D 102, T1D for 9 days, ZnT8 0.21, GAD65 411, T1D 044, ZnT8 0.183, GAD65 86, T1D 023, GAD65 644, T1D 057, ZnT8 0.078, GAD65 286, IA-2 403, T1D 043, ZnT8 0.952, GAD65 84, IA-2 217, T1D 081, ZnT8 0.187, GAD65 859, T1D 015, ZnT8 1.087, T1D 094\n",
      "\n",
      "mustjoki-2017-natcomms\n",
      "HLA MHC Class I, CD8+, Peripheral blood lymphocytes (PBL), T cells, Rheumatoid Arthritis, Synovial fluid, CD4+, CD4-, CD8-, T Cell Large Granular Lymphocyte Leukemia (LGL)\n",
      "\n",
      "ramesh-2015-ci\n",
      "PBMC, HLA MHC Class I, Blood\n",
      "\n",
      "towlerton-2022-hiv\n",
      "4.906849315 years_rel_to_art, 40.48219178 age_at_min_evidence, 47.56438356 age_at_art, 52 age_at_collection, <= 5 years Post-ART time_group, A*020101 aAllele1, A*030101 aAllele2, B*400102 bAllele1, B*440201 bAllele2, C*030401 cAllele1, C*050101 cAllele2, DPA1*010301 DPA1allele1, DPA1*010301 DPA1allele2, DPB1*030101 DPB1allele1, DPB1*040101 DPB1allele2, DQA1*030301 DQA1allele1, DQA1*040101 DQA1allele2, DQB1*030101 DQB1allele1, DQB1*04020101 DQB1allele2, DRB1*040101 DRB1allele1, DRB1*080101 DRB1allele2, DRB345*Not_Present DRB345allele1, DRB4*010301 DRB345allele2, Female, NA age_at_death, Patient ID 1016, Positive HIV_status, Post-ART timePoint, 22.38630137 age_at_min_evidence, 25.95890411 age_at_art, 31 age_at_collection, 5.312328767 years_rel_to_art, <= 6 years Post-ART time_group, A*330101 aAllele1, A*680102 aAllele2, B*140201 bAllele1, B*390501 bAllele2, C*070201 cAllele1, C*080201 cAllele2, DPB1*040101 DPB1allele1, DPB1*040201 DPB1allele2, DQA1*030101 DQA1allele1, DQA1*050101 DQA1allele2, DQB1*020101 DQB1allele1, DQB1*030201 DQB1allele2, DRB1*030101 DRB1allele1, DRB1*040301 DRB1allele2, DRB3*020201 DRB345allele1, Male, Patient ID 1024, 39.70491803 age_at_min_evidence, 43.71038251 age_at_art, 50 age_at_collection, 50 age_at_death, 6.917808219 years_rel_to_art, <= 7 years Post-ART time_group, A*250101 aAllele1, A*680201 aAllele2, B*180101 bAllele1, B*530101 bAllele2, C*040101 cAllele1, C*120301 cAllele2, DPB1*1040101 DPB1allele2, DQA1*010201 DQA1allele1, DQA1*010201 DQA1allele2, DQB1*060201 DQB1allele1, DQB1*060401 DQB1allele2, DRB1*130201 DRB1allele1, DRB1*150101 DRB1allele2, DRB3*030101 DRB345allele1, DRB5*010101 DRB345allele2, Patient ID 1009, 40.1420765 age_at_min_evidence, 42.09041096 age_at_art, 47 age_at_collection, 5.150684932 years_rel_to_art, A*110101 aAllele2, B*400201 bAllele1, C*020202 cAllele1, DPB1*020201 DPB1allele1, DPB1*030101 DPB1allele2, DQA1*010301 DQA1allele1, DQB1*060301 DQB1allele2, DRB1*130101 DRB1allele2, DRB3*020201 DRB345allele2, Patient ID 1023, 40.60547945 age_at_min_evidence, 45.6147541 age_at_art, 53 age_at_collection, 8.180821918 years_rel_to_art, <= 9 years Post-ART time_group, A*010101 aAllele1, A*2317 aAllele2, B*440301 bAllele1, B*8101 bAllele2, C*1801 cAllele2, DPA1*020202 DPA1allele1, DPA1*030101 DPA1allele2, DPB1*010101 DPB1allele1, DPB1*1050101 DPB1allele2, DQA1*040101 DQA1allele1, DQA1*050501 DQA1allele2, DQB1*031901 DQB1allele1, DRB1*030201 DRB1allele1, DRB1*110201 DRB1allele2, DRB3*010102 DRB345allele1, Patient ID 1007, 41.89589041 age_at_min_evidence, 47.7890411 age_at_art, 55 age_at_collection, 7.353424658 years_rel_to_art, <= 8 years Post-ART time_group, A*240301 aAllele1, B*450101 bAllele2, C*160101 cAllele2, DPB1*1801 DPB1allele2, DRB1*150301 DRB1allele2, Patient ID 1001, 5.252054795 years_rel_to_art, 7.202185792 years_rel_to_art, 35.66027397 age_at_min_evidence, 39.49589041 age_at_art, 45 age_at_collection, 6.193989071 years_rel_to_art, A*240201 aAllele2, B*270502 bAllele1, B*400102 bAllele2, C*010201 cAllele1, C*030401 cAllele2, DPB1*020102 DPB1allele1, DQB1*04020101 DQB1allele1, DQB1*060201 DQB1allele2, DRB1*080101 DRB1allele1, Patient ID 1015, 34.04383562 age_at_min_evidence, 41.24931507 age_at_art, 6.535519126 years_rel_to_art, B*070201 bAllele1, C*070201 cAllele2, DPA1*020101 DPA1allele1, DPA1*020101 DPA1allele2, DPB1*010102 DPB1allele1, DPB1*170101 DPB1allele2, DQA1*040102 DQA1allele2, DQB1*030201 DQB1allele1, DQB1*031901 DQB1allele2, DRB1*040401 DRB1allele1, DRB1*080401 DRB1allele2, Patient ID 1003, 22.85245902 age_at_min_evidence, 26.30874317 age_at_art, 32 age_at_collection, 6.578082192 years_rel_to_art, A*020101 aAllele2, B*520101 bAllele1, B*520101 bAllele2, C*120202 cAllele1, C*120202 cAllele2, DQA1*010501 DQA1allele2, DQB1*050101 DQB1allele1, DRB1*100101 DRB1allele1, Patient ID 1004, 33 age_at_collection, 6.895890411 years_rel_to_art, 23.42739726 age_at_min_evidence, 23.88767123 age_at_art, 9.183561644 years_rel_to_art, <= 10 years Post-ART time_group, A*320101 aAllele2, B*080101 bAllele1, C*070101 cAllele2, DPA1*020102 DPA1allele2, DPB1*020102 DPB1allele2, DRB1*040401 DRB1allele2, Patient ID 1012, 3.717808219 years_rel_to_art, 50.19726027 age_at_min_evidence, 50.42739726 age_at_art, 54 age_at_collection, <= 4 years Post-ART time_group, A*030101 aAllele1, A*250101 aAllele2, C*050101 cAllele1, DQA1*010101 DQA1allele1, DQA1*010301 DQA1allele2, DRB1*010101 DRB1allele1, DRB345*Not_Present DRB345allele2, Patient ID 1029, 49 age_at_collection, 6.224657534 years_rel_to_art, 4.06557377 years_rel_to_art, 6.4 years_rel_to_art, 48 age_at_collection, 7.602739726 years_rel_to_art, 6.278688525 years_rel_to_art, 5.065753425 years_rel_to_art, 10.39344262 years_rel_to_art, 34 age_at_collection, > 10 years Post-ART time_group, 4.387978142 years_rel_to_art, 48.06027397 age_at_min_evidence, 49.00821918 age_at_art, 58 age_at_collection, 9.346994536 years_rel_to_art, A*680101 aAllele2, B*440201 bAllele1, C*060201 cAllele2, DPB1*110101 DPB1allele1, DPB1*110101 DPB1allele2, DQA1*030301 DQA1allele2, DRB1*040501 DRB1allele1, DRB4*010301 DRB345allele1, Patient ID 1017, 7.642076503 years_rel_to_art, 0.273224044 years_rel_to_art, <= 3 months Post-ART time_group, 0.710382514 years_rel_to_art, 51 age_at_collection, <= 1 year Post-ART time_group, 0.860655738 years_rel_to_art, 32.06575342 age_at_min_evidence, 32.61917808 age_at_art, B*350101 bAllele2, DQA1*020101 DQA1allele2, DQB1*030302 DQB1allele1, DQB1*050101 DQB1allele2, DRB1*010301 DRB1allele1, DRB1*070101 DRB1allele2, DRB4*01030102N DRB345allele2, Patient ID 1028, -0.345205479 years_rel_to_art, <= 3 months Pre-ART time_group, Pre-ART timePoint, 1.320547945 years_rel_to_art, 27.13972603 age_at_min_evidence, 27.42739726 age_at_art, 28 age_at_collection, <= 2 years Post-ART time_group, DPB1*060101 DPB1allele2, Patient ID 1030, 3.106849315 years_rel_to_art, 30 age_at_collection, 4.326027397 years_rel_to_art, 0.567123288 years_rel_to_art, 27 age_at_collection, <= 6 months Post-ART time_group, 0 years_rel_to_art, <= 0 month Pre-ART time_group, 1.194520548 years_rel_to_art, 2.237704918 years_rel_to_art, <= 3 years Post-ART time_group, -0.090410959 years_rel_to_art, 1.095890411 years_rel_to_art, 0.517808219 years_rel_to_art, 28.74863388 age_at_min_evidence, 28.93442623 age_at_art, 29 age_at_collection, A*260101 aAllele2, B*380101 bAllele1, B*490101 bAllele2, C*120301 cAllele1, DRB1*110101 DRB1allele1, Patient ID 1025, 0.454794521 years_rel_to_art, 27.55890411 age_at_min_evidence, 27.75068493 age_at_art, B*150101 bAllele2, C*030301 cAllele1, Patient ID 1027, 0.778082192 years_rel_to_art, 1.810958904 years_rel_to_art, -0.054794521 years_rel_to_art, 0.325136612 years_rel_to_art, 1.249315068 years_rel_to_art, 2.249315068 years_rel_to_art, 5.876712329 years_rel_to_art, 2.890710383 years_rel_to_art, 46.58630137 age_at_min_evidence, 50.33972603 age_at_art, B*440301 bAllele2, DPB1*050101 DPB1allele2, DQB1*020201 DQB1allele1, DQB1*060901 DQB1allele2, DRB1*070101 DRB1allele1, DRB1*130201 DRB1allele2, DRB4*01010101 DRB345allele2, Patient ID 1026, 1.22739726 years_rel_to_art, 0.421917808 years_rel_to_art, -0.04109589 years_rel_to_art, -0.520547945 years_rel_to_art, <= 6 months Pre-ART time_group, -1.461748634 years_rel_to_art, 24 age_at_collection, <= 1 year Pre-ART time_group, -0.144808743 years_rel_to_art, -0.756164384 years_rel_to_art, 25 age_at_collection, 1.542465753 years_rel_to_art, 26.65753425 age_at_min_evidence, 27.76712329 age_at_art, C*070101 cAllele1, DRB4*01010101 DRB345allele1, Patient ID 1022, -0.443835616 years_rel_to_art, -0.504109589 years_rel_to_art, 41 age_at_collection, -0.101369863 years_rel_to_art, 0.301369863 years_rel_to_art, 26 age_at_collection, 3.449315068 years_rel_to_art, 2.284153005 years_rel_to_art, 44 age_at_collection, 4.117808219 years_rel_to_art, 46 age_at_collection, -2.531506849 years_rel_to_art, 46.96712329 age_at_min_evidence, 53.2630137 age_at_art, <= 2 years Pre-ART time_group, A*290201 aAllele1, A*300201 aAllele2, C*080201 cAllele1, Patient ID 1013, 0.56010929 years_rel_to_art, -1.68852459 years_rel_to_art, 40 age_at_collection, -1.057377049 years_rel_to_art, 2.010958904 years_rel_to_art, 49.6739726 age_at_min_evidence, 50.78630137 age_at_art, A*300201 aAllele1, A*310102 aAllele2, B*150301 bAllele1, C*040101 cAllele2, DPA1*020701 DPA1allele2, DPB1*850101 DPB1allele2, DQA1*010501 DQA1allele1, DRB1*090102 DRB1allele1, DRB1*100101 DRB1allele2, Patient ID 1021, 0.191780822 years_rel_to_art, 42 age_at_collection, 3.032876712 years_rel_to_art, 1.183561644 years_rel_to_art, 43 age_at_collection, -0.476712329 years_rel_to_art, 24.84931507 age_at_min_evidence, 25.84657534 age_at_art, B*150101 bAllele1, DPA1*020202 DPA1allele2, DQA1*010202 DQA1allele1, DQB1*050201 DQB1allele1, DRB1*130101 DRB1allele1, DRB1*160101 DRB1allele2, DRB5*020201 DRB345allele2, Patient ID 1020, -0.301369863 years_rel_to_art, 0.038251366 years_rel_to_art, 0.349726776 years_rel_to_art, 1.595628415 years_rel_to_art, 37.01369863 age_at_min_evidence, 40.54520548 age_at_art, A*230101 aAllele1, B*510101 bAllele1, B*580201 bAllele2, C*060201 cAllele1, DPB1*1050101 DPB1allele1, DRB1*120101 DRB1allele1, DRB1*130301 DRB1allele2, Patient ID 1019, 0.382513661 years_rel_to_art, 2.260273973 years_rel_to_art, 1.021917808 years_rel_to_art, 3.008219178 years_rel_to_art, 3.323287671 years_rel_to_art, 4.366120219 years_rel_to_art, -2.797814208 years_rel_to_art, 29.49589041 age_at_min_evidence, 34.20821918 age_at_art, B*270502 bAllele2, C*150502 cAllele2, DRB1*1304 DRB1allele2, Patient ID 1018, -1.750684932 years_rel_to_art, 2.276712329 years_rel_to_art, 2.967123288 years_rel_to_art, 56 age_at_collection, 7.304109589 years_rel_to_art, -0.01369863 years_rel_to_art, 0.317808219 years_rel_to_art, 1.010928962 years_rel_to_art, -1.109589041 years_rel_to_art, 0.819178082 years_rel_to_art, 35 age_at_collection, 4.635616438 years_rel_to_art, 1.87704918 years_rel_to_art, 36 age_at_collection, 2.890410959 years_rel_to_art, 37 age_at_collection, -2.090410959 years_rel_to_art, -1.456284153 years_rel_to_art, 1.942465753 years_rel_to_art, -0.076712329 years_rel_to_art, 45.68579235 age_at_min_evidence, 49.15027322 age_at_art, 52 age_at_death, A*230101 aAllele2, DQA1*020101 DQA1allele1, DQB1*020201 DQB1allele2, Patient ID 1014, 0.31147541 years_rel_to_art, -0.490410959 years_rel_to_art, 2.950819672 years_rel_to_art, 0.273972603 years_rel_to_art, 4.005479452 years_rel_to_art, 2.183561644 years_rel_to_art, 1.263013699 years_rel_to_art, 2.978082192 years_rel_to_art, 2.021857923 years_rel_to_art, 3.112328767 years_rel_to_art, -0.284931507 years_rel_to_art, 34.51780822 age_at_min_evidence, 37.42465753 age_at_art, A*240201 aAllele1, B*350101 bAllele1, C*160401 cAllele2, DPB1*100101 DPB1allele2, DRB1*110401 DRB1allele2, Patient ID 1010, 2.687671233 years_rel_to_art, 1.778082192 years_rel_to_art, 3.355191257 years_rel_to_art, -0.61369863 years_rel_to_art, -1.8 years_rel_to_art, 39 age_at_collection, -0.775342466 years_rel_to_art, 0.534246575 years_rel_to_art, -0.394520548 years_rel_to_art, 0.468493151 years_rel_to_art, 1.155737705 years_rel_to_art, 2.393442623 years_rel_to_art, 3.049315068 years_rel_to_art, 2.556164384 years_rel_to_art, 4.753424658 years_rel_to_art, -1.210382514 years_rel_to_art, 38 age_at_collection, 23 age_at_collection, 48.70410959 age_at_min_evidence, 48.91506849 age_at_art, 57 age_at_death, B*351201 bAllele1, B*570101 bAllele2, DPB1*040201 DPB1allele1, DQA1*050301 DQA1allele1, DQB1*030101 DQB1allele2, DRB1*130301 DRB1allele1, DRB1*140201 DRB1allele2, DRB3*010102 DRB345allele2, Patient ID 1011, 0.290410959 years_rel_to_art, 0.452054795 years_rel_to_art, 0.019178082 years_rel_to_art, 0.865753425 years_rel_to_art, 0.405479452 years_rel_to_art, 1.936986301 years_rel_to_art, 1.480874317 years_rel_to_art, 0.652054795 years_rel_to_art, -3.18630137 years_rel_to_art, <= 3 years Pre-ART time_group, -0.317808219 years_rel_to_art, 36.36065574 age_at_min_evidence, 37.87945205 age_at_art, Patient ID 1008, -3.071232877 years_rel_to_art, -1.824657534 years_rel_to_art, -1.783561644 years_rel_to_art, -1.257534247 years_rel_to_art, 1.580821918 years_rel_to_art, 2.079234973 years_rel_to_art, -0.571038251 years_rel_to_art, 0.167123288 years_rel_to_art, 3.38630137 years_rel_to_art, 2.071232877 years_rel_to_art, 3.103825137 years_rel_to_art, -4.97260274 years_rel_to_art, 49.24657534 age_at_min_evidence, 59.8 age_at_art, <= 4 years Pre-ART time_group, A*110101 aAllele1, B*080101 bAllele2, DQA1*050101 DQA1allele1, DQB1*020101 DQB1allele2, DRB1*030101 DRB1allele2, Patient ID 1006, -0.844262295 years_rel_to_art, -0.745901639 years_rel_to_art, 31.7890411 age_at_min_evidence, 36.47540984 age_at_art, B*350301 bAllele2, DRB1*150101 DRB1allele1, DRB5*010101 DRB345allele1, Patient ID 1005, 0.115068493 years_rel_to_art, 2.095890411 years_rel_to_art, -0.712328767 years_rel_to_art, 59 age_at_collection, 2.78630137 years_rel_to_art, 0.62295082 years_rel_to_art, 60 age_at_collection, 2.347945205 years_rel_to_art, 62 age_at_collection, 0.306010929 years_rel_to_art, -0.390710383 years_rel_to_art, -3.161202186 years_rel_to_art, -1.843835616 years_rel_to_art, 57 age_at_collection, 2.087671233 years_rel_to_art, 3.087431694 years_rel_to_art, 4.482191781 years_rel_to_art, 1.336986301 years_rel_to_art, 61 age_at_collection, -4.42739726 years_rel_to_art, -2.306849315 years_rel_to_art, 0.75890411 years_rel_to_art, 37.20821918 age_at_min_evidence, 50.03013699 age_at_art, A*330101 aAllele2, DRB1*040101 DRB1allele2, Patient ID 1002, -0.193989071 years_rel_to_art, 1.387978142 years_rel_to_art, -0.216438356 years_rel_to_art, 0.643835616 years_rel_to_art, 4.246575342 years_rel_to_art, 2.139344262 years_rel_to_art, 3.230136986 years_rel_to_art, 0.010958904 years_rel_to_art, -1.295890411 years_rel_to_art, -1.005479452 years_rel_to_art, -0.643835616 years_rel_to_art, -2.630136986 years_rel_to_art, -0.873972603 years_rel_to_art, -0.539726027 years_rel_to_art, 0.073972603 years_rel_to_art, 1.2 years_rel_to_art, 4.224657534 years_rel_to_art\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for k, v in sample_tags.groupby(\"study\")[\"sample_tags\"].unique().iteritems():\n",
    "    print(k)\n",
    "    print(\", \".join(v))\n",
    "    print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Let's annotate each study manually, one by one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs_adaptive_manual = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>total_templates</th>\n",
       "      <th>total_reads</th>\n",
       "      <th>productive_templates</th>\n",
       "      <th>total_productive_reads</th>\n",
       "      <th>fraction_productive</th>\n",
       "      <th>total_rearrangements</th>\n",
       "      <th>productive_rearrangements</th>\n",
       "      <th>productive_clonality</th>\n",
       "      <th>max_productive_frequency</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>study</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>HIP00110</td>\n",
       "      <td>224859</td>\n",
       "      <td>4059338</td>\n",
       "      <td>179411</td>\n",
       "      <td>3238889</td>\n",
       "      <td>0.7979</td>\n",
       "      <td>130940</td>\n",
       "      <td>104850</td>\n",
       "      <td>0.1007</td>\n",
       "      <td>1.371890</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>55 Years, Cohort 01, Cytomegalovirus -, HLA-A*...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>HIP00169</td>\n",
       "      <td>164698</td>\n",
       "      <td>1338914</td>\n",
       "      <td>138676</td>\n",
       "      <td>1127369</td>\n",
       "      <td>0.8420</td>\n",
       "      <td>115919</td>\n",
       "      <td>96869</td>\n",
       "      <td>0.0690</td>\n",
       "      <td>2.641903</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>41 Years, Cohort 01, Cytomegalovirus -, HLA-A*...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>HIP00594</td>\n",
       "      <td>356131</td>\n",
       "      <td>2918797</td>\n",
       "      <td>289407</td>\n",
       "      <td>2371942</td>\n",
       "      <td>0.8126</td>\n",
       "      <td>198830</td>\n",
       "      <td>161024</td>\n",
       "      <td>0.1787</td>\n",
       "      <td>2.311060</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>21 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>HIP00602</td>\n",
       "      <td>427705</td>\n",
       "      <td>9588747</td>\n",
       "      <td>335488</td>\n",
       "      <td>7521347</td>\n",
       "      <td>0.7844</td>\n",
       "      <td>248959</td>\n",
       "      <td>197856</td>\n",
       "      <td>0.1145</td>\n",
       "      <td>0.673071</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>45 Years, Cohort 01, Cytomegalovirus -, Female...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>773</th>\n",
       "      <td>HIP00614</td>\n",
       "      <td>na</td>\n",
       "      <td>1937837</td>\n",
       "      <td>na</td>\n",
       "      <td>1585816</td>\n",
       "      <td>0.8183</td>\n",
       "      <td>128274</td>\n",
       "      <td>105597</td>\n",
       "      <td>0.0944</td>\n",
       "      <td>1.085183</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>27 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Keck0116_MC1</td>\n",
       "      <td>611470</td>\n",
       "      <td>na</td>\n",
       "      <td>494052</td>\n",
       "      <td>na</td>\n",
       "      <td>0.8080</td>\n",
       "      <td>438842</td>\n",
       "      <td>357423</td>\n",
       "      <td>0.0495</td>\n",
       "      <td>0.627464</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>19 Years, Asian or Pacific Islander, Cohort 02...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Keck0117_MC1</td>\n",
       "      <td>638224</td>\n",
       "      <td>na</td>\n",
       "      <td>513475</td>\n",
       "      <td>na</td>\n",
       "      <td>0.8045</td>\n",
       "      <td>331272</td>\n",
       "      <td>272565</td>\n",
       "      <td>0.1364</td>\n",
       "      <td>5.619358</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>34 Years, Caucasian, Cohort 02, Cytomegaloviru...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>Keck0118_MC1</td>\n",
       "      <td>848716</td>\n",
       "      <td>na</td>\n",
       "      <td>660975</td>\n",
       "      <td>na</td>\n",
       "      <td>0.7788</td>\n",
       "      <td>397497</td>\n",
       "      <td>315296</td>\n",
       "      <td>0.1585</td>\n",
       "      <td>1.969061</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>28 Years, Caucasian, Cohort 02, Cytomegaloviru...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>Keck0119_MC1</td>\n",
       "      <td>590033</td>\n",
       "      <td>na</td>\n",
       "      <td>487792</td>\n",
       "      <td>na</td>\n",
       "      <td>0.8267</td>\n",
       "      <td>421166</td>\n",
       "      <td>353573</td>\n",
       "      <td>0.0425</td>\n",
       "      <td>1.139010</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>24 Years, Asian or Pacific Islander, Cohort 02...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Keck0120_MC1</td>\n",
       "      <td>584420</td>\n",
       "      <td>na</td>\n",
       "      <td>474402</td>\n",
       "      <td>na</td>\n",
       "      <td>0.8117</td>\n",
       "      <td>397185</td>\n",
       "      <td>325366</td>\n",
       "      <td>0.0459</td>\n",
       "      <td>0.382587</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>26 Years, Caucasian, Cohort 02, Cytomegaloviru...</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>786 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_name total_templates total_reads productive_templates  \\\n",
       "315      HIP00110          224859     4059338               179411   \n",
       "239      HIP00169          164698     1338914               138676   \n",
       "206      HIP00594          356131     2918797               289407   \n",
       "542      HIP00602          427705     9588747               335488   \n",
       "773      HIP00614              na     1937837                   na   \n",
       "..            ...             ...         ...                  ...   \n",
       "92   Keck0116_MC1          611470          na               494052   \n",
       "54   Keck0117_MC1          638224          na               513475   \n",
       "93   Keck0118_MC1          848716          na               660975   \n",
       "94   Keck0119_MC1          590033          na               487792   \n",
       "55   Keck0120_MC1          584420          na               474402   \n",
       "\n",
       "    total_productive_reads  fraction_productive  total_rearrangements  \\\n",
       "315                3238889               0.7979                130940   \n",
       "239                1127369               0.8420                115919   \n",
       "206                2371942               0.8126                198830   \n",
       "542                7521347               0.7844                248959   \n",
       "773                1585816               0.8183                128274   \n",
       "..                     ...                  ...                   ...   \n",
       "92                      na               0.8080                438842   \n",
       "54                      na               0.8045                331272   \n",
       "93                      na               0.7788                397497   \n",
       "94                      na               0.8267                421166   \n",
       "55                      na               0.8117                397185   \n",
       "\n",
       "     productive_rearrangements  productive_clonality  \\\n",
       "315                     104850                0.1007   \n",
       "239                      96869                0.0690   \n",
       "206                     161024                0.1787   \n",
       "542                     197856                0.1145   \n",
       "773                     105597                0.0944   \n",
       "..                         ...                   ...   \n",
       "92                      357423                0.0495   \n",
       "54                      272565                0.1364   \n",
       "93                      315296                0.1585   \n",
       "94                      353573                0.0425   \n",
       "55                      325366                0.0459   \n",
       "\n",
       "     max_productive_frequency locus  \\\n",
       "315                  1.371890  TCRB   \n",
       "239                  2.641903  TCRB   \n",
       "206                  2.311060  TCRB   \n",
       "542                  0.673071  TCRB   \n",
       "773                  1.085183  TCRB   \n",
       "..                        ...   ...   \n",
       "92                   0.627464  TCRB   \n",
       "54                   5.619358  TCRB   \n",
       "93                   1.969061  TCRB   \n",
       "94                   1.139010  TCRB   \n",
       "55                   0.382587  TCRB   \n",
       "\n",
       "                                           sample_tags                study  \n",
       "315  55 Years, Cohort 01, Cytomegalovirus -, HLA-A*...  emerson-2017-natgen  \n",
       "239  41 Years, Cohort 01, Cytomegalovirus -, HLA-A*...  emerson-2017-natgen  \n",
       "206  21 Years, Caucasian, Cohort 01, Cytomegaloviru...  emerson-2017-natgen  \n",
       "542  45 Years, Cohort 01, Cytomegalovirus -, Female...  emerson-2017-natgen  \n",
       "773  27 Years, Caucasian, Cohort 01, Cytomegaloviru...  emerson-2017-natgen  \n",
       "..                                                 ...                  ...  \n",
       "92   19 Years, Asian or Pacific Islander, Cohort 02...  emerson-2017-natgen  \n",
       "54   34 Years, Caucasian, Cohort 02, Cytomegaloviru...  emerson-2017-natgen  \n",
       "93   28 Years, Caucasian, Cohort 02, Cytomegaloviru...  emerson-2017-natgen  \n",
       "94   24 Years, Asian or Pacific Islander, Cohort 02...  emerson-2017-natgen  \n",
       "55   26 Years, Caucasian, Cohort 02, Cytomegaloviru...  emerson-2017-natgen  \n",
       "\n",
       "[786 rows x 13 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Emerson healthy TCR specimens, CMV+ and CMV-:\n",
    "\n",
    "ImmuneAccess says not to run \"Export Sample\", and instead use their dedicated zip file link. That sounds fine, and we could use ImmuneAccess's metadata from the Sample Overview screen as usual...\n",
    "\n",
    "except ImmuneAccess has somehow lost the \"HIP\"-prefixed labels for cohort 1, which are necessary to match up to CMV positivity metadata from SI Table 1. Instead ImmuneAccess now labels cohort 1 samples from 1 to N, basically.\n",
    "\n",
    "But we found an older zip file online that has the correct \"HIP-\" labels: https://s3-us-west-2.amazonaws.com/publishedproject-supplements/emerson-2017-natgen/emerson-2017-natgen.zip\n",
    "\n",
    "And someone kindly exported the Sample Overview ImmuneAccess metadata a while back and uploaded it here: https://github.com/jostmey/dkm/blob/7839937413af23203a442d5291e311ccb034dce7/repertoire-classification-problem/dataset/downloads/SampleOverview_08-19-2017_12-59-19_PM.tsv\n",
    "This is saved in our repo as emerson-2017-natgen.tsv.\n",
    "\n",
    "From brief inspection, all that has changed is the sample labels are now mangled in ImmuneAccess. And the order is different. Bummer.\n",
    "\n",
    "We're going to use the older exports, but merge in some important fields from the newer metadata export.\n",
    "\n",
    "One more important note: We have split the source files into two study names corresponding to the Emerson training and validation cohorts.\n",
    "The metadata files here are under the master study name, but the source files are under the two split study names.\n",
    "\"HIP\" samples: emerson-2017-natgen_train\n",
    "\"Keck\" samples: emerson-2017-natgen_validation\n",
    "\"\"\"\n",
    "\n",
    "df = _load_metadata_df(\"emerson-2017-natgen\", filter_cols=False)\n",
    "df.sort_values(\"sample_name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Inferred CMV -                        796\n",
       "gDNA                                  786\n",
       "T cells                               786\n",
       "Peripheral blood lymphocytes (PBL)    786\n",
       "Human                                 786\n",
       "PBMC                                  786\n",
       "Cohort 01                             666\n",
       "Inferred CMV +                        606\n",
       "Non-Hispanic or Latino                529\n",
       "Caucasian                             465\n",
       "Cytomegalovirus -                     421\n",
       "Male                                  392\n",
       "Female                                370\n",
       "Cytomegalovirus +                     340\n",
       "HLA-A*02                              294\n",
       "Inferred HLA-A*02                     282\n",
       "Unknown racial group                  252\n",
       "Unknown Ethnicity                     226\n",
       "Inferred HLA-A*01                     195\n",
       "HLA-A*01                              186\n",
       "Inferred HLA-B*07                     159\n",
       "Inferred HLA-B*44                     158\n",
       "HLA-B*07                              156\n",
       "Inferred HLA-A*03                     154\n",
       "HLA-A*03                              153\n",
       "HLA-B*44                              151\n",
       "Inferred HLA-B*08                     143\n",
       "HLA-B*08                              136\n",
       "HLA-A*24                              129\n",
       "Inferred HLA-A*24                     124\n",
       "Cohort 02                             120\n",
       "HLA-B*35                              115\n",
       "Inferred HLA-B*35                     113\n",
       "HLA-B*51                               74\n",
       "Inferred HLA-B*51                      67\n",
       "HLA-A*11                               61\n",
       "HLA-B*18                               55\n",
       "Inferred HLA-A*11                      51\n",
       "Inferred HLA-B*18                      51\n",
       "HLA-A*26                               50\n",
       "HLA-A*29                               50\n",
       "Inferred HLA-A*29                      49\n",
       "Inferred HLA-B*14                      48\n",
       "Asian or Pacific Islander              47\n",
       "HLA-A*68                               47\n",
       "Inferred HLA-B*62                      46\n",
       "HLA-B*14                               45\n",
       "HLA-B*27                               44\n",
       "HLA-B*15                               43\n",
       "HLA-B*62                               42\n",
       "Name: sample_tags, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"sample_tags\"].str.split(\",\").explode().str.strip().value_counts().head(n=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "v2    590\n",
       "v3    120\n",
       "v1     76\n",
       "Name: counting_method, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# counting_method is not available in the older immuneaccess sampleoverview we found online (see note above)\n",
    "# we'll use the wrong-labeled one to check this:\n",
    "df_m_wrong_labels = _load_metadata_df(\n",
    "    \"emerson-2017-natgen.wrong_labels\", filter_cols=False\n",
    ")\n",
    "df_m_wrong_labels[\"counting_method\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((786, 22), (786, 13))"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# same shape, different orders\n",
    "df_m_wrong_labels.shape, df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['sample_name',\n",
       " 'locus',\n",
       " 'sample_tags',\n",
       " 'counting_method',\n",
       " 'primer_set',\n",
       " 'species',\n",
       " 'study']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols_desired"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(786, 13)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(786, 16)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.merge(\n",
    "    df,\n",
    "    df_m_wrong_labels[\n",
    "        [\n",
    "            \"total_templates\",\n",
    "            \"total_reads\",\n",
    "            \"productive_templates\",\n",
    "            \"total_productive_reads\",\n",
    "            \"total_rearrangements\",\n",
    "            \"productive_rearrangements\",\n",
    "            \"counting_method\",\n",
    "            \"primer_set\",\n",
    "            \"species\",\n",
    "        ]\n",
    "    ],\n",
    "    on=[\n",
    "        \"total_templates\",\n",
    "        \"total_reads\",\n",
    "        \"productive_templates\",\n",
    "        \"total_productive_reads\",\n",
    "        \"total_rearrangements\",\n",
    "        \"productive_rearrangements\",\n",
    "    ],\n",
    "    how=\"inner\",\n",
    "    validate=\"1:1\",\n",
    ")\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIP15860</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP14363</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*02, HLA-A*11, HLA-B*35, Human...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIP14178</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*01, HLA-A*03, HLA-B*08, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIP13944</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*02, HLA-A*03, HLA-B*18, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP13911</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>22 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>HIP01765</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>48 Years, Cohort 01, Cytomegalovirus +, Female...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>782</th>\n",
       "      <td>HIP13760</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, Cytomegalovirus -, HLA-A*11, HLA-A*...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783</th>\n",
       "      <td>HIP08821</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>55 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>784</th>\n",
       "      <td>HIP05552</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>53 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>785</th>\n",
       "      <td>HIP03484</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>46 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>786 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_name locus                                        sample_tags  \\\n",
       "0      HIP15860  TCRB  Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B...   \n",
       "1      HIP14363  TCRB  Cohort 01, HLA-A*02, HLA-A*11, HLA-B*35, Human...   \n",
       "2      HIP14178  TCRB  Cohort 01, HLA-A*01, HLA-A*03, HLA-B*08, HLA-B...   \n",
       "3      HIP13944  TCRB  Cohort 01, HLA-A*02, HLA-A*03, HLA-B*18, HLA-B...   \n",
       "4      HIP13911  TCRB  22 Years, Caucasian, Cohort 01, Cytomegaloviru...   \n",
       "..          ...   ...                                                ...   \n",
       "781    HIP01765  TCRB  48 Years, Cohort 01, Cytomegalovirus +, Female...   \n",
       "782    HIP13760  TCRB  Cohort 01, Cytomegalovirus -, HLA-A*11, HLA-A*...   \n",
       "783    HIP08821  TCRB  55 Years, Caucasian, Cohort 01, Cytomegaloviru...   \n",
       "784    HIP05552  TCRB  53 Years, Caucasian, Cohort 01, Cytomegaloviru...   \n",
       "785    HIP03484  TCRB  46 Years, Caucasian, Cohort 01, Cytomegaloviru...   \n",
       "\n",
       "    counting_method       primer_set species                study  \n",
       "0                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen  \n",
       "1                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen  \n",
       "2                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen  \n",
       "3                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen  \n",
       "4                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen  \n",
       "..              ...              ...     ...                  ...  \n",
       "781              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen  \n",
       "782              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen  \n",
       "783              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen  \n",
       "784              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen  \n",
       "785              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen  \n",
       "\n",
       "[786 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[cols_desired]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "v2    590\n",
       "v3    120\n",
       "v1     76\n",
       "Name: counting_method, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sanity check\n",
    "df[\"counting_method\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "315    HIP00110\n",
       "239    HIP00169\n",
       "206    HIP00594\n",
       "542    HIP00602\n",
       "773    HIP00614\n",
       "         ...   \n",
       "92     Keck0116\n",
       "54     Keck0117\n",
       "93     Keck0118\n",
       "94     Keck0119\n",
       "55     Keck0120\n",
       "Name: participant_label, Length: 786, dtype: object"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"sequencing_type\"] = \"gDNA\"\n",
    "# Split e.g. Keck0116_MC1 into Keck0116\n",
    "df[\"specimen_label\"] = df[\"sample_name\"]\n",
    "df[\"participant_label\"] = df[\"sample_name\"].str.split(\"_\").str[0]\n",
    "assert (df[\"participant_label\"].value_counts() == 1).all()\n",
    "df[\"participant_label\"].sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Race and ethnicity</th>\n",
       "      <th>Known CMV status</th>\n",
       "      <th>Inferred CMV status (all data)</th>\n",
       "      <th>Inferred CMV status (cross validation)</th>\n",
       "      <th>Known HLA alleles</th>\n",
       "      <th>Inferred HLA alleles               (cross validation)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIP00110</td>\n",
       "      <td>Male</td>\n",
       "      <td>54.557153</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A24, A3, B7, B7</td>\n",
       "      <td>A24, A3, B7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP00169</td>\n",
       "      <td>Male</td>\n",
       "      <td>40.681725</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A1, A2, B37, B27</td>\n",
       "      <td>A2, B27, B7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIP00594</td>\n",
       "      <td>Male</td>\n",
       "      <td>20.632444</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>A2, A32, B7, B61</td>\n",
       "      <td>A2, A3, B44, B7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIP00602</td>\n",
       "      <td>Female</td>\n",
       "      <td>44.678987</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A11, A2, B35, B51</td>\n",
       "      <td>A11, A2, B35, B51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP00614</td>\n",
       "      <td>Male</td>\n",
       "      <td>26.688569</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A3, A29, B7, B7</td>\n",
       "      <td>A3, B7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661</th>\n",
       "      <td>HIP17887</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>ND</td>\n",
       "      <td>ND</td>\n",
       "      <td>A2, A29, B7, B44</td>\n",
       "      <td>A29, A2, B7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>662</th>\n",
       "      <td>HIP19048</td>\n",
       "      <td>Female</td>\n",
       "      <td>57.404517</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A1, A2, B57, B51</td>\n",
       "      <td>A1, A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663</th>\n",
       "      <td>HIP19089</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>ND</td>\n",
       "      <td>ND</td>\n",
       "      <td>A1, A80, B13, B44</td>\n",
       "      <td>A1, A29, B44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>664</th>\n",
       "      <td>HIP19716</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>ND</td>\n",
       "      <td>ND</td>\n",
       "      <td>A1, A1, B8, B8</td>\n",
       "      <td>A1, B8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>665</th>\n",
       "      <td>HIP19717</td>\n",
       "      <td>Male</td>\n",
       "      <td>47.912389</td>\n",
       "      <td>Unknown, Hispanic or Latino</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>A1, A1, B18, B35</td>\n",
       "      <td>A1, B18, B35, B8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>666 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Subject ID      Sex        Age            Race and ethnicity   \\\n",
       "0     HIP00110     Male  54.557153                        Unknown   \n",
       "1     HIP00169     Male  40.681725                        Unknown   \n",
       "2     HIP00594     Male  20.632444  White, Not Hispanic or Latino   \n",
       "3     HIP00602   Female  44.678987                        Unknown   \n",
       "4     HIP00614     Male  26.688569  White, Not Hispanic or Latino   \n",
       "..         ...      ...        ...                            ...   \n",
       "661   HIP17887  Unknown    Unknown                        Unknown   \n",
       "662   HIP19048   Female  57.404517                        Unknown   \n",
       "663   HIP19089  Unknown    Unknown                        Unknown   \n",
       "664   HIP19716  Unknown    Unknown                        Unknown   \n",
       "665   HIP19717     Male  47.912389    Unknown, Hispanic or Latino   \n",
       "\n",
       "    Known CMV status Inferred CMV status (all data)  \\\n",
       "0                  -                              -   \n",
       "1                  -                              -   \n",
       "2                  +                              +   \n",
       "3                  -                              -   \n",
       "4                  -                              -   \n",
       "..               ...                            ...   \n",
       "661          Unknown                             ND   \n",
       "662                -                              -   \n",
       "663          Unknown                             ND   \n",
       "664          Unknown                             ND   \n",
       "665                +                              +   \n",
       "\n",
       "    Inferred CMV status (cross validation)  Known HLA alleles  \\\n",
       "0                                        -    A24, A3, B7, B7   \n",
       "1                                        -   A1, A2, B37, B27   \n",
       "2                                        +   A2, A32, B7, B61   \n",
       "3                                        -  A11, A2, B35, B51   \n",
       "4                                        -    A3, A29, B7, B7   \n",
       "..                                     ...                ...   \n",
       "661                                     ND   A2, A29, B7, B44   \n",
       "662                                      -   A1, A2, B57, B51   \n",
       "663                                     ND  A1, A80, B13, B44   \n",
       "664                                     ND     A1, A1, B8, B8   \n",
       "665                                      +   A1, A1, B18, B35   \n",
       "\n",
       "    Inferred HLA alleles               (cross validation)  \n",
       "0                                          A24, A3, B7     \n",
       "1                                          A2, B27, B7     \n",
       "2                                      A2, A3, B44, B7     \n",
       "3                                    A11, A2, B35, B51     \n",
       "4                                               A3, B7     \n",
       "..                                                 ...     \n",
       "661                                        A29, A2, B7     \n",
       "662                                             A1, A2     \n",
       "663                                       A1, A29, B44     \n",
       "664                                             A1, B8     \n",
       "665                                   A1, B18, B35, B8     \n",
       "\n",
       "[666 rows x 9 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# let's add metadata from SI Table 1\n",
    "df_m1 = pd.read_excel(\n",
    "    config.paths.metadata_dir / \"adaptive\" / \"emerson-2017-natgen.si_table_1.xlsx\",\n",
    "    sheet_name=0,\n",
    ")\n",
    "df_m2 = pd.read_excel(\n",
    "    config.paths.metadata_dir / \"adaptive\" / \"emerson-2017-natgen.si_table_1.xlsx\",\n",
    "    sheet_name=1,\n",
    ")\n",
    "df_m1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Race and ethnicity</th>\n",
       "      <th>Known CMV status</th>\n",
       "      <th>Inferred CMV status (trained on Cohort 1)</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>Unnamed: 7</th>\n",
       "      <th></th>\n",
       "      <th>.1</th>\n",
       "      <th>.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Keck0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>22</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Keck0002</td>\n",
       "      <td>Female</td>\n",
       "      <td>21</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Keck0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>25</td>\n",
       "      <td>American Indian or Alaska Native, Not Hispanic...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Keck0004</td>\n",
       "      <td>Female</td>\n",
       "      <td>24</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Keck0005</td>\n",
       "      <td>Male</td>\n",
       "      <td>24</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>Keck0116</td>\n",
       "      <td>Female</td>\n",
       "      <td>19</td>\n",
       "      <td>Asian, Not Hispanic or Latino</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>Keck0117</td>\n",
       "      <td>Male</td>\n",
       "      <td>34</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>Keck0118</td>\n",
       "      <td>Male</td>\n",
       "      <td>28</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>Keck0119</td>\n",
       "      <td>Female</td>\n",
       "      <td>24</td>\n",
       "      <td>Asian, Not Hispanic or Latino</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>Keck0120</td>\n",
       "      <td>Female</td>\n",
       "      <td>26</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Subject ID     Sex  Age  \\\n",
       "0     Keck0001  Female   22   \n",
       "1     Keck0002  Female   21   \n",
       "2     Keck0003  Female   25   \n",
       "3     Keck0004  Female   24   \n",
       "4     Keck0005    Male   24   \n",
       "..         ...     ...  ...   \n",
       "115   Keck0116  Female   19   \n",
       "116   Keck0117    Male   34   \n",
       "117   Keck0118    Male   28   \n",
       "118   Keck0119  Female   24   \n",
       "119   Keck0120  Female   26   \n",
       "\n",
       "                                   Race and ethnicity  Known CMV status  \\\n",
       "0                        White, Not Hispanic or Latino                +   \n",
       "1                        White, Not Hispanic or Latino                -   \n",
       "2    American Indian or Alaska Native, Not Hispanic...                -   \n",
       "3                        White, Not Hispanic or Latino                +   \n",
       "4                        White, Not Hispanic or Latino                -   \n",
       "..                                                 ...              ...   \n",
       "115                      Asian, Not Hispanic or Latino                -   \n",
       "116                      White, Not Hispanic or Latino                +   \n",
       "117                      White, Not Hispanic or Latino                +   \n",
       "118                      Asian, Not Hispanic or Latino                +   \n",
       "119                      White, Not Hispanic or Latino                -   \n",
       "\n",
       "    Inferred CMV status (trained on Cohort 1)  Unnamed: 6  Unnamed: 7      \\\n",
       "0                                           -         NaN         NaN NaN   \n",
       "1                                           -         NaN         NaN NaN   \n",
       "2                                           -         NaN         NaN NaN   \n",
       "3                                           +         NaN         NaN NaN   \n",
       "4                                           -         NaN         NaN NaN   \n",
       "..                                        ...         ...         ...  ..   \n",
       "115                                         -         NaN         NaN NaN   \n",
       "116                                         +         NaN         NaN NaN   \n",
       "117                                         +         NaN         NaN NaN   \n",
       "118                                         +         NaN         NaN NaN   \n",
       "119                                         -         NaN         NaN NaN   \n",
       "\n",
       "      .1   .2  \n",
       "0    NaN  NaN  \n",
       "1    NaN  NaN  \n",
       "2    NaN  NaN  \n",
       "3    NaN  NaN  \n",
       "4    NaN  NaN  \n",
       "..   ...  ...  \n",
       "115  NaN  NaN  \n",
       "116  NaN  NaN  \n",
       "117  NaN  NaN  \n",
       "118  NaN  NaN  \n",
       "119  NaN  NaN  \n",
       "\n",
       "[120 rows x 11 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_m2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subject ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Race and ethnicity</th>\n",
       "      <th>Known CMV status</th>\n",
       "      <th>Inferred CMV status (all data)</th>\n",
       "      <th>Inferred CMV status (cross validation)</th>\n",
       "      <th>Known HLA alleles</th>\n",
       "      <th>Inferred HLA alleles               (cross validation)</th>\n",
       "      <th>Inferred CMV status (trained on Cohort 1)</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>Unnamed: 7</th>\n",
       "      <th></th>\n",
       "      <th>.1</th>\n",
       "      <th>.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIP00110</td>\n",
       "      <td>Male</td>\n",
       "      <td>54.557153</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A24, A3, B7, B7</td>\n",
       "      <td>A24, A3, B7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP00169</td>\n",
       "      <td>Male</td>\n",
       "      <td>40.681725</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A1, A2, B37, B27</td>\n",
       "      <td>A2, B27, B7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIP00594</td>\n",
       "      <td>Male</td>\n",
       "      <td>20.632444</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>A2, A32, B7, B61</td>\n",
       "      <td>A2, A3, B44, B7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIP00602</td>\n",
       "      <td>Female</td>\n",
       "      <td>44.678987</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A11, A2, B35, B51</td>\n",
       "      <td>A11, A2, B35, B51</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP00614</td>\n",
       "      <td>Male</td>\n",
       "      <td>26.688569</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A3, A29, B7, B7</td>\n",
       "      <td>A3, B7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>Keck0116</td>\n",
       "      <td>Female</td>\n",
       "      <td>19</td>\n",
       "      <td>Asian, Not Hispanic or Latino</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>Keck0117</td>\n",
       "      <td>Male</td>\n",
       "      <td>34</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>Keck0118</td>\n",
       "      <td>Male</td>\n",
       "      <td>28</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>Keck0119</td>\n",
       "      <td>Female</td>\n",
       "      <td>24</td>\n",
       "      <td>Asian, Not Hispanic or Latino</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>Keck0120</td>\n",
       "      <td>Female</td>\n",
       "      <td>26</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>786 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Subject ID     Sex        Age            Race and ethnicity   \\\n",
       "0     HIP00110    Male  54.557153                        Unknown   \n",
       "1     HIP00169    Male  40.681725                        Unknown   \n",
       "2     HIP00594    Male  20.632444  White, Not Hispanic or Latino   \n",
       "3     HIP00602  Female  44.678987                        Unknown   \n",
       "4     HIP00614    Male  26.688569  White, Not Hispanic or Latino   \n",
       "..         ...     ...        ...                            ...   \n",
       "115   Keck0116  Female         19  Asian, Not Hispanic or Latino   \n",
       "116   Keck0117    Male         34  White, Not Hispanic or Latino   \n",
       "117   Keck0118    Male         28  White, Not Hispanic or Latino   \n",
       "118   Keck0119  Female         24  Asian, Not Hispanic or Latino   \n",
       "119   Keck0120  Female         26  White, Not Hispanic or Latino   \n",
       "\n",
       "    Known CMV status Inferred CMV status (all data)  \\\n",
       "0                  -                              -   \n",
       "1                  -                              -   \n",
       "2                  +                              +   \n",
       "3                  -                              -   \n",
       "4                  -                              -   \n",
       "..               ...                            ...   \n",
       "115                -                            NaN   \n",
       "116                +                            NaN   \n",
       "117                +                            NaN   \n",
       "118                +                            NaN   \n",
       "119                -                            NaN   \n",
       "\n",
       "    Inferred CMV status (cross validation)  Known HLA alleles  \\\n",
       "0                                        -    A24, A3, B7, B7   \n",
       "1                                        -   A1, A2, B37, B27   \n",
       "2                                        +   A2, A32, B7, B61   \n",
       "3                                        -  A11, A2, B35, B51   \n",
       "4                                        -    A3, A29, B7, B7   \n",
       "..                                     ...                ...   \n",
       "115                                    NaN                NaN   \n",
       "116                                    NaN                NaN   \n",
       "117                                    NaN                NaN   \n",
       "118                                    NaN                NaN   \n",
       "119                                    NaN                NaN   \n",
       "\n",
       "    Inferred HLA alleles               (cross validation)  \\\n",
       "0                                          A24, A3, B7      \n",
       "1                                          A2, B27, B7      \n",
       "2                                      A2, A3, B44, B7      \n",
       "3                                    A11, A2, B35, B51      \n",
       "4                                               A3, B7      \n",
       "..                                                 ...      \n",
       "115                                                NaN      \n",
       "116                                                NaN      \n",
       "117                                                NaN      \n",
       "118                                                NaN      \n",
       "119                                                NaN      \n",
       "\n",
       "    Inferred CMV status (trained on Cohort 1)  Unnamed: 6  Unnamed: 7      \\\n",
       "0                                         NaN         NaN         NaN NaN   \n",
       "1                                         NaN         NaN         NaN NaN   \n",
       "2                                         NaN         NaN         NaN NaN   \n",
       "3                                         NaN         NaN         NaN NaN   \n",
       "4                                         NaN         NaN         NaN NaN   \n",
       "..                                        ...         ...         ...  ..   \n",
       "115                                         -         NaN         NaN NaN   \n",
       "116                                         +         NaN         NaN NaN   \n",
       "117                                         +         NaN         NaN NaN   \n",
       "118                                         +         NaN         NaN NaN   \n",
       "119                                         -         NaN         NaN NaN   \n",
       "\n",
       "      .1   .2  \n",
       "0    NaN  NaN  \n",
       "1    NaN  NaN  \n",
       "2    NaN  NaN  \n",
       "3    NaN  NaN  \n",
       "4    NaN  NaN  \n",
       "..   ...  ...  \n",
       "115  NaN  NaN  \n",
       "116  NaN  NaN  \n",
       "117  NaN  NaN  \n",
       "118  NaN  NaN  \n",
       "119  NaN  NaN  \n",
       "\n",
       "[786 rows x 15 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_m = pd.concat([df_m1, df_m2], axis=0)\n",
    "df_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-          421\n",
       "+          340\n",
       "Unknown     25\n",
       "Name: Known CMV status, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_m[\"Known CMV status\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>cmv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIP00110</td>\n",
       "      <td>Male</td>\n",
       "      <td>54.557153</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CMV-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP00169</td>\n",
       "      <td>Male</td>\n",
       "      <td>40.681725</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CMV-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIP00594</td>\n",
       "      <td>Male</td>\n",
       "      <td>20.632444</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIP00602</td>\n",
       "      <td>Female</td>\n",
       "      <td>44.678987</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CMV-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP00614</td>\n",
       "      <td>Male</td>\n",
       "      <td>26.688569</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>Keck0116</td>\n",
       "      <td>Female</td>\n",
       "      <td>19</td>\n",
       "      <td>Asian, Not Hispanic or Latino</td>\n",
       "      <td>CMV-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>Keck0117</td>\n",
       "      <td>Male</td>\n",
       "      <td>34</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>Keck0118</td>\n",
       "      <td>Male</td>\n",
       "      <td>28</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>Keck0119</td>\n",
       "      <td>Female</td>\n",
       "      <td>24</td>\n",
       "      <td>Asian, Not Hispanic or Latino</td>\n",
       "      <td>CMV+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>Keck0120</td>\n",
       "      <td>Female</td>\n",
       "      <td>26</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>786 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    participant_label     Sex        Age                      Ethnicity   cmv\n",
       "0            HIP00110    Male  54.557153                        Unknown  CMV-\n",
       "1            HIP00169    Male  40.681725                        Unknown  CMV-\n",
       "2            HIP00594    Male  20.632444  White, Not Hispanic or Latino  CMV+\n",
       "3            HIP00602  Female  44.678987                        Unknown  CMV-\n",
       "4            HIP00614    Male  26.688569  White, Not Hispanic or Latino  CMV-\n",
       "..                ...     ...        ...                            ...   ...\n",
       "115          Keck0116  Female         19  Asian, Not Hispanic or Latino  CMV-\n",
       "116          Keck0117    Male         34  White, Not Hispanic or Latino  CMV+\n",
       "117          Keck0118    Male         28  White, Not Hispanic or Latino  CMV+\n",
       "118          Keck0119  Female         24  Asian, Not Hispanic or Latino  CMV+\n",
       "119          Keck0120  Female         26  White, Not Hispanic or Latino  CMV-\n",
       "\n",
       "[786 rows x 5 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_m = df_m.rename(\n",
    "    columns={\n",
    "        \"Subject ID\": \"participant_label\",\n",
    "        \"Race and ethnicity \": \"Ethnicity\",\n",
    "    }\n",
    ").assign(cmv=df_m[\"Known CMV status\"].map({\"+\": \"CMV+\", \"-\": \"CMV-\"}))[\n",
    "    [\"participant_label\", \"Sex\", \"Age\", \"Ethnicity\", \"cmv\"]\n",
    "]\n",
    "df_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>cmv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIP15860</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP15860</td>\n",
       "      <td>HIP15860</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP14363</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*02, HLA-A*11, HLA-B*35, Human...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP14363</td>\n",
       "      <td>HIP14363</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIP14178</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*01, HLA-A*03, HLA-B*08, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP14178</td>\n",
       "      <td>HIP14178</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIP13944</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*02, HLA-A*03, HLA-B*18, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP13944</td>\n",
       "      <td>HIP13944</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP13911</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>22 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP13911</td>\n",
       "      <td>HIP13911</td>\n",
       "      <td>Male</td>\n",
       "      <td>22.469541</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>HIP01765</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>48 Years, Cohort 01, Cytomegalovirus +, Female...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP01765</td>\n",
       "      <td>HIP01765</td>\n",
       "      <td>Female</td>\n",
       "      <td>47.843943</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CMV+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>782</th>\n",
       "      <td>HIP13760</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, Cytomegalovirus -, HLA-A*11, HLA-A*...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP13760</td>\n",
       "      <td>HIP13760</td>\n",
       "      <td>Male</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CMV-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783</th>\n",
       "      <td>HIP08821</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>55 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP08821</td>\n",
       "      <td>HIP08821</td>\n",
       "      <td>Female</td>\n",
       "      <td>54.631075</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>784</th>\n",
       "      <td>HIP05552</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>53 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP05552</td>\n",
       "      <td>HIP05552</td>\n",
       "      <td>Female</td>\n",
       "      <td>52.835044</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>785</th>\n",
       "      <td>HIP03484</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>46 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP03484</td>\n",
       "      <td>HIP03484</td>\n",
       "      <td>Male</td>\n",
       "      <td>46.49692</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>786 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_name locus                                        sample_tags  \\\n",
       "0      HIP15860  TCRB  Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B...   \n",
       "1      HIP14363  TCRB  Cohort 01, HLA-A*02, HLA-A*11, HLA-B*35, Human...   \n",
       "2      HIP14178  TCRB  Cohort 01, HLA-A*01, HLA-A*03, HLA-B*08, HLA-B...   \n",
       "3      HIP13944  TCRB  Cohort 01, HLA-A*02, HLA-A*03, HLA-B*18, HLA-B...   \n",
       "4      HIP13911  TCRB  22 Years, Caucasian, Cohort 01, Cytomegaloviru...   \n",
       "..          ...   ...                                                ...   \n",
       "781    HIP01765  TCRB  48 Years, Cohort 01, Cytomegalovirus +, Female...   \n",
       "782    HIP13760  TCRB  Cohort 01, Cytomegalovirus -, HLA-A*11, HLA-A*...   \n",
       "783    HIP08821  TCRB  55 Years, Caucasian, Cohort 01, Cytomegaloviru...   \n",
       "784    HIP05552  TCRB  53 Years, Caucasian, Cohort 01, Cytomegaloviru...   \n",
       "785    HIP03484  TCRB  46 Years, Caucasian, Cohort 01, Cytomegaloviru...   \n",
       "\n",
       "    counting_method       primer_set species                study  \\\n",
       "0                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen   \n",
       "1                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen   \n",
       "2                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen   \n",
       "3                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen   \n",
       "4                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen   \n",
       "..              ...              ...     ...                  ...   \n",
       "781              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen   \n",
       "782              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen   \n",
       "783              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen   \n",
       "784              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen   \n",
       "785              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen   \n",
       "\n",
       "    sequencing_type specimen_label participant_label      Sex        Age  \\\n",
       "0              gDNA       HIP15860          HIP15860  Unknown    Unknown   \n",
       "1              gDNA       HIP14363          HIP14363  Unknown    Unknown   \n",
       "2              gDNA       HIP14178          HIP14178  Unknown    Unknown   \n",
       "3              gDNA       HIP13944          HIP13944  Unknown    Unknown   \n",
       "4              gDNA       HIP13911          HIP13911     Male  22.469541   \n",
       "..              ...            ...               ...      ...        ...   \n",
       "781            gDNA       HIP01765          HIP01765   Female  47.843943   \n",
       "782            gDNA       HIP13760          HIP13760     Male    Unknown   \n",
       "783            gDNA       HIP08821          HIP08821   Female  54.631075   \n",
       "784            gDNA       HIP05552          HIP05552   Female  52.835044   \n",
       "785            gDNA       HIP03484          HIP03484     Male   46.49692   \n",
       "\n",
       "                         Ethnicity   cmv  \n",
       "0                          Unknown   NaN  \n",
       "1                          Unknown   NaN  \n",
       "2                          Unknown   NaN  \n",
       "3                          Unknown   NaN  \n",
       "4    White, Not Hispanic or Latino  CMV+  \n",
       "..                             ...   ...  \n",
       "781                        Unknown  CMV+  \n",
       "782                        Unknown  CMV-  \n",
       "783  White, Not Hispanic or Latino  CMV+  \n",
       "784  White, Not Hispanic or Latino  CMV-  \n",
       "785  White, Not Hispanic or Latino  CMV-  \n",
       "\n",
       "[786 rows x 14 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.merge(\n",
    "    df,\n",
    "    df_m,\n",
    "    how=\"left\",\n",
    "    on=\"participant_label\",\n",
    "    validate=\"1:1\",\n",
    ")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    760\n",
       "True      26\n",
       "Name: cmv, dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"cmv\"].isna().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CMV-    420\n",
       "CMV+    340\n",
       "Name: cmv, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"cmv\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Healthy/Background - CMV-           420\n",
       "Healthy/Background - CMV+           340\n",
       "Healthy/Background - CMV Unknown     26\n",
       "Name: disease_subtype, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"disease\"] = healthy_label\n",
    "df[\"disease_subtype\"] = f\"{healthy_label} - \" + df[\"cmv\"].fillna(\"CMV Unknown\")\n",
    "df[\"disease_subtype\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "emerson-2017-natgen_train         666\n",
       "emerson-2017-natgen_validation    120\n",
       "Name: study, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# One more important note: We have split the source files into two study names corresponding to the Emerson training and validation cohorts.\n",
    "# The metadata files here are under the master study name, but the source files are under the two split study names.\n",
    "# \"HIP\" samples: emerson-2017-natgen_train\n",
    "# \"Keck\" samples: emerson-2017-natgen_validation\n",
    "\n",
    "# So we must update the study name.\n",
    "df[\"study\"] = (\n",
    "    df[\"study\"]\n",
    "    + \"_\"\n",
    "    + df[\"participant_label\"]\n",
    "    .str.startswith(\"Keck\")\n",
    "    .map({True: \"validation\", False: \"train\"})\n",
    ")\n",
    "df[\"study\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>cmv</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIP15860</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP15860</td>\n",
       "      <td>HIP15860</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CMV Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP14363</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*02, HLA-A*11, HLA-B*35, Human...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP14363</td>\n",
       "      <td>HIP14363</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CMV Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIP14178</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*01, HLA-A*03, HLA-B*08, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP14178</td>\n",
       "      <td>HIP14178</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CMV Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIP13944</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*02, HLA-A*03, HLA-B*18, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP13944</td>\n",
       "      <td>HIP13944</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CMV Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP13911</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>22 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP13911</td>\n",
       "      <td>HIP13911</td>\n",
       "      <td>Male</td>\n",
       "      <td>22.469541</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV+</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CMV+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>HIP01765</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>48 Years, Cohort 01, Cytomegalovirus +, Female...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP01765</td>\n",
       "      <td>HIP01765</td>\n",
       "      <td>Female</td>\n",
       "      <td>47.843943</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CMV+</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CMV+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>782</th>\n",
       "      <td>HIP13760</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, Cytomegalovirus -, HLA-A*11, HLA-A*...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP13760</td>\n",
       "      <td>HIP13760</td>\n",
       "      <td>Male</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CMV-</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CMV-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783</th>\n",
       "      <td>HIP08821</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>55 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP08821</td>\n",
       "      <td>HIP08821</td>\n",
       "      <td>Female</td>\n",
       "      <td>54.631075</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV+</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CMV+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>784</th>\n",
       "      <td>HIP05552</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>53 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP05552</td>\n",
       "      <td>HIP05552</td>\n",
       "      <td>Female</td>\n",
       "      <td>52.835044</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV-</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CMV-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>785</th>\n",
       "      <td>HIP03484</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>46 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP03484</td>\n",
       "      <td>HIP03484</td>\n",
       "      <td>Male</td>\n",
       "      <td>46.49692</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>CMV-</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CMV-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>786 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_name locus                                        sample_tags  \\\n",
       "0      HIP15860  TCRB  Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B...   \n",
       "1      HIP14363  TCRB  Cohort 01, HLA-A*02, HLA-A*11, HLA-B*35, Human...   \n",
       "2      HIP14178  TCRB  Cohort 01, HLA-A*01, HLA-A*03, HLA-B*08, HLA-B...   \n",
       "3      HIP13944  TCRB  Cohort 01, HLA-A*02, HLA-A*03, HLA-B*18, HLA-B...   \n",
       "4      HIP13911  TCRB  22 Years, Caucasian, Cohort 01, Cytomegaloviru...   \n",
       "..          ...   ...                                                ...   \n",
       "781    HIP01765  TCRB  48 Years, Cohort 01, Cytomegalovirus +, Female...   \n",
       "782    HIP13760  TCRB  Cohort 01, Cytomegalovirus -, HLA-A*11, HLA-A*...   \n",
       "783    HIP08821  TCRB  55 Years, Caucasian, Cohort 01, Cytomegaloviru...   \n",
       "784    HIP05552  TCRB  53 Years, Caucasian, Cohort 01, Cytomegaloviru...   \n",
       "785    HIP03484  TCRB  46 Years, Caucasian, Cohort 01, Cytomegaloviru...   \n",
       "\n",
       "    counting_method       primer_set species                      study  \\\n",
       "0                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen_train   \n",
       "1                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen_train   \n",
       "2                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen_train   \n",
       "3                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen_train   \n",
       "4                v2  Human-TCRB-PD1x   Human  emerson-2017-natgen_train   \n",
       "..              ...              ...     ...                        ...   \n",
       "781              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen_train   \n",
       "782              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen_train   \n",
       "783              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen_train   \n",
       "784              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen_train   \n",
       "785              v1  Human-TCRB-PD1x   Human  emerson-2017-natgen_train   \n",
       "\n",
       "    sequencing_type specimen_label participant_label      Sex        Age  \\\n",
       "0              gDNA       HIP15860          HIP15860  Unknown    Unknown   \n",
       "1              gDNA       HIP14363          HIP14363  Unknown    Unknown   \n",
       "2              gDNA       HIP14178          HIP14178  Unknown    Unknown   \n",
       "3              gDNA       HIP13944          HIP13944  Unknown    Unknown   \n",
       "4              gDNA       HIP13911          HIP13911     Male  22.469541   \n",
       "..              ...            ...               ...      ...        ...   \n",
       "781            gDNA       HIP01765          HIP01765   Female  47.843943   \n",
       "782            gDNA       HIP13760          HIP13760     Male    Unknown   \n",
       "783            gDNA       HIP08821          HIP08821   Female  54.631075   \n",
       "784            gDNA       HIP05552          HIP05552   Female  52.835044   \n",
       "785            gDNA       HIP03484          HIP03484     Male   46.49692   \n",
       "\n",
       "                         Ethnicity   cmv             disease  \\\n",
       "0                          Unknown   NaN  Healthy/Background   \n",
       "1                          Unknown   NaN  Healthy/Background   \n",
       "2                          Unknown   NaN  Healthy/Background   \n",
       "3                          Unknown   NaN  Healthy/Background   \n",
       "4    White, Not Hispanic or Latino  CMV+  Healthy/Background   \n",
       "..                             ...   ...                 ...   \n",
       "781                        Unknown  CMV+  Healthy/Background   \n",
       "782                        Unknown  CMV-  Healthy/Background   \n",
       "783  White, Not Hispanic or Latino  CMV+  Healthy/Background   \n",
       "784  White, Not Hispanic or Latino  CMV-  Healthy/Background   \n",
       "785  White, Not Hispanic or Latino  CMV-  Healthy/Background   \n",
       "\n",
       "                      disease_subtype  \n",
       "0    Healthy/Background - CMV Unknown  \n",
       "1    Healthy/Background - CMV Unknown  \n",
       "2    Healthy/Background - CMV Unknown  \n",
       "3    Healthy/Background - CMV Unknown  \n",
       "4           Healthy/Background - CMV+  \n",
       "..                                ...  \n",
       "781         Healthy/Background - CMV+  \n",
       "782         Healthy/Background - CMV-  \n",
       "783         Healthy/Background - CMV+  \n",
       "784         Healthy/Background - CMV-  \n",
       "785         Healthy/Background - CMV-  \n",
       "\n",
       "[786 rows x 16 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual.append(df)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>310114_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>03 Years, 12.6575342465753 Years at diagnosis,...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>310251_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>06 Years, 6 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>310113_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>03 Years, 11.5095890410959 Years at diagnosis,...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>310205_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>0 Years at visit, 09 Months, Caucasian, Contro...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>310283_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>08 Years, 12.6575342465753 Years at diagnosis,...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>DenverT1D-170_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>17 Years, 17 Years at diagnosis, 18-24 Years (...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>DenverT1D-106_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 16 Years, 16 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>DenverT1D-256_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 16 Years, 16 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>DenverT1D-055_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 07 Years, 7 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>DenverT1D-271_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 09 Years, 9 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>359 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample_name locus  \\\n",
       "0           310114_TCRB  TCRB   \n",
       "1           310251_TCRB  TCRB   \n",
       "2           310113_TCRB  TCRB   \n",
       "3           310205_TCRB  TCRB   \n",
       "4           310283_TCRB  TCRB   \n",
       "..                  ...   ...   \n",
       "354  DenverT1D-170_TCRB  TCRB   \n",
       "355  DenverT1D-106_TCRB  TCRB   \n",
       "356  DenverT1D-256_TCRB  TCRB   \n",
       "357  DenverT1D-055_TCRB  TCRB   \n",
       "358  DenverT1D-271_TCRB  TCRB   \n",
       "\n",
       "                                           sample_tags counting_method  \\\n",
       "0    03 Years, 12.6575342465753 Years at diagnosis,...              v4   \n",
       "1    06 Years, 6 Years at visit, Caucasian, Control...              v4   \n",
       "2    03 Years, 11.5095890410959 Years at diagnosis,...              v4   \n",
       "3    0 Years at visit, 09 Months, Caucasian, Contro...              v4   \n",
       "4    08 Years, 12.6575342465753 Years at diagnosis,...              v4   \n",
       "..                                                 ...             ...   \n",
       "354  17 Years, 17 Years at diagnosis, 18-24 Years (...              v4   \n",
       "355  12-18 Years (Adolescence), 16 Years, 16 Years ...              v4   \n",
       "356  12-18 Years (Adolescence), 16 Years, 16 Years ...              v4   \n",
       "357  02-12 (Childhood), 07 Years, 7 Years at diagno...              v4   \n",
       "358  02-12 (Childhood), 09 Years, 9 Years at diagno...              v4   \n",
       "\n",
       "           primer_set species               study sequencing_type  \n",
       "0    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA  \n",
       "1    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA  \n",
       "2    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA  \n",
       "3    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA  \n",
       "4    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA  \n",
       "..                ...     ...                 ...             ...  \n",
       "354   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA  \n",
       "355   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA  \n",
       "356   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA  \n",
       "357   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA  \n",
       "358   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA  \n",
       "\n",
       "[359 rows x 8 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = _load_metadata_df(\"mitchell-2022-jcii\")\n",
    "\n",
    "\"\"\"\n",
    "https://insight.jci.org/articles/view/161885\n",
    "Childhood T1D:\n",
    "\n",
    "143 new-onset T1D: \"As an independent validation, we sequenced and analyzed TCRÎ² repertoires from a cohort of new-onset T1D patients (n=143)\"\n",
    "\n",
    "25 HHC (4 samples each)\n",
    "29 children that progress to T1D (4 samples each)\n",
    "\"longitudinal peripheral blood DNA samples at four time points beginning early in life (median age of 1.4 years) from children that progressed to T1D (n=29) and age/sex-matched islet autoantibody negative controls (n=25)\"\n",
    "\"\"\"\n",
    "\n",
    "df[\"sequencing_type\"] = \"gDNA\"\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0             310114\n",
       "1             310251\n",
       "2             310113\n",
       "3             310205\n",
       "4             310283\n",
       "           ...      \n",
       "354    DenverT1D-170\n",
       "355    DenverT1D-106\n",
       "356    DenverT1D-256\n",
       "357    DenverT1D-055\n",
       "358    DenverT1D-271\n",
       "Name: specimen_label, Length: 359, dtype: object"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"specimen_label\"] = df[\"sample_name\"].str.split(\"_TCRB\").str[0]\n",
    "df[\"specimen_label\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    216\n",
       "True     143\n",
       "Name: sample_name, dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# new onset n=143 cohort is \"Denver\"\n",
    "df[\"sample_name\"].str.contains(\"Denver\").value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "216"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The other 216 samples\n",
    "29 * 4 + 25 * 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Caucasian                 354\n",
       "Non-Hispanic or Latino    310\n",
       "HLA-DQB1*0302             238\n",
       "HLA-DQA1*0301             216\n",
       "Male                      214\n",
       "HLA-DPA1*0103             212\n",
       "IA-2 0                    182\n",
       "HLA-DQB1*0201             178\n",
       "HLA-DRB1*0301             178\n",
       "HLA-DQA1*0501             178\n",
       "GAD65 0                   161\n",
       "HLA-DRB1*0401             150\n",
       "HLA-DPB1*0401             148\n",
       "Female                    145\n",
       "Type 1 Diabetes           143\n",
       "Case                      116\n",
       "ZnT8 NR                   109\n",
       "HLA-A*0201                108\n",
       "IAA 0                     108\n",
       "Whole Blood               104\n",
       "Name: sample_tags, dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"sample_tags\"].str.split(\",\").explode().str.strip().value_counts().head(n=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now: import subject IDs, age, sex, ancestry from https://insight.jci.org/articles/view/161885/sd/2\n",
    "# Notice that the 29 children that progress to T1D don't necessarily have T1D diagnosis prior to the last sample\n",
    "# So only include the new-onset T1D samples and the 25 healthy control children (with repeated samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adaptive ID</th>\n",
       "      <th>Age at Visit (years)</th>\n",
       "      <th>Sample Timepoint</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>GAD65 (nl &lt; 20 DK Units)</th>\n",
       "      <th>IA-2 (nl &lt; 5 DK Units)</th>\n",
       "      <th>IAA (nl &lt; 0.011)*</th>\n",
       "      <th>ZnT8 (nl &lt; 0.021)</th>\n",
       "      <th>...</th>\n",
       "      <th>PPI TCR 9</th>\n",
       "      <th>PPI TCR 10</th>\n",
       "      <th>PPI TCR 11</th>\n",
       "      <th>PPI TCR 12</th>\n",
       "      <th>PPI TCR 13</th>\n",
       "      <th>PPI TCR 14</th>\n",
       "      <th>PPI TCR 15</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>310101</td>\n",
       "      <td>2.913073</td>\n",
       "      <td>1</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0.006</td>\n",
       "      <td>NR</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>control_1</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>310103</td>\n",
       "      <td>3.961670</td>\n",
       "      <td>2</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.002</td>\n",
       "      <td>NR</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>control_1</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>310216</td>\n",
       "      <td>8.570000</td>\n",
       "      <td>3</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NR</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>control_1</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>310146</td>\n",
       "      <td>12.720000</td>\n",
       "      <td>4</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NR</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>control_1</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>310104</td>\n",
       "      <td>1.256673</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.009</td>\n",
       "      <td>NR</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>control_2</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>310163</td>\n",
       "      <td>10.570000</td>\n",
       "      <td>4</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NR</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>control_24</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>310219</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NR</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>control_25</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>310172</td>\n",
       "      <td>9.875427</td>\n",
       "      <td>2</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.002</td>\n",
       "      <td>NR</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>control_25</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>310182</td>\n",
       "      <td>11.767282</td>\n",
       "      <td>3</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.004</td>\n",
       "      <td>NR</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>control_25</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>310198</td>\n",
       "      <td>19.843942</td>\n",
       "      <td>4</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001</td>\n",
       "      <td>NR</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>control_25</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Adaptive ID  Age at Visit (years)  Sample Timepoint     Sex   Race  \\\n",
       "0        310101              2.913073                 1  Female  White   \n",
       "1        310103              3.961670                 2  Female  White   \n",
       "2        310216              8.570000                 3  Female  White   \n",
       "3        310146             12.720000                 4  Female  White   \n",
       "4        310104              1.256673                 1    Male  White   \n",
       "..          ...                   ...               ...     ...    ...   \n",
       "95       310163             10.570000                 4    Male  White   \n",
       "96       310219              0.860000                 1    Male  White   \n",
       "97       310172              9.875427                 2    Male  White   \n",
       "98       310182             11.767282                 3    Male  White   \n",
       "99       310198             19.843942                 4    Male  White   \n",
       "\n",
       "       Ethnicity  GAD65 (nl < 20 DK Units)  IA-2 (nl < 5 DK Units)  \\\n",
       "0   Non-Hispanic                         0                       5   \n",
       "1   Non-Hispanic                         3                       0   \n",
       "2   Non-Hispanic                         0                       0   \n",
       "3   Non-Hispanic                         0                       0   \n",
       "4   Non-Hispanic                         0                       0   \n",
       "..           ...                       ...                     ...   \n",
       "95  Non-Hispanic                         0                       0   \n",
       "96  Non-Hispanic                         0                       0   \n",
       "97  Non-Hispanic                         0                       0   \n",
       "98  Non-Hispanic                         0                       3   \n",
       "99  Non-Hispanic                         0                       0   \n",
       "\n",
       "    IAA (nl < 0.011)* ZnT8 (nl < 0.021)  ... PPI TCR 9 PPI TCR 10 PPI TCR 11  \\\n",
       "0               0.006                NR  ...         0          0          0   \n",
       "1               0.002                NR  ...         0          0          0   \n",
       "2               0.000                NR  ...         0          0          0   \n",
       "3               0.000                NR  ...         0          0          0   \n",
       "4               0.009                NR  ...         0          0          0   \n",
       "..                ...               ...  ...       ...        ...        ...   \n",
       "95              0.000                NR  ...         0          1          0   \n",
       "96              0.000                NR  ...         0          0          0   \n",
       "97              0.002                NR  ...         0          0          0   \n",
       "98              0.004                NR  ...         0          0          0   \n",
       "99              0.001                NR  ...         0          0          0   \n",
       "\n",
       "   PPI TCR 12 PPI TCR 13 PPI TCR 14 PPI TCR 15 participant_label  \\\n",
       "0           0          0          0          0         control_1   \n",
       "1           0          0          0          0         control_1   \n",
       "2           0          0          0          0         control_1   \n",
       "3           0          0          0          0         control_1   \n",
       "4           0          0          0          0         control_2   \n",
       "..        ...        ...        ...        ...               ...   \n",
       "95          0          0          0          0        control_24   \n",
       "96          0          0          0          0        control_25   \n",
       "97          0          0          0          0        control_25   \n",
       "98          0          0          0          0        control_25   \n",
       "99          0          0          0          0        control_25   \n",
       "\n",
       "               disease                    disease_subtype  \n",
       "0   Healthy/Background  Healthy/Background - T1D negative  \n",
       "1   Healthy/Background  Healthy/Background - T1D negative  \n",
       "2   Healthy/Background  Healthy/Background - T1D negative  \n",
       "3   Healthy/Background  Healthy/Background - T1D negative  \n",
       "4   Healthy/Background  Healthy/Background - T1D negative  \n",
       "..                 ...                                ...  \n",
       "95  Healthy/Background  Healthy/Background - T1D negative  \n",
       "96  Healthy/Background  Healthy/Background - T1D negative  \n",
       "97  Healthy/Background  Healthy/Background - T1D negative  \n",
       "98  Healthy/Background  Healthy/Background - T1D negative  \n",
       "99  Healthy/Background  Healthy/Background - T1D negative  \n",
       "\n",
       "[100 rows x 45 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_samples = pd.read_excel(\n",
    "    config.paths.metadata_dir / \"adaptive\" / \"jci.insight.161885.sdd1.xlsx\",\n",
    "    sheet_name=\"DAISY Control\",\n",
    ")\n",
    "\n",
    "# first row per person has Sex/Race/Ethnicity not null\n",
    "not_null_demographics = control_samples[\"Sex\"].notna()\n",
    "\n",
    "# double check that all those rows have sample timepoint = 1\n",
    "assert (control_samples.loc[not_null_demographics, \"Sample Timepoint\"] == 1).all()\n",
    "\n",
    "\n",
    "# Assign 'participant_label' as the cumulative sum of 'not_null_demographics', starting from 1\n",
    "control_samples[\n",
    "    \"participant_label\"\n",
    "] = \"control_\" + not_null_demographics.cumsum().astype(str)\n",
    "control_samples[\"disease\"] = healthy_label\n",
    "control_samples[\"disease_subtype\"] = f\"{healthy_label} - T1D negative\"\n",
    "\n",
    "# Transfer demographics to rest of rows per person\n",
    "# Fill null column values in each group with the first non-null column value in the group (ffill() is the key part, but bfill() saves us if for some reason the first row isn't the one with the entry - not the case for us here but just in case)\n",
    "for col in [\"Sex\", \"Race\", \"Ethnicity\"]:\n",
    "    control_samples[col] = (\n",
    "        control_samples.groupby(\"participant_label\")[col].bfill().ffill()\n",
    "    )\n",
    "\n",
    "control_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adaptive ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sample Timepoint</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>310101</td>\n",
       "      <td>2.913073</td>\n",
       "      <td>1</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_1</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>310103</td>\n",
       "      <td>3.961670</td>\n",
       "      <td>2</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_1</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>310216</td>\n",
       "      <td>8.570000</td>\n",
       "      <td>3</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_1</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>310146</td>\n",
       "      <td>12.720000</td>\n",
       "      <td>4</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_1</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>310104</td>\n",
       "      <td>1.256673</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_2</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>310163</td>\n",
       "      <td>10.570000</td>\n",
       "      <td>4</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_24</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>310219</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_25</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>310172</td>\n",
       "      <td>9.875427</td>\n",
       "      <td>2</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_25</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>310182</td>\n",
       "      <td>11.767282</td>\n",
       "      <td>3</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_25</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>310198</td>\n",
       "      <td>19.843942</td>\n",
       "      <td>4</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_25</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Adaptive ID        Age  Sample Timepoint     Sex   Race     Ethnicity  \\\n",
       "0       310101   2.913073                 1  Female  White  Non-Hispanic   \n",
       "1       310103   3.961670                 2  Female  White  Non-Hispanic   \n",
       "2       310216   8.570000                 3  Female  White  Non-Hispanic   \n",
       "3       310146  12.720000                 4  Female  White  Non-Hispanic   \n",
       "4       310104   1.256673                 1    Male  White  Non-Hispanic   \n",
       "..         ...        ...               ...     ...    ...           ...   \n",
       "95      310163  10.570000                 4    Male  White  Non-Hispanic   \n",
       "96      310219   0.860000                 1    Male  White  Non-Hispanic   \n",
       "97      310172   9.875427                 2    Male  White  Non-Hispanic   \n",
       "98      310182  11.767282                 3    Male  White  Non-Hispanic   \n",
       "99      310198  19.843942                 4    Male  White  Non-Hispanic   \n",
       "\n",
       "   participant_label             disease                    disease_subtype  \n",
       "0          control_1  Healthy/Background  Healthy/Background - T1D negative  \n",
       "1          control_1  Healthy/Background  Healthy/Background - T1D negative  \n",
       "2          control_1  Healthy/Background  Healthy/Background - T1D negative  \n",
       "3          control_1  Healthy/Background  Healthy/Background - T1D negative  \n",
       "4          control_2  Healthy/Background  Healthy/Background - T1D negative  \n",
       "..               ...                 ...                                ...  \n",
       "95        control_24  Healthy/Background  Healthy/Background - T1D negative  \n",
       "96        control_25  Healthy/Background  Healthy/Background - T1D negative  \n",
       "97        control_25  Healthy/Background  Healthy/Background - T1D negative  \n",
       "98        control_25  Healthy/Background  Healthy/Background - T1D negative  \n",
       "99        control_25  Healthy/Background  Healthy/Background - T1D negative  \n",
       "\n",
       "[100 rows x 9 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_samples = control_samples[\n",
    "    [\n",
    "        \"Adaptive ID\",\n",
    "        \"Age at Visit (years)\",\n",
    "        \"Sample Timepoint\",\n",
    "        \"Sex\",\n",
    "        \"Race\",\n",
    "        \"Ethnicity\",\n",
    "        \"participant_label\",\n",
    "        \"disease\",\n",
    "        \"disease_subtype\",\n",
    "    ]\n",
    "].rename(columns={\"Age at Visit (years)\": \"Age\"})\n",
    "control_samples[\"Adaptive ID\"] = control_samples[\"Adaptive ID\"].astype(str)\n",
    "control_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sample Timepoint</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>310251_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>06 Years, 6 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310251</td>\n",
       "      <td>6.661190</td>\n",
       "      <td>4</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>control_16</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>310205_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>0 Years at visit, 09 Months, Caucasian, Contro...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310205</td>\n",
       "      <td>0.829568</td>\n",
       "      <td>1</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_6</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>310168_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>05 Years, 5 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310168</td>\n",
       "      <td>5.670088</td>\n",
       "      <td>3</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_13</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>310170_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>08 Years, 8 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310170</td>\n",
       "      <td>8.407939</td>\n",
       "      <td>2</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_12</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>310148_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>06 Years, 6 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310148</td>\n",
       "      <td>6.269678</td>\n",
       "      <td>4</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_9</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>310137_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>07 Years, 7 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310137</td>\n",
       "      <td>7.140314</td>\n",
       "      <td>3</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_14</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>310217_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>03 Years, 3 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310217</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>3</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_20</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>310188_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>09 Years, 9 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310188</td>\n",
       "      <td>9.607118</td>\n",
       "      <td>4</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_22</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>310106_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>01 Years, 1 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310106</td>\n",
       "      <td>1.212867</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_23</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>310167_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>13 Years, 13 Years at visit, Caucasian, Contro...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310167</td>\n",
       "      <td>13.492128</td>\n",
       "      <td>4</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_2</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_name locus                                        sample_tags  \\\n",
       "0   310251_TCRB  TCRB  06 Years, 6 Years at visit, Caucasian, Control...   \n",
       "1   310205_TCRB  TCRB  0 Years at visit, 09 Months, Caucasian, Contro...   \n",
       "2   310168_TCRB  TCRB  05 Years, 5 Years at visit, Caucasian, Control...   \n",
       "3   310170_TCRB  TCRB  08 Years, 8 Years at visit, Caucasian, Control...   \n",
       "4   310148_TCRB  TCRB  06 Years, 6 Years at visit, Caucasian, Control...   \n",
       "..          ...   ...                                                ...   \n",
       "95  310137_TCRB  TCRB  07 Years, 7 Years at visit, Caucasian, Control...   \n",
       "96  310217_TCRB  TCRB  03 Years, 3 Years at visit, Caucasian, Control...   \n",
       "97  310188_TCRB  TCRB  09 Years, 9 Years at visit, Caucasian, Control...   \n",
       "98  310106_TCRB  TCRB  01 Years, 1 Years at visit, Caucasian, Control...   \n",
       "99  310167_TCRB  TCRB  13 Years, 13 Years at visit, Caucasian, Contro...   \n",
       "\n",
       "   counting_method        primer_set species               study  \\\n",
       "0               v4  Human-TCRB-PD4bx   Human  mitchell-2022-jcii   \n",
       "1               v4  Human-TCRB-PD4bx   Human  mitchell-2022-jcii   \n",
       "2               v4  Human-TCRB-PD4bx   Human  mitchell-2022-jcii   \n",
       "3               v4  Human-TCRB-PD4bx   Human  mitchell-2022-jcii   \n",
       "4               v4  Human-TCRB-PD4bx   Human  mitchell-2022-jcii   \n",
       "..             ...               ...     ...                 ...   \n",
       "95              v4  Human-TCRB-PD4bx   Human  mitchell-2022-jcii   \n",
       "96              v4  Human-TCRB-PD4bx   Human  mitchell-2022-jcii   \n",
       "97              v4  Human-TCRB-PD4bx   Human  mitchell-2022-jcii   \n",
       "98              v4  Human-TCRB-PD4bx   Human  mitchell-2022-jcii   \n",
       "99              v4  Human-TCRB-PD4bx   Human  mitchell-2022-jcii   \n",
       "\n",
       "   sequencing_type specimen_label        Age  Sample Timepoint     Sex   Race  \\\n",
       "0             gDNA         310251   6.661190                 4  Female  White   \n",
       "1             gDNA         310205   0.829568                 1  Female  White   \n",
       "2             gDNA         310168   5.670088                 3    Male  White   \n",
       "3             gDNA         310170   8.407939                 2    Male  White   \n",
       "4             gDNA         310148   6.269678                 4    Male  White   \n",
       "..             ...            ...        ...               ...     ...    ...   \n",
       "95            gDNA         310137   7.140314                 3  Female  White   \n",
       "96            gDNA         310217   3.300000                 3    Male  White   \n",
       "97            gDNA         310188   9.607118                 4    Male  White   \n",
       "98            gDNA         310106   1.212867                 1    Male  White   \n",
       "99            gDNA         310167  13.492128                 4    Male  White   \n",
       "\n",
       "       Ethnicity participant_label             disease  \\\n",
       "0       Hispanic        control_16  Healthy/Background   \n",
       "1   Non-Hispanic         control_6  Healthy/Background   \n",
       "2   Non-Hispanic        control_13  Healthy/Background   \n",
       "3   Non-Hispanic        control_12  Healthy/Background   \n",
       "4   Non-Hispanic         control_9  Healthy/Background   \n",
       "..           ...               ...                 ...   \n",
       "95  Non-Hispanic        control_14  Healthy/Background   \n",
       "96  Non-Hispanic        control_20  Healthy/Background   \n",
       "97  Non-Hispanic        control_22  Healthy/Background   \n",
       "98  Non-Hispanic        control_23  Healthy/Background   \n",
       "99  Non-Hispanic         control_2  Healthy/Background   \n",
       "\n",
       "                      disease_subtype  \n",
       "0   Healthy/Background - T1D negative  \n",
       "1   Healthy/Background - T1D negative  \n",
       "2   Healthy/Background - T1D negative  \n",
       "3   Healthy/Background - T1D negative  \n",
       "4   Healthy/Background - T1D negative  \n",
       "..                                ...  \n",
       "95  Healthy/Background - T1D negative  \n",
       "96  Healthy/Background - T1D negative  \n",
       "97  Healthy/Background - T1D negative  \n",
       "98  Healthy/Background - T1D negative  \n",
       "99  Healthy/Background - T1D negative  \n",
       "\n",
       "[100 rows x 17 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_samples = pd.merge(\n",
    "    df[~df[\"sample_name\"].str.contains(\"Denver\")],\n",
    "    control_samples,\n",
    "    how=\"inner\",\n",
    "    left_on=\"specimen_label\",\n",
    "    right_on=\"Adaptive ID\",\n",
    "    validate=\"1:1\",\n",
    ").drop(columns=\"Adaptive ID\")\n",
    "# left with 25*4 as expected\n",
    "control_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "control_16    4\n",
       "control_7     4\n",
       "control_14    4\n",
       "control_23    4\n",
       "control_24    4\n",
       "control_2     4\n",
       "control_25    4\n",
       "control_4     4\n",
       "control_10    4\n",
       "control_22    4\n",
       "control_5     4\n",
       "control_15    4\n",
       "control_1     4\n",
       "control_6     4\n",
       "control_20    4\n",
       "control_18    4\n",
       "control_17    4\n",
       "control_11    4\n",
       "control_19    4\n",
       "control_3     4\n",
       "control_21    4\n",
       "control_9     4\n",
       "control_12    4\n",
       "control_13    4\n",
       "control_8     4\n",
       "Name: participant_label, dtype: int64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_samples[\"participant_label\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adaptive ID</th>\n",
       "      <th>Age at Diagnosis (years)</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>T1D Duration (days)</th>\n",
       "      <th>GAD65 (nl &lt; 20 DK Units)</th>\n",
       "      <th>IA-2 (nl &lt; 5 DK Units)</th>\n",
       "      <th>IAA (nl &lt; 0.011)</th>\n",
       "      <th>ZnT8 (nl &lt; 0.021)</th>\n",
       "      <th>...</th>\n",
       "      <th>PPI TCR 9</th>\n",
       "      <th>PPI TCR 10</th>\n",
       "      <th>PPI TCR 11</th>\n",
       "      <th>PPI TCR 12</th>\n",
       "      <th>PPI TCR 13</th>\n",
       "      <th>PPI TCR 14</th>\n",
       "      <th>PPI TCR 15</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>13.5</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>1</td>\n",
       "      <td>23.0</td>\n",
       "      <td>212.0</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.621</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DenverT1D-009</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>41</td>\n",
       "      <td>14.9</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>15</td>\n",
       "      <td>36.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.005</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DenverT1D-041</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42</td>\n",
       "      <td>9.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>0</td>\n",
       "      <td>703.0</td>\n",
       "      <td>196.0</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.398</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DenverT1D-042</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>43</td>\n",
       "      <td>5.5</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>246.0</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.030</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DenverT1D-043</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>44</td>\n",
       "      <td>16.5</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>239.0</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.270</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DenverT1D-044</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>382</td>\n",
       "      <td>11.6</td>\n",
       "      <td>Male</td>\n",
       "      <td>Black</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>55</td>\n",
       "      <td>168.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.021</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DenverT1D-382</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>383</td>\n",
       "      <td>10.8</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>2</td>\n",
       "      <td>68.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.458</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DenverT1D-383</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>384</td>\n",
       "      <td>10.9</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>3</td>\n",
       "      <td>613.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>0.031</td>\n",
       "      <td>0.208</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DenverT1D-384</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>386</td>\n",
       "      <td>10.1</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>3</td>\n",
       "      <td>279.0</td>\n",
       "      <td>271.0</td>\n",
       "      <td>0.096</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DenverT1D-386</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>397</td>\n",
       "      <td>9.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>37</td>\n",
       "      <td>14.0</td>\n",
       "      <td>324.0</td>\n",
       "      <td>0.089</td>\n",
       "      <td>0.495</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DenverT1D-397</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>143 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Adaptive ID  Age at Diagnosis (years)     Sex   Race     Ethnicity  \\\n",
       "0              9                      13.5  Female  White  Non-Hispanic   \n",
       "1             41                      14.9  Female  White  Non-Hispanic   \n",
       "2             42                       9.0  Female  White  Non-Hispanic   \n",
       "3             43                       5.5  Female  White  Non-Hispanic   \n",
       "4             44                      16.5    Male  White  Non-Hispanic   \n",
       "..           ...                       ...     ...    ...           ...   \n",
       "138          382                      11.6    Male  Black  Non-Hispanic   \n",
       "139          383                      10.8  Female  White  Non-Hispanic   \n",
       "140          384                      10.9    Male  White  Non-Hispanic   \n",
       "141          386                      10.1  Female  White  Non-Hispanic   \n",
       "142          397                       9.0    Male  White  Non-Hispanic   \n",
       "\n",
       "     T1D Duration (days)  GAD65 (nl < 20 DK Units)  IA-2 (nl < 5 DK Units)  \\\n",
       "0                      1                      23.0                   212.0   \n",
       "1                     15                      36.0                     2.0   \n",
       "2                      0                     703.0                   196.0   \n",
       "3                      1                       9.0                   246.0   \n",
       "4                      1                       5.0                   239.0   \n",
       "..                   ...                       ...                     ...   \n",
       "138                   55                     168.0                   381.0   \n",
       "139                    2                      68.0                     0.0   \n",
       "140                    3                     613.0                   348.0   \n",
       "141                    3                     279.0                   271.0   \n",
       "142                   37                      14.0                   324.0   \n",
       "\n",
       "     IAA (nl < 0.011)  ZnT8 (nl < 0.021)  ...  PPI TCR 9  PPI TCR 10  \\\n",
       "0               0.007              0.621  ...        NaN           0   \n",
       "1               0.000              0.005  ...        NaN           0   \n",
       "2               0.011              0.398  ...        NaN           0   \n",
       "3               0.006              0.030  ...        NaN           0   \n",
       "4               0.001              0.270  ...        NaN           0   \n",
       "..                ...                ...  ...        ...         ...   \n",
       "138             0.011              0.021  ...        NaN           0   \n",
       "139             0.000              0.458  ...        NaN           0   \n",
       "140             0.031              0.208  ...        NaN           0   \n",
       "141             0.096              0.000  ...        NaN           0   \n",
       "142             0.089              0.495  ...        NaN           0   \n",
       "\n",
       "     PPI TCR 11  PPI TCR 12  PPI TCR 13  PPI TCR 14  PPI TCR 15  \\\n",
       "0             0         NaN         NaN           0           0   \n",
       "1             0         NaN         NaN           0           0   \n",
       "2             0         NaN         NaN           0           0   \n",
       "3             0         NaN         NaN           0           0   \n",
       "4             0         NaN         NaN           0           0   \n",
       "..          ...         ...         ...         ...         ...   \n",
       "138           0         NaN         NaN           0           0   \n",
       "139           0         NaN         NaN           0           0   \n",
       "140           0         NaN         NaN           0           0   \n",
       "141           0         NaN         NaN           0           0   \n",
       "142           0         NaN         NaN           0           0   \n",
       "\n",
       "     participant_label  disease  disease_subtype  \n",
       "0        DenverT1D-009      T1D  T1D - new onset  \n",
       "1        DenverT1D-041      T1D  T1D - new onset  \n",
       "2        DenverT1D-042      T1D  T1D - new onset  \n",
       "3        DenverT1D-043      T1D  T1D - new onset  \n",
       "4        DenverT1D-044      T1D  T1D - new onset  \n",
       "..                 ...      ...              ...  \n",
       "138      DenverT1D-382      T1D  T1D - new onset  \n",
       "139      DenverT1D-383      T1D  T1D - new onset  \n",
       "140      DenverT1D-384      T1D  T1D - new onset  \n",
       "141      DenverT1D-386      T1D  T1D - new onset  \n",
       "142      DenverT1D-397      T1D  T1D - new onset  \n",
       "\n",
       "[143 rows x 35 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_onset_samples = pd.read_excel(\n",
    "    config.paths.metadata_dir / \"adaptive\" / \"jci.insight.161885.sdd1.xlsx\",\n",
    "    sheet_name=\"New-onset T1D\",\n",
    ")\n",
    "# Left zero pad\n",
    "new_onset_samples[\"participant_label\"] = \"DenverT1D-\" + new_onset_samples[\n",
    "    \"Adaptive ID\"\n",
    "].astype(str).str.zfill(3)\n",
    "new_onset_samples[\"disease\"] = \"T1D\"\n",
    "new_onset_samples[\"disease_subtype\"] = \"T1D - new onset\"\n",
    "new_onset_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    143.000000\n",
       "mean       4.881119\n",
       "std        9.059653\n",
       "min        0.000000\n",
       "25%        1.000000\n",
       "50%        2.000000\n",
       "75%        5.000000\n",
       "max       55.000000\n",
       "Name: T1D Duration (days), dtype: float64"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# these are really new onset\n",
    "new_onset_samples[\"T1D Duration (days)\"].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DenverT1D-009</td>\n",
       "      <td>13.5</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DenverT1D-041</td>\n",
       "      <td>14.9</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DenverT1D-042</td>\n",
       "      <td>9.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DenverT1D-043</td>\n",
       "      <td>5.5</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DenverT1D-044</td>\n",
       "      <td>16.5</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>DenverT1D-382</td>\n",
       "      <td>11.6</td>\n",
       "      <td>Male</td>\n",
       "      <td>Black</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>DenverT1D-383</td>\n",
       "      <td>10.8</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>DenverT1D-384</td>\n",
       "      <td>10.9</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>DenverT1D-386</td>\n",
       "      <td>10.1</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>DenverT1D-397</td>\n",
       "      <td>9.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>143 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    participant_label   Age     Sex   Race     Ethnicity disease  \\\n",
       "0       DenverT1D-009  13.5  Female  White  Non-Hispanic     T1D   \n",
       "1       DenverT1D-041  14.9  Female  White  Non-Hispanic     T1D   \n",
       "2       DenverT1D-042   9.0  Female  White  Non-Hispanic     T1D   \n",
       "3       DenverT1D-043   5.5  Female  White  Non-Hispanic     T1D   \n",
       "4       DenverT1D-044  16.5    Male  White  Non-Hispanic     T1D   \n",
       "..                ...   ...     ...    ...           ...     ...   \n",
       "138     DenverT1D-382  11.6    Male  Black  Non-Hispanic     T1D   \n",
       "139     DenverT1D-383  10.8  Female  White  Non-Hispanic     T1D   \n",
       "140     DenverT1D-384  10.9    Male  White  Non-Hispanic     T1D   \n",
       "141     DenverT1D-386  10.1  Female  White  Non-Hispanic     T1D   \n",
       "142     DenverT1D-397   9.0    Male  White  Non-Hispanic     T1D   \n",
       "\n",
       "     disease_subtype  \n",
       "0    T1D - new onset  \n",
       "1    T1D - new onset  \n",
       "2    T1D - new onset  \n",
       "3    T1D - new onset  \n",
       "4    T1D - new onset  \n",
       "..               ...  \n",
       "138  T1D - new onset  \n",
       "139  T1D - new onset  \n",
       "140  T1D - new onset  \n",
       "141  T1D - new onset  \n",
       "142  T1D - new onset  \n",
       "\n",
       "[143 rows x 7 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_onset_samples = new_onset_samples[\n",
    "    [\n",
    "        \"participant_label\",\n",
    "        \"Age at Diagnosis (years)\",\n",
    "        \"Sex\",\n",
    "        \"Race\",\n",
    "        \"Ethnicity\",\n",
    "        \"disease\",\n",
    "        \"disease_subtype\",\n",
    "    ]\n",
    "].rename(columns={\"Age at Diagnosis (years)\": \"Age\"})\n",
    "new_onset_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>specimen_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>DenverT1D-045_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 15 Years, 15 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>DenverT1D-041_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 14 Years, 14 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>DenverT1D-103_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 09 Years, 9 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>DenverT1D-332_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 08 Years, 8 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>DenverT1D-179_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12 Years, 12 Years at diagnosis, 12-18 Years (...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>DenverT1D-170_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>17 Years, 17 Years at diagnosis, 18-24 Years (...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>DenverT1D-106_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 16 Years, 16 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>DenverT1D-256_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 16 Years, 16 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>DenverT1D-055_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 07 Years, 7 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>DenverT1D-271_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 09 Years, 9 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-271</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>143 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample_name locus  \\\n",
       "216  DenverT1D-045_TCRB  TCRB   \n",
       "217  DenverT1D-041_TCRB  TCRB   \n",
       "218  DenverT1D-103_TCRB  TCRB   \n",
       "219  DenverT1D-332_TCRB  TCRB   \n",
       "220  DenverT1D-179_TCRB  TCRB   \n",
       "..                  ...   ...   \n",
       "354  DenverT1D-170_TCRB  TCRB   \n",
       "355  DenverT1D-106_TCRB  TCRB   \n",
       "356  DenverT1D-256_TCRB  TCRB   \n",
       "357  DenverT1D-055_TCRB  TCRB   \n",
       "358  DenverT1D-271_TCRB  TCRB   \n",
       "\n",
       "                                           sample_tags counting_method  \\\n",
       "216  12-18 Years (Adolescence), 15 Years, 15 Years ...              v4   \n",
       "217  12-18 Years (Adolescence), 14 Years, 14 Years ...              v4   \n",
       "218  02-12 (Childhood), 09 Years, 9 Years at diagno...              v4   \n",
       "219  02-12 (Childhood), 08 Years, 8 Years at diagno...              v4   \n",
       "220  12 Years, 12 Years at diagnosis, 12-18 Years (...              v4   \n",
       "..                                                 ...             ...   \n",
       "354  17 Years, 17 Years at diagnosis, 18-24 Years (...              v4   \n",
       "355  12-18 Years (Adolescence), 16 Years, 16 Years ...              v4   \n",
       "356  12-18 Years (Adolescence), 16 Years, 16 Years ...              v4   \n",
       "357  02-12 (Childhood), 07 Years, 7 Years at diagno...              v4   \n",
       "358  02-12 (Childhood), 09 Years, 9 Years at diagno...              v4   \n",
       "\n",
       "          primer_set species               study sequencing_type  \\\n",
       "216  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "217  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "218  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "219  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "220  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "..               ...     ...                 ...             ...   \n",
       "354  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "355  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "356  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "357  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "358  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "\n",
       "    specimen_label  \n",
       "216  DenverT1D-045  \n",
       "217  DenverT1D-041  \n",
       "218  DenverT1D-103  \n",
       "219  DenverT1D-332  \n",
       "220  DenverT1D-179  \n",
       "..             ...  \n",
       "354  DenverT1D-170  \n",
       "355  DenverT1D-106  \n",
       "356  DenverT1D-256  \n",
       "357  DenverT1D-055  \n",
       "358  DenverT1D-271  \n",
       "\n",
       "[143 rows x 9 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"sample_name\"].str.contains(\"Denver\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DenverT1D-045_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 15 Years, 15 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-045</td>\n",
       "      <td>DenverT1D-045</td>\n",
       "      <td>15.6</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DenverT1D-041_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 14 Years, 14 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-041</td>\n",
       "      <td>DenverT1D-041</td>\n",
       "      <td>14.9</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DenverT1D-103_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 09 Years, 9 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-103</td>\n",
       "      <td>DenverT1D-103</td>\n",
       "      <td>9.2</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DenverT1D-332_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 08 Years, 8 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-332</td>\n",
       "      <td>DenverT1D-332</td>\n",
       "      <td>8.1</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DenverT1D-179_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12 Years, 12 Years at diagnosis, 12-18 Years (...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-179</td>\n",
       "      <td>DenverT1D-179</td>\n",
       "      <td>12.8</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>DenverT1D-170_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>17 Years, 17 Years at diagnosis, 18-24 Years (...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-170</td>\n",
       "      <td>DenverT1D-170</td>\n",
       "      <td>17.9</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>DenverT1D-106_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 16 Years, 16 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-106</td>\n",
       "      <td>DenverT1D-106</td>\n",
       "      <td>16.4</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>DenverT1D-256_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 16 Years, 16 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-256</td>\n",
       "      <td>DenverT1D-256</td>\n",
       "      <td>16.1</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>DenverT1D-055_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 07 Years, 7 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-055</td>\n",
       "      <td>DenverT1D-055</td>\n",
       "      <td>7.5</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>DenverT1D-271_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 09 Years, 9 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-271</td>\n",
       "      <td>DenverT1D-271</td>\n",
       "      <td>9.8</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>143 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample_name locus  \\\n",
       "0    DenverT1D-045_TCRB  TCRB   \n",
       "1    DenverT1D-041_TCRB  TCRB   \n",
       "2    DenverT1D-103_TCRB  TCRB   \n",
       "3    DenverT1D-332_TCRB  TCRB   \n",
       "4    DenverT1D-179_TCRB  TCRB   \n",
       "..                  ...   ...   \n",
       "138  DenverT1D-170_TCRB  TCRB   \n",
       "139  DenverT1D-106_TCRB  TCRB   \n",
       "140  DenverT1D-256_TCRB  TCRB   \n",
       "141  DenverT1D-055_TCRB  TCRB   \n",
       "142  DenverT1D-271_TCRB  TCRB   \n",
       "\n",
       "                                           sample_tags counting_method  \\\n",
       "0    12-18 Years (Adolescence), 15 Years, 15 Years ...              v4   \n",
       "1    12-18 Years (Adolescence), 14 Years, 14 Years ...              v4   \n",
       "2    02-12 (Childhood), 09 Years, 9 Years at diagno...              v4   \n",
       "3    02-12 (Childhood), 08 Years, 8 Years at diagno...              v4   \n",
       "4    12 Years, 12 Years at diagnosis, 12-18 Years (...              v4   \n",
       "..                                                 ...             ...   \n",
       "138  17 Years, 17 Years at diagnosis, 18-24 Years (...              v4   \n",
       "139  12-18 Years (Adolescence), 16 Years, 16 Years ...              v4   \n",
       "140  12-18 Years (Adolescence), 16 Years, 16 Years ...              v4   \n",
       "141  02-12 (Childhood), 07 Years, 7 Years at diagno...              v4   \n",
       "142  02-12 (Childhood), 09 Years, 9 Years at diagno...              v4   \n",
       "\n",
       "          primer_set species               study sequencing_type  \\\n",
       "0    Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "1    Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "2    Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "3    Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "4    Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "..               ...     ...                 ...             ...   \n",
       "138  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "139  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "140  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "141  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "142  Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "\n",
       "    specimen_label participant_label   Age     Sex   Race     Ethnicity  \\\n",
       "0    DenverT1D-045     DenverT1D-045  15.6  Female  White  Non-Hispanic   \n",
       "1    DenverT1D-041     DenverT1D-041  14.9  Female  White  Non-Hispanic   \n",
       "2    DenverT1D-103     DenverT1D-103   9.2  Female  White     Hispanic    \n",
       "3    DenverT1D-332     DenverT1D-332   8.1  Female  White  Non-Hispanic   \n",
       "4    DenverT1D-179     DenverT1D-179  12.8    Male  White  Non-Hispanic   \n",
       "..             ...               ...   ...     ...    ...           ...   \n",
       "138  DenverT1D-170     DenverT1D-170  17.9    Male  White  Non-Hispanic   \n",
       "139  DenverT1D-106     DenverT1D-106  16.4    Male  White  Non-Hispanic   \n",
       "140  DenverT1D-256     DenverT1D-256  16.1    Male  White  Non-Hispanic   \n",
       "141  DenverT1D-055     DenverT1D-055   7.5  Female  White  Non-Hispanic   \n",
       "142  DenverT1D-271     DenverT1D-271   9.8    Male  White  Non-Hispanic   \n",
       "\n",
       "    disease  disease_subtype  \n",
       "0       T1D  T1D - new onset  \n",
       "1       T1D  T1D - new onset  \n",
       "2       T1D  T1D - new onset  \n",
       "3       T1D  T1D - new onset  \n",
       "4       T1D  T1D - new onset  \n",
       "..      ...              ...  \n",
       "138     T1D  T1D - new onset  \n",
       "139     T1D  T1D - new onset  \n",
       "140     T1D  T1D - new onset  \n",
       "141     T1D  T1D - new onset  \n",
       "142     T1D  T1D - new onset  \n",
       "\n",
       "[143 rows x 16 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_onset_samples = pd.merge(\n",
    "    df[df[\"sample_name\"].str.contains(\"Denver\")],\n",
    "    new_onset_samples,\n",
    "    how=\"inner\",\n",
    "    left_on=\"specimen_label\",\n",
    "    right_on=\"participant_label\",\n",
    "    validate=\"1:1\",\n",
    ")\n",
    "# left with 143 as expected\n",
    "new_onset_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sample Timepoint</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>310251_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>06 Years, 6 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310251</td>\n",
       "      <td>6.661190</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>control_16</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>310205_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>0 Years at visit, 09 Months, Caucasian, Contro...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310205</td>\n",
       "      <td>0.829568</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_6</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>310168_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>05 Years, 5 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310168</td>\n",
       "      <td>5.670088</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_13</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>310170_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>08 Years, 8 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310170</td>\n",
       "      <td>8.407939</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_12</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>310148_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>06 Years, 6 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310148</td>\n",
       "      <td>6.269678</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_9</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>DenverT1D-170_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>17 Years, 17 Years at diagnosis, 18-24 Years (...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-170</td>\n",
       "      <td>17.900000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>DenverT1D-170</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>DenverT1D-106_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 16 Years, 16 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-106</td>\n",
       "      <td>16.400000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>DenverT1D-106</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>DenverT1D-256_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 16 Years, 16 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-256</td>\n",
       "      <td>16.100000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>DenverT1D-256</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>DenverT1D-055_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 07 Years, 7 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-055</td>\n",
       "      <td>7.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>DenverT1D-055</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>DenverT1D-271_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 09 Years, 9 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-271</td>\n",
       "      <td>9.800000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>DenverT1D-271</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>243 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample_name locus  \\\n",
       "0           310251_TCRB  TCRB   \n",
       "1           310205_TCRB  TCRB   \n",
       "2           310168_TCRB  TCRB   \n",
       "3           310170_TCRB  TCRB   \n",
       "4           310148_TCRB  TCRB   \n",
       "..                  ...   ...   \n",
       "138  DenverT1D-170_TCRB  TCRB   \n",
       "139  DenverT1D-106_TCRB  TCRB   \n",
       "140  DenverT1D-256_TCRB  TCRB   \n",
       "141  DenverT1D-055_TCRB  TCRB   \n",
       "142  DenverT1D-271_TCRB  TCRB   \n",
       "\n",
       "                                           sample_tags counting_method  \\\n",
       "0    06 Years, 6 Years at visit, Caucasian, Control...              v4   \n",
       "1    0 Years at visit, 09 Months, Caucasian, Contro...              v4   \n",
       "2    05 Years, 5 Years at visit, Caucasian, Control...              v4   \n",
       "3    08 Years, 8 Years at visit, Caucasian, Control...              v4   \n",
       "4    06 Years, 6 Years at visit, Caucasian, Control...              v4   \n",
       "..                                                 ...             ...   \n",
       "138  17 Years, 17 Years at diagnosis, 18-24 Years (...              v4   \n",
       "139  12-18 Years (Adolescence), 16 Years, 16 Years ...              v4   \n",
       "140  12-18 Years (Adolescence), 16 Years, 16 Years ...              v4   \n",
       "141  02-12 (Childhood), 07 Years, 7 Years at diagno...              v4   \n",
       "142  02-12 (Childhood), 09 Years, 9 Years at diagno...              v4   \n",
       "\n",
       "           primer_set species               study sequencing_type  \\\n",
       "0    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA   \n",
       "1    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA   \n",
       "2    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA   \n",
       "3    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA   \n",
       "4    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA   \n",
       "..                ...     ...                 ...             ...   \n",
       "138   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "139   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "140   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "141   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "142   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "\n",
       "    specimen_label        Age  Sample Timepoint     Sex   Race     Ethnicity  \\\n",
       "0           310251   6.661190               4.0  Female  White      Hispanic   \n",
       "1           310205   0.829568               1.0  Female  White  Non-Hispanic   \n",
       "2           310168   5.670088               3.0    Male  White  Non-Hispanic   \n",
       "3           310170   8.407939               2.0    Male  White  Non-Hispanic   \n",
       "4           310148   6.269678               4.0    Male  White  Non-Hispanic   \n",
       "..             ...        ...               ...     ...    ...           ...   \n",
       "138  DenverT1D-170  17.900000               NaN    Male  White  Non-Hispanic   \n",
       "139  DenverT1D-106  16.400000               NaN    Male  White  Non-Hispanic   \n",
       "140  DenverT1D-256  16.100000               NaN    Male  White  Non-Hispanic   \n",
       "141  DenverT1D-055   7.500000               NaN  Female  White  Non-Hispanic   \n",
       "142  DenverT1D-271   9.800000               NaN    Male  White  Non-Hispanic   \n",
       "\n",
       "    participant_label             disease                    disease_subtype  \n",
       "0          control_16  Healthy/Background  Healthy/Background - T1D negative  \n",
       "1           control_6  Healthy/Background  Healthy/Background - T1D negative  \n",
       "2          control_13  Healthy/Background  Healthy/Background - T1D negative  \n",
       "3          control_12  Healthy/Background  Healthy/Background - T1D negative  \n",
       "4           control_9  Healthy/Background  Healthy/Background - T1D negative  \n",
       "..                ...                 ...                                ...  \n",
       "138     DenverT1D-170                 T1D                    T1D - new onset  \n",
       "139     DenverT1D-106                 T1D                    T1D - new onset  \n",
       "140     DenverT1D-256                 T1D                    T1D - new onset  \n",
       "141     DenverT1D-055                 T1D                    T1D - new onset  \n",
       "142     DenverT1D-271                 T1D                    T1D - new onset  \n",
       "\n",
       "[243 rows x 17 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Recombine\n",
    "df = pd.concat([control_samples, new_onset_samples], axis=0)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sample Timepoint</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>310251_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>06 Years, 6 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310251</td>\n",
       "      <td>6.661190</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>control_16</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>310205_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>0 Years at visit, 09 Months, Caucasian, Contro...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310205</td>\n",
       "      <td>0.829568</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_6</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>310168_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>05 Years, 5 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310168</td>\n",
       "      <td>5.670088</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_13</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>310170_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>08 Years, 8 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310170</td>\n",
       "      <td>8.407939</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_12</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>310148_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>06 Years, 6 Years at visit, Caucasian, Control...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>310148</td>\n",
       "      <td>6.269678</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>control_9</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - T1D negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>DenverT1D-170_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>17 Years, 17 Years at diagnosis, 18-24 Years (...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-170</td>\n",
       "      <td>17.900000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>DenverT1D-170</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>DenverT1D-106_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 16 Years, 16 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-106</td>\n",
       "      <td>16.400000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>DenverT1D-106</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>DenverT1D-256_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>12-18 Years (Adolescence), 16 Years, 16 Years ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-256</td>\n",
       "      <td>16.100000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>DenverT1D-256</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>DenverT1D-055_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 07 Years, 7 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-055</td>\n",
       "      <td>7.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>DenverT1D-055</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>DenverT1D-271_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>02-12 (Childhood), 09 Years, 9 Years at diagno...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mitchell-2022-jcii</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>DenverT1D-271</td>\n",
       "      <td>9.800000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Non-Hispanic</td>\n",
       "      <td>DenverT1D-271</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D - new onset</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>243 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample_name locus  \\\n",
       "0           310251_TCRB  TCRB   \n",
       "1           310205_TCRB  TCRB   \n",
       "2           310168_TCRB  TCRB   \n",
       "3           310170_TCRB  TCRB   \n",
       "4           310148_TCRB  TCRB   \n",
       "..                  ...   ...   \n",
       "138  DenverT1D-170_TCRB  TCRB   \n",
       "139  DenverT1D-106_TCRB  TCRB   \n",
       "140  DenverT1D-256_TCRB  TCRB   \n",
       "141  DenverT1D-055_TCRB  TCRB   \n",
       "142  DenverT1D-271_TCRB  TCRB   \n",
       "\n",
       "                                           sample_tags counting_method  \\\n",
       "0    06 Years, 6 Years at visit, Caucasian, Control...              v4   \n",
       "1    0 Years at visit, 09 Months, Caucasian, Contro...              v4   \n",
       "2    05 Years, 5 Years at visit, Caucasian, Control...              v4   \n",
       "3    08 Years, 8 Years at visit, Caucasian, Control...              v4   \n",
       "4    06 Years, 6 Years at visit, Caucasian, Control...              v4   \n",
       "..                                                 ...             ...   \n",
       "138  17 Years, 17 Years at diagnosis, 18-24 Years (...              v4   \n",
       "139  12-18 Years (Adolescence), 16 Years, 16 Years ...              v4   \n",
       "140  12-18 Years (Adolescence), 16 Years, 16 Years ...              v4   \n",
       "141  02-12 (Childhood), 07 Years, 7 Years at diagno...              v4   \n",
       "142  02-12 (Childhood), 09 Years, 9 Years at diagno...              v4   \n",
       "\n",
       "           primer_set species               study sequencing_type  \\\n",
       "0    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA   \n",
       "1    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA   \n",
       "2    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA   \n",
       "3    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA   \n",
       "4    Human-TCRB-PD4bx   Human  mitchell-2022-jcii            gDNA   \n",
       "..                ...     ...                 ...             ...   \n",
       "138   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "139   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "140   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "141   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "142   Human-TCRB-PD4x   Human  mitchell-2022-jcii            gDNA   \n",
       "\n",
       "    specimen_label        Age  Sample Timepoint     Sex   Race     Ethnicity  \\\n",
       "0           310251   6.661190               4.0  Female  White      Hispanic   \n",
       "1           310205   0.829568               1.0  Female  White  Non-Hispanic   \n",
       "2           310168   5.670088               3.0    Male  White  Non-Hispanic   \n",
       "3           310170   8.407939               2.0    Male  White  Non-Hispanic   \n",
       "4           310148   6.269678               4.0    Male  White  Non-Hispanic   \n",
       "..             ...        ...               ...     ...    ...           ...   \n",
       "138  DenverT1D-170  17.900000               NaN    Male  White  Non-Hispanic   \n",
       "139  DenverT1D-106  16.400000               NaN    Male  White  Non-Hispanic   \n",
       "140  DenverT1D-256  16.100000               NaN    Male  White  Non-Hispanic   \n",
       "141  DenverT1D-055   7.500000               NaN  Female  White  Non-Hispanic   \n",
       "142  DenverT1D-271   9.800000               NaN    Male  White  Non-Hispanic   \n",
       "\n",
       "    participant_label             disease                    disease_subtype  \n",
       "0          control_16  Healthy/Background  Healthy/Background - T1D negative  \n",
       "1           control_6  Healthy/Background  Healthy/Background - T1D negative  \n",
       "2          control_13  Healthy/Background  Healthy/Background - T1D negative  \n",
       "3          control_12  Healthy/Background  Healthy/Background - T1D negative  \n",
       "4           control_9  Healthy/Background  Healthy/Background - T1D negative  \n",
       "..                ...                 ...                                ...  \n",
       "138     DenverT1D-170                 T1D                    T1D - new onset  \n",
       "139     DenverT1D-106                 T1D                    T1D - new onset  \n",
       "140     DenverT1D-256                 T1D                    T1D - new onset  \n",
       "141     DenverT1D-055                 T1D                    T1D - new onset  \n",
       "142     DenverT1D-271                 T1D                    T1D - new onset  \n",
       "\n",
       "[243 rows x 17 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual.append(df)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RA3-3years</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RA1-7months</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RA2-18months</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC9</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>RA7</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>RA8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>RA14</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>RA4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>RA10</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_name locus                                        sample_tags  \\\n",
       "0     RA3-3years  TCRB                                                NaN   \n",
       "1    RA1-7months  TCRB                                                NaN   \n",
       "2   RA2-18months  TCRB                                                NaN   \n",
       "3            HC8  TCRB                                                NaN   \n",
       "4            HC9  TCRB                                                NaN   \n",
       "..           ...   ...                                                ...   \n",
       "89           RA7  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "90           RA8  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "91          RA14  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "92           RA4  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "93          RA10  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "\n",
       "   counting_method       primer_set species                   study  \\\n",
       "0               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "1               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "2               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "3               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "4               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "..             ...              ...     ...                     ...   \n",
       "89              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "90              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "91              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "92              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "93              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "\n",
       "   sequencing_type participant_label  \n",
       "0             gDNA               RA3  \n",
       "1             gDNA               RA1  \n",
       "2             gDNA               RA2  \n",
       "3             gDNA               HC8  \n",
       "4             gDNA               HC9  \n",
       "..             ...               ...  \n",
       "89            gDNA               RA7  \n",
       "90            gDNA               RA8  \n",
       "91            gDNA              RA14  \n",
       "92            gDNA               RA4  \n",
       "93            gDNA              RA10  \n",
       "\n",
       "[94 rows x 9 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = _load_metadata_df(\"mustjoki-2017-natcomms\")\n",
    "\n",
    "\"\"\"\n",
    "https://www.nature.com/articles/ncomms15869\n",
    "RA, newly diagnosed. Some seronegative, some seropositive. A few patients have on-treatment followup\n",
    "See kelkka-2020-jci above for a follow-up study.\n",
    "\n",
    "Peripheral-blood mononuclear cells (PBMCs) were separated from EDTA blood using Ficoll gradient separation (Ficoll-Paque PLUS, GE Healthcare, cat. no 17-1440-03).\n",
    "CD4+ and CD8+ cells were separated with magnetic bead sorting using positive selection for both fractions\n",
    "\n",
    "DNA was extracted from CD4- and CD8-enriched samples or from the PBMC fraction\n",
    "TCRB deep sequencing was performed from 65 patients, accompanied by sequencing of 20 healthy controls\n",
    "Genomic DNA was used in all cases.\n",
    "\n",
    "Totals: 82 newly diagnosed RA patients, among whom 25 patients were sequenced with the immunogene panel. The 20 healthy controls were also included in the immunogene panel.\n",
    "\n",
    "Table 1: Immunogene panel sequencing was performed on both CD4+ and CD8+ cells of 25 newly diagnosed RA patients and 20 healthy controls (HCs)\n",
    "\n",
    "More info at https://static-content.springer.com/esm/art%3A10.1038%2Fncomms15869/MediaObjects/41467_2017_BFncomms15869_MOESM218_ESM.pdf SI Table 2\n",
    "\"\"\"\n",
    "\n",
    "df[\"sequencing_type\"] = \"gDNA\"\n",
    "df[\"participant_label\"] = df[\"sample_name\"].str.split(\"-\").str[0]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['HC1', 'HC10', 'HC11', 'HC12', 'HC13', 'HC14', 'HC15', 'HC16',\n",
       "       'HC17', 'HC18', 'HC19', 'HC2', 'HC20', 'HC3', 'HC4', 'HC5', 'HC6',\n",
       "       'HC7', 'HC8', 'HC9', 'RA1', 'RA10', 'RA11', 'RA12', 'RA13', 'RA14',\n",
       "       'RA15', 'RA16', 'RA17', 'RA18', 'RA19', 'RA2', 'RA20', 'RA21',\n",
       "       'RA22', 'RA23', 'RA24', 'RA25', 'RA26', 'RA28', 'RA29', 'RA3',\n",
       "       'RA30', 'RA31', 'RA32', 'RA33', 'RA37', 'RA4', 'RA40', 'RA41',\n",
       "       'RA46', 'RA47', 'RA48', 'RA49', 'RA5', 'RA50', 'RA51', 'RA52',\n",
       "       'RA53', 'RA54', 'RA55', 'RA56', 'RA57', 'RA58', 'RA59', 'RA6',\n",
       "       'RA60', 'RA62', 'RA63', 'RA65', 'RA66', 'RA68', 'RA69', 'RA7',\n",
       "       'RA70', 'RA72', 'RA74', 'RA75', 'RA76', 'RA79', 'RA8', 'RA80',\n",
       "       'RA81', 'RA82', 'RA9'], dtype=object)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"participant_label\"].sort_values().unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "T cells                                            35\n",
       "Rheumatoid Arthritis                               27\n",
       "CD8+                                               24\n",
       "HLA MHC Class I                                    12\n",
       "Peripheral blood lymphocytes (PBL)                 11\n",
       "CD4-                                                8\n",
       "Synovial fluid                                      3\n",
       "CD4+                                                2\n",
       "T Cell Large Granular Lymphocyte Leukemia (LGL)     2\n",
       "CD8-                                                1\n",
       "Name: sample_tags, dtype: int64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"sample_tags\"].str.split(\",\").explode().str.strip().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RA3-3years</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RA1-7months</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RA2-18months</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC9</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>RA7</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>RA8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>RA14</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>RA4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>RA10</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_name locus                                        sample_tags  \\\n",
       "0     RA3-3years  TCRB                                                NaN   \n",
       "1    RA1-7months  TCRB                                                NaN   \n",
       "2   RA2-18months  TCRB                                                NaN   \n",
       "3            HC8  TCRB                                                NaN   \n",
       "4            HC9  TCRB                                                NaN   \n",
       "..           ...   ...                                                ...   \n",
       "89           RA7  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "90           RA8  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "91          RA14  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "92           RA4  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "93          RA10  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "\n",
       "   counting_method       primer_set species                   study  \\\n",
       "0               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "1               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "2               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "3               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "4               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "..             ...              ...     ...                     ...   \n",
       "89              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "90              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "91              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "92              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "93              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "\n",
       "   sequencing_type participant_label  \n",
       "0             gDNA               RA3  \n",
       "1             gDNA               RA1  \n",
       "2             gDNA               RA2  \n",
       "3             gDNA               HC8  \n",
       "4             gDNA               HC9  \n",
       "..             ...               ...  \n",
       "89            gDNA               RA7  \n",
       "90            gDNA               RA8  \n",
       "91            gDNA              RA14  \n",
       "92            gDNA               RA4  \n",
       "93            gDNA              RA10  \n",
       "\n",
       "[91 rows x 9 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[~df[\"sample_tags\"].str.contains(\"Synovial fluid\").fillna(False)].copy()\n",
    "\n",
    "# sanity check\n",
    "assert not df[\"sample_name\"].str.contains(\"-SF-\").any()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age at diagnosis</th>\n",
       "      <th>Seropositive</th>\n",
       "      <th>DAS28</th>\n",
       "      <th>Shared epitope</th>\n",
       "      <th>Clonality</th>\n",
       "      <th>Largest CD8+ clone (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>75</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3.5</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.63</td>\n",
       "      <td>33.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>F</td>\n",
       "      <td>72</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3.5</td>\n",
       "      <td>Yes/yes</td>\n",
       "      <td>0.28</td>\n",
       "      <td>7.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>M</td>\n",
       "      <td>44</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.60</td>\n",
       "      <td>51.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>F</td>\n",
       "      <td>74</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.24</td>\n",
       "      <td>13.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>F</td>\n",
       "      <td>59</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.44</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>M</td>\n",
       "      <td>75</td>\n",
       "      <td>Yes</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Yes/yes</td>\n",
       "      <td>0.41</td>\n",
       "      <td>11.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>M</td>\n",
       "      <td>66</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3.9</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.35</td>\n",
       "      <td>13.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>F</td>\n",
       "      <td>61</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3.1</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.08</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>M</td>\n",
       "      <td>45</td>\n",
       "      <td>Yes</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.28</td>\n",
       "      <td>17.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>M</td>\n",
       "      <td>69</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3.3</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.46</td>\n",
       "      <td>16.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>F</td>\n",
       "      <td>52</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2.1</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.43</td>\n",
       "      <td>27.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>M</td>\n",
       "      <td>51</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3.6</td>\n",
       "      <td>Yes/yes</td>\n",
       "      <td>0.38</td>\n",
       "      <td>15.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>F</td>\n",
       "      <td>66</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2.8</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.27</td>\n",
       "      <td>9.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>F</td>\n",
       "      <td>74</td>\n",
       "      <td>Yes</td>\n",
       "      <td>4.5</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.20</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>F</td>\n",
       "      <td>25</td>\n",
       "      <td>No</td>\n",
       "      <td>4.8</td>\n",
       "      <td>Yes/yes</td>\n",
       "      <td>0.15</td>\n",
       "      <td>4.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>F</td>\n",
       "      <td>62</td>\n",
       "      <td>Yes</td>\n",
       "      <td>4.2</td>\n",
       "      <td>Yes/yes</td>\n",
       "      <td>0.30</td>\n",
       "      <td>6.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>F</td>\n",
       "      <td>71</td>\n",
       "      <td>No</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.39</td>\n",
       "      <td>20.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>F</td>\n",
       "      <td>65</td>\n",
       "      <td>Yes</td>\n",
       "      <td>4.5</td>\n",
       "      <td>Yes/yes</td>\n",
       "      <td>0.28</td>\n",
       "      <td>10.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>F</td>\n",
       "      <td>58</td>\n",
       "      <td>Yes</td>\n",
       "      <td>5.2</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.17</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>F</td>\n",
       "      <td>46</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2.1</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.37</td>\n",
       "      <td>16.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>F</td>\n",
       "      <td>48</td>\n",
       "      <td>No</td>\n",
       "      <td>5.0</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.11</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>F</td>\n",
       "      <td>68</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1.9</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.21</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>F</td>\n",
       "      <td>24</td>\n",
       "      <td>No</td>\n",
       "      <td>4.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.13</td>\n",
       "      <td>5.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>24</td>\n",
       "      <td>F</td>\n",
       "      <td>57</td>\n",
       "      <td>No</td>\n",
       "      <td>4.7</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.18</td>\n",
       "      <td>5.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>M</td>\n",
       "      <td>19</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3.8</td>\n",
       "      <td>Yes/yes</td>\n",
       "      <td>0.09</td>\n",
       "      <td>4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HC1</td>\n",
       "      <td>F</td>\n",
       "      <td>58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.38</td>\n",
       "      <td>11.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HC2</td>\n",
       "      <td>M</td>\n",
       "      <td>61</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/yes</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HC3</td>\n",
       "      <td>F</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.36</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HC4</td>\n",
       "      <td>M</td>\n",
       "      <td>58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.26</td>\n",
       "      <td>5.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HC5</td>\n",
       "      <td>F</td>\n",
       "      <td>65</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.35</td>\n",
       "      <td>24.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HC6</td>\n",
       "      <td>F</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.24</td>\n",
       "      <td>4.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>HC7</td>\n",
       "      <td>F</td>\n",
       "      <td>62</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.23</td>\n",
       "      <td>4.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>HC8</td>\n",
       "      <td>F</td>\n",
       "      <td>44</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.12</td>\n",
       "      <td>2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>HC9</td>\n",
       "      <td>M</td>\n",
       "      <td>55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.18</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>HC10</td>\n",
       "      <td>M</td>\n",
       "      <td>52</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/yes</td>\n",
       "      <td>0.08</td>\n",
       "      <td>1.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>HC11</td>\n",
       "      <td>M</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>HC12</td>\n",
       "      <td>F</td>\n",
       "      <td>57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.15</td>\n",
       "      <td>6.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>HC13</td>\n",
       "      <td>F</td>\n",
       "      <td>56</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.13</td>\n",
       "      <td>4.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>HC14</td>\n",
       "      <td>M</td>\n",
       "      <td>70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.17</td>\n",
       "      <td>2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>HC15</td>\n",
       "      <td>F</td>\n",
       "      <td>47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.30</td>\n",
       "      <td>7.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>HC16</td>\n",
       "      <td>M</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.18</td>\n",
       "      <td>11.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>HC17</td>\n",
       "      <td>F</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.08</td>\n",
       "      <td>7.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>HC18</td>\n",
       "      <td>M</td>\n",
       "      <td>60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No/no</td>\n",
       "      <td>0.38</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>HC19</td>\n",
       "      <td>M</td>\n",
       "      <td>43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.38</td>\n",
       "      <td>17.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>HC20</td>\n",
       "      <td>F</td>\n",
       "      <td>66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes/no</td>\n",
       "      <td>0.26</td>\n",
       "      <td>7.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Patient ID Sex  Age at diagnosis Seropositive  DAS28 Shared epitope  \\\n",
       "0           1   M                75          Yes    3.5         Yes/no   \n",
       "1           2   F                72          Yes    3.5        Yes/yes   \n",
       "2           3   M                44          Yes    NaN         Yes/no   \n",
       "3           4   F                74           No    NaN          No/no   \n",
       "4           5   F                59          Yes    NaN          No/no   \n",
       "5           6   M                75          Yes    4.0        Yes/yes   \n",
       "6           7   M                66          Yes    3.9          No/no   \n",
       "7           8   F                61          Yes    3.1          No/no   \n",
       "8           9   M                45          Yes    4.0         Yes/no   \n",
       "9          10   M                69          Yes    3.3         Yes/no   \n",
       "10         11   F                52          Yes    2.1         Yes/no   \n",
       "11         12   M                51          Yes    3.6        Yes/yes   \n",
       "12         13   F                66          Yes    2.8         Yes/no   \n",
       "13         14   F                74          Yes    4.5         Yes/no   \n",
       "14         15   F                25           No    4.8        Yes/yes   \n",
       "15         16   F                62          Yes    4.2        Yes/yes   \n",
       "16         17   F                71           No    6.0         Yes/no   \n",
       "17         18   F                65          Yes    4.5        Yes/yes   \n",
       "18         19   F                58          Yes    5.2          No/no   \n",
       "19         20   F                46          Yes    2.1         Yes/no   \n",
       "20         21   F                48           No    5.0          No/no   \n",
       "21         22   F                68          Yes    1.9          No/no   \n",
       "22         23   F                24           No    4.2            NaN   \n",
       "23         24   F                57           No    4.7          No/no   \n",
       "24         25   M                19          Yes    3.8        Yes/yes   \n",
       "25        HC1   F                58          NaN    NaN         Yes/no   \n",
       "26        HC2   M                61          NaN    NaN        Yes/yes   \n",
       "27        HC3   F                50          NaN    NaN          No/no   \n",
       "28        HC4   M                58          NaN    NaN          No/no   \n",
       "29        HC5   F                65          NaN    NaN         Yes/no   \n",
       "30        HC6   F                50          NaN    NaN          No/no   \n",
       "31        HC7   F                62          NaN    NaN         Yes/no   \n",
       "32        HC8   F                44          NaN    NaN         Yes/no   \n",
       "33        HC9   M                55          NaN    NaN          No/no   \n",
       "34       HC10   M                52          NaN    NaN        Yes/yes   \n",
       "35       HC11   M                48          NaN    NaN          No/no   \n",
       "36       HC12   F                57          NaN    NaN         Yes/no   \n",
       "37       HC13   F                56          NaN    NaN         Yes/no   \n",
       "38       HC14   M                70          NaN    NaN         Yes/no   \n",
       "39       HC15   F                47          NaN    NaN          No/no   \n",
       "40       HC16   M                48          NaN    NaN         Yes/no   \n",
       "41       HC17   F                21          NaN    NaN         Yes/no   \n",
       "42       HC18   M                60          NaN    NaN          No/no   \n",
       "43       HC19   M                43          NaN    NaN         Yes/no   \n",
       "44       HC20   F                66          NaN    NaN         Yes/no   \n",
       "\n",
       "    Clonality  Largest CD8+ clone (%)  \n",
       "0        0.63                    33.9  \n",
       "1        0.28                     7.8  \n",
       "2        0.60                    51.0  \n",
       "3        0.24                    13.2  \n",
       "4        0.44                    29.0  \n",
       "5        0.41                    11.1  \n",
       "6        0.35                    13.8  \n",
       "7        0.08                     2.0  \n",
       "8        0.28                    17.2  \n",
       "9        0.46                    16.2  \n",
       "10       0.43                    27.5  \n",
       "11       0.38                    15.7  \n",
       "12       0.27                     9.2  \n",
       "13       0.20                     8.0  \n",
       "14       0.15                     4.5  \n",
       "15       0.30                     6.8  \n",
       "16       0.39                    20.3  \n",
       "17       0.28                    10.2  \n",
       "18       0.17                     8.0  \n",
       "19       0.37                    16.2  \n",
       "20       0.11                     3.0  \n",
       "21       0.21                     7.0  \n",
       "22       0.13                     5.8  \n",
       "23       0.18                     5.3  \n",
       "24       0.09                     4.2  \n",
       "25       0.38                    11.8  \n",
       "26       0.09                     0.8  \n",
       "27       0.36                    26.0  \n",
       "28       0.26                     5.4  \n",
       "29       0.35                    24.7  \n",
       "30       0.24                     4.1  \n",
       "31       0.23                     4.4  \n",
       "32       0.12                     2.7  \n",
       "33       0.18                     3.0  \n",
       "34       0.08                     1.7  \n",
       "35       0.04                     0.6  \n",
       "36       0.15                     6.2  \n",
       "37       0.13                     4.6  \n",
       "38       0.17                     2.4  \n",
       "39       0.30                     7.1  \n",
       "40       0.18                    11.1  \n",
       "41       0.08                     7.1  \n",
       "42       0.38                    12.0  \n",
       "43       0.38                    17.5  \n",
       "44       0.26                     7.1  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# table 1 has the HCs\n",
    "df1 = pd.read_csv(\n",
    "    config.paths.metadata_dir / \"adaptive\" / \"mustjoki-2017-natcomms.table_1.csv\"\n",
    ")\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HC1</td>\n",
       "      <td>F</td>\n",
       "      <td>58</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HC2</td>\n",
       "      <td>M</td>\n",
       "      <td>61</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HC3</td>\n",
       "      <td>F</td>\n",
       "      <td>50</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HC4</td>\n",
       "      <td>M</td>\n",
       "      <td>58</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HC5</td>\n",
       "      <td>F</td>\n",
       "      <td>65</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HC6</td>\n",
       "      <td>F</td>\n",
       "      <td>50</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>HC7</td>\n",
       "      <td>F</td>\n",
       "      <td>62</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>HC8</td>\n",
       "      <td>F</td>\n",
       "      <td>44</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>HC9</td>\n",
       "      <td>M</td>\n",
       "      <td>55</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>HC10</td>\n",
       "      <td>M</td>\n",
       "      <td>52</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>HC11</td>\n",
       "      <td>M</td>\n",
       "      <td>48</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>HC12</td>\n",
       "      <td>F</td>\n",
       "      <td>57</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>HC13</td>\n",
       "      <td>F</td>\n",
       "      <td>56</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>HC14</td>\n",
       "      <td>M</td>\n",
       "      <td>70</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>HC15</td>\n",
       "      <td>F</td>\n",
       "      <td>47</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>HC16</td>\n",
       "      <td>M</td>\n",
       "      <td>48</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>HC17</td>\n",
       "      <td>F</td>\n",
       "      <td>21</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>HC18</td>\n",
       "      <td>M</td>\n",
       "      <td>60</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>HC19</td>\n",
       "      <td>M</td>\n",
       "      <td>43</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>HC20</td>\n",
       "      <td>F</td>\n",
       "      <td>66</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Patient ID Sex  Age             disease                   disease_subtype\n",
       "25        HC1   F   58  Healthy/Background  Healthy/Background - RA negative\n",
       "26        HC2   M   61  Healthy/Background  Healthy/Background - RA negative\n",
       "27        HC3   F   50  Healthy/Background  Healthy/Background - RA negative\n",
       "28        HC4   M   58  Healthy/Background  Healthy/Background - RA negative\n",
       "29        HC5   F   65  Healthy/Background  Healthy/Background - RA negative\n",
       "30        HC6   F   50  Healthy/Background  Healthy/Background - RA negative\n",
       "31        HC7   F   62  Healthy/Background  Healthy/Background - RA negative\n",
       "32        HC8   F   44  Healthy/Background  Healthy/Background - RA negative\n",
       "33        HC9   M   55  Healthy/Background  Healthy/Background - RA negative\n",
       "34       HC10   M   52  Healthy/Background  Healthy/Background - RA negative\n",
       "35       HC11   M   48  Healthy/Background  Healthy/Background - RA negative\n",
       "36       HC12   F   57  Healthy/Background  Healthy/Background - RA negative\n",
       "37       HC13   F   56  Healthy/Background  Healthy/Background - RA negative\n",
       "38       HC14   M   70  Healthy/Background  Healthy/Background - RA negative\n",
       "39       HC15   F   47  Healthy/Background  Healthy/Background - RA negative\n",
       "40       HC16   M   48  Healthy/Background  Healthy/Background - RA negative\n",
       "41       HC17   F   21  Healthy/Background  Healthy/Background - RA negative\n",
       "42       HC18   M   60  Healthy/Background  Healthy/Background - RA negative\n",
       "43       HC19   M   43  Healthy/Background  Healthy/Background - RA negative\n",
       "44       HC20   F   66  Healthy/Background  Healthy/Background - RA negative"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = df1[df1[\"Patient ID\"].str.startswith(\"HC\")]\n",
    "df1 = df1.rename(columns={\"Age at diagnosis\": \"Age\"})[[\"Patient ID\", \"Sex\", \"Age\"]]\n",
    "df1[\"disease\"] = healthy_label\n",
    "df1[\"disease_subtype\"] = f\"{healthy_label} - RA negative\"\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient ID</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Largest CD8+ clone (% of CD8+)</th>\n",
       "      <th>Age at diagnosis</th>\n",
       "      <th>Hb</th>\n",
       "      <th>ESR</th>\n",
       "      <th>Leucocytes</th>\n",
       "      <th>Lymphocytes</th>\n",
       "      <th>Neutrophils</th>\n",
       "      <th>Monocytes</th>\n",
       "      <th>...</th>\n",
       "      <th>RF</th>\n",
       "      <th>Seropositive</th>\n",
       "      <th>ACPA</th>\n",
       "      <th>DAS28</th>\n",
       "      <th>tender joints</th>\n",
       "      <th>swollen joints</th>\n",
       "      <th>pat_global</th>\n",
       "      <th>rtg_erosions</th>\n",
       "      <th>HAQ</th>\n",
       "      <th>Duration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>33.9</td>\n",
       "      <td>74.7</td>\n",
       "      <td>132</td>\n",
       "      <td>33</td>\n",
       "      <td>6.4</td>\n",
       "      <td>1.1</td>\n",
       "      <td>4.4</td>\n",
       "      <td>0.8</td>\n",
       "      <td>...</td>\n",
       "      <td>189.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>160</td>\n",
       "      <td>3.46</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>no</td>\n",
       "      <td>2.000</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>F</td>\n",
       "      <td>7.8</td>\n",
       "      <td>72.0</td>\n",
       "      <td>126</td>\n",
       "      <td>47</td>\n",
       "      <td>7.1</td>\n",
       "      <td>2.1</td>\n",
       "      <td>4.1</td>\n",
       "      <td>0.6</td>\n",
       "      <td>...</td>\n",
       "      <td>19.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>300</td>\n",
       "      <td>3.54</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.300</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>M</td>\n",
       "      <td>51.0</td>\n",
       "      <td>44.3</td>\n",
       "      <td>156</td>\n",
       "      <td>12</td>\n",
       "      <td>8.3</td>\n",
       "      <td>3.2</td>\n",
       "      <td>4.2</td>\n",
       "      <td>0.7</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.625</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>F</td>\n",
       "      <td>13.2</td>\n",
       "      <td>73.6</td>\n",
       "      <td>150</td>\n",
       "      <td>31</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.2</td>\n",
       "      <td>4.1</td>\n",
       "      <td>0.6</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0</td>\n",
       "      <td>no</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.630</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>F</td>\n",
       "      <td>29.0</td>\n",
       "      <td>58.8</td>\n",
       "      <td>134</td>\n",
       "      <td>5</td>\n",
       "      <td>9.1</td>\n",
       "      <td>3.2</td>\n",
       "      <td>4.6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>204</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>78</td>\n",
       "      <td>F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>39.1</td>\n",
       "      <td>142</td>\n",
       "      <td>8</td>\n",
       "      <td>9.2</td>\n",
       "      <td>2.1</td>\n",
       "      <td>6.1</td>\n",
       "      <td>0.8</td>\n",
       "      <td>...</td>\n",
       "      <td>120.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>340</td>\n",
       "      <td>4.65</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.380</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>79</td>\n",
       "      <td>F</td>\n",
       "      <td>10.3</td>\n",
       "      <td>76.6</td>\n",
       "      <td>113</td>\n",
       "      <td>68</td>\n",
       "      <td>8.1</td>\n",
       "      <td>1.7</td>\n",
       "      <td>5.5</td>\n",
       "      <td>0.7</td>\n",
       "      <td>...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>no</td>\n",
       "      <td>6</td>\n",
       "      <td>5.17</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>50</td>\n",
       "      <td>no</td>\n",
       "      <td>0.880</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>80</td>\n",
       "      <td>F</td>\n",
       "      <td>2.0</td>\n",
       "      <td>44.2</td>\n",
       "      <td>146</td>\n",
       "      <td>27</td>\n",
       "      <td>7.8</td>\n",
       "      <td>2.1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>...</td>\n",
       "      <td>18.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>301</td>\n",
       "      <td>3.87</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>yes</td>\n",
       "      <td>1.000</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>81</td>\n",
       "      <td>F</td>\n",
       "      <td>19.1</td>\n",
       "      <td>59.0</td>\n",
       "      <td>122</td>\n",
       "      <td>24</td>\n",
       "      <td>5.1</td>\n",
       "      <td>1.5</td>\n",
       "      <td>2.9</td>\n",
       "      <td>0.5</td>\n",
       "      <td>...</td>\n",
       "      <td>216.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>120</td>\n",
       "      <td>5.28</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>92</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.500</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>82</td>\n",
       "      <td>F</td>\n",
       "      <td>11.2</td>\n",
       "      <td>52.3</td>\n",
       "      <td>128</td>\n",
       "      <td>19</td>\n",
       "      <td>7.1</td>\n",
       "      <td>1.7</td>\n",
       "      <td>4.7</td>\n",
       "      <td>0.4</td>\n",
       "      <td>...</td>\n",
       "      <td>79.0</td>\n",
       "      <td>yes</td>\n",
       "      <td>180</td>\n",
       "      <td>5.12</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Patient ID Gender  Largest CD8+ clone (% of CD8+)  Age at diagnosis   Hb  \\\n",
       "0            1      M                            33.9              74.7  132   \n",
       "1            2      F                             7.8              72.0  126   \n",
       "2            3      M                            51.0              44.3  156   \n",
       "3            4      F                            13.2              73.6  150   \n",
       "4            5      F                            29.0              58.8  134   \n",
       "..         ...    ...                             ...               ...  ...   \n",
       "77          78      F                             NaN              39.1  142   \n",
       "78          79      F                            10.3              76.6  113   \n",
       "79          80      F                             2.0              44.2  146   \n",
       "80          81      F                            19.1              59.0  122   \n",
       "81          82      F                            11.2              52.3  128   \n",
       "\n",
       "    ESR  Leucocytes  Lymphocytes  Neutrophils  Monocytes  ...     RF  \\\n",
       "0    33         6.4          1.1          4.4        0.8  ...  189.0   \n",
       "1    47         7.1          2.1          4.1        0.6  ...   19.0   \n",
       "2    12         8.3          3.2          4.2        0.7  ...    NaN   \n",
       "3    31         7.0          2.2          4.1        0.6  ...   10.0   \n",
       "4     5         9.1          3.2          4.6        1.0  ...    5.0   \n",
       "..  ...         ...          ...          ...        ...  ...    ...   \n",
       "77    8         9.2          2.1          6.1        0.8  ...  120.0   \n",
       "78   68         8.1          1.7          5.5        0.7  ...   11.0   \n",
       "79   27         7.8          2.1          5.0        0.6  ...   18.0   \n",
       "80   24         5.1          1.5          2.9        0.5  ...  216.0   \n",
       "81   19         7.1          1.7          4.7        0.4  ...   79.0   \n",
       "\n",
       "    Seropositive  ACPA  DAS28  tender joints swollen joints  pat_global  \\\n",
       "0            yes   160   3.46              5              0           3   \n",
       "1            yes   300   3.54              1              3          40   \n",
       "2            yes   200    NaN             11              1          55   \n",
       "3             no     6    NaN              1             15          18   \n",
       "4            yes   204    NaN              4              1          47   \n",
       "..           ...   ...    ...            ...            ...         ...   \n",
       "77           yes   340   4.65             14              5          68   \n",
       "78            no     6   5.17              1              5          50   \n",
       "79           yes   301   3.87              2              2          27   \n",
       "80           yes   120   5.28             13              6          92   \n",
       "81           yes   180   5.12             19             18          39   \n",
       "\n",
       "    rtg_erosions    HAQ  Duration  \n",
       "0             no  2.000       5.0  \n",
       "1            NaN  1.300       NaN  \n",
       "2            yes  0.625       NaN  \n",
       "3            NaN  0.630       NaN  \n",
       "4            NaN    NaN      12.0  \n",
       "..           ...    ...       ...  \n",
       "77           NaN  1.380       NaN  \n",
       "78            no  0.880       8.0  \n",
       "79           yes  1.000      14.0  \n",
       "80           NaN  0.500       NaN  \n",
       "81           NaN  1.000       NaN  \n",
       "\n",
       "[82 rows x 24 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SI table 2 has all the RAs\n",
    "df2 = pd.read_csv(\n",
    "    config.paths.metadata_dir / \"adaptive\" / \"mustjoki-2017-natcomms.si_table_2.csv\"\n",
    ")\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Seropositive</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RA1</td>\n",
       "      <td>M</td>\n",
       "      <td>yes</td>\n",
       "      <td>74.7</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RA2</td>\n",
       "      <td>F</td>\n",
       "      <td>yes</td>\n",
       "      <td>72.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RA3</td>\n",
       "      <td>M</td>\n",
       "      <td>yes</td>\n",
       "      <td>44.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RA4</td>\n",
       "      <td>F</td>\n",
       "      <td>no</td>\n",
       "      <td>73.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RA5</td>\n",
       "      <td>F</td>\n",
       "      <td>yes</td>\n",
       "      <td>58.8</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>RA78</td>\n",
       "      <td>F</td>\n",
       "      <td>yes</td>\n",
       "      <td>39.1</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>RA79</td>\n",
       "      <td>F</td>\n",
       "      <td>no</td>\n",
       "      <td>76.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>RA80</td>\n",
       "      <td>F</td>\n",
       "      <td>yes</td>\n",
       "      <td>44.2</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>RA81</td>\n",
       "      <td>F</td>\n",
       "      <td>yes</td>\n",
       "      <td>59.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>RA82</td>\n",
       "      <td>F</td>\n",
       "      <td>yes</td>\n",
       "      <td>52.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Patient ID Sex Seropositive   Age disease     disease_subtype\n",
       "0         RA1   M          yes  74.7      RA  RA - sero-positive\n",
       "1         RA2   F          yes  72.0      RA  RA - sero-positive\n",
       "2         RA3   M          yes  44.3      RA  RA - sero-positive\n",
       "3         RA4   F           no  73.6      RA  RA - sero-negative\n",
       "4         RA5   F          yes  58.8      RA  RA - sero-positive\n",
       "..        ...  ..          ...   ...     ...                 ...\n",
       "77       RA78   F          yes  39.1      RA  RA - sero-positive\n",
       "78       RA79   F           no  76.6      RA  RA - sero-negative\n",
       "79       RA80   F          yes  44.2      RA  RA - sero-positive\n",
       "80       RA81   F          yes  59.0      RA  RA - sero-positive\n",
       "81       RA82   F          yes  52.3      RA  RA - sero-positive\n",
       "\n",
       "[82 rows x 6 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = df2.rename(columns={\"Age at diagnosis\": \"Age\", \"Gender\": \"Sex\"})[\n",
    "    [\"Patient ID\", \"Sex\", \"Seropositive\", \"Age\"]\n",
    "]\n",
    "df2[\"Patient ID\"] = \"RA\" + df2[\"Patient ID\"].astype(str)\n",
    "df2[\"disease\"] = \"RA\"\n",
    "df2[\"disease_subtype\"] = \"RA - \" + df2[\"Seropositive\"].map(\n",
    "    {\"yes\": \"sero-positive\", \"no\": \"sero-negative\"}\n",
    ")\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RA3-3years</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA3</td>\n",
       "      <td>M</td>\n",
       "      <td>44.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RA1-7months</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA1</td>\n",
       "      <td>M</td>\n",
       "      <td>74.7</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RA2-18months</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA2</td>\n",
       "      <td>F</td>\n",
       "      <td>72.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC8</td>\n",
       "      <td>F</td>\n",
       "      <td>44.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC9</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC9</td>\n",
       "      <td>M</td>\n",
       "      <td>55.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>RA7</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA7</td>\n",
       "      <td>M</td>\n",
       "      <td>66.1</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>RA8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA8</td>\n",
       "      <td>F</td>\n",
       "      <td>60.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>RA14</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA14</td>\n",
       "      <td>F</td>\n",
       "      <td>73.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>RA4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA4</td>\n",
       "      <td>F</td>\n",
       "      <td>73.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>RA10</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA10</td>\n",
       "      <td>M</td>\n",
       "      <td>69.2</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_name locus                                        sample_tags  \\\n",
       "0     RA3-3years  TCRB                                                NaN   \n",
       "1    RA1-7months  TCRB                                                NaN   \n",
       "2   RA2-18months  TCRB                                                NaN   \n",
       "3            HC8  TCRB                                                NaN   \n",
       "4            HC9  TCRB                                                NaN   \n",
       "..           ...   ...                                                ...   \n",
       "86           RA7  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "87           RA8  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "88          RA14  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "89           RA4  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "90          RA10  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "\n",
       "   counting_method       primer_set species                   study  \\\n",
       "0               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "1               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "2               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "3               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "4               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "..             ...              ...     ...                     ...   \n",
       "86              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "87              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "88              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "89              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "90              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "\n",
       "   sequencing_type participant_label Sex   Age             disease  \\\n",
       "0             gDNA               RA3   M  44.3                  RA   \n",
       "1             gDNA               RA1   M  74.7                  RA   \n",
       "2             gDNA               RA2   F  72.0                  RA   \n",
       "3             gDNA               HC8   F  44.0  Healthy/Background   \n",
       "4             gDNA               HC9   M  55.0  Healthy/Background   \n",
       "..             ...               ...  ..   ...                 ...   \n",
       "86            gDNA               RA7   M  66.1                  RA   \n",
       "87            gDNA               RA8   F  60.5                  RA   \n",
       "88            gDNA              RA14   F  73.5                  RA   \n",
       "89            gDNA               RA4   F  73.6                  RA   \n",
       "90            gDNA              RA10   M  69.2                  RA   \n",
       "\n",
       "                     disease_subtype  \n",
       "0                 RA - sero-positive  \n",
       "1                 RA - sero-positive  \n",
       "2                 RA - sero-positive  \n",
       "3   Healthy/Background - RA negative  \n",
       "4   Healthy/Background - RA negative  \n",
       "..                               ...  \n",
       "86                RA - sero-positive  \n",
       "87                RA - sero-positive  \n",
       "88                RA - sero-positive  \n",
       "89                RA - sero-negative  \n",
       "90                RA - sero-positive  \n",
       "\n",
       "[91 rows x 13 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.merge(\n",
    "    df,\n",
    "    pd.concat([df1, df2], axis=0).drop(columns=\"Seropositive\"),\n",
    "    how=\"left\",\n",
    "    left_on=\"participant_label\",\n",
    "    right_on=\"Patient ID\",\n",
    "    validate=\"m:1\",\n",
    ").drop(columns=\"Patient ID\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert not df[\"disease\"].isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RA2     3\n",
       "RA3     2\n",
       "RA6     2\n",
       "RA23    2\n",
       "RA1     2\n",
       "RA47    1\n",
       "HC5     1\n",
       "HC6     1\n",
       "HC7     1\n",
       "RA51    1\n",
       "Name: participant_label, dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"participant_label\"].value_counts().head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>RA1</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA1</td>\n",
       "      <td>M</td>\n",
       "      <td>74.7</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RA1-7months</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA1</td>\n",
       "      <td>M</td>\n",
       "      <td>74.7</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>RA2</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA2</td>\n",
       "      <td>F</td>\n",
       "      <td>72.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RA2-18months</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA2</td>\n",
       "      <td>F</td>\n",
       "      <td>72.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>RA2-CD8+Vb1+</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, T Cell Large Granular Lymphocyte Leukemi...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA2</td>\n",
       "      <td>F</td>\n",
       "      <td>72.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>RA23</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA23</td>\n",
       "      <td>F</td>\n",
       "      <td>24.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>RA23-CD4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA23</td>\n",
       "      <td>F</td>\n",
       "      <td>24.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>RA3</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA3</td>\n",
       "      <td>M</td>\n",
       "      <td>44.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RA3-3years</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA3</td>\n",
       "      <td>M</td>\n",
       "      <td>44.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>RA6</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA6</td>\n",
       "      <td>M</td>\n",
       "      <td>74.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>RA6-CD4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA6</td>\n",
       "      <td>M</td>\n",
       "      <td>74.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_name locus                                        sample_tags  \\\n",
       "80           RA1  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "1    RA1-7months  TCRB                                                NaN   \n",
       "85           RA2  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "2   RA2-18months  TCRB                                                NaN   \n",
       "79  RA2-CD8+Vb1+  TCRB  CD8+, T Cell Large Granular Lymphocyte Leukemi...   \n",
       "65          RA23  TCRB  CD8+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "66      RA23-CD4  TCRB  CD4+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "84           RA3  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "0     RA3-3years  TCRB                                                NaN   \n",
       "77           RA6  TCRB                CD4-, Rheumatoid Arthritis, T cells   \n",
       "76       RA6-CD4  TCRB                CD8-, Rheumatoid Arthritis, T cells   \n",
       "\n",
       "   counting_method       primer_set species                   study  \\\n",
       "80              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "1               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "85              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "2               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "79              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "65              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "66              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "84              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "0               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "77              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "76              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "\n",
       "   sequencing_type participant_label Sex   Age disease     disease_subtype  \n",
       "80            gDNA               RA1   M  74.7      RA  RA - sero-positive  \n",
       "1             gDNA               RA1   M  74.7      RA  RA - sero-positive  \n",
       "85            gDNA               RA2   F  72.0      RA  RA - sero-positive  \n",
       "2             gDNA               RA2   F  72.0      RA  RA - sero-positive  \n",
       "79            gDNA               RA2   F  72.0      RA  RA - sero-positive  \n",
       "65            gDNA              RA23   F  24.4      RA  RA - sero-negative  \n",
       "66            gDNA              RA23   F  24.4      RA  RA - sero-negative  \n",
       "84            gDNA               RA3   M  44.3      RA  RA - sero-positive  \n",
       "0             gDNA               RA3   M  44.3      RA  RA - sero-positive  \n",
       "77            gDNA               RA6   M  74.5      RA  RA - sero-positive  \n",
       "76            gDNA               RA6   M  74.5      RA  RA - sero-positive  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# people with multiple samples\n",
    "df[\n",
    "    df[\"participant_label\"].isin(\n",
    "        df[\"participant_label\"]\n",
    "        .value_counts()[df[\"participant_label\"].value_counts() > 1]\n",
    "        .index\n",
    "    )\n",
    "].sort_values(\"sample_name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove late timepoints from RA 1, 2, and 3. We want newly diagnosed, treatment naive.\n",
    "df = df[~df[\"sample_name\"].isin([\"RA1-7months\", \"RA2-18months\", \"RA3-3years\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove RA patient 1 2's special extra sorts:\n",
    "# Patient 2 was a 72-year-old female who also had palindromic rheumatism and a previous history of other inflammatory disorders: asthma, lichen ruber and atrophic gastritis. At the time of RA diagnosis, flow cytometric screening identified two unusually large populations of CD8+ T cells: VÎ²1+ (14%) and VÎ²13.6+ (11%)\n",
    "df = df[~df[\"sample_name\"].isin([\"RA2-CD8+Vb1+\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>RA23</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA23</td>\n",
       "      <td>F</td>\n",
       "      <td>24.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>RA23-CD4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA23</td>\n",
       "      <td>F</td>\n",
       "      <td>24.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>RA6</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA6</td>\n",
       "      <td>M</td>\n",
       "      <td>74.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>RA6-CD4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA6</td>\n",
       "      <td>M</td>\n",
       "      <td>74.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name locus                                        sample_tags  \\\n",
       "65        RA23  TCRB  CD8+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "66    RA23-CD4  TCRB  CD4+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "77         RA6  TCRB                CD4-, Rheumatoid Arthritis, T cells   \n",
       "76     RA6-CD4  TCRB                CD8-, Rheumatoid Arthritis, T cells   \n",
       "\n",
       "   counting_method       primer_set species                   study  \\\n",
       "65              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "66              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "77              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "76              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "\n",
       "   sequencing_type participant_label Sex   Age disease     disease_subtype  \n",
       "65            gDNA              RA23   F  24.4      RA  RA - sero-negative  \n",
       "66            gDNA              RA23   F  24.4      RA  RA - sero-negative  \n",
       "77            gDNA               RA6   M  74.5      RA  RA - sero-positive  \n",
       "76            gDNA               RA6   M  74.5      RA  RA - sero-positive  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Recheck: people with multiple samples\n",
    "# We see some CD4 separated out of CD8\n",
    "df[\n",
    "    df[\"participant_label\"].isin(\n",
    "        df[\"participant_label\"]\n",
    "        .value_counts()[df[\"participant_label\"].value_counts() > 1]\n",
    "        .index\n",
    "    )\n",
    "].sort_values(\"sample_name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>RA23-CD4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA23</td>\n",
       "      <td>F</td>\n",
       "      <td>24.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name locus                                        sample_tags  \\\n",
       "66    RA23-CD4  TCRB  CD4+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "\n",
       "   counting_method       primer_set species                   study  \\\n",
       "66              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "\n",
       "   sequencing_type participant_label Sex   Age disease     disease_subtype  \n",
       "66            gDNA              RA23   F  24.4      RA  RA - sero-negative  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"sample_tags\"].str.contains(\"CD4\\+\").fillna(False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>RA6-CD4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA6</td>\n",
       "      <td>M</td>\n",
       "      <td>74.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name locus                          sample_tags counting_method  \\\n",
       "76     RA6-CD4  TCRB  CD8-, Rheumatoid Arthritis, T cells              v2   \n",
       "\n",
       "         primer_set species                   study sequencing_type  \\\n",
       "76  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms            gDNA   \n",
       "\n",
       "   participant_label Sex   Age disease     disease_subtype  \n",
       "76               RA6   M  74.5      RA  RA - sero-positive  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"sample_tags\"].str.contains(\"CD8-\").fillna(False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>RA20</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA20</td>\n",
       "      <td>F</td>\n",
       "      <td>46.1</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>RA19</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA19</td>\n",
       "      <td>F</td>\n",
       "      <td>57.9</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>RA18</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA18</td>\n",
       "      <td>F</td>\n",
       "      <td>65.1</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>RA17</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA17</td>\n",
       "      <td>F</td>\n",
       "      <td>70.7</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>RA15</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA15</td>\n",
       "      <td>F</td>\n",
       "      <td>25.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>RA13</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA13</td>\n",
       "      <td>F</td>\n",
       "      <td>65.7</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>RA12</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA12</td>\n",
       "      <td>M</td>\n",
       "      <td>50.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>RA6</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA6</td>\n",
       "      <td>M</td>\n",
       "      <td>74.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name locus                          sample_tags counting_method  \\\n",
       "69        RA20  TCRB  CD4-, Rheumatoid Arthritis, T cells              v2   \n",
       "70        RA19  TCRB  CD4-, Rheumatoid Arthritis, T cells              v2   \n",
       "71        RA18  TCRB  CD4-, Rheumatoid Arthritis, T cells              v2   \n",
       "72        RA17  TCRB  CD4-, Rheumatoid Arthritis, T cells              v2   \n",
       "73        RA15  TCRB  CD4-, Rheumatoid Arthritis, T cells              v2   \n",
       "74        RA13  TCRB  CD4-, Rheumatoid Arthritis, T cells              v2   \n",
       "75        RA12  TCRB  CD4-, Rheumatoid Arthritis, T cells              v2   \n",
       "77         RA6  TCRB  CD4-, Rheumatoid Arthritis, T cells              v2   \n",
       "\n",
       "         primer_set species                   study sequencing_type  \\\n",
       "69  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms            gDNA   \n",
       "70  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms            gDNA   \n",
       "71  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms            gDNA   \n",
       "72  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms            gDNA   \n",
       "73  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms            gDNA   \n",
       "74  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms            gDNA   \n",
       "75  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms            gDNA   \n",
       "77  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms            gDNA   \n",
       "\n",
       "   participant_label Sex   Age disease     disease_subtype  \n",
       "69              RA20   F  46.1      RA  RA - sero-positive  \n",
       "70              RA19   F  57.9      RA  RA - sero-positive  \n",
       "71              RA18   F  65.1      RA  RA - sero-positive  \n",
       "72              RA17   F  70.7      RA  RA - sero-negative  \n",
       "73              RA15   F  25.0      RA  RA - sero-negative  \n",
       "74              RA13   F  65.7      RA  RA - sero-positive  \n",
       "75              RA12   M  50.6      RA  RA - sero-positive  \n",
       "77               RA6   M  74.5      RA  RA - sero-positive  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"sample_tags\"].str.contains(\"CD4-\").fillna(False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>HC7</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC7</td>\n",
       "      <td>F</td>\n",
       "      <td>62.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>HC6</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC6</td>\n",
       "      <td>F</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>HC5</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC5</td>\n",
       "      <td>F</td>\n",
       "      <td>65.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>HC4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC4</td>\n",
       "      <td>M</td>\n",
       "      <td>58.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>HC3</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC3</td>\n",
       "      <td>F</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>HC2</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC2</td>\n",
       "      <td>M</td>\n",
       "      <td>61.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>HC1</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC1</td>\n",
       "      <td>F</td>\n",
       "      <td>58.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>RA23</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA23</td>\n",
       "      <td>F</td>\n",
       "      <td>24.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>RA22</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Peripheral blood lympho...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA22</td>\n",
       "      <td>F</td>\n",
       "      <td>68.2</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>RA21</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA21</td>\n",
       "      <td>F</td>\n",
       "      <td>48.2</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>RA5</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, T Cell Large Granular Lymphocyte Leukemi...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA5</td>\n",
       "      <td>F</td>\n",
       "      <td>58.8</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>RA1</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA1</td>\n",
       "      <td>M</td>\n",
       "      <td>74.7</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>RA16</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA16</td>\n",
       "      <td>F</td>\n",
       "      <td>62.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>RA11</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA11</td>\n",
       "      <td>F</td>\n",
       "      <td>52.2</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>RA9</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA9</td>\n",
       "      <td>M</td>\n",
       "      <td>45.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>RA3</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA3</td>\n",
       "      <td>M</td>\n",
       "      <td>44.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>RA2</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA2</td>\n",
       "      <td>F</td>\n",
       "      <td>72.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>RA7</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA7</td>\n",
       "      <td>M</td>\n",
       "      <td>66.1</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>RA8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA8</td>\n",
       "      <td>F</td>\n",
       "      <td>60.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>RA14</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA14</td>\n",
       "      <td>F</td>\n",
       "      <td>73.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>RA4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA4</td>\n",
       "      <td>F</td>\n",
       "      <td>73.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>RA10</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA10</td>\n",
       "      <td>M</td>\n",
       "      <td>69.2</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name locus                                        sample_tags  \\\n",
       "58         HC7  TCRB  CD8+, Peripheral blood lymphocytes (PBL), T cells   \n",
       "59         HC6  TCRB  CD8+, Peripheral blood lymphocytes (PBL), T cells   \n",
       "60         HC5  TCRB  CD8+, Peripheral blood lymphocytes (PBL), T cells   \n",
       "61         HC4  TCRB  CD8+, Peripheral blood lymphocytes (PBL), T cells   \n",
       "62         HC3  TCRB  CD8+, Peripheral blood lymphocytes (PBL), T cells   \n",
       "63         HC2  TCRB  CD8+, Peripheral blood lymphocytes (PBL), T cells   \n",
       "64         HC1  TCRB  CD8+, Peripheral blood lymphocytes (PBL), T cells   \n",
       "65        RA23  TCRB  CD8+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "67        RA22  TCRB  CD8+, HLA MHC Class I, Peripheral blood lympho...   \n",
       "68        RA21  TCRB  CD8+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "78         RA5  TCRB  CD8+, T Cell Large Granular Lymphocyte Leukemi...   \n",
       "80         RA1  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "81        RA16  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "82        RA11  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "83         RA9  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "84         RA3  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "85         RA2  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "86         RA7  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "87         RA8  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "88        RA14  TCRB                CD8+, Rheumatoid Arthritis, T cells   \n",
       "89         RA4  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "90        RA10  TCRB  CD8+, HLA MHC Class I, Rheumatoid Arthritis, T...   \n",
       "\n",
       "   counting_method       primer_set species                   study  \\\n",
       "58              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "59              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "60              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "61              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "62              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "63              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "64              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "65              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "67              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "68              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "78              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "80              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "81              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "82              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "83              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "84              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "85              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "86              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "87              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "88              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "89              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "90              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "\n",
       "   sequencing_type participant_label Sex   Age             disease  \\\n",
       "58            gDNA               HC7   F  62.0  Healthy/Background   \n",
       "59            gDNA               HC6   F  50.0  Healthy/Background   \n",
       "60            gDNA               HC5   F  65.0  Healthy/Background   \n",
       "61            gDNA               HC4   M  58.0  Healthy/Background   \n",
       "62            gDNA               HC3   F  50.0  Healthy/Background   \n",
       "63            gDNA               HC2   M  61.0  Healthy/Background   \n",
       "64            gDNA               HC1   F  58.0  Healthy/Background   \n",
       "65            gDNA              RA23   F  24.4                  RA   \n",
       "67            gDNA              RA22   F  68.2                  RA   \n",
       "68            gDNA              RA21   F  48.2                  RA   \n",
       "78            gDNA               RA5   F  58.8                  RA   \n",
       "80            gDNA               RA1   M  74.7                  RA   \n",
       "81            gDNA              RA16   F  62.4                  RA   \n",
       "82            gDNA              RA11   F  52.2                  RA   \n",
       "83            gDNA               RA9   M  45.3                  RA   \n",
       "84            gDNA               RA3   M  44.3                  RA   \n",
       "85            gDNA               RA2   F  72.0                  RA   \n",
       "86            gDNA               RA7   M  66.1                  RA   \n",
       "87            gDNA               RA8   F  60.5                  RA   \n",
       "88            gDNA              RA14   F  73.5                  RA   \n",
       "89            gDNA               RA4   F  73.6                  RA   \n",
       "90            gDNA              RA10   M  69.2                  RA   \n",
       "\n",
       "                     disease_subtype  \n",
       "58  Healthy/Background - RA negative  \n",
       "59  Healthy/Background - RA negative  \n",
       "60  Healthy/Background - RA negative  \n",
       "61  Healthy/Background - RA negative  \n",
       "62  Healthy/Background - RA negative  \n",
       "63  Healthy/Background - RA negative  \n",
       "64  Healthy/Background - RA negative  \n",
       "65                RA - sero-negative  \n",
       "67                RA - sero-positive  \n",
       "68                RA - sero-negative  \n",
       "78                RA - sero-positive  \n",
       "80                RA - sero-positive  \n",
       "81                RA - sero-positive  \n",
       "82                RA - sero-positive  \n",
       "83                RA - sero-positive  \n",
       "84                RA - sero-positive  \n",
       "85                RA - sero-positive  \n",
       "86                RA - sero-positive  \n",
       "87                RA - sero-positive  \n",
       "88                RA - sero-positive  \n",
       "89                RA - sero-negative  \n",
       "90                RA - sero-positive  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"sample_tags\"].str.contains(\"CD8\\+\").fillna(False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC8</td>\n",
       "      <td>F</td>\n",
       "      <td>44.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC9</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC9</td>\n",
       "      <td>M</td>\n",
       "      <td>55.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HC11</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC11</td>\n",
       "      <td>M</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HC17</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC17</td>\n",
       "      <td>F</td>\n",
       "      <td>21.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HC10</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC10</td>\n",
       "      <td>M</td>\n",
       "      <td>52.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HC18</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC18</td>\n",
       "      <td>M</td>\n",
       "      <td>60.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HC19</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC19</td>\n",
       "      <td>M</td>\n",
       "      <td>43.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HC20</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC20</td>\n",
       "      <td>F</td>\n",
       "      <td>66.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>RA72</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA72</td>\n",
       "      <td>F</td>\n",
       "      <td>61.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>RA74</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA74</td>\n",
       "      <td>F</td>\n",
       "      <td>43.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RA75</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA75</td>\n",
       "      <td>M</td>\n",
       "      <td>54.8</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>RA76</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA76</td>\n",
       "      <td>F</td>\n",
       "      <td>27.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>RA79</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA79</td>\n",
       "      <td>F</td>\n",
       "      <td>76.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>RA80</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA80</td>\n",
       "      <td>F</td>\n",
       "      <td>44.2</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>RA24</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA24</td>\n",
       "      <td>F</td>\n",
       "      <td>56.9</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>RA81</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA81</td>\n",
       "      <td>F</td>\n",
       "      <td>59.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>RA82</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA82</td>\n",
       "      <td>F</td>\n",
       "      <td>52.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>RA25</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA25</td>\n",
       "      <td>M</td>\n",
       "      <td>18.8</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HC12</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC12</td>\n",
       "      <td>F</td>\n",
       "      <td>57.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HC13</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC13</td>\n",
       "      <td>F</td>\n",
       "      <td>56.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HC14</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC14</td>\n",
       "      <td>M</td>\n",
       "      <td>70.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HC15</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC15</td>\n",
       "      <td>F</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HC16</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC16</td>\n",
       "      <td>M</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>RA52</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA52</td>\n",
       "      <td>F</td>\n",
       "      <td>33.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>RA53</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA53</td>\n",
       "      <td>F</td>\n",
       "      <td>32.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>RA54</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA54</td>\n",
       "      <td>F</td>\n",
       "      <td>35.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>RA55</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA55</td>\n",
       "      <td>F</td>\n",
       "      <td>31.1</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>RA56</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA56</td>\n",
       "      <td>F</td>\n",
       "      <td>69.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>RA57</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA57</td>\n",
       "      <td>F</td>\n",
       "      <td>70.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>RA58</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA58</td>\n",
       "      <td>F</td>\n",
       "      <td>49.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>RA59</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA59</td>\n",
       "      <td>F</td>\n",
       "      <td>54.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>RA60</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA60</td>\n",
       "      <td>F</td>\n",
       "      <td>62.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>RA62</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA62</td>\n",
       "      <td>F</td>\n",
       "      <td>36.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>RA63</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA63</td>\n",
       "      <td>F</td>\n",
       "      <td>63.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>RA65</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA65</td>\n",
       "      <td>F</td>\n",
       "      <td>37.8</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>RA66</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA66</td>\n",
       "      <td>F</td>\n",
       "      <td>65.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>RA68</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA68</td>\n",
       "      <td>F</td>\n",
       "      <td>64.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>RA69</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA69</td>\n",
       "      <td>F</td>\n",
       "      <td>81.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>RA70</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA70</td>\n",
       "      <td>M</td>\n",
       "      <td>47.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>RA26</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA26</td>\n",
       "      <td>F</td>\n",
       "      <td>47.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>RA28</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA28</td>\n",
       "      <td>F</td>\n",
       "      <td>30.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>RA29</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA29</td>\n",
       "      <td>F</td>\n",
       "      <td>32.9</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>RA30</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA30</td>\n",
       "      <td>F</td>\n",
       "      <td>65.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>RA31</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA31</td>\n",
       "      <td>F</td>\n",
       "      <td>44.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>RA32</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA32</td>\n",
       "      <td>F</td>\n",
       "      <td>39.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>RA33</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA33</td>\n",
       "      <td>F</td>\n",
       "      <td>53.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>RA37</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA37</td>\n",
       "      <td>F</td>\n",
       "      <td>47.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>RA40</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA40</td>\n",
       "      <td>F</td>\n",
       "      <td>64.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>RA41</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA41</td>\n",
       "      <td>F</td>\n",
       "      <td>38.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>RA46</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA46</td>\n",
       "      <td>F</td>\n",
       "      <td>45.2</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>RA47</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA47</td>\n",
       "      <td>F</td>\n",
       "      <td>67.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>RA48</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA48</td>\n",
       "      <td>F</td>\n",
       "      <td>46.1</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>RA49</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA49</td>\n",
       "      <td>M</td>\n",
       "      <td>29.9</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>RA50</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA50</td>\n",
       "      <td>F</td>\n",
       "      <td>32.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>RA51</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA51</td>\n",
       "      <td>F</td>\n",
       "      <td>49.9</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name locus      sample_tags counting_method       primer_set  \\\n",
       "3          HC8  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "4          HC9  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "5         HC11  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "6         HC17  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "7         HC10  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "8         HC18  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "9         HC19  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "10        HC20  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "11        RA72  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "12        RA74  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "13        RA75  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "14        RA76  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "15        RA79  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "16        RA80  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "17        RA24  TCRB  HLA MHC Class I              v3  Human-TCRB-PD1x   \n",
       "18        RA81  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "19        RA82  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "20        RA25  TCRB  HLA MHC Class I              v3  Human-TCRB-PD1x   \n",
       "21        HC12  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "22        HC13  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "23        HC14  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "24        HC15  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "25        HC16  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "26        RA52  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "27        RA53  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "28        RA54  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "29        RA55  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "30        RA56  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "31        RA57  TCRB  HLA MHC Class I              v3  Human-TCRB-PD1x   \n",
       "32        RA58  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "33        RA59  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "34        RA60  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "35        RA62  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "36        RA63  TCRB  HLA MHC Class I              v3  Human-TCRB-PD1x   \n",
       "37        RA65  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "38        RA66  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "39        RA68  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "40        RA69  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "41        RA70  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "42        RA26  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "43        RA28  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "44        RA29  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "45        RA30  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "46        RA31  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "47        RA32  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "48        RA33  TCRB  HLA MHC Class I              v3  Human-TCRB-PD1x   \n",
       "49        RA37  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "50        RA40  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "51        RA41  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "52        RA46  TCRB  HLA MHC Class I              v3  Human-TCRB-PD1x   \n",
       "53        RA47  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "54        RA48  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "55        RA49  TCRB  HLA MHC Class I              v3  Human-TCRB-PD1x   \n",
       "56        RA50  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "57        RA51  TCRB              NaN              v3  Human-TCRB-PD1x   \n",
       "\n",
       "   species                   study sequencing_type participant_label Sex  \\\n",
       "3    Human  mustjoki-2017-natcomms            gDNA               HC8   F   \n",
       "4    Human  mustjoki-2017-natcomms            gDNA               HC9   M   \n",
       "5    Human  mustjoki-2017-natcomms            gDNA              HC11   M   \n",
       "6    Human  mustjoki-2017-natcomms            gDNA              HC17   F   \n",
       "7    Human  mustjoki-2017-natcomms            gDNA              HC10   M   \n",
       "8    Human  mustjoki-2017-natcomms            gDNA              HC18   M   \n",
       "9    Human  mustjoki-2017-natcomms            gDNA              HC19   M   \n",
       "10   Human  mustjoki-2017-natcomms            gDNA              HC20   F   \n",
       "11   Human  mustjoki-2017-natcomms            gDNA              RA72   F   \n",
       "12   Human  mustjoki-2017-natcomms            gDNA              RA74   F   \n",
       "13   Human  mustjoki-2017-natcomms            gDNA              RA75   M   \n",
       "14   Human  mustjoki-2017-natcomms            gDNA              RA76   F   \n",
       "15   Human  mustjoki-2017-natcomms            gDNA              RA79   F   \n",
       "16   Human  mustjoki-2017-natcomms            gDNA              RA80   F   \n",
       "17   Human  mustjoki-2017-natcomms            gDNA              RA24   F   \n",
       "18   Human  mustjoki-2017-natcomms            gDNA              RA81   F   \n",
       "19   Human  mustjoki-2017-natcomms            gDNA              RA82   F   \n",
       "20   Human  mustjoki-2017-natcomms            gDNA              RA25   M   \n",
       "21   Human  mustjoki-2017-natcomms            gDNA              HC12   F   \n",
       "22   Human  mustjoki-2017-natcomms            gDNA              HC13   F   \n",
       "23   Human  mustjoki-2017-natcomms            gDNA              HC14   M   \n",
       "24   Human  mustjoki-2017-natcomms            gDNA              HC15   F   \n",
       "25   Human  mustjoki-2017-natcomms            gDNA              HC16   M   \n",
       "26   Human  mustjoki-2017-natcomms            gDNA              RA52   F   \n",
       "27   Human  mustjoki-2017-natcomms            gDNA              RA53   F   \n",
       "28   Human  mustjoki-2017-natcomms            gDNA              RA54   F   \n",
       "29   Human  mustjoki-2017-natcomms            gDNA              RA55   F   \n",
       "30   Human  mustjoki-2017-natcomms            gDNA              RA56   F   \n",
       "31   Human  mustjoki-2017-natcomms            gDNA              RA57   F   \n",
       "32   Human  mustjoki-2017-natcomms            gDNA              RA58   F   \n",
       "33   Human  mustjoki-2017-natcomms            gDNA              RA59   F   \n",
       "34   Human  mustjoki-2017-natcomms            gDNA              RA60   F   \n",
       "35   Human  mustjoki-2017-natcomms            gDNA              RA62   F   \n",
       "36   Human  mustjoki-2017-natcomms            gDNA              RA63   F   \n",
       "37   Human  mustjoki-2017-natcomms            gDNA              RA65   F   \n",
       "38   Human  mustjoki-2017-natcomms            gDNA              RA66   F   \n",
       "39   Human  mustjoki-2017-natcomms            gDNA              RA68   F   \n",
       "40   Human  mustjoki-2017-natcomms            gDNA              RA69   F   \n",
       "41   Human  mustjoki-2017-natcomms            gDNA              RA70   M   \n",
       "42   Human  mustjoki-2017-natcomms            gDNA              RA26   F   \n",
       "43   Human  mustjoki-2017-natcomms            gDNA              RA28   F   \n",
       "44   Human  mustjoki-2017-natcomms            gDNA              RA29   F   \n",
       "45   Human  mustjoki-2017-natcomms            gDNA              RA30   F   \n",
       "46   Human  mustjoki-2017-natcomms            gDNA              RA31   F   \n",
       "47   Human  mustjoki-2017-natcomms            gDNA              RA32   F   \n",
       "48   Human  mustjoki-2017-natcomms            gDNA              RA33   F   \n",
       "49   Human  mustjoki-2017-natcomms            gDNA              RA37   F   \n",
       "50   Human  mustjoki-2017-natcomms            gDNA              RA40   F   \n",
       "51   Human  mustjoki-2017-natcomms            gDNA              RA41   F   \n",
       "52   Human  mustjoki-2017-natcomms            gDNA              RA46   F   \n",
       "53   Human  mustjoki-2017-natcomms            gDNA              RA47   F   \n",
       "54   Human  mustjoki-2017-natcomms            gDNA              RA48   F   \n",
       "55   Human  mustjoki-2017-natcomms            gDNA              RA49   M   \n",
       "56   Human  mustjoki-2017-natcomms            gDNA              RA50   F   \n",
       "57   Human  mustjoki-2017-natcomms            gDNA              RA51   F   \n",
       "\n",
       "     Age             disease                   disease_subtype  \n",
       "3   44.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "4   55.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "5   48.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "6   21.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "7   52.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "8   60.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "9   43.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "10  66.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "11  61.6                  RA                RA - sero-positive  \n",
       "12  43.4                  RA                RA - sero-positive  \n",
       "13  54.8                  RA                RA - sero-positive  \n",
       "14  27.5                  RA                RA - sero-positive  \n",
       "15  76.6                  RA                RA - sero-negative  \n",
       "16  44.2                  RA                RA - sero-positive  \n",
       "17  56.9                  RA                RA - sero-negative  \n",
       "18  59.0                  RA                RA - sero-positive  \n",
       "19  52.3                  RA                RA - sero-positive  \n",
       "20  18.8                  RA                RA - sero-positive  \n",
       "21  57.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "22  56.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "23  70.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "24  47.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "25  48.0  Healthy/Background  Healthy/Background - RA negative  \n",
       "26  33.3                  RA                RA - sero-positive  \n",
       "27  32.6                  RA                RA - sero-positive  \n",
       "28  35.3                  RA                RA - sero-negative  \n",
       "29  31.1                  RA                RA - sero-positive  \n",
       "30  69.6                  RA                RA - sero-positive  \n",
       "31  70.4                  RA                RA - sero-negative  \n",
       "32  49.3                  RA                RA - sero-positive  \n",
       "33  54.5                  RA                RA - sero-positive  \n",
       "34  62.6                  RA                RA - sero-positive  \n",
       "35  36.6                  RA                RA - sero-positive  \n",
       "36  63.0                  RA                RA - sero-positive  \n",
       "37  37.8                  RA                RA - sero-positive  \n",
       "38  65.5                  RA                RA - sero-positive  \n",
       "39  64.3                  RA                RA - sero-negative  \n",
       "40  81.6                  RA                RA - sero-positive  \n",
       "41  47.0                  RA                RA - sero-positive  \n",
       "42  47.5                  RA                RA - sero-positive  \n",
       "43  30.6                  RA                RA - sero-positive  \n",
       "44  32.9                  RA                RA - sero-negative  \n",
       "45  65.4                  RA                RA - sero-positive  \n",
       "46  44.3                  RA                RA - sero-positive  \n",
       "47  39.3                  RA                RA - sero-positive  \n",
       "48  53.4                  RA                RA - sero-positive  \n",
       "49  47.0                  RA                RA - sero-positive  \n",
       "50  64.0                  RA                RA - sero-positive  \n",
       "51  38.4                  RA                RA - sero-negative  \n",
       "52  45.2                  RA                RA - sero-positive  \n",
       "53  67.6                  RA                RA - sero-positive  \n",
       "54  46.1                  RA                RA - sero-positive  \n",
       "55  29.9                  RA                RA - sero-positive  \n",
       "56  32.0                  RA                RA - sero-positive  \n",
       "57  49.9                  RA                RA - sero-positive  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Many samples are full PBMC though\n",
    "df[\n",
    "    (~df[\"sample_tags\"].str.contains(\"CD8\").fillna(False))\n",
    "    & (~df[\"sample_tags\"].str.contains(\"CD4\").fillna(False))\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>replicate_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>RA23</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA23</td>\n",
       "      <td>F</td>\n",
       "      <td>24.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA23</td>\n",
       "      <td>RA23_CD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>RA23-CD4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA23</td>\n",
       "      <td>F</td>\n",
       "      <td>24.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA23</td>\n",
       "      <td>RA23_CD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>RA6</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA6</td>\n",
       "      <td>M</td>\n",
       "      <td>74.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA6</td>\n",
       "      <td>RA6_CD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>RA6-CD4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA6</td>\n",
       "      <td>M</td>\n",
       "      <td>74.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA6</td>\n",
       "      <td>RA6_CD4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name locus                                        sample_tags  \\\n",
       "65        RA23  TCRB  CD8+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "66    RA23-CD4  TCRB  CD4+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "77         RA6  TCRB                CD4-, Rheumatoid Arthritis, T cells   \n",
       "76     RA6-CD4  TCRB                CD8-, Rheumatoid Arthritis, T cells   \n",
       "\n",
       "   counting_method       primer_set species                   study  \\\n",
       "65              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "66              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "77              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "76              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "\n",
       "   sequencing_type participant_label Sex   Age disease     disease_subtype  \\\n",
       "65            gDNA              RA23   F  24.4      RA  RA - sero-negative   \n",
       "66            gDNA              RA23   F  24.4      RA  RA - sero-negative   \n",
       "77            gDNA               RA6   M  74.5      RA  RA - sero-positive   \n",
       "76            gDNA               RA6   M  74.5      RA  RA - sero-positive   \n",
       "\n",
       "   specimen_label replicate_label  \n",
       "65           RA23        RA23_CD8  \n",
       "66           RA23        RA23_CD4  \n",
       "77            RA6         RA6_CD8  \n",
       "76            RA6         RA6_CD4  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Mark the CD8/CD4 separations\n",
    "# keep the cell type inside the replicate label, to be merged within the ETL flow later.\n",
    "df[\"specimen_label\"] = df[\"participant_label\"]\n",
    "df[\"replicate_label\"] = df[\"sample_name\"]\n",
    "df.loc[df[\"sample_name\"] == \"RA23-CD4\", \"replicate_label\"] = \"RA23_CD4\"\n",
    "df.loc[df[\"sample_name\"] == \"RA23\", \"replicate_label\"] = \"RA23_CD8\"\n",
    "df.loc[df[\"sample_name\"] == \"RA6-CD4\", \"replicate_label\"] = \"RA6_CD4\"\n",
    "df.loc[df[\"sample_name\"] == \"RA6\", \"replicate_label\"] = \"RA6_CD8\"\n",
    "\n",
    "# Recheck: people with multiple samples\n",
    "# We see some CD4 separated out of CD8\n",
    "df[\n",
    "    df[\"participant_label\"].isin(\n",
    "        df[\"participant_label\"]\n",
    "        .value_counts()[df[\"participant_label\"].value_counts() > 1]\n",
    "        .index\n",
    "    )\n",
    "].sort_values(\"sample_name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>replicate_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HC10</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC10</td>\n",
       "      <td>M</td>\n",
       "      <td>52.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC10</td>\n",
       "      <td>HC10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HC11</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC11</td>\n",
       "      <td>M</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC11</td>\n",
       "      <td>HC11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HC12</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC12</td>\n",
       "      <td>F</td>\n",
       "      <td>57.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC12</td>\n",
       "      <td>HC12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HC13</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC13</td>\n",
       "      <td>F</td>\n",
       "      <td>56.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC13</td>\n",
       "      <td>HC13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HC14</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC14</td>\n",
       "      <td>M</td>\n",
       "      <td>70.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC14</td>\n",
       "      <td>HC14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HC15</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC15</td>\n",
       "      <td>F</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC15</td>\n",
       "      <td>HC15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HC16</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC16</td>\n",
       "      <td>M</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC16</td>\n",
       "      <td>HC16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HC17</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC17</td>\n",
       "      <td>F</td>\n",
       "      <td>21.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC17</td>\n",
       "      <td>HC17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HC18</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC18</td>\n",
       "      <td>M</td>\n",
       "      <td>60.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC18</td>\n",
       "      <td>HC18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HC19</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC19</td>\n",
       "      <td>M</td>\n",
       "      <td>43.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC19</td>\n",
       "      <td>HC19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HC20</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC20</td>\n",
       "      <td>F</td>\n",
       "      <td>66.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC20</td>\n",
       "      <td>HC20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC8</td>\n",
       "      <td>F</td>\n",
       "      <td>44.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC8</td>\n",
       "      <td>HC8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC9</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC9</td>\n",
       "      <td>M</td>\n",
       "      <td>55.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC9</td>\n",
       "      <td>HC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>RA23-CD4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA23</td>\n",
       "      <td>F</td>\n",
       "      <td>24.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA23</td>\n",
       "      <td>RA23_CD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>RA23</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA23</td>\n",
       "      <td>F</td>\n",
       "      <td>24.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA23</td>\n",
       "      <td>RA23_CD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>RA26</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA26</td>\n",
       "      <td>F</td>\n",
       "      <td>47.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA26</td>\n",
       "      <td>RA26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>RA28</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA28</td>\n",
       "      <td>F</td>\n",
       "      <td>30.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA28</td>\n",
       "      <td>RA28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>RA29</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA29</td>\n",
       "      <td>F</td>\n",
       "      <td>32.9</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA29</td>\n",
       "      <td>RA29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>RA30</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA30</td>\n",
       "      <td>F</td>\n",
       "      <td>65.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA30</td>\n",
       "      <td>RA30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>RA31</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA31</td>\n",
       "      <td>F</td>\n",
       "      <td>44.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA31</td>\n",
       "      <td>RA31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>RA32</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA32</td>\n",
       "      <td>F</td>\n",
       "      <td>39.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA32</td>\n",
       "      <td>RA32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>RA37</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA37</td>\n",
       "      <td>F</td>\n",
       "      <td>47.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA37</td>\n",
       "      <td>RA37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>RA40</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA40</td>\n",
       "      <td>F</td>\n",
       "      <td>64.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA40</td>\n",
       "      <td>RA40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>RA41</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA41</td>\n",
       "      <td>F</td>\n",
       "      <td>38.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA41</td>\n",
       "      <td>RA41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>RA47</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA47</td>\n",
       "      <td>F</td>\n",
       "      <td>67.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA47</td>\n",
       "      <td>RA47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>RA48</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA48</td>\n",
       "      <td>F</td>\n",
       "      <td>46.1</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA48</td>\n",
       "      <td>RA48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>RA50</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA50</td>\n",
       "      <td>F</td>\n",
       "      <td>32.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA50</td>\n",
       "      <td>RA50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>RA51</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA51</td>\n",
       "      <td>F</td>\n",
       "      <td>49.9</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA51</td>\n",
       "      <td>RA51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>RA52</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA52</td>\n",
       "      <td>F</td>\n",
       "      <td>33.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA52</td>\n",
       "      <td>RA52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>RA53</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA53</td>\n",
       "      <td>F</td>\n",
       "      <td>32.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA53</td>\n",
       "      <td>RA53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>RA54</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA54</td>\n",
       "      <td>F</td>\n",
       "      <td>35.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA54</td>\n",
       "      <td>RA54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>RA55</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA55</td>\n",
       "      <td>F</td>\n",
       "      <td>31.1</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA55</td>\n",
       "      <td>RA55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>RA56</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA56</td>\n",
       "      <td>F</td>\n",
       "      <td>69.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA56</td>\n",
       "      <td>RA56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>RA58</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA58</td>\n",
       "      <td>F</td>\n",
       "      <td>49.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA58</td>\n",
       "      <td>RA58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>RA59</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA59</td>\n",
       "      <td>F</td>\n",
       "      <td>54.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA59</td>\n",
       "      <td>RA59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>RA60</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA60</td>\n",
       "      <td>F</td>\n",
       "      <td>62.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA60</td>\n",
       "      <td>RA60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>RA62</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA62</td>\n",
       "      <td>F</td>\n",
       "      <td>36.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA62</td>\n",
       "      <td>RA62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>RA65</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA65</td>\n",
       "      <td>F</td>\n",
       "      <td>37.8</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA65</td>\n",
       "      <td>RA65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>RA66</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA66</td>\n",
       "      <td>F</td>\n",
       "      <td>65.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA66</td>\n",
       "      <td>RA66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>RA68</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA68</td>\n",
       "      <td>F</td>\n",
       "      <td>64.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA68</td>\n",
       "      <td>RA68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>RA69</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA69</td>\n",
       "      <td>F</td>\n",
       "      <td>81.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA69</td>\n",
       "      <td>RA69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>RA6-CD4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA6</td>\n",
       "      <td>M</td>\n",
       "      <td>74.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA6</td>\n",
       "      <td>RA6_CD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>RA6</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA6</td>\n",
       "      <td>M</td>\n",
       "      <td>74.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA6</td>\n",
       "      <td>RA6_CD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>RA70</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA70</td>\n",
       "      <td>M</td>\n",
       "      <td>47.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA70</td>\n",
       "      <td>RA70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>RA72</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA72</td>\n",
       "      <td>F</td>\n",
       "      <td>61.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA72</td>\n",
       "      <td>RA72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>RA74</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA74</td>\n",
       "      <td>F</td>\n",
       "      <td>43.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA74</td>\n",
       "      <td>RA74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RA75</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA75</td>\n",
       "      <td>M</td>\n",
       "      <td>54.8</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA75</td>\n",
       "      <td>RA75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>RA76</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA76</td>\n",
       "      <td>F</td>\n",
       "      <td>27.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA76</td>\n",
       "      <td>RA76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>RA79</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA79</td>\n",
       "      <td>F</td>\n",
       "      <td>76.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA79</td>\n",
       "      <td>RA79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>RA80</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA80</td>\n",
       "      <td>F</td>\n",
       "      <td>44.2</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA80</td>\n",
       "      <td>RA80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>RA81</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA81</td>\n",
       "      <td>F</td>\n",
       "      <td>59.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA81</td>\n",
       "      <td>RA81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>RA82</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA82</td>\n",
       "      <td>F</td>\n",
       "      <td>52.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA82</td>\n",
       "      <td>RA82</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name locus                                        sample_tags  \\\n",
       "7         HC10  TCRB                                                NaN   \n",
       "5         HC11  TCRB                                                NaN   \n",
       "21        HC12  TCRB                                                NaN   \n",
       "22        HC13  TCRB                                                NaN   \n",
       "23        HC14  TCRB                                                NaN   \n",
       "24        HC15  TCRB                                                NaN   \n",
       "25        HC16  TCRB                                                NaN   \n",
       "6         HC17  TCRB                                                NaN   \n",
       "8         HC18  TCRB                                                NaN   \n",
       "9         HC19  TCRB                                                NaN   \n",
       "10        HC20  TCRB                                                NaN   \n",
       "3          HC8  TCRB                                                NaN   \n",
       "4          HC9  TCRB                                                NaN   \n",
       "66    RA23-CD4  TCRB  CD4+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "65        RA23  TCRB  CD8+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "42        RA26  TCRB                                                NaN   \n",
       "43        RA28  TCRB                                                NaN   \n",
       "44        RA29  TCRB                                                NaN   \n",
       "45        RA30  TCRB                                                NaN   \n",
       "46        RA31  TCRB                                                NaN   \n",
       "47        RA32  TCRB                                                NaN   \n",
       "49        RA37  TCRB                                                NaN   \n",
       "50        RA40  TCRB                                                NaN   \n",
       "51        RA41  TCRB                                                NaN   \n",
       "53        RA47  TCRB                                                NaN   \n",
       "54        RA48  TCRB                                                NaN   \n",
       "56        RA50  TCRB                                                NaN   \n",
       "57        RA51  TCRB                                                NaN   \n",
       "26        RA52  TCRB                                                NaN   \n",
       "27        RA53  TCRB                                                NaN   \n",
       "28        RA54  TCRB                                                NaN   \n",
       "29        RA55  TCRB                                                NaN   \n",
       "30        RA56  TCRB                                                NaN   \n",
       "32        RA58  TCRB                                                NaN   \n",
       "33        RA59  TCRB                                                NaN   \n",
       "34        RA60  TCRB                                                NaN   \n",
       "35        RA62  TCRB                                                NaN   \n",
       "37        RA65  TCRB                                                NaN   \n",
       "38        RA66  TCRB                                                NaN   \n",
       "39        RA68  TCRB                                                NaN   \n",
       "40        RA69  TCRB                                                NaN   \n",
       "76     RA6-CD4  TCRB                CD8-, Rheumatoid Arthritis, T cells   \n",
       "77         RA6  TCRB                CD4-, Rheumatoid Arthritis, T cells   \n",
       "41        RA70  TCRB                                                NaN   \n",
       "11        RA72  TCRB                                                NaN   \n",
       "12        RA74  TCRB                                                NaN   \n",
       "13        RA75  TCRB                                                NaN   \n",
       "14        RA76  TCRB                                                NaN   \n",
       "15        RA79  TCRB                                                NaN   \n",
       "16        RA80  TCRB                                                NaN   \n",
       "18        RA81  TCRB                                                NaN   \n",
       "19        RA82  TCRB                                                NaN   \n",
       "\n",
       "   counting_method       primer_set species                   study  \\\n",
       "7               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "5               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "21              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "22              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "23              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "24              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "25              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "6               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "8               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "9               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "10              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "3               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "4               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "66              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "65              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "42              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "43              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "44              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "45              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "46              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "47              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "49              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "50              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "51              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "53              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "54              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "56              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "57              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "26              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "27              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "28              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "29              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "30              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "32              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "33              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "34              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "35              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "37              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "38              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "39              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "40              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "76              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "77              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "41              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "11              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "12              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "13              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "14              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "15              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "16              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "18              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "19              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "\n",
       "   sequencing_type participant_label Sex   Age             disease  \\\n",
       "7             gDNA              HC10   M  52.0  Healthy/Background   \n",
       "5             gDNA              HC11   M  48.0  Healthy/Background   \n",
       "21            gDNA              HC12   F  57.0  Healthy/Background   \n",
       "22            gDNA              HC13   F  56.0  Healthy/Background   \n",
       "23            gDNA              HC14   M  70.0  Healthy/Background   \n",
       "24            gDNA              HC15   F  47.0  Healthy/Background   \n",
       "25            gDNA              HC16   M  48.0  Healthy/Background   \n",
       "6             gDNA              HC17   F  21.0  Healthy/Background   \n",
       "8             gDNA              HC18   M  60.0  Healthy/Background   \n",
       "9             gDNA              HC19   M  43.0  Healthy/Background   \n",
       "10            gDNA              HC20   F  66.0  Healthy/Background   \n",
       "3             gDNA               HC8   F  44.0  Healthy/Background   \n",
       "4             gDNA               HC9   M  55.0  Healthy/Background   \n",
       "66            gDNA              RA23   F  24.4                  RA   \n",
       "65            gDNA              RA23   F  24.4                  RA   \n",
       "42            gDNA              RA26   F  47.5                  RA   \n",
       "43            gDNA              RA28   F  30.6                  RA   \n",
       "44            gDNA              RA29   F  32.9                  RA   \n",
       "45            gDNA              RA30   F  65.4                  RA   \n",
       "46            gDNA              RA31   F  44.3                  RA   \n",
       "47            gDNA              RA32   F  39.3                  RA   \n",
       "49            gDNA              RA37   F  47.0                  RA   \n",
       "50            gDNA              RA40   F  64.0                  RA   \n",
       "51            gDNA              RA41   F  38.4                  RA   \n",
       "53            gDNA              RA47   F  67.6                  RA   \n",
       "54            gDNA              RA48   F  46.1                  RA   \n",
       "56            gDNA              RA50   F  32.0                  RA   \n",
       "57            gDNA              RA51   F  49.9                  RA   \n",
       "26            gDNA              RA52   F  33.3                  RA   \n",
       "27            gDNA              RA53   F  32.6                  RA   \n",
       "28            gDNA              RA54   F  35.3                  RA   \n",
       "29            gDNA              RA55   F  31.1                  RA   \n",
       "30            gDNA              RA56   F  69.6                  RA   \n",
       "32            gDNA              RA58   F  49.3                  RA   \n",
       "33            gDNA              RA59   F  54.5                  RA   \n",
       "34            gDNA              RA60   F  62.6                  RA   \n",
       "35            gDNA              RA62   F  36.6                  RA   \n",
       "37            gDNA              RA65   F  37.8                  RA   \n",
       "38            gDNA              RA66   F  65.5                  RA   \n",
       "39            gDNA              RA68   F  64.3                  RA   \n",
       "40            gDNA              RA69   F  81.6                  RA   \n",
       "76            gDNA               RA6   M  74.5                  RA   \n",
       "77            gDNA               RA6   M  74.5                  RA   \n",
       "41            gDNA              RA70   M  47.0                  RA   \n",
       "11            gDNA              RA72   F  61.6                  RA   \n",
       "12            gDNA              RA74   F  43.4                  RA   \n",
       "13            gDNA              RA75   M  54.8                  RA   \n",
       "14            gDNA              RA76   F  27.5                  RA   \n",
       "15            gDNA              RA79   F  76.6                  RA   \n",
       "16            gDNA              RA80   F  44.2                  RA   \n",
       "18            gDNA              RA81   F  59.0                  RA   \n",
       "19            gDNA              RA82   F  52.3                  RA   \n",
       "\n",
       "                     disease_subtype specimen_label replicate_label  \n",
       "7   Healthy/Background - RA negative           HC10            HC10  \n",
       "5   Healthy/Background - RA negative           HC11            HC11  \n",
       "21  Healthy/Background - RA negative           HC12            HC12  \n",
       "22  Healthy/Background - RA negative           HC13            HC13  \n",
       "23  Healthy/Background - RA negative           HC14            HC14  \n",
       "24  Healthy/Background - RA negative           HC15            HC15  \n",
       "25  Healthy/Background - RA negative           HC16            HC16  \n",
       "6   Healthy/Background - RA negative           HC17            HC17  \n",
       "8   Healthy/Background - RA negative           HC18            HC18  \n",
       "9   Healthy/Background - RA negative           HC19            HC19  \n",
       "10  Healthy/Background - RA negative           HC20            HC20  \n",
       "3   Healthy/Background - RA negative            HC8             HC8  \n",
       "4   Healthy/Background - RA negative            HC9             HC9  \n",
       "66                RA - sero-negative           RA23        RA23_CD4  \n",
       "65                RA - sero-negative           RA23        RA23_CD8  \n",
       "42                RA - sero-positive           RA26            RA26  \n",
       "43                RA - sero-positive           RA28            RA28  \n",
       "44                RA - sero-negative           RA29            RA29  \n",
       "45                RA - sero-positive           RA30            RA30  \n",
       "46                RA - sero-positive           RA31            RA31  \n",
       "47                RA - sero-positive           RA32            RA32  \n",
       "49                RA - sero-positive           RA37            RA37  \n",
       "50                RA - sero-positive           RA40            RA40  \n",
       "51                RA - sero-negative           RA41            RA41  \n",
       "53                RA - sero-positive           RA47            RA47  \n",
       "54                RA - sero-positive           RA48            RA48  \n",
       "56                RA - sero-positive           RA50            RA50  \n",
       "57                RA - sero-positive           RA51            RA51  \n",
       "26                RA - sero-positive           RA52            RA52  \n",
       "27                RA - sero-positive           RA53            RA53  \n",
       "28                RA - sero-negative           RA54            RA54  \n",
       "29                RA - sero-positive           RA55            RA55  \n",
       "30                RA - sero-positive           RA56            RA56  \n",
       "32                RA - sero-positive           RA58            RA58  \n",
       "33                RA - sero-positive           RA59            RA59  \n",
       "34                RA - sero-positive           RA60            RA60  \n",
       "35                RA - sero-positive           RA62            RA62  \n",
       "37                RA - sero-positive           RA65            RA65  \n",
       "38                RA - sero-positive           RA66            RA66  \n",
       "39                RA - sero-negative           RA68            RA68  \n",
       "40                RA - sero-positive           RA69            RA69  \n",
       "76                RA - sero-positive            RA6         RA6_CD4  \n",
       "77                RA - sero-positive            RA6         RA6_CD8  \n",
       "41                RA - sero-positive           RA70            RA70  \n",
       "11                RA - sero-positive           RA72            RA72  \n",
       "12                RA - sero-positive           RA74            RA74  \n",
       "13                RA - sero-positive           RA75            RA75  \n",
       "14                RA - sero-positive           RA76            RA76  \n",
       "15                RA - sero-negative           RA79            RA79  \n",
       "16                RA - sero-positive           RA80            RA80  \n",
       "18                RA - sero-positive           RA81            RA81  \n",
       "19                RA - sero-positive           RA82            RA82  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Keep full PBMC samples only, for consistency\n",
    "df = pd.concat(\n",
    "    [\n",
    "        df[\n",
    "            (~df[\"sample_tags\"].str.contains(\"CD8\").fillna(False))\n",
    "            & (~df[\"sample_tags\"].str.contains(\"CD4\").fillna(False))\n",
    "            # this would indicate CD8:\n",
    "            & (~df[\"sample_tags\"].str.contains(\"HLA MHC Class I\").fillna(False))\n",
    "        ],\n",
    "        # also keep the ones where we have both fractions - see previous cell\n",
    "        df[df[\"replicate_label\"].isin([\"RA23_CD8\", \"RA23_CD4\", \"RA6_CD8\", \"RA6_CD4\"])],\n",
    "    ],\n",
    "    axis=0,\n",
    ")\n",
    "df = df.sort_values(\"replicate_label\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>replicate_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HC10</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC10</td>\n",
       "      <td>M</td>\n",
       "      <td>52.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC10</td>\n",
       "      <td>HC10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HC11</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC11</td>\n",
       "      <td>M</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC11</td>\n",
       "      <td>HC11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HC12</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC12</td>\n",
       "      <td>F</td>\n",
       "      <td>57.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC12</td>\n",
       "      <td>HC12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HC13</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC13</td>\n",
       "      <td>F</td>\n",
       "      <td>56.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC13</td>\n",
       "      <td>HC13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HC14</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC14</td>\n",
       "      <td>M</td>\n",
       "      <td>70.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC14</td>\n",
       "      <td>HC14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HC15</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC15</td>\n",
       "      <td>F</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC15</td>\n",
       "      <td>HC15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HC16</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC16</td>\n",
       "      <td>M</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC16</td>\n",
       "      <td>HC16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HC17</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC17</td>\n",
       "      <td>F</td>\n",
       "      <td>21.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC17</td>\n",
       "      <td>HC17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HC18</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC18</td>\n",
       "      <td>M</td>\n",
       "      <td>60.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC18</td>\n",
       "      <td>HC18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HC19</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC19</td>\n",
       "      <td>M</td>\n",
       "      <td>43.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC19</td>\n",
       "      <td>HC19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HC20</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC20</td>\n",
       "      <td>F</td>\n",
       "      <td>66.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC20</td>\n",
       "      <td>HC20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC8</td>\n",
       "      <td>F</td>\n",
       "      <td>44.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC8</td>\n",
       "      <td>HC8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC9</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HC9</td>\n",
       "      <td>M</td>\n",
       "      <td>55.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - RA negative</td>\n",
       "      <td>HC9</td>\n",
       "      <td>HC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>RA23-CD4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA23</td>\n",
       "      <td>F</td>\n",
       "      <td>24.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA23</td>\n",
       "      <td>RA23_CD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>RA23</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8+, Peripheral blood lymphocytes (PBL), Rheu...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA23</td>\n",
       "      <td>F</td>\n",
       "      <td>24.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA23</td>\n",
       "      <td>RA23_CD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>RA26</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA26</td>\n",
       "      <td>F</td>\n",
       "      <td>47.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA26</td>\n",
       "      <td>RA26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>RA28</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA28</td>\n",
       "      <td>F</td>\n",
       "      <td>30.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA28</td>\n",
       "      <td>RA28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>RA29</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA29</td>\n",
       "      <td>F</td>\n",
       "      <td>32.9</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA29</td>\n",
       "      <td>RA29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>RA30</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA30</td>\n",
       "      <td>F</td>\n",
       "      <td>65.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA30</td>\n",
       "      <td>RA30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>RA31</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA31</td>\n",
       "      <td>F</td>\n",
       "      <td>44.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA31</td>\n",
       "      <td>RA31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>RA32</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA32</td>\n",
       "      <td>F</td>\n",
       "      <td>39.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA32</td>\n",
       "      <td>RA32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>RA37</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA37</td>\n",
       "      <td>F</td>\n",
       "      <td>47.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA37</td>\n",
       "      <td>RA37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>RA40</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA40</td>\n",
       "      <td>F</td>\n",
       "      <td>64.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA40</td>\n",
       "      <td>RA40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>RA41</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA41</td>\n",
       "      <td>F</td>\n",
       "      <td>38.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA41</td>\n",
       "      <td>RA41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>RA47</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA47</td>\n",
       "      <td>F</td>\n",
       "      <td>67.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA47</td>\n",
       "      <td>RA47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>RA48</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA48</td>\n",
       "      <td>F</td>\n",
       "      <td>46.1</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA48</td>\n",
       "      <td>RA48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>RA50</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA50</td>\n",
       "      <td>F</td>\n",
       "      <td>32.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA50</td>\n",
       "      <td>RA50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>RA51</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA51</td>\n",
       "      <td>F</td>\n",
       "      <td>49.9</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA51</td>\n",
       "      <td>RA51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>RA52</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA52</td>\n",
       "      <td>F</td>\n",
       "      <td>33.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA52</td>\n",
       "      <td>RA52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>RA53</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA53</td>\n",
       "      <td>F</td>\n",
       "      <td>32.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA53</td>\n",
       "      <td>RA53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>RA54</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA54</td>\n",
       "      <td>F</td>\n",
       "      <td>35.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA54</td>\n",
       "      <td>RA54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>RA55</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA55</td>\n",
       "      <td>F</td>\n",
       "      <td>31.1</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA55</td>\n",
       "      <td>RA55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>RA56</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA56</td>\n",
       "      <td>F</td>\n",
       "      <td>69.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA56</td>\n",
       "      <td>RA56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>RA58</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA58</td>\n",
       "      <td>F</td>\n",
       "      <td>49.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA58</td>\n",
       "      <td>RA58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>RA59</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA59</td>\n",
       "      <td>F</td>\n",
       "      <td>54.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA59</td>\n",
       "      <td>RA59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>RA60</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA60</td>\n",
       "      <td>F</td>\n",
       "      <td>62.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA60</td>\n",
       "      <td>RA60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>RA62</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA62</td>\n",
       "      <td>F</td>\n",
       "      <td>36.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA62</td>\n",
       "      <td>RA62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>RA65</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA65</td>\n",
       "      <td>F</td>\n",
       "      <td>37.8</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA65</td>\n",
       "      <td>RA65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>RA66</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA66</td>\n",
       "      <td>F</td>\n",
       "      <td>65.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA66</td>\n",
       "      <td>RA66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>RA68</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA68</td>\n",
       "      <td>F</td>\n",
       "      <td>64.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA68</td>\n",
       "      <td>RA68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>RA69</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA69</td>\n",
       "      <td>F</td>\n",
       "      <td>81.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA69</td>\n",
       "      <td>RA69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>RA6-CD4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD8-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA6</td>\n",
       "      <td>M</td>\n",
       "      <td>74.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA6</td>\n",
       "      <td>RA6_CD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>RA6</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>CD4-, Rheumatoid Arthritis, T cells</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA6</td>\n",
       "      <td>M</td>\n",
       "      <td>74.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA6</td>\n",
       "      <td>RA6_CD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>RA70</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA70</td>\n",
       "      <td>M</td>\n",
       "      <td>47.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA70</td>\n",
       "      <td>RA70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>RA72</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA72</td>\n",
       "      <td>F</td>\n",
       "      <td>61.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA72</td>\n",
       "      <td>RA72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>RA74</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA74</td>\n",
       "      <td>F</td>\n",
       "      <td>43.4</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA74</td>\n",
       "      <td>RA74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RA75</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA75</td>\n",
       "      <td>M</td>\n",
       "      <td>54.8</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA75</td>\n",
       "      <td>RA75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>RA76</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA76</td>\n",
       "      <td>F</td>\n",
       "      <td>27.5</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA76</td>\n",
       "      <td>RA76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>RA79</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA79</td>\n",
       "      <td>F</td>\n",
       "      <td>76.6</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-negative</td>\n",
       "      <td>RA79</td>\n",
       "      <td>RA79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>RA80</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA80</td>\n",
       "      <td>F</td>\n",
       "      <td>44.2</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA80</td>\n",
       "      <td>RA80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>RA81</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA81</td>\n",
       "      <td>F</td>\n",
       "      <td>59.0</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA81</td>\n",
       "      <td>RA81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>RA82</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>mustjoki-2017-natcomms</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>RA82</td>\n",
       "      <td>F</td>\n",
       "      <td>52.3</td>\n",
       "      <td>RA</td>\n",
       "      <td>RA - sero-positive</td>\n",
       "      <td>RA82</td>\n",
       "      <td>RA82</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name locus                                        sample_tags  \\\n",
       "7         HC10  TCRB                                                NaN   \n",
       "5         HC11  TCRB                                                NaN   \n",
       "21        HC12  TCRB                                                NaN   \n",
       "22        HC13  TCRB                                                NaN   \n",
       "23        HC14  TCRB                                                NaN   \n",
       "24        HC15  TCRB                                                NaN   \n",
       "25        HC16  TCRB                                                NaN   \n",
       "6         HC17  TCRB                                                NaN   \n",
       "8         HC18  TCRB                                                NaN   \n",
       "9         HC19  TCRB                                                NaN   \n",
       "10        HC20  TCRB                                                NaN   \n",
       "3          HC8  TCRB                                                NaN   \n",
       "4          HC9  TCRB                                                NaN   \n",
       "66    RA23-CD4  TCRB  CD4+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "65        RA23  TCRB  CD8+, Peripheral blood lymphocytes (PBL), Rheu...   \n",
       "42        RA26  TCRB                                                NaN   \n",
       "43        RA28  TCRB                                                NaN   \n",
       "44        RA29  TCRB                                                NaN   \n",
       "45        RA30  TCRB                                                NaN   \n",
       "46        RA31  TCRB                                                NaN   \n",
       "47        RA32  TCRB                                                NaN   \n",
       "49        RA37  TCRB                                                NaN   \n",
       "50        RA40  TCRB                                                NaN   \n",
       "51        RA41  TCRB                                                NaN   \n",
       "53        RA47  TCRB                                                NaN   \n",
       "54        RA48  TCRB                                                NaN   \n",
       "56        RA50  TCRB                                                NaN   \n",
       "57        RA51  TCRB                                                NaN   \n",
       "26        RA52  TCRB                                                NaN   \n",
       "27        RA53  TCRB                                                NaN   \n",
       "28        RA54  TCRB                                                NaN   \n",
       "29        RA55  TCRB                                                NaN   \n",
       "30        RA56  TCRB                                                NaN   \n",
       "32        RA58  TCRB                                                NaN   \n",
       "33        RA59  TCRB                                                NaN   \n",
       "34        RA60  TCRB                                                NaN   \n",
       "35        RA62  TCRB                                                NaN   \n",
       "37        RA65  TCRB                                                NaN   \n",
       "38        RA66  TCRB                                                NaN   \n",
       "39        RA68  TCRB                                                NaN   \n",
       "40        RA69  TCRB                                                NaN   \n",
       "76     RA6-CD4  TCRB                CD8-, Rheumatoid Arthritis, T cells   \n",
       "77         RA6  TCRB                CD4-, Rheumatoid Arthritis, T cells   \n",
       "41        RA70  TCRB                                                NaN   \n",
       "11        RA72  TCRB                                                NaN   \n",
       "12        RA74  TCRB                                                NaN   \n",
       "13        RA75  TCRB                                                NaN   \n",
       "14        RA76  TCRB                                                NaN   \n",
       "15        RA79  TCRB                                                NaN   \n",
       "16        RA80  TCRB                                                NaN   \n",
       "18        RA81  TCRB                                                NaN   \n",
       "19        RA82  TCRB                                                NaN   \n",
       "\n",
       "   counting_method       primer_set species                   study  \\\n",
       "7               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "5               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "21              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "22              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "23              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "24              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "25              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "6               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "8               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "9               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "10              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "3               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "4               v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "66              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "65              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "42              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "43              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "44              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "45              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "46              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "47              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "49              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "50              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "51              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "53              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "54              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "56              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "57              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "26              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "27              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "28              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "29              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "30              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "32              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "33              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "34              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "35              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "37              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "38              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "39              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "40              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "76              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "77              v2  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "41              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "11              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "12              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "13              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "14              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "15              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "16              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "18              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "19              v3  Human-TCRB-PD1x   Human  mustjoki-2017-natcomms   \n",
       "\n",
       "   sequencing_type participant_label Sex   Age             disease  \\\n",
       "7             gDNA              HC10   M  52.0  Healthy/Background   \n",
       "5             gDNA              HC11   M  48.0  Healthy/Background   \n",
       "21            gDNA              HC12   F  57.0  Healthy/Background   \n",
       "22            gDNA              HC13   F  56.0  Healthy/Background   \n",
       "23            gDNA              HC14   M  70.0  Healthy/Background   \n",
       "24            gDNA              HC15   F  47.0  Healthy/Background   \n",
       "25            gDNA              HC16   M  48.0  Healthy/Background   \n",
       "6             gDNA              HC17   F  21.0  Healthy/Background   \n",
       "8             gDNA              HC18   M  60.0  Healthy/Background   \n",
       "9             gDNA              HC19   M  43.0  Healthy/Background   \n",
       "10            gDNA              HC20   F  66.0  Healthy/Background   \n",
       "3             gDNA               HC8   F  44.0  Healthy/Background   \n",
       "4             gDNA               HC9   M  55.0  Healthy/Background   \n",
       "66            gDNA              RA23   F  24.4                  RA   \n",
       "65            gDNA              RA23   F  24.4                  RA   \n",
       "42            gDNA              RA26   F  47.5                  RA   \n",
       "43            gDNA              RA28   F  30.6                  RA   \n",
       "44            gDNA              RA29   F  32.9                  RA   \n",
       "45            gDNA              RA30   F  65.4                  RA   \n",
       "46            gDNA              RA31   F  44.3                  RA   \n",
       "47            gDNA              RA32   F  39.3                  RA   \n",
       "49            gDNA              RA37   F  47.0                  RA   \n",
       "50            gDNA              RA40   F  64.0                  RA   \n",
       "51            gDNA              RA41   F  38.4                  RA   \n",
       "53            gDNA              RA47   F  67.6                  RA   \n",
       "54            gDNA              RA48   F  46.1                  RA   \n",
       "56            gDNA              RA50   F  32.0                  RA   \n",
       "57            gDNA              RA51   F  49.9                  RA   \n",
       "26            gDNA              RA52   F  33.3                  RA   \n",
       "27            gDNA              RA53   F  32.6                  RA   \n",
       "28            gDNA              RA54   F  35.3                  RA   \n",
       "29            gDNA              RA55   F  31.1                  RA   \n",
       "30            gDNA              RA56   F  69.6                  RA   \n",
       "32            gDNA              RA58   F  49.3                  RA   \n",
       "33            gDNA              RA59   F  54.5                  RA   \n",
       "34            gDNA              RA60   F  62.6                  RA   \n",
       "35            gDNA              RA62   F  36.6                  RA   \n",
       "37            gDNA              RA65   F  37.8                  RA   \n",
       "38            gDNA              RA66   F  65.5                  RA   \n",
       "39            gDNA              RA68   F  64.3                  RA   \n",
       "40            gDNA              RA69   F  81.6                  RA   \n",
       "76            gDNA               RA6   M  74.5                  RA   \n",
       "77            gDNA               RA6   M  74.5                  RA   \n",
       "41            gDNA              RA70   M  47.0                  RA   \n",
       "11            gDNA              RA72   F  61.6                  RA   \n",
       "12            gDNA              RA74   F  43.4                  RA   \n",
       "13            gDNA              RA75   M  54.8                  RA   \n",
       "14            gDNA              RA76   F  27.5                  RA   \n",
       "15            gDNA              RA79   F  76.6                  RA   \n",
       "16            gDNA              RA80   F  44.2                  RA   \n",
       "18            gDNA              RA81   F  59.0                  RA   \n",
       "19            gDNA              RA82   F  52.3                  RA   \n",
       "\n",
       "                     disease_subtype specimen_label replicate_label  \n",
       "7   Healthy/Background - RA negative           HC10            HC10  \n",
       "5   Healthy/Background - RA negative           HC11            HC11  \n",
       "21  Healthy/Background - RA negative           HC12            HC12  \n",
       "22  Healthy/Background - RA negative           HC13            HC13  \n",
       "23  Healthy/Background - RA negative           HC14            HC14  \n",
       "24  Healthy/Background - RA negative           HC15            HC15  \n",
       "25  Healthy/Background - RA negative           HC16            HC16  \n",
       "6   Healthy/Background - RA negative           HC17            HC17  \n",
       "8   Healthy/Background - RA negative           HC18            HC18  \n",
       "9   Healthy/Background - RA negative           HC19            HC19  \n",
       "10  Healthy/Background - RA negative           HC20            HC20  \n",
       "3   Healthy/Background - RA negative            HC8             HC8  \n",
       "4   Healthy/Background - RA negative            HC9             HC9  \n",
       "66                RA - sero-negative           RA23        RA23_CD4  \n",
       "65                RA - sero-negative           RA23        RA23_CD8  \n",
       "42                RA - sero-positive           RA26            RA26  \n",
       "43                RA - sero-positive           RA28            RA28  \n",
       "44                RA - sero-negative           RA29            RA29  \n",
       "45                RA - sero-positive           RA30            RA30  \n",
       "46                RA - sero-positive           RA31            RA31  \n",
       "47                RA - sero-positive           RA32            RA32  \n",
       "49                RA - sero-positive           RA37            RA37  \n",
       "50                RA - sero-positive           RA40            RA40  \n",
       "51                RA - sero-negative           RA41            RA41  \n",
       "53                RA - sero-positive           RA47            RA47  \n",
       "54                RA - sero-positive           RA48            RA48  \n",
       "56                RA - sero-positive           RA50            RA50  \n",
       "57                RA - sero-positive           RA51            RA51  \n",
       "26                RA - sero-positive           RA52            RA52  \n",
       "27                RA - sero-positive           RA53            RA53  \n",
       "28                RA - sero-negative           RA54            RA54  \n",
       "29                RA - sero-positive           RA55            RA55  \n",
       "30                RA - sero-positive           RA56            RA56  \n",
       "32                RA - sero-positive           RA58            RA58  \n",
       "33                RA - sero-positive           RA59            RA59  \n",
       "34                RA - sero-positive           RA60            RA60  \n",
       "35                RA - sero-positive           RA62            RA62  \n",
       "37                RA - sero-positive           RA65            RA65  \n",
       "38                RA - sero-positive           RA66            RA66  \n",
       "39                RA - sero-negative           RA68            RA68  \n",
       "40                RA - sero-positive           RA69            RA69  \n",
       "76                RA - sero-positive            RA6         RA6_CD4  \n",
       "77                RA - sero-positive            RA6         RA6_CD8  \n",
       "41                RA - sero-positive           RA70            RA70  \n",
       "11                RA - sero-positive           RA72            RA72  \n",
       "12                RA - sero-positive           RA74            RA74  \n",
       "13                RA - sero-positive           RA75            RA75  \n",
       "14                RA - sero-positive           RA76            RA76  \n",
       "15                RA - sero-negative           RA79            RA79  \n",
       "16                RA - sero-positive           RA80            RA80  \n",
       "18                RA - sero-positive           RA81            RA81  \n",
       "19                RA - sero-positive           RA82            RA82  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual.append(df)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>155</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>213</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>374</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>320</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>228</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>N18</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>N7</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>N9</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>14_Mt. Sinai-Cunningham-Rundles-P01</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>2_Mt. Sinai-Cunningham-Rundles-P01</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            sample_name locus sample_tags counting_method  \\\n",
       "0                                   155  TCRB        PBMC              v1   \n",
       "1                                   213  TCRB        PBMC              v1   \n",
       "2                                   374  TCRB        PBMC              v1   \n",
       "3                                   320  TCRB        PBMC              v1   \n",
       "4                                   228  TCRB        PBMC              v1   \n",
       "..                                  ...   ...         ...             ...   \n",
       "56                                  N18  TCRB        PBMC              v1   \n",
       "57                                   N7  TCRB        PBMC              v1   \n",
       "58                                   N9  TCRB        PBMC              v1   \n",
       "59  14_Mt. Sinai-Cunningham-Rundles-P01  TCRB        PBMC              v1   \n",
       "60   2_Mt. Sinai-Cunningham-Rundles-P01  TCRB        PBMC              v1   \n",
       "\n",
       "         primer_set species           study sequencing_type  \n",
       "0   Human-TCRB-none   Human  ramesh-2015-ci            gDNA  \n",
       "1   Human-TCRB-none   Human  ramesh-2015-ci            gDNA  \n",
       "2   Human-TCRB-none   Human  ramesh-2015-ci            gDNA  \n",
       "3   Human-TCRB-none   Human  ramesh-2015-ci            gDNA  \n",
       "4   Human-TCRB-none   Human  ramesh-2015-ci            gDNA  \n",
       "..              ...     ...             ...             ...  \n",
       "56  Human-TCRB-none   Human  ramesh-2015-ci            gDNA  \n",
       "57  Human-TCRB-none   Human  ramesh-2015-ci            gDNA  \n",
       "58  Human-TCRB-none   Human  ramesh-2015-ci            gDNA  \n",
       "59  Human-TCRB-none   Human  ramesh-2015-ci            gDNA  \n",
       "60  Human-TCRB-none   Human  ramesh-2015-ci            gDNA  \n",
       "\n",
       "[61 rows x 8 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = _load_metadata_df(\"ramesh-2015-ci\")\n",
    "\n",
    "\"\"\"\n",
    "CVID as a T cell defect, not just B cell\n",
    "https://www.sciencedirect.com/science/article/abs/pii/S1521661615000042?via%3Dihub\n",
    "up to 44 CVID, 22 HC? see https://ars.els-cdn.com/content/image/1-s2.0-S1521661615000042-gr6_lrg.jpg for the participant names\n",
    "\n",
    "Peripheral blood DNA samples of 44 CVID subjects, 15 females and 29 males of ages 9 to 64 years with a mean of 40,\n",
    "and 22 healthy adult volunteers, 12 females and 10 males of ages 23 to 66 years with a mean of 34\n",
    "\n",
    "Genomic DNA\n",
    "\"\"\"\n",
    "\n",
    "df[\"sequencing_type\"] = \"gDNA\"\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PBMC               61\n",
       "HLA MHC Class I     6\n",
       "Blood               1\n",
       "Name: sample_tags, dtype: int64"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"sample_tags\"].str.split(\",\").explode().str.strip().value_counts().head(n=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['155', '213', '374', '320', '228', '130', '345', '131', '211',\n",
       "       '289', '279', '335', '334', '386', '317', '378', '441', '172',\n",
       "       '139', '110', '299', '203', '462', '329', '272', '102', 'N12',\n",
       "       '400', '251', '218', '268', '161', '366', '248', '263', '232',\n",
       "       '311', '175', 'N8', 'N17', 'N13', 'N14', 'N4', 'N16', 'N15', 'N2',\n",
       "       'N11', 'N5', 'N20', 'N10', 'N1', 'N3', 'N21', 'N6', 'N22', 'N19',\n",
       "       'N18', 'N7', 'N9', '14_Mt. Sinai-Cunningham-Rundles-P01',\n",
       "       '2_Mt. Sinai-Cunningham-Rundles-P01'], dtype=object)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Not sure: What are the last two?\n",
    "df[\"sample_name\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>155</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>213</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>374</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>320</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>228</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>130</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>345</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>131</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>211</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>289</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>279</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>335</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>334</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>386</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>317</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>378</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>441</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>172</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>139</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>110</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>299</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>203</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>462</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>329</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>272</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>102</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Blood, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>N12</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>400</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>251</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>218</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>268</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>161</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>366</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>248</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>263</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>232</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>311</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>175</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>N8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>N17</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>N13</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>N14</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>N4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>N16</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>N15</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>N2</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>N11</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>N5</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>N20</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>N10</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>N1</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>N3</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>N21</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>N6</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>N22</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>N19</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>N18</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>N7</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>N9</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name locus            sample_tags counting_method       primer_set  \\\n",
       "0          155  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "1          213  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "2          374  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "3          320  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "4          228  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "5          130  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "6          345  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "7          131  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "8          211  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "9          289  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "10         279  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "11         335  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "12         334  TCRB  HLA MHC Class I, PBMC              v1  Human-TCRB-none   \n",
       "13         386  TCRB  HLA MHC Class I, PBMC              v1  Human-TCRB-none   \n",
       "14         317  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "15         378  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "16         441  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "17         172  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "18         139  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "19         110  TCRB  HLA MHC Class I, PBMC              v1  Human-TCRB-none   \n",
       "20         299  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "21         203  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "22         462  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "23         329  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "24         272  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "25         102  TCRB            Blood, PBMC              v1  Human-TCRB-none   \n",
       "26         N12  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "27         400  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "28         251  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "29         218  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "30         268  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "31         161  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "32         366  TCRB  HLA MHC Class I, PBMC              v1  Human-TCRB-none   \n",
       "33         248  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "34         263  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "35         232  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "36         311  TCRB  HLA MHC Class I, PBMC              v1  Human-TCRB-none   \n",
       "37         175  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "38          N8  TCRB  HLA MHC Class I, PBMC              v1  Human-TCRB-none   \n",
       "39         N17  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "40         N13  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "41         N14  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "42          N4  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "43         N16  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "44         N15  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "45          N2  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "46         N11  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "47          N5  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "48         N20  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "49         N10  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "50          N1  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "51          N3  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "52         N21  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "53          N6  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "54         N22  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "55         N19  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "56         N18  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "57          N7  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "58          N9  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "\n",
       "   species           study sequencing_type  \n",
       "0    Human  ramesh-2015-ci            gDNA  \n",
       "1    Human  ramesh-2015-ci            gDNA  \n",
       "2    Human  ramesh-2015-ci            gDNA  \n",
       "3    Human  ramesh-2015-ci            gDNA  \n",
       "4    Human  ramesh-2015-ci            gDNA  \n",
       "5    Human  ramesh-2015-ci            gDNA  \n",
       "6    Human  ramesh-2015-ci            gDNA  \n",
       "7    Human  ramesh-2015-ci            gDNA  \n",
       "8    Human  ramesh-2015-ci            gDNA  \n",
       "9    Human  ramesh-2015-ci            gDNA  \n",
       "10   Human  ramesh-2015-ci            gDNA  \n",
       "11   Human  ramesh-2015-ci            gDNA  \n",
       "12   Human  ramesh-2015-ci            gDNA  \n",
       "13   Human  ramesh-2015-ci            gDNA  \n",
       "14   Human  ramesh-2015-ci            gDNA  \n",
       "15   Human  ramesh-2015-ci            gDNA  \n",
       "16   Human  ramesh-2015-ci            gDNA  \n",
       "17   Human  ramesh-2015-ci            gDNA  \n",
       "18   Human  ramesh-2015-ci            gDNA  \n",
       "19   Human  ramesh-2015-ci            gDNA  \n",
       "20   Human  ramesh-2015-ci            gDNA  \n",
       "21   Human  ramesh-2015-ci            gDNA  \n",
       "22   Human  ramesh-2015-ci            gDNA  \n",
       "23   Human  ramesh-2015-ci            gDNA  \n",
       "24   Human  ramesh-2015-ci            gDNA  \n",
       "25   Human  ramesh-2015-ci            gDNA  \n",
       "26   Human  ramesh-2015-ci            gDNA  \n",
       "27   Human  ramesh-2015-ci            gDNA  \n",
       "28   Human  ramesh-2015-ci            gDNA  \n",
       "29   Human  ramesh-2015-ci            gDNA  \n",
       "30   Human  ramesh-2015-ci            gDNA  \n",
       "31   Human  ramesh-2015-ci            gDNA  \n",
       "32   Human  ramesh-2015-ci            gDNA  \n",
       "33   Human  ramesh-2015-ci            gDNA  \n",
       "34   Human  ramesh-2015-ci            gDNA  \n",
       "35   Human  ramesh-2015-ci            gDNA  \n",
       "36   Human  ramesh-2015-ci            gDNA  \n",
       "37   Human  ramesh-2015-ci            gDNA  \n",
       "38   Human  ramesh-2015-ci            gDNA  \n",
       "39   Human  ramesh-2015-ci            gDNA  \n",
       "40   Human  ramesh-2015-ci            gDNA  \n",
       "41   Human  ramesh-2015-ci            gDNA  \n",
       "42   Human  ramesh-2015-ci            gDNA  \n",
       "43   Human  ramesh-2015-ci            gDNA  \n",
       "44   Human  ramesh-2015-ci            gDNA  \n",
       "45   Human  ramesh-2015-ci            gDNA  \n",
       "46   Human  ramesh-2015-ci            gDNA  \n",
       "47   Human  ramesh-2015-ci            gDNA  \n",
       "48   Human  ramesh-2015-ci            gDNA  \n",
       "49   Human  ramesh-2015-ci            gDNA  \n",
       "50   Human  ramesh-2015-ci            gDNA  \n",
       "51   Human  ramesh-2015-ci            gDNA  \n",
       "52   Human  ramesh-2015-ci            gDNA  \n",
       "53   Human  ramesh-2015-ci            gDNA  \n",
       "54   Human  ramesh-2015-ci            gDNA  \n",
       "55   Human  ramesh-2015-ci            gDNA  \n",
       "56   Human  ramesh-2015-ci            gDNA  \n",
       "57   Human  ramesh-2015-ci            gDNA  \n",
       "58   Human  ramesh-2015-ci            gDNA  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop those to be safe.\n",
    "df = df[~df[\"sample_name\"].str.contains(\"Rundles\")].copy()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"specimen_label\"] = df[\"sample_name\"]\n",
    "df[\"participant_label\"] = df[\"sample_name\"]\n",
    "df[\"disease\"] = (\n",
    "    df[\"sample_name\"].str.startswith(\"N\").map({True: healthy_label, False: \"CVID\"})\n",
    ")\n",
    "df[\"disease_subtype\"] = df[\"disease\"].replace(\n",
    "    {healthy_label: healthy_label + \" - CVID negative\"}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>155</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>155</td>\n",
       "      <td>155</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>213</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>213</td>\n",
       "      <td>213</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>374</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>374</td>\n",
       "      <td>374</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>320</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>320</td>\n",
       "      <td>320</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>228</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>228</td>\n",
       "      <td>228</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>130</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>130</td>\n",
       "      <td>130</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>345</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>345</td>\n",
       "      <td>345</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>131</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>131</td>\n",
       "      <td>131</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>211</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>211</td>\n",
       "      <td>211</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>289</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>289</td>\n",
       "      <td>289</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>279</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>279</td>\n",
       "      <td>279</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>335</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>335</td>\n",
       "      <td>335</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>334</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>334</td>\n",
       "      <td>334</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>386</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>386</td>\n",
       "      <td>386</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>317</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>317</td>\n",
       "      <td>317</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>378</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>378</td>\n",
       "      <td>378</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>441</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>441</td>\n",
       "      <td>441</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>172</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>172</td>\n",
       "      <td>172</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>139</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>139</td>\n",
       "      <td>139</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>110</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>299</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>299</td>\n",
       "      <td>299</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>203</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>203</td>\n",
       "      <td>203</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>462</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>462</td>\n",
       "      <td>462</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>329</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>329</td>\n",
       "      <td>329</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>272</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>272</td>\n",
       "      <td>272</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>102</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Blood, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>N12</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N12</td>\n",
       "      <td>N12</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>400</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>400</td>\n",
       "      <td>400</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>251</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>251</td>\n",
       "      <td>251</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>218</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>218</td>\n",
       "      <td>218</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>268</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>268</td>\n",
       "      <td>268</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>161</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>161</td>\n",
       "      <td>161</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>366</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>366</td>\n",
       "      <td>366</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>248</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>248</td>\n",
       "      <td>248</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>263</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>263</td>\n",
       "      <td>263</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>232</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>232</td>\n",
       "      <td>232</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>311</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>311</td>\n",
       "      <td>311</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>175</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>175</td>\n",
       "      <td>175</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>N8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>HLA MHC Class I, PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N8</td>\n",
       "      <td>N8</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>N17</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N17</td>\n",
       "      <td>N17</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>N13</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N13</td>\n",
       "      <td>N13</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>N14</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N14</td>\n",
       "      <td>N14</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>N4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N4</td>\n",
       "      <td>N4</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>N16</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N16</td>\n",
       "      <td>N16</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>N15</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N15</td>\n",
       "      <td>N15</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>N2</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N2</td>\n",
       "      <td>N2</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>N11</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N11</td>\n",
       "      <td>N11</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>N5</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N5</td>\n",
       "      <td>N5</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>N20</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N20</td>\n",
       "      <td>N20</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>N10</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N10</td>\n",
       "      <td>N10</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>N1</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N1</td>\n",
       "      <td>N1</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>N3</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N3</td>\n",
       "      <td>N3</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>N21</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N21</td>\n",
       "      <td>N21</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>N6</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N6</td>\n",
       "      <td>N6</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>N22</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N22</td>\n",
       "      <td>N22</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>N19</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N19</td>\n",
       "      <td>N19</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>N18</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N18</td>\n",
       "      <td>N18</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>N7</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N7</td>\n",
       "      <td>N7</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>N9</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>v1</td>\n",
       "      <td>Human-TCRB-none</td>\n",
       "      <td>Human</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>N9</td>\n",
       "      <td>N9</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - CVID negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name locus            sample_tags counting_method       primer_set  \\\n",
       "0          155  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "1          213  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "2          374  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "3          320  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "4          228  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "5          130  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "6          345  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "7          131  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "8          211  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "9          289  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "10         279  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "11         335  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "12         334  TCRB  HLA MHC Class I, PBMC              v1  Human-TCRB-none   \n",
       "13         386  TCRB  HLA MHC Class I, PBMC              v1  Human-TCRB-none   \n",
       "14         317  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "15         378  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "16         441  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "17         172  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "18         139  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "19         110  TCRB  HLA MHC Class I, PBMC              v1  Human-TCRB-none   \n",
       "20         299  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "21         203  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "22         462  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "23         329  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "24         272  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "25         102  TCRB            Blood, PBMC              v1  Human-TCRB-none   \n",
       "26         N12  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "27         400  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "28         251  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "29         218  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "30         268  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "31         161  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "32         366  TCRB  HLA MHC Class I, PBMC              v1  Human-TCRB-none   \n",
       "33         248  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "34         263  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "35         232  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "36         311  TCRB  HLA MHC Class I, PBMC              v1  Human-TCRB-none   \n",
       "37         175  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "38          N8  TCRB  HLA MHC Class I, PBMC              v1  Human-TCRB-none   \n",
       "39         N17  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "40         N13  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "41         N14  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "42          N4  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "43         N16  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "44         N15  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "45          N2  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "46         N11  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "47          N5  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "48         N20  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "49         N10  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "50          N1  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "51          N3  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "52         N21  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "53          N6  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "54         N22  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "55         N19  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "56         N18  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "57          N7  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "58          N9  TCRB                   PBMC              v1  Human-TCRB-none   \n",
       "\n",
       "   species           study sequencing_type specimen_label participant_label  \\\n",
       "0    Human  ramesh-2015-ci            gDNA            155               155   \n",
       "1    Human  ramesh-2015-ci            gDNA            213               213   \n",
       "2    Human  ramesh-2015-ci            gDNA            374               374   \n",
       "3    Human  ramesh-2015-ci            gDNA            320               320   \n",
       "4    Human  ramesh-2015-ci            gDNA            228               228   \n",
       "5    Human  ramesh-2015-ci            gDNA            130               130   \n",
       "6    Human  ramesh-2015-ci            gDNA            345               345   \n",
       "7    Human  ramesh-2015-ci            gDNA            131               131   \n",
       "8    Human  ramesh-2015-ci            gDNA            211               211   \n",
       "9    Human  ramesh-2015-ci            gDNA            289               289   \n",
       "10   Human  ramesh-2015-ci            gDNA            279               279   \n",
       "11   Human  ramesh-2015-ci            gDNA            335               335   \n",
       "12   Human  ramesh-2015-ci            gDNA            334               334   \n",
       "13   Human  ramesh-2015-ci            gDNA            386               386   \n",
       "14   Human  ramesh-2015-ci            gDNA            317               317   \n",
       "15   Human  ramesh-2015-ci            gDNA            378               378   \n",
       "16   Human  ramesh-2015-ci            gDNA            441               441   \n",
       "17   Human  ramesh-2015-ci            gDNA            172               172   \n",
       "18   Human  ramesh-2015-ci            gDNA            139               139   \n",
       "19   Human  ramesh-2015-ci            gDNA            110               110   \n",
       "20   Human  ramesh-2015-ci            gDNA            299               299   \n",
       "21   Human  ramesh-2015-ci            gDNA            203               203   \n",
       "22   Human  ramesh-2015-ci            gDNA            462               462   \n",
       "23   Human  ramesh-2015-ci            gDNA            329               329   \n",
       "24   Human  ramesh-2015-ci            gDNA            272               272   \n",
       "25   Human  ramesh-2015-ci            gDNA            102               102   \n",
       "26   Human  ramesh-2015-ci            gDNA            N12               N12   \n",
       "27   Human  ramesh-2015-ci            gDNA            400               400   \n",
       "28   Human  ramesh-2015-ci            gDNA            251               251   \n",
       "29   Human  ramesh-2015-ci            gDNA            218               218   \n",
       "30   Human  ramesh-2015-ci            gDNA            268               268   \n",
       "31   Human  ramesh-2015-ci            gDNA            161               161   \n",
       "32   Human  ramesh-2015-ci            gDNA            366               366   \n",
       "33   Human  ramesh-2015-ci            gDNA            248               248   \n",
       "34   Human  ramesh-2015-ci            gDNA            263               263   \n",
       "35   Human  ramesh-2015-ci            gDNA            232               232   \n",
       "36   Human  ramesh-2015-ci            gDNA            311               311   \n",
       "37   Human  ramesh-2015-ci            gDNA            175               175   \n",
       "38   Human  ramesh-2015-ci            gDNA             N8                N8   \n",
       "39   Human  ramesh-2015-ci            gDNA            N17               N17   \n",
       "40   Human  ramesh-2015-ci            gDNA            N13               N13   \n",
       "41   Human  ramesh-2015-ci            gDNA            N14               N14   \n",
       "42   Human  ramesh-2015-ci            gDNA             N4                N4   \n",
       "43   Human  ramesh-2015-ci            gDNA            N16               N16   \n",
       "44   Human  ramesh-2015-ci            gDNA            N15               N15   \n",
       "45   Human  ramesh-2015-ci            gDNA             N2                N2   \n",
       "46   Human  ramesh-2015-ci            gDNA            N11               N11   \n",
       "47   Human  ramesh-2015-ci            gDNA             N5                N5   \n",
       "48   Human  ramesh-2015-ci            gDNA            N20               N20   \n",
       "49   Human  ramesh-2015-ci            gDNA            N10               N10   \n",
       "50   Human  ramesh-2015-ci            gDNA             N1                N1   \n",
       "51   Human  ramesh-2015-ci            gDNA             N3                N3   \n",
       "52   Human  ramesh-2015-ci            gDNA            N21               N21   \n",
       "53   Human  ramesh-2015-ci            gDNA             N6                N6   \n",
       "54   Human  ramesh-2015-ci            gDNA            N22               N22   \n",
       "55   Human  ramesh-2015-ci            gDNA            N19               N19   \n",
       "56   Human  ramesh-2015-ci            gDNA            N18               N18   \n",
       "57   Human  ramesh-2015-ci            gDNA             N7                N7   \n",
       "58   Human  ramesh-2015-ci            gDNA             N9                N9   \n",
       "\n",
       "               disease                     disease_subtype  \n",
       "0                 CVID                                CVID  \n",
       "1                 CVID                                CVID  \n",
       "2                 CVID                                CVID  \n",
       "3                 CVID                                CVID  \n",
       "4                 CVID                                CVID  \n",
       "5                 CVID                                CVID  \n",
       "6                 CVID                                CVID  \n",
       "7                 CVID                                CVID  \n",
       "8                 CVID                                CVID  \n",
       "9                 CVID                                CVID  \n",
       "10                CVID                                CVID  \n",
       "11                CVID                                CVID  \n",
       "12                CVID                                CVID  \n",
       "13                CVID                                CVID  \n",
       "14                CVID                                CVID  \n",
       "15                CVID                                CVID  \n",
       "16                CVID                                CVID  \n",
       "17                CVID                                CVID  \n",
       "18                CVID                                CVID  \n",
       "19                CVID                                CVID  \n",
       "20                CVID                                CVID  \n",
       "21                CVID                                CVID  \n",
       "22                CVID                                CVID  \n",
       "23                CVID                                CVID  \n",
       "24                CVID                                CVID  \n",
       "25                CVID                                CVID  \n",
       "26  Healthy/Background  Healthy/Background - CVID negative  \n",
       "27                CVID                                CVID  \n",
       "28                CVID                                CVID  \n",
       "29                CVID                                CVID  \n",
       "30                CVID                                CVID  \n",
       "31                CVID                                CVID  \n",
       "32                CVID                                CVID  \n",
       "33                CVID                                CVID  \n",
       "34                CVID                                CVID  \n",
       "35                CVID                                CVID  \n",
       "36                CVID                                CVID  \n",
       "37                CVID                                CVID  \n",
       "38  Healthy/Background  Healthy/Background - CVID negative  \n",
       "39  Healthy/Background  Healthy/Background - CVID negative  \n",
       "40  Healthy/Background  Healthy/Background - CVID negative  \n",
       "41  Healthy/Background  Healthy/Background - CVID negative  \n",
       "42  Healthy/Background  Healthy/Background - CVID negative  \n",
       "43  Healthy/Background  Healthy/Background - CVID negative  \n",
       "44  Healthy/Background  Healthy/Background - CVID negative  \n",
       "45  Healthy/Background  Healthy/Background - CVID negative  \n",
       "46  Healthy/Background  Healthy/Background - CVID negative  \n",
       "47  Healthy/Background  Healthy/Background - CVID negative  \n",
       "48  Healthy/Background  Healthy/Background - CVID negative  \n",
       "49  Healthy/Background  Healthy/Background - CVID negative  \n",
       "50  Healthy/Background  Healthy/Background - CVID negative  \n",
       "51  Healthy/Background  Healthy/Background - CVID negative  \n",
       "52  Healthy/Background  Healthy/Background - CVID negative  \n",
       "53  Healthy/Background  Healthy/Background - CVID negative  \n",
       "54  Healthy/Background  Healthy/Background - CVID negative  \n",
       "55  Healthy/Background  Healthy/Background - CVID negative  \n",
       "56  Healthy/Background  Healthy/Background - CVID negative  \n",
       "57  Healthy/Background  Healthy/Background - CVID negative  \n",
       "58  Healthy/Background  Healthy/Background - CVID negative  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual.append(df)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Subject_76</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>67 Years, Blood, Cohort 3, Healthy, Hispanic, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Subject_57</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>60 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Subject_53</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>40 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Subject_104</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>29 Years, Black or African American, Blood, Co...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Subject_146</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>61 Years, Caucasian, Head and Neck Cancer, Mal...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>Subject_141</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>60 Years, Caucasian, Colorectal Cancer, Male, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>Subject_49</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>72 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>Subject_58</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>40 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>Subject_111</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>81 Years, Ex-smoker, Lung Cancer, Male, Native...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>Subject_30</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>66 Years, Black or African American, Colorecta...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>158 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_name locus                                        sample_tags  \\\n",
       "0     Subject_76  TCRB  67 Years, Blood, Cohort 3, Healthy, Hispanic, ...   \n",
       "1     Subject_57  TCRB  60 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "2     Subject_53  TCRB  40 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "3    Subject_104  TCRB  29 Years, Black or African American, Blood, Co...   \n",
       "4    Subject_146  TCRB  61 Years, Caucasian, Head and Neck Cancer, Mal...   \n",
       "..           ...   ...                                                ...   \n",
       "153  Subject_141  TCRB  60 Years, Caucasian, Colorectal Cancer, Male, ...   \n",
       "154   Subject_49  TCRB  72 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "155   Subject_58  TCRB  40 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "156  Subject_111  TCRB  81 Years, Ex-smoker, Lung Cancer, Male, Native...   \n",
       "157   Subject_30  TCRB  66 Years, Black or African American, Colorecta...   \n",
       "\n",
       "    counting_method        primer_set species           study sequencing_type  \n",
       "0                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "1                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "2                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "3                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "4                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "..              ...               ...     ...             ...             ...  \n",
       "153              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "154              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "155              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "156              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "157              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "\n",
       "[158 rows x 8 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = _load_metadata_df(\"TCRBv4-control\")\n",
    "\n",
    "\"\"\"\n",
    "colorectal cancer, lung cancer, head and neck cancer, and healthy control\n",
    "58 cancer, 88 hhc, pre-covid\n",
    "ignore the 11 FFPE samples\n",
    "add the HHCs, don't do cancer for now\n",
    "\"\"\"\n",
    "\n",
    "df[\"sequencing_type\"] = \"gDNA\"\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Male                                94\n",
       "Blood                               88\n",
       "Cohort 3                            88\n",
       "Healthy                             88\n",
       "Inactive                            88\n",
       "Caucasian                           70\n",
       "Female                              64\n",
       "PBMC                                59\n",
       "Non-smoker                          53\n",
       "Black or African American           43\n",
       "Smoker                              34\n",
       "Ex-smoker                           27\n",
       "Never Smoked                        25\n",
       "Hispanic                            22\n",
       "Lung Cancer                         22\n",
       "Colorectal Cancer                   20\n",
       "Head and Neck Cancer                20\n",
       "Unknown Smoking Status              17\n",
       "Asian or Pacific Islander           16\n",
       "FFPE                                11\n",
       "Melanoma                             8\n",
       "66 Years                             8\n",
       "30 Years                             7\n",
       "57 Years                             7\n",
       "51 Years                             6\n",
       "63 Years                             6\n",
       "Native American or Alaska Native     6\n",
       "32 Years                             5\n",
       "31 Years                             5\n",
       "55 Years                             5\n",
       "73 Years                             5\n",
       "48 Years                             4\n",
       "40 Years                             4\n",
       "68 Years                             4\n",
       "59 Years                             4\n",
       "67 Years                             3\n",
       "50 Years                             3\n",
       "34 Years                             3\n",
       "21 Years                             3\n",
       "49 Years                             3\n",
       "75 Years                             3\n",
       "54 Years                             3\n",
       "29 Years                             3\n",
       "61 Years                             3\n",
       "38 Years                             3\n",
       "24 Years                             3\n",
       "20 Years                             3\n",
       "36 Years                             3\n",
       "43 Years                             3\n",
       "60 Years                             3\n",
       "Name: sample_tags, dtype: int64"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"sample_tags\"].str.split(\",\").explode().str.strip().value_counts().head(n=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Subject_3</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Caucasian, FFPE, Male, Melanoma, Unknown Smoki...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Subject_6</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Caucasian, FFPE, Male, Melanoma, Never Smoked</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Subject_4</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Caucasian, FFPE, Male, Melanoma, Unknown Smoki...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>Subject_7</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Caucasian, FFPE, Male, Melanoma, Unknown Smoki...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Subject_5</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Caucasian, FFPE, Male, Melanoma, Unknown Smoki...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Subject_2</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Caucasian, FFPE, Female, Melanoma, Unknown Smo...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>Subject_8</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Caucasian, FFPE, Male, Melanoma, Unknown Smoki...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>Subject_9</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Caucasian, FFPE, Female, Melanoma, Unknown Smo...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_name locus                                        sample_tags  \\\n",
       "39    Subject_3  TCRB  Caucasian, FFPE, Male, Melanoma, Unknown Smoki...   \n",
       "40    Subject_6  TCRB      Caucasian, FFPE, Male, Melanoma, Never Smoked   \n",
       "48    Subject_4  TCRB  Caucasian, FFPE, Male, Melanoma, Unknown Smoki...   \n",
       "66    Subject_7  TCRB  Caucasian, FFPE, Male, Melanoma, Unknown Smoki...   \n",
       "76    Subject_5  TCRB  Caucasian, FFPE, Male, Melanoma, Unknown Smoki...   \n",
       "83    Subject_2  TCRB  Caucasian, FFPE, Female, Melanoma, Unknown Smo...   \n",
       "89    Subject_8  TCRB  Caucasian, FFPE, Male, Melanoma, Unknown Smoki...   \n",
       "112   Subject_9  TCRB  Caucasian, FFPE, Female, Melanoma, Unknown Smo...   \n",
       "\n",
       "    counting_method        primer_set species           study sequencing_type  \n",
       "39               v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "40               v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "48               v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "66               v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "76               v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "83               v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "89               v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "112              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# almost always, a row's sample tag will contain EITHER cancer or healthy\n",
    "# except for these FFPE rows\n",
    "df[\n",
    "    df[\"sample_tags\"].str.contains(\"Cancer\")\n",
    "    != ~df[\"sample_tags\"].str.contains(\"Healthy\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove FFPE\n",
    "df = df[~df[\"sample_tags\"].str.contains(\"FFPE\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# All remaining rows either marked Blood or PBMC (or both)\n",
    "assert (\n",
    "    df[\"sample_tags\"].str.contains(\"Blood\") | df[\"sample_tags\"].str.contains(\"PBMC\")\n",
    ").all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now that we filtered to PBMC: a row's sample tag will contain EITHER cancer or healthy, NEVER both\n",
    "assert (\n",
    "    df[\"sample_tags\"].str.contains(\"Cancer\")\n",
    "    != df[\"sample_tags\"].str.contains(\"Healthy\")\n",
    ").all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Subject_76</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>67 Years, Blood, Cohort 3, Healthy, Hispanic, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Subject_57</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>60 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Subject_53</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>40 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Subject_104</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>29 Years, Black or African American, Blood, Co...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Subject_152</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>38 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Subject_63</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>57 Years, Blood, Cohort 3, Healthy, Hispanic, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Subject_18</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>38 Years, Asian or Pacific Islander, Blood, Co...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Subject_60</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>57 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>Subject_49</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>72 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>Subject_58</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>40 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_name locus                                        sample_tags  \\\n",
       "0     Subject_76  TCRB  67 Years, Blood, Cohort 3, Healthy, Hispanic, ...   \n",
       "1     Subject_57  TCRB  60 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "2     Subject_53  TCRB  40 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "3    Subject_104  TCRB  29 Years, Black or African American, Blood, Co...   \n",
       "5    Subject_152  TCRB  38 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "..           ...   ...                                                ...   \n",
       "149   Subject_63  TCRB  57 Years, Blood, Cohort 3, Healthy, Hispanic, ...   \n",
       "150   Subject_18  TCRB  38 Years, Asian or Pacific Islander, Blood, Co...   \n",
       "151   Subject_60  TCRB  57 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "154   Subject_49  TCRB  72 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "155   Subject_58  TCRB  40 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "\n",
       "    counting_method        primer_set species           study sequencing_type  \n",
       "0                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "1                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "2                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "3                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "5                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "..              ...               ...     ...             ...             ...  \n",
       "149              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "150              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "151              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "154              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "155              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA  \n",
       "\n",
       "[88 rows x 8 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter to healthy only. remove cancer\n",
    "df = df[df[\"sample_tags\"].str.contains(\"Healthy\")]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Pull out age, sex, ancestry\n",
    "# df['sample_tags'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"disease\"] = healthy_label\n",
    "df[\"disease_subtype\"] = f\"{healthy_label} - TCRBv4-control\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Subject_76', 'Subject_57', 'Subject_53', 'Subject_104',\n",
       "       'Subject_152', 'Subject_158', 'Subject_73', 'Subject_98',\n",
       "       'Subject_151', 'Subject_92', 'Subject_52', 'Subject_75',\n",
       "       'Subject_85', 'Subject_56', 'Subject_54', 'Subject_15',\n",
       "       'Subject_82', 'Subject_97', 'Subject_87', 'Subject_65',\n",
       "       'Subject_48', 'Subject_41', 'Subject_26', 'Subject_16',\n",
       "       'Subject_68', 'Subject_86', 'Subject_102', 'Subject_47',\n",
       "       'Subject_19', 'Subject_39', 'Subject_72', 'Subject_70',\n",
       "       'Subject_78', 'Subject_74', 'Subject_69', 'Subject_99',\n",
       "       'Subject_44', 'Subject_12', 'Subject_66', 'Subject_88',\n",
       "       'Subject_46', 'Subject_50', 'Subject_91', 'Subject_93',\n",
       "       'Subject_27', 'Subject_17', 'Subject_40', 'Subject_90',\n",
       "       'Subject_103', 'Subject_101', 'Subject_80', 'Subject_43',\n",
       "       'Subject_96', 'Subject_59', 'Subject_71', 'Subject_14',\n",
       "       'Subject_81', 'Subject_64', 'Subject_94', 'Subject_55',\n",
       "       'Subject_155', 'Subject_61', 'Subject_42', 'Subject_13',\n",
       "       'Subject_95', 'Subject_83', 'Subject_157', 'Subject_105',\n",
       "       'Subject_154', 'Subject_100', 'Subject_153', 'Subject_67',\n",
       "       'Subject_79', 'Subject_84', 'Subject_156', 'Subject_106',\n",
       "       'Subject_77', 'Subject_89', 'Subject_51', 'Subject_24',\n",
       "       'Subject_45', 'Subject_62', 'Subject_22', 'Subject_63',\n",
       "       'Subject_18', 'Subject_60', 'Subject_49', 'Subject_58'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"participant_label\"] = df[\"sample_name\"]\n",
    "df[\"participant_label\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"specimen_label\"] = df[\"sample_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>specimen_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Subject_76</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>67 Years, Blood, Cohort 3, Healthy, Hispanic, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - TCRBv4-control</td>\n",
       "      <td>Subject_76</td>\n",
       "      <td>Subject_76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Subject_57</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>60 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - TCRBv4-control</td>\n",
       "      <td>Subject_57</td>\n",
       "      <td>Subject_57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Subject_53</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>40 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - TCRBv4-control</td>\n",
       "      <td>Subject_53</td>\n",
       "      <td>Subject_53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Subject_104</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>29 Years, Black or African American, Blood, Co...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - TCRBv4-control</td>\n",
       "      <td>Subject_104</td>\n",
       "      <td>Subject_104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Subject_152</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>38 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - TCRBv4-control</td>\n",
       "      <td>Subject_152</td>\n",
       "      <td>Subject_152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Subject_63</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>57 Years, Blood, Cohort 3, Healthy, Hispanic, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - TCRBv4-control</td>\n",
       "      <td>Subject_63</td>\n",
       "      <td>Subject_63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Subject_18</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>38 Years, Asian or Pacific Islander, Blood, Co...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - TCRBv4-control</td>\n",
       "      <td>Subject_18</td>\n",
       "      <td>Subject_18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Subject_60</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>57 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - TCRBv4-control</td>\n",
       "      <td>Subject_60</td>\n",
       "      <td>Subject_60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>Subject_49</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>72 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - TCRBv4-control</td>\n",
       "      <td>Subject_49</td>\n",
       "      <td>Subject_49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>Subject_58</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>40 Years, Blood, Caucasian, Cohort 3, Female, ...</td>\n",
       "      <td>v4</td>\n",
       "      <td>Human-TCRB-PD4bx</td>\n",
       "      <td>Human</td>\n",
       "      <td>TCRBv4-control</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background - TCRBv4-control</td>\n",
       "      <td>Subject_58</td>\n",
       "      <td>Subject_58</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_name locus                                        sample_tags  \\\n",
       "0     Subject_76  TCRB  67 Years, Blood, Cohort 3, Healthy, Hispanic, ...   \n",
       "1     Subject_57  TCRB  60 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "2     Subject_53  TCRB  40 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "3    Subject_104  TCRB  29 Years, Black or African American, Blood, Co...   \n",
       "5    Subject_152  TCRB  38 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "..           ...   ...                                                ...   \n",
       "149   Subject_63  TCRB  57 Years, Blood, Cohort 3, Healthy, Hispanic, ...   \n",
       "150   Subject_18  TCRB  38 Years, Asian or Pacific Islander, Blood, Co...   \n",
       "151   Subject_60  TCRB  57 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "154   Subject_49  TCRB  72 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "155   Subject_58  TCRB  40 Years, Blood, Caucasian, Cohort 3, Female, ...   \n",
       "\n",
       "    counting_method        primer_set species           study sequencing_type  \\\n",
       "0                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA   \n",
       "1                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA   \n",
       "2                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA   \n",
       "3                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA   \n",
       "5                v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA   \n",
       "..              ...               ...     ...             ...             ...   \n",
       "149              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA   \n",
       "150              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA   \n",
       "151              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA   \n",
       "154              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA   \n",
       "155              v4  Human-TCRB-PD4bx   Human  TCRBv4-control            gDNA   \n",
       "\n",
       "                disease                      disease_subtype  \\\n",
       "0    Healthy/Background  Healthy/Background - TCRBv4-control   \n",
       "1    Healthy/Background  Healthy/Background - TCRBv4-control   \n",
       "2    Healthy/Background  Healthy/Background - TCRBv4-control   \n",
       "3    Healthy/Background  Healthy/Background - TCRBv4-control   \n",
       "5    Healthy/Background  Healthy/Background - TCRBv4-control   \n",
       "..                  ...                                  ...   \n",
       "149  Healthy/Background  Healthy/Background - TCRBv4-control   \n",
       "150  Healthy/Background  Healthy/Background - TCRBv4-control   \n",
       "151  Healthy/Background  Healthy/Background - TCRBv4-control   \n",
       "154  Healthy/Background  Healthy/Background - TCRBv4-control   \n",
       "155  Healthy/Background  Healthy/Background - TCRBv4-control   \n",
       "\n",
       "    participant_label specimen_label  \n",
       "0          Subject_76     Subject_76  \n",
       "1          Subject_57     Subject_57  \n",
       "2          Subject_53     Subject_53  \n",
       "3         Subject_104    Subject_104  \n",
       "5         Subject_152    Subject_152  \n",
       "..                ...            ...  \n",
       "149        Subject_63     Subject_63  \n",
       "150        Subject_18     Subject_18  \n",
       "151        Subject_60     Subject_60  \n",
       "154        Subject_49     Subject_49  \n",
       "155        Subject_58     Subject_58  \n",
       "\n",
       "[88 rows x 12 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual.append(df)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>015V15001607_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>4.906849315 years_rel_to_art, 40.48219178 age_...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>015V15000803_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>22.38630137 age_at_min_evidence, 25.95890411 a...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>015V15009714_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>39.70491803 age_at_min_evidence, 43.71038251 a...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>015V15000746_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>40.1420765 age_at_min_evidence, 42.09041096 ag...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>015V16003146_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>40.60547945 age_at_min_evidence, 45.6147541 ag...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>015V09001205_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.539726027 years_rel_to_art, 41.89589041 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>015V09003735_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>0.073972603 years_rel_to_art, 41.89589041 age_...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>015V10006273_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>1.2 years_rel_to_art, 41.89589041 age_at_min_e...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>015V12002996_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>2.950819672 years_rel_to_art, 41.89589041 age_...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>015V14000020_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>4.224657534 years_rel_to_art, 41.89589041 age_...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>192 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_name locus  \\\n",
       "0    015V15001607_CFAR  TCRB   \n",
       "1    015V15000803_CFAR  TCRB   \n",
       "2    015V15009714_CFAR  TCRB   \n",
       "3    015V15000746_CFAR  TCRB   \n",
       "4    015V16003146_CFAR  TCRB   \n",
       "..                 ...   ...   \n",
       "187  015V09001205_CFAR  TCRB   \n",
       "188  015V09003735_CFAR  TCRB   \n",
       "189  015V10006273_CFAR  TCRB   \n",
       "190  015V12002996_CFAR  TCRB   \n",
       "191  015V14000020_CFAR  TCRB   \n",
       "\n",
       "                                           sample_tags counting_method  \\\n",
       "0    4.906849315 years_rel_to_art, 40.48219178 age_...              v3   \n",
       "1    22.38630137 age_at_min_evidence, 25.95890411 a...              v3   \n",
       "2    39.70491803 age_at_min_evidence, 43.71038251 a...              v3   \n",
       "3    40.1420765 age_at_min_evidence, 42.09041096 ag...              v3   \n",
       "4    40.60547945 age_at_min_evidence, 45.6147541 ag...              v3   \n",
       "..                                                 ...             ...   \n",
       "187  -0.539726027 years_rel_to_art, 41.89589041 age...              v3   \n",
       "188  0.073972603 years_rel_to_art, 41.89589041 age_...              v3   \n",
       "189  1.2 years_rel_to_art, 41.89589041 age_at_min_e...              v3   \n",
       "190  2.950819672 years_rel_to_art, 41.89589041 age_...              v3   \n",
       "191  4.224657534 years_rel_to_art, 41.89589041 age_...              v3   \n",
       "\n",
       "          primer_set species               study sequencing_type  \n",
       "0    Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA  \n",
       "1    Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA  \n",
       "2    Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA  \n",
       "3    Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA  \n",
       "4    Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA  \n",
       "..               ...     ...                 ...             ...  \n",
       "187  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA  \n",
       "188  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA  \n",
       "189  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA  \n",
       "190  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA  \n",
       "191  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA  \n",
       "\n",
       "[192 rows x 8 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = _load_metadata_df(\"towlerton-2022-hiv\")\n",
    "\n",
    "\"\"\"\n",
    "HIV on long term therapy, before and after\n",
    "30 adults, 192 samples?\n",
    "30 adults with HIV infection before and after the initiation of ART\n",
    "\n",
    "A total of 192 cryopreserved PBMC samples serially collected over a mean of 6 (range, 1-12) years from 30 adults with confirmed HIV infection (median, 7 samples per subject) were received from the University of Washington CFAR Biorepository.\n",
    "PBMC samples collected at 1-4 timepoints before and 2-8 time points after the initiation of ART were available from each subject.\n",
    "genomic DNA was extracted from unsorted PBMC\n",
    "\n",
    "The blood samples from PLHIV were obtained between 2004 and 2017: pre-pandemic\n",
    "\n",
    "Metadata:\n",
    "https://www.frontiersin.org/articles/10.3389/fimmu.2022.879190/full#supplementary-material\n",
    "https://github.com/shashidhar22/airr_seq_pipelines/blob/main/metadata/cfar/CFAR_Dean_metadata.xlsx\n",
    "(identical?)\n",
    "\"\"\"\n",
    "\n",
    "df[\"sequencing_type\"] = \"gDNA\"\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Positive HIV_status                 192\n",
       "NA age_at_death                     178\n",
       "DPA1*010301 DPA1allele1             156\n",
       "Post-ART timePoint                  131\n",
       "DPA1*010301 DPA1allele2             123\n",
       "Male                                117\n",
       "DPB1*040101 DPB1allele1              76\n",
       "Female                               75\n",
       "Pre-ART timePoint                    61\n",
       "DQA1*010201 DQA1allele1              61\n",
       "DRB3*010102 DRB345allele1            56\n",
       "DRB5*010101 DRB345allele2            54\n",
       "C*040101 cAllele1                    52\n",
       "A*020101 aAllele1                    52\n",
       "DPB1*040101 DPB1allele2              51\n",
       "DRB345*Not_Present DRB345allele1     49\n",
       "DRB4*010301 DRB345allele2            46\n",
       "DRB1*030101 DRB1allele1              45\n",
       "DQB1*020101 DQB1allele1              45\n",
       "DRB1*150101 DRB1allele2              45\n",
       "A*010101 aAllele1                    45\n",
       "DQA1*050101 DQA1allele2              45\n",
       "DQB1*060201 DQB1allele2              38\n",
       "DPA1*020101 DPA1allele2              37\n",
       "DPB1*030101 DPB1allele1              37\n",
       "DQB1*060301 DQB1allele2              36\n",
       "C*070101 cAllele2                    36\n",
       "B*440201 bAllele2                    35\n",
       "DRB3*020201 DRB345allele1            35\n",
       "B*070201 bAllele1                    34\n",
       "A*030101 aAllele2                    34\n",
       "DRB3*020201 DRB345allele2            32\n",
       "DQB1*030101 DQB1allele1              32\n",
       "DQA1*050501 DQA1allele2              32\n",
       "DRB1*130101 DRB1allele2              31\n",
       "DPB1*040201 DPB1allele2              31\n",
       "B*080101 bAllele1                    31\n",
       "C*030401 cAllele1                    29\n",
       "DQA1*030101 DQA1allele1              29\n",
       "B*530101 bAllele2                    27\n",
       "DRB3*030101 DRB345allele1            27\n",
       "C*070201 cAllele2                    27\n",
       "DQB1*050101 DQB1allele2              27\n",
       "<= 2 years Post-ART time_group       25\n",
       "DQA1*010201 DQA1allele2              24\n",
       "DPB1*020102 DPB1allele1              24\n",
       "C*160101 cAllele2                    24\n",
       "DQA1*010101 DQA1allele1              24\n",
       "<= 3 years Post-ART time_group       23\n",
       "DPB1*010101 DPB1allele1              23\n",
       "Name: sample_tags, dtype: int64"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"sample_tags\"].str.split(\",\").explode().str.strip().value_counts().head(n=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# each entry has a pre or post ART timepoint identified\n",
    "df[\"sample_tags\"].str.contains(\"ART\").all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32       <= 0 month Pre-ART time_group\n",
       "25       <= 1 year Post-ART time_group\n",
       "51        <= 1 year Pre-ART time_group\n",
       "12     <= 10 years Post-ART time_group\n",
       "28      <= 2 years Post-ART time_group\n",
       "62       <= 2 years Pre-ART time_group\n",
       "24     <= 3 months Post-ART time_group\n",
       "27      <= 3 months Pre-ART time_group\n",
       "34      <= 3 years Post-ART time_group\n",
       "139      <= 3 years Pre-ART time_group\n",
       "13      <= 4 years Post-ART time_group\n",
       "153      <= 4 years Pre-ART time_group\n",
       "0       <= 5 years Post-ART time_group\n",
       "31     <= 6 months Post-ART time_group\n",
       "50      <= 6 months Pre-ART time_group\n",
       "1       <= 6 years Post-ART time_group\n",
       "2       <= 7 years Post-ART time_group\n",
       "5       <= 8 years Post-ART time_group\n",
       "4       <= 9 years Post-ART time_group\n",
       "20      > 10 years Post-ART time_group\n",
       "0                   Post-ART timePoint\n",
       "27                   Pre-ART timePoint\n",
       "Name: sample_tags, dtype: object"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_tags = df[\"sample_tags\"].str.split(\",\").explode().str.strip().drop_duplicates()\n",
    "sample_tags[sample_tags.str.contains(\"ART\")].sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>patientID</th>\n",
       "      <th>years_rel_to_art</th>\n",
       "      <th>timePoint</th>\n",
       "      <th>time_group</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>015V07002801_CFAR</td>\n",
       "      <td>1001</td>\n",
       "      <td>-2.630137</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 2 years Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>015V08002889_CFAR</td>\n",
       "      <td>1001</td>\n",
       "      <td>-0.873973</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 6 months Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>015V09001205_CFAR</td>\n",
       "      <td>1001</td>\n",
       "      <td>-0.539726</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 6 months Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>015V09003735_CFAR</td>\n",
       "      <td>1001</td>\n",
       "      <td>0.073973</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>015V10006273_CFAR</td>\n",
       "      <td>1001</td>\n",
       "      <td>1.200000</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 2 years Post-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>015V15015350_CFAR</td>\n",
       "      <td>1029</td>\n",
       "      <td>4.387978</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 5 years Post-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>015V16002103_CFAR</td>\n",
       "      <td>1029</td>\n",
       "      <td>5.065753</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 5 years Post-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>015V13001607_CFAR</td>\n",
       "      <td>1030</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>015V13009411_CFAR</td>\n",
       "      <td>1030</td>\n",
       "      <td>0.567123</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 6 months Post-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>015V14001903_CFAR</td>\n",
       "      <td>1030</td>\n",
       "      <td>1.320548</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 2 years Post-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>192 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              sampleID  patientID  years_rel_to_art timePoint  \\\n",
       "0    015V07002801_CFAR       1001         -2.630137   Pre-ART   \n",
       "1    015V08002889_CFAR       1001         -0.873973   Pre-ART   \n",
       "2    015V09001205_CFAR       1001         -0.539726   Pre-ART   \n",
       "3    015V09003735_CFAR       1001          0.073973  Post-ART   \n",
       "4    015V10006273_CFAR       1001          1.200000  Post-ART   \n",
       "..                 ...        ...               ...       ...   \n",
       "187  015V15015350_CFAR       1029          4.387978  Post-ART   \n",
       "188  015V16002103_CFAR       1029          5.065753  Post-ART   \n",
       "189  015V13001607_CFAR       1030          0.000000   Pre-ART   \n",
       "190  015V13009411_CFAR       1030          0.567123  Post-ART   \n",
       "191  015V14001903_CFAR       1030          1.320548  Post-ART   \n",
       "\n",
       "               time_group sex  age_at_collection  \n",
       "0      <= 2 years Pre-ART   M                 45  \n",
       "1     <= 6 months Pre-ART   M                 46  \n",
       "2     <= 6 months Pre-ART   M                 47  \n",
       "3      <= 0 month Pre-ART   M                 47  \n",
       "4     <= 2 years Post-ART   M                 48  \n",
       "..                    ...  ..                ...  \n",
       "187   <= 5 years Post-ART   F                 54  \n",
       "188   <= 5 years Post-ART   F                 55  \n",
       "189    <= 0 month Pre-ART   F                 27  \n",
       "190  <= 6 months Post-ART   F                 27  \n",
       "191   <= 2 years Post-ART   F                 28  \n",
       "\n",
       "[192 rows x 7 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_m = pd.read_excel(\n",
    "    config.paths.metadata_dir\n",
    "    / \"adaptive\"\n",
    "    / \"towlerton-2022-hiv.CFAR_Dean_metadata.xlsx\",\n",
    "    sheet_name=\"CFAR_Sample_info\",\n",
    ")\n",
    "df_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(192, 8)"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>patientID</th>\n",
       "      <th>years_rel_to_art</th>\n",
       "      <th>timePoint</th>\n",
       "      <th>time_group</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>015V15001607_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>4.906849315 years_rel_to_art, 40.48219178 age_...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1016</td>\n",
       "      <td>4.906849</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 5 years Post-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>015V15000803_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>22.38630137 age_at_min_evidence, 25.95890411 a...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1024</td>\n",
       "      <td>5.312329</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 6 years Post-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>015V15009714_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>39.70491803 age_at_min_evidence, 43.71038251 a...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1009</td>\n",
       "      <td>6.917808</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 7 years Post-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>015V15000746_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>40.1420765 age_at_min_evidence, 42.09041096 ag...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1023</td>\n",
       "      <td>5.150685</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 6 years Post-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>015V16003146_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>40.60547945 age_at_min_evidence, 45.6147541 ag...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1007</td>\n",
       "      <td>8.180822</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 9 years Post-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>015V09001205_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.539726027 years_rel_to_art, 41.89589041 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>-0.539726</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 6 months Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>015V09003735_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>0.073972603 years_rel_to_art, 41.89589041 age_...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>0.073973</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>015V10006273_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>1.2 years_rel_to_art, 41.89589041 age_at_min_e...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>1.200000</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 2 years Post-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>015V12002996_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>2.950819672 years_rel_to_art, 41.89589041 age_...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>2.950820</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 3 years Post-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>015V14000020_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>4.224657534 years_rel_to_art, 41.89589041 age_...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>4.224658</td>\n",
       "      <td>Post-ART</td>\n",
       "      <td>&lt;= 5 years Post-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>192 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_name locus  \\\n",
       "0    015V15001607_CFAR  TCRB   \n",
       "1    015V15000803_CFAR  TCRB   \n",
       "2    015V15009714_CFAR  TCRB   \n",
       "3    015V15000746_CFAR  TCRB   \n",
       "4    015V16003146_CFAR  TCRB   \n",
       "..                 ...   ...   \n",
       "187  015V09001205_CFAR  TCRB   \n",
       "188  015V09003735_CFAR  TCRB   \n",
       "189  015V10006273_CFAR  TCRB   \n",
       "190  015V12002996_CFAR  TCRB   \n",
       "191  015V14000020_CFAR  TCRB   \n",
       "\n",
       "                                           sample_tags counting_method  \\\n",
       "0    4.906849315 years_rel_to_art, 40.48219178 age_...              v3   \n",
       "1    22.38630137 age_at_min_evidence, 25.95890411 a...              v3   \n",
       "2    39.70491803 age_at_min_evidence, 43.71038251 a...              v3   \n",
       "3    40.1420765 age_at_min_evidence, 42.09041096 ag...              v3   \n",
       "4    40.60547945 age_at_min_evidence, 45.6147541 ag...              v3   \n",
       "..                                                 ...             ...   \n",
       "187  -0.539726027 years_rel_to_art, 41.89589041 age...              v3   \n",
       "188  0.073972603 years_rel_to_art, 41.89589041 age_...              v3   \n",
       "189  1.2 years_rel_to_art, 41.89589041 age_at_min_e...              v3   \n",
       "190  2.950819672 years_rel_to_art, 41.89589041 age_...              v3   \n",
       "191  4.224657534 years_rel_to_art, 41.89589041 age_...              v3   \n",
       "\n",
       "          primer_set species               study sequencing_type  patientID  \\\n",
       "0    Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1016   \n",
       "1    Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1024   \n",
       "2    Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1009   \n",
       "3    Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1023   \n",
       "4    Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1007   \n",
       "..               ...     ...                 ...             ...        ...   \n",
       "187  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1001   \n",
       "188  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1001   \n",
       "189  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1001   \n",
       "190  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1001   \n",
       "191  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1001   \n",
       "\n",
       "     years_rel_to_art timePoint           time_group sex  age_at_collection  \n",
       "0            4.906849  Post-ART  <= 5 years Post-ART   F                 52  \n",
       "1            5.312329  Post-ART  <= 6 years Post-ART   M                 31  \n",
       "2            6.917808  Post-ART  <= 7 years Post-ART   M                 50  \n",
       "3            5.150685  Post-ART  <= 6 years Post-ART   F                 47  \n",
       "4            8.180822  Post-ART  <= 9 years Post-ART   M                 53  \n",
       "..                ...       ...                  ...  ..                ...  \n",
       "187         -0.539726   Pre-ART  <= 6 months Pre-ART   M                 47  \n",
       "188          0.073973  Post-ART   <= 0 month Pre-ART   M                 47  \n",
       "189          1.200000  Post-ART  <= 2 years Post-ART   M                 48  \n",
       "190          2.950820  Post-ART  <= 3 years Post-ART   M                 50  \n",
       "191          4.224658  Post-ART  <= 5 years Post-ART   M                 52  \n",
       "\n",
       "[192 rows x 14 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.merge(\n",
    "    df, df_m, left_on=\"sample_name\", right_on=\"sampleID\", validate=\"1:1\", how=\"left\"\n",
    ").drop(columns=\"sampleID\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Post-ART    131\n",
       "Pre-ART      61\n",
       "Name: timePoint, dtype: int64"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"timePoint\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>patientID</th>\n",
       "      <th>years_rel_to_art</th>\n",
       "      <th>timePoint</th>\n",
       "      <th>time_group</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>015V11001386_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.345205479 years_rel_to_art, 32 age_at_colle...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1028</td>\n",
       "      <td>-0.345205</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 3 months Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>015V13001607_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>0 years_rel_to_art, 27 age_at_collection, 27.1...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1030</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>015V11001839_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.090410959 years_rel_to_art, 50 age_at_colle...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1029</td>\n",
       "      <td>-0.090411</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>015V11002805_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.054794521 years_rel_to_art, 32 age_at_colle...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1028</td>\n",
       "      <td>-0.054795</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>015V09002862_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.04109589 years_rel_to_art, 27 age_at_collec...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1027</td>\n",
       "      <td>-0.041096</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>015V06004712_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.643835616 years_rel_to_art, 37.20821918 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1002</td>\n",
       "      <td>-0.643836</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 6 months Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>015V06009959_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.317808219 years_rel_to_art, 37.20821918 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1002</td>\n",
       "      <td>-0.317808</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 3 months Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>015V07002801_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-2.630136986 years_rel_to_art, 41.89589041 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>-2.630137</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 2 years Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>015V08002889_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.873972603 years_rel_to_art, 41.89589041 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>-0.873973</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 6 months Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>015V09001205_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.539726027 years_rel_to_art, 41.89589041 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>-0.539726</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 6 months Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_name locus  \\\n",
       "27   015V11001386_CFAR  TCRB   \n",
       "32   015V13001607_CFAR  TCRB   \n",
       "35   015V11001839_CFAR  TCRB   \n",
       "41   015V11002805_CFAR  TCRB   \n",
       "49   015V09002862_CFAR  TCRB   \n",
       "..                 ...   ...   \n",
       "183  015V06004712_CFAR  TCRB   \n",
       "184  015V06009959_CFAR  TCRB   \n",
       "185  015V07002801_CFAR  TCRB   \n",
       "186  015V08002889_CFAR  TCRB   \n",
       "187  015V09001205_CFAR  TCRB   \n",
       "\n",
       "                                           sample_tags counting_method  \\\n",
       "27   -0.345205479 years_rel_to_art, 32 age_at_colle...              v3   \n",
       "32   0 years_rel_to_art, 27 age_at_collection, 27.1...              v3   \n",
       "35   -0.090410959 years_rel_to_art, 50 age_at_colle...              v3   \n",
       "41   -0.054794521 years_rel_to_art, 32 age_at_colle...              v3   \n",
       "49   -0.04109589 years_rel_to_art, 27 age_at_collec...              v3   \n",
       "..                                                 ...             ...   \n",
       "183  -0.643835616 years_rel_to_art, 37.20821918 age...              v3   \n",
       "184  -0.317808219 years_rel_to_art, 37.20821918 age...              v3   \n",
       "185  -2.630136986 years_rel_to_art, 41.89589041 age...              v3   \n",
       "186  -0.873972603 years_rel_to_art, 41.89589041 age...              v3   \n",
       "187  -0.539726027 years_rel_to_art, 41.89589041 age...              v3   \n",
       "\n",
       "          primer_set species               study sequencing_type  patientID  \\\n",
       "27   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1028   \n",
       "32   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1030   \n",
       "35   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1029   \n",
       "41   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1028   \n",
       "49   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1027   \n",
       "..               ...     ...                 ...             ...        ...   \n",
       "183  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1002   \n",
       "184  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1002   \n",
       "185  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1001   \n",
       "186  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1001   \n",
       "187  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA       1001   \n",
       "\n",
       "     years_rel_to_art timePoint           time_group sex  age_at_collection  \n",
       "27          -0.345205   Pre-ART  <= 3 months Pre-ART   M                 32  \n",
       "32           0.000000   Pre-ART   <= 0 month Pre-ART   F                 27  \n",
       "35          -0.090411   Pre-ART   <= 0 month Pre-ART   F                 50  \n",
       "41          -0.054795   Pre-ART   <= 0 month Pre-ART   M                 32  \n",
       "49          -0.041096   Pre-ART   <= 0 month Pre-ART   F                 27  \n",
       "..                ...       ...                  ...  ..                ...  \n",
       "183         -0.643836   Pre-ART  <= 6 months Pre-ART   F                 49  \n",
       "184         -0.317808   Pre-ART  <= 3 months Pre-ART   F                 49  \n",
       "185         -2.630137   Pre-ART   <= 2 years Pre-ART   M                 45  \n",
       "186         -0.873973   Pre-ART  <= 6 months Pre-ART   M                 46  \n",
       "187         -0.539726   Pre-ART  <= 6 months Pre-ART   M                 47  \n",
       "\n",
       "[61 rows x 14 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[df[\"timePoint\"] == \"Pre-ART\"]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<= 0 month Pre-ART     15\n",
       "<= 1 year Pre-ART      14\n",
       "<= 6 months Pre-ART    12\n",
       "<= 3 months Pre-ART    10\n",
       "<= 2 years Pre-ART      5\n",
       "<= 3 years Pre-ART      3\n",
       "<= 4 years Pre-ART      2\n",
       "Name: time_group, dtype: int64"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"time_group\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.rename(\n",
    "    columns={\"patientID\": \"participant_label\", \"age_at_collection\": \"Age\", \"sex\": \"Sex\"}\n",
    ")\n",
    "df[\"specimen_label\"] = df[\"sample_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>years_rel_to_art</th>\n",
       "      <th>timePoint</th>\n",
       "      <th>time_group</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>015V11001386_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.345205479 years_rel_to_art, 32 age_at_colle...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1028</td>\n",
       "      <td>-0.345205</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 3 months Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>32</td>\n",
       "      <td>015V11001386_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>015V13001607_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>0 years_rel_to_art, 27 age_at_collection, 27.1...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1030</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>27</td>\n",
       "      <td>015V13001607_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>015V11001839_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.090410959 years_rel_to_art, 50 age_at_colle...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1029</td>\n",
       "      <td>-0.090411</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>50</td>\n",
       "      <td>015V11001839_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>015V11002805_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.054794521 years_rel_to_art, 32 age_at_colle...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1028</td>\n",
       "      <td>-0.054795</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>32</td>\n",
       "      <td>015V11002805_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>015V09002862_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.04109589 years_rel_to_art, 27 age_at_collec...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1027</td>\n",
       "      <td>-0.041096</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>27</td>\n",
       "      <td>015V09002862_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>015V06004712_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.643835616 years_rel_to_art, 37.20821918 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1002</td>\n",
       "      <td>-0.643836</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 6 months Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>49</td>\n",
       "      <td>015V06004712_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>015V06009959_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.317808219 years_rel_to_art, 37.20821918 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1002</td>\n",
       "      <td>-0.317808</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 3 months Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>49</td>\n",
       "      <td>015V06009959_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>015V07002801_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-2.630136986 years_rel_to_art, 41.89589041 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>-2.630137</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 2 years Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>45</td>\n",
       "      <td>015V07002801_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>015V08002889_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.873972603 years_rel_to_art, 41.89589041 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>-0.873973</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 6 months Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>46</td>\n",
       "      <td>015V08002889_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>015V09001205_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.539726027 years_rel_to_art, 41.89589041 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>-0.539726</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 6 months Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>47</td>\n",
       "      <td>015V09001205_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_name locus  \\\n",
       "27   015V11001386_CFAR  TCRB   \n",
       "32   015V13001607_CFAR  TCRB   \n",
       "35   015V11001839_CFAR  TCRB   \n",
       "41   015V11002805_CFAR  TCRB   \n",
       "49   015V09002862_CFAR  TCRB   \n",
       "..                 ...   ...   \n",
       "183  015V06004712_CFAR  TCRB   \n",
       "184  015V06009959_CFAR  TCRB   \n",
       "185  015V07002801_CFAR  TCRB   \n",
       "186  015V08002889_CFAR  TCRB   \n",
       "187  015V09001205_CFAR  TCRB   \n",
       "\n",
       "                                           sample_tags counting_method  \\\n",
       "27   -0.345205479 years_rel_to_art, 32 age_at_colle...              v3   \n",
       "32   0 years_rel_to_art, 27 age_at_collection, 27.1...              v3   \n",
       "35   -0.090410959 years_rel_to_art, 50 age_at_colle...              v3   \n",
       "41   -0.054794521 years_rel_to_art, 32 age_at_colle...              v3   \n",
       "49   -0.04109589 years_rel_to_art, 27 age_at_collec...              v3   \n",
       "..                                                 ...             ...   \n",
       "183  -0.643835616 years_rel_to_art, 37.20821918 age...              v3   \n",
       "184  -0.317808219 years_rel_to_art, 37.20821918 age...              v3   \n",
       "185  -2.630136986 years_rel_to_art, 41.89589041 age...              v3   \n",
       "186  -0.873972603 years_rel_to_art, 41.89589041 age...              v3   \n",
       "187  -0.539726027 years_rel_to_art, 41.89589041 age...              v3   \n",
       "\n",
       "          primer_set species               study sequencing_type  \\\n",
       "27   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "32   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "35   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "41   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "49   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "..               ...     ...                 ...             ...   \n",
       "183  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "184  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "185  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "186  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "187  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "\n",
       "     participant_label  years_rel_to_art timePoint           time_group Sex  \\\n",
       "27                1028         -0.345205   Pre-ART  <= 3 months Pre-ART   M   \n",
       "32                1030          0.000000   Pre-ART   <= 0 month Pre-ART   F   \n",
       "35                1029         -0.090411   Pre-ART   <= 0 month Pre-ART   F   \n",
       "41                1028         -0.054795   Pre-ART   <= 0 month Pre-ART   M   \n",
       "49                1027         -0.041096   Pre-ART   <= 0 month Pre-ART   F   \n",
       "..                 ...               ...       ...                  ...  ..   \n",
       "183               1002         -0.643836   Pre-ART  <= 6 months Pre-ART   F   \n",
       "184               1002         -0.317808   Pre-ART  <= 3 months Pre-ART   F   \n",
       "185               1001         -2.630137   Pre-ART   <= 2 years Pre-ART   M   \n",
       "186               1001         -0.873973   Pre-ART  <= 6 months Pre-ART   M   \n",
       "187               1001         -0.539726   Pre-ART  <= 6 months Pre-ART   M   \n",
       "\n",
       "     Age     specimen_label disease                       disease_subtype  \n",
       "27    32  015V11001386_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "32    27  015V13001607_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "35    50  015V11001839_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "41    32  015V11002805_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "49    27  015V09002862_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "..   ...                ...     ...                                   ...  \n",
       "183   49  015V06004712_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "184   49  015V06009959_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "185   45  015V07002801_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "186   46  015V08002889_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "187   47  015V09001205_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "\n",
       "[61 rows x 17 columns]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"disease\"] = \"HIV\"\n",
    "df[\"disease_subtype\"] = \"HIV - before anti-retroviral therapy\"\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    14\n",
       "3     7\n",
       "2     5\n",
       "4     4\n",
       "Name: participant_label, dtype: int64"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many samples per person?\n",
    "df[\"participant_label\"].value_counts().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>years_rel_to_art</th>\n",
       "      <th>timePoint</th>\n",
       "      <th>time_group</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>015V11001386_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.345205479 years_rel_to_art, 32 age_at_colle...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1028</td>\n",
       "      <td>-0.345205</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 3 months Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>32</td>\n",
       "      <td>015V11001386_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>015V13001607_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>0 years_rel_to_art, 27 age_at_collection, 27.1...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1030</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>27</td>\n",
       "      <td>015V13001607_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>015V11001839_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.090410959 years_rel_to_art, 50 age_at_colle...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1029</td>\n",
       "      <td>-0.090411</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>50</td>\n",
       "      <td>015V11001839_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>015V11002805_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.054794521 years_rel_to_art, 32 age_at_colle...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1028</td>\n",
       "      <td>-0.054795</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>32</td>\n",
       "      <td>015V11002805_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>015V09002862_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.04109589 years_rel_to_art, 27 age_at_collec...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1027</td>\n",
       "      <td>-0.041096</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 0 month Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>27</td>\n",
       "      <td>015V09002862_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>015V06004712_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.643835616 years_rel_to_art, 37.20821918 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1002</td>\n",
       "      <td>-0.643836</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 6 months Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>49</td>\n",
       "      <td>015V06004712_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>015V06009959_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.317808219 years_rel_to_art, 37.20821918 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1002</td>\n",
       "      <td>-0.317808</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 3 months Pre-ART</td>\n",
       "      <td>F</td>\n",
       "      <td>49</td>\n",
       "      <td>015V06009959_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>015V07002801_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-2.630136986 years_rel_to_art, 41.89589041 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>-2.630137</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 2 years Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>45</td>\n",
       "      <td>015V07002801_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>015V08002889_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.873972603 years_rel_to_art, 41.89589041 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>-0.873973</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 6 months Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>46</td>\n",
       "      <td>015V08002889_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>015V09001205_CFAR</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>-0.539726027 years_rel_to_art, 41.89589041 age...</td>\n",
       "      <td>v3</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>towlerton-2022-hiv</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>-0.539726</td>\n",
       "      <td>Pre-ART</td>\n",
       "      <td>&lt;= 6 months Pre-ART</td>\n",
       "      <td>M</td>\n",
       "      <td>47</td>\n",
       "      <td>015V09001205_CFAR</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV - before anti-retroviral therapy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_name locus  \\\n",
       "27   015V11001386_CFAR  TCRB   \n",
       "32   015V13001607_CFAR  TCRB   \n",
       "35   015V11001839_CFAR  TCRB   \n",
       "41   015V11002805_CFAR  TCRB   \n",
       "49   015V09002862_CFAR  TCRB   \n",
       "..                 ...   ...   \n",
       "183  015V06004712_CFAR  TCRB   \n",
       "184  015V06009959_CFAR  TCRB   \n",
       "185  015V07002801_CFAR  TCRB   \n",
       "186  015V08002889_CFAR  TCRB   \n",
       "187  015V09001205_CFAR  TCRB   \n",
       "\n",
       "                                           sample_tags counting_method  \\\n",
       "27   -0.345205479 years_rel_to_art, 32 age_at_colle...              v3   \n",
       "32   0 years_rel_to_art, 27 age_at_collection, 27.1...              v3   \n",
       "35   -0.090410959 years_rel_to_art, 50 age_at_colle...              v3   \n",
       "41   -0.054794521 years_rel_to_art, 32 age_at_colle...              v3   \n",
       "49   -0.04109589 years_rel_to_art, 27 age_at_collec...              v3   \n",
       "..                                                 ...             ...   \n",
       "183  -0.643835616 years_rel_to_art, 37.20821918 age...              v3   \n",
       "184  -0.317808219 years_rel_to_art, 37.20821918 age...              v3   \n",
       "185  -2.630136986 years_rel_to_art, 41.89589041 age...              v3   \n",
       "186  -0.873972603 years_rel_to_art, 41.89589041 age...              v3   \n",
       "187  -0.539726027 years_rel_to_art, 41.89589041 age...              v3   \n",
       "\n",
       "          primer_set species               study sequencing_type  \\\n",
       "27   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "32   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "35   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "41   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "49   Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "..               ...     ...                 ...             ...   \n",
       "183  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "184  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "185  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "186  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "187  Human-TCRB-PD1x   Human  towlerton-2022-hiv            gDNA   \n",
       "\n",
       "     participant_label  years_rel_to_art timePoint           time_group Sex  \\\n",
       "27                1028         -0.345205   Pre-ART  <= 3 months Pre-ART   M   \n",
       "32                1030          0.000000   Pre-ART   <= 0 month Pre-ART   F   \n",
       "35                1029         -0.090411   Pre-ART   <= 0 month Pre-ART   F   \n",
       "41                1028         -0.054795   Pre-ART   <= 0 month Pre-ART   M   \n",
       "49                1027         -0.041096   Pre-ART   <= 0 month Pre-ART   F   \n",
       "..                 ...               ...       ...                  ...  ..   \n",
       "183               1002         -0.643836   Pre-ART  <= 6 months Pre-ART   F   \n",
       "184               1002         -0.317808   Pre-ART  <= 3 months Pre-ART   F   \n",
       "185               1001         -2.630137   Pre-ART   <= 2 years Pre-ART   M   \n",
       "186               1001         -0.873973   Pre-ART  <= 6 months Pre-ART   M   \n",
       "187               1001         -0.539726   Pre-ART  <= 6 months Pre-ART   M   \n",
       "\n",
       "     Age     specimen_label disease                       disease_subtype  \n",
       "27    32  015V11001386_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "32    27  015V13001607_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "35    50  015V11001839_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "41    32  015V11002805_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "49    27  015V09002862_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "..   ...                ...     ...                                   ...  \n",
       "183   49  015V06004712_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "184   49  015V06009959_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "185   45  015V07002801_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "186   46  015V08002889_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "187   47  015V09001205_CFAR     HIV  HIV - before anti-retroviral therapy  \n",
       "\n",
       "[61 rows x 17 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual.append(df)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Combine all manually annotated studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dfs_adaptive_manual)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Also add ImmuneCode from the very top\n",
    "dfs_adaptive_manual.append(adaptive_tcr_covid_specimens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "dfs_adaptive_manual = pd.concat(dfs_adaptive_manual, axis=0).reset_index(drop=True)\n",
    "dfs_adaptive_manual_bak = dfs_adaptive_manual.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>...</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>Sample Timepoint</th>\n",
       "      <th>Race</th>\n",
       "      <th>replicate_label</th>\n",
       "      <th>years_rel_to_art</th>\n",
       "      <th>timePoint</th>\n",
       "      <th>time_group</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>ethnicity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIP15860</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP15860</td>\n",
       "      <td>HIP15860</td>\n",
       "      <td>...</td>\n",
       "      <td>Healthy/Background - CMV Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP14363</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*02, HLA-A*11, HLA-B*35, Human...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP14363</td>\n",
       "      <td>HIP14363</td>\n",
       "      <td>...</td>\n",
       "      <td>Healthy/Background - CMV Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIP14178</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*01, HLA-A*03, HLA-B*08, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP14178</td>\n",
       "      <td>HIP14178</td>\n",
       "      <td>...</td>\n",
       "      <td>Healthy/Background - CMV Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIP13944</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*02, HLA-A*03, HLA-B*18, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP13944</td>\n",
       "      <td>HIP13944</td>\n",
       "      <td>...</td>\n",
       "      <td>Healthy/Background - CMV Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP13911</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>22 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>HIP13911</td>\n",
       "      <td>HIP13911</td>\n",
       "      <td>...</td>\n",
       "      <td>Healthy/Background - CMV+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1372</th>\n",
       "      <td>860011117_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>860011117_TCRB</td>\n",
       "      <td>ImmuneCode-304752</td>\n",
       "      <td>...</td>\n",
       "      <td>COVID-19-HUniv12Oct - Hospitalized - ICU</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Caucasian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1373</th>\n",
       "      <td>860011106_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>860011106_TCRB</td>\n",
       "      <td>ImmuneCode-775827</td>\n",
       "      <td>...</td>\n",
       "      <td>COVID-19-HUniv12Oct - Hospitalized</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57.0</td>\n",
       "      <td>F</td>\n",
       "      <td>Caucasian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1374</th>\n",
       "      <td>BS-GIGI_10-replacement_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>BS-GIGI_10-replacement_TCRB</td>\n",
       "      <td>ImmuneCode-0000446</td>\n",
       "      <td>...</td>\n",
       "      <td>COVID-19-NIH/NIAID - Hospitalized</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Caucasian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1375</th>\n",
       "      <td>BS-EQ-25-T1_BS-GIGI-71-replacement_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>BS-EQ-25-T1_BS-GIGI-71-replacement_TCRB</td>\n",
       "      <td>ImmuneCode-0000160</td>\n",
       "      <td>...</td>\n",
       "      <td>COVID-19-NIH/NIAID - Hospitalized - ICU</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Caucasian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1376</th>\n",
       "      <td>860011240_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>860011240_TCRB</td>\n",
       "      <td>ImmuneCode-506345</td>\n",
       "      <td>...</td>\n",
       "      <td>COVID-19-HUniv12Oct - Hospitalized</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>78.0</td>\n",
       "      <td>F</td>\n",
       "      <td>Caucasian</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1377 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  sample_name locus  \\\n",
       "0                                    HIP15860  TCRB   \n",
       "1                                    HIP14363  TCRB   \n",
       "2                                    HIP14178  TCRB   \n",
       "3                                    HIP13944  TCRB   \n",
       "4                                    HIP13911  TCRB   \n",
       "...                                       ...   ...   \n",
       "1372                           860011117_TCRB  TCRB   \n",
       "1373                           860011106_TCRB  TCRB   \n",
       "1374              BS-GIGI_10-replacement_TCRB  TCRB   \n",
       "1375  BS-EQ-25-T1_BS-GIGI-71-replacement_TCRB  TCRB   \n",
       "1376                           860011240_TCRB  TCRB   \n",
       "\n",
       "                                            sample_tags counting_method  \\\n",
       "0     Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B...              v2   \n",
       "1     Cohort 01, HLA-A*02, HLA-A*11, HLA-B*35, Human...              v2   \n",
       "2     Cohort 01, HLA-A*01, HLA-A*03, HLA-B*08, HLA-B...              v2   \n",
       "3     Cohort 01, HLA-A*02, HLA-A*03, HLA-B*18, HLA-B...              v2   \n",
       "4     22 Years, Caucasian, Cohort 01, Cytomegaloviru...              v2   \n",
       "...                                                 ...             ...   \n",
       "1372                                                NaN             NaN   \n",
       "1373                                                NaN             NaN   \n",
       "1374                                                NaN             NaN   \n",
       "1375                                                NaN             NaN   \n",
       "1376                                                NaN             NaN   \n",
       "\n",
       "           primer_set species                      study sequencing_type  \\\n",
       "0     Human-TCRB-PD1x   Human  emerson-2017-natgen_train            gDNA   \n",
       "1     Human-TCRB-PD1x   Human  emerson-2017-natgen_train            gDNA   \n",
       "2     Human-TCRB-PD1x   Human  emerson-2017-natgen_train            gDNA   \n",
       "3     Human-TCRB-PD1x   Human  emerson-2017-natgen_train            gDNA   \n",
       "4     Human-TCRB-PD1x   Human  emerson-2017-natgen_train            gDNA   \n",
       "...               ...     ...                        ...             ...   \n",
       "1372              NaN   Human                 immunecode            gDNA   \n",
       "1373              NaN   Human                 immunecode            gDNA   \n",
       "1374              NaN   Human                 immunecode            gDNA   \n",
       "1375              NaN   Human                 immunecode            gDNA   \n",
       "1376              NaN   Human                 immunecode            gDNA   \n",
       "\n",
       "                               specimen_label   participant_label  ...  \\\n",
       "0                                    HIP15860            HIP15860  ...   \n",
       "1                                    HIP14363            HIP14363  ...   \n",
       "2                                    HIP14178            HIP14178  ...   \n",
       "3                                    HIP13944            HIP13944  ...   \n",
       "4                                    HIP13911            HIP13911  ...   \n",
       "...                                       ...                 ...  ...   \n",
       "1372                           860011117_TCRB   ImmuneCode-304752  ...   \n",
       "1373                           860011106_TCRB   ImmuneCode-775827  ...   \n",
       "1374              BS-GIGI_10-replacement_TCRB  ImmuneCode-0000446  ...   \n",
       "1375  BS-EQ-25-T1_BS-GIGI-71-replacement_TCRB  ImmuneCode-0000160  ...   \n",
       "1376                           860011240_TCRB   ImmuneCode-506345  ...   \n",
       "\n",
       "                               disease_subtype Sample Timepoint Race  \\\n",
       "0             Healthy/Background - CMV Unknown              NaN  NaN   \n",
       "1             Healthy/Background - CMV Unknown              NaN  NaN   \n",
       "2             Healthy/Background - CMV Unknown              NaN  NaN   \n",
       "3             Healthy/Background - CMV Unknown              NaN  NaN   \n",
       "4                    Healthy/Background - CMV+              NaN  NaN   \n",
       "...                                        ...              ...  ...   \n",
       "1372  COVID-19-HUniv12Oct - Hospitalized - ICU              NaN  NaN   \n",
       "1373        COVID-19-HUniv12Oct - Hospitalized              NaN  NaN   \n",
       "1374         COVID-19-NIH/NIAID - Hospitalized              NaN  NaN   \n",
       "1375   COVID-19-NIH/NIAID - Hospitalized - ICU              NaN  NaN   \n",
       "1376        COVID-19-HUniv12Oct - Hospitalized              NaN  NaN   \n",
       "\n",
       "     replicate_label years_rel_to_art timePoint  time_group   age  sex  \\\n",
       "0                NaN              NaN       NaN         NaN   NaN  NaN   \n",
       "1                NaN              NaN       NaN         NaN   NaN  NaN   \n",
       "2                NaN              NaN       NaN         NaN   NaN  NaN   \n",
       "3                NaN              NaN       NaN         NaN   NaN  NaN   \n",
       "4                NaN              NaN       NaN         NaN   NaN  NaN   \n",
       "...              ...              ...       ...         ...   ...  ...   \n",
       "1372             NaN              NaN       NaN         NaN  60.0    M   \n",
       "1373             NaN              NaN       NaN         NaN  57.0    F   \n",
       "1374             NaN              NaN       NaN         NaN  75.0    M   \n",
       "1375             NaN              NaN       NaN         NaN  72.0    M   \n",
       "1376             NaN              NaN       NaN         NaN  78.0    F   \n",
       "\n",
       "      ethnicity  \n",
       "0           NaN  \n",
       "1           NaN  \n",
       "2           NaN  \n",
       "3           NaN  \n",
       "4           NaN  \n",
       "...         ...  \n",
       "1372  Caucasian  \n",
       "1373  Caucasian  \n",
       "1374  Caucasian  \n",
       "1375  Caucasian  \n",
       "1376  Caucasian  \n",
       "\n",
       "[1377 rows x 25 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual = dfs_adaptive_manual_bak.copy()  # for easy reset when debugging\n",
    "dfs_adaptive_manual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "del df  # unused variable now - let's get it out of scope to be safe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gDNA    1377\n",
       "Name: sequencing_type, dtype: int64"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual[\"sequencing_type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert not dfs_adaptive_manual[\"sequencing_type\"].isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Healthy/Background    1009\n",
       "T1D                    143\n",
       "Covid19                 88\n",
       "HIV                     61\n",
       "RA                      39\n",
       "CVID                    37\n",
       "Name: disease, dtype: int64"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual[\"disease\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert not dfs_adaptive_manual[\"study\"].isna().any()\n",
    "assert not dfs_adaptive_manual[\"participant_label\"].isna().any()\n",
    "assert not dfs_adaptive_manual[\"specimen_label\"].isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To be consistent with boydlab columns, we'll add amplification_label, which here will always equal specimen_label.\n",
    "# See sample_sequences.py for more details on how this gets used.\n",
    "if \"amplification_label\" not in dfs_adaptive_manual.columns:\n",
    "    dfs_adaptive_manual[\"amplification_label\"] = dfs_adaptive_manual[\"specimen_label\"]\n",
    "else:\n",
    "    # fill NA\n",
    "    dfs_adaptive_manual[\"amplification_label\"].fillna(\n",
    "        dfs_adaptive_manual[\"specimen_label\"], inplace=True\n",
    "    )\n",
    "\n",
    "# Fill replicate_label\n",
    "dfs_adaptive_manual[\"replicate_label\"].fillna(\n",
    "    dfs_adaptive_manual[\"specimen_label\"], inplace=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add study prefixes to make these labels unique to study:\n",
    "for col in [\n",
    "    \"participant_label\",\n",
    "    \"specimen_label\",\n",
    "    \"amplification_label\",\n",
    "    \"replicate_label\",\n",
    "]:\n",
    "    dfs_adaptive_manual[col] = (\n",
    "        dfs_adaptive_manual[\"study\"] + \"_\" + dfs_adaptive_manual[col].astype(str)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "# confirm one entry per replicate label per locus, at most!\n",
    "# (specimens can have multiple replicates, e.g. cell type subsets that get merged.)\n",
    "# (participants can have multiple specimens, e.g. separate time points)\n",
    "assert (dfs_adaptive_manual.groupby([\"locus\", \"replicate_label\"]).size() == 1).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>participant_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>locus</th>\n",
       "      <th>disease</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">gDNA</th>\n",
       "      <th rowspan=\"6\" valign=\"top\">TCRB</th>\n",
       "      <th>HIV</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CVID</th>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA</th>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Covid19</th>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D</th>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background</th>\n",
       "      <td>934</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          participant_label\n",
       "sequencing_type locus disease                              \n",
       "gDNA            TCRB  HIV                                30\n",
       "                      CVID                               37\n",
       "                      RA                                 37\n",
       "                      Covid19                            88\n",
       "                      T1D                               143\n",
       "                      Healthy/Background                934"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual.groupby([\"sequencing_type\", \"locus\", \"disease\"], observed=True)[\n",
    "    \"participant_label\"\n",
    "].nunique().to_frame().sort_values(\"participant_label\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>participant_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>locus</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>disease</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"9\" valign=\"top\">gDNA</th>\n",
       "      <th rowspan=\"9\" valign=\"top\">TCRB</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">v1</th>\n",
       "      <th>CVID</th>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background</th>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">v2</th>\n",
       "      <th>Healthy/Background</th>\n",
       "      <td>590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">v3</th>\n",
       "      <th>HIV</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background</th>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA</th>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">v4</th>\n",
       "      <th>Healthy/Background</th>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D</th>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          participant_label\n",
       "sequencing_type locus counting_method disease                              \n",
       "gDNA            TCRB  v1              CVID                               37\n",
       "                                      Healthy/Background                 98\n",
       "                      v2              Healthy/Background                590\n",
       "                                      RA                                  2\n",
       "                      v3              HIV                                30\n",
       "                                      Healthy/Background                133\n",
       "                                      RA                                 35\n",
       "                      v4              Healthy/Background                113\n",
       "                                      T1D                               143"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual.groupby(\n",
    "    [\"sequencing_type\", \"locus\", \"counting_method\", \"disease\"], observed=True\n",
    ")[\"participant_label\"].nunique().to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual[\"disease_subtype\"].isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual[\"disease\"].isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs_adaptive_manual[\"disease_subtype\"].fillna(\n",
    "    dfs_adaptive_manual[\"disease\"], inplace=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample_tags         True\n",
       "counting_method     True\n",
       "primer_set          True\n",
       "Sex                 True\n",
       "Age                 True\n",
       "Ethnicity           True\n",
       "cmv                 True\n",
       "Sample Timepoint    True\n",
       "Race                True\n",
       "years_rel_to_art    True\n",
       "timePoint           True\n",
       "time_group          True\n",
       "age                 True\n",
       "sex                 True\n",
       "ethnicity           True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual.isna().any()[dfs_adaptive_manual.isna().any()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Healthy/Background - CMV-                   420\n",
       "Healthy/Background - CMV+                   340\n",
       "T1D - new onset                             143\n",
       "Healthy/Background - T1D negative           100\n",
       "Healthy/Background - TCRBv4-control          88\n",
       "HIV - before anti-retroviral therapy         61\n",
       "COVID-19-NIH/NIAID - Hospitalized            39\n",
       "CVID                                         37\n",
       "RA - sero-positive                           32\n",
       "Healthy/Background - CMV Unknown             26\n",
       "COVID-19-HUniv12Oct - Hospitalized           26\n",
       "Healthy/Background - CVID negative           22\n",
       "Healthy/Background - RA negative             13\n",
       "COVID-19-ISB                                  9\n",
       "RA - sero-negative                            7\n",
       "COVID-19-NIH/NIAID - Hospitalized - ICU       7\n",
       "COVID-19-HUniv12Oct - Hospitalized - ICU      6\n",
       "COVID-19-HUniv12Oct                           1\n",
       "Name: disease_subtype, dtype: int64"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual[\"disease_subtype\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: study, dtype: int64)"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual[dfs_adaptive_manual[\"disease_subtype\"] == healthy_label][\n",
    "    \"study\"\n",
    "].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>participant_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"18\" valign=\"top\">TCRB</th>\n",
       "      <th>CVID</th>\n",
       "      <th>CVID</th>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">Covid19</th>\n",
       "      <th>COVID-19-HUniv12Oct</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COVID-19-HUniv12Oct - Hospitalized</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COVID-19-HUniv12Oct - Hospitalized - ICU</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COVID-19-ISB</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COVID-19-NIH/NIAID - Hospitalized</th>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COVID-19-NIH/NIAID - Hospitalized - ICU</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIV</th>\n",
       "      <th>HIV - before anti-retroviral therapy</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">Healthy/Background</th>\n",
       "      <th>Healthy/Background - CMV Unknown</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - CMV+</th>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - CMV-</th>\n",
       "      <td>420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - CVID negative</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - RA negative</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - T1D negative</th>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - TCRBv4-control</th>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">RA</th>\n",
       "      <th>RA - sero-negative</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA - sero-positive</th>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D</th>\n",
       "      <th>T1D - new onset</th>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                   participant_label\n",
       "locus disease            disease_subtype                                            \n",
       "TCRB  CVID               CVID                                                     37\n",
       "      Covid19            COVID-19-HUniv12Oct                                       1\n",
       "                         COVID-19-HUniv12Oct - Hospitalized                       26\n",
       "                         COVID-19-HUniv12Oct - Hospitalized - ICU                  6\n",
       "                         COVID-19-ISB                                              9\n",
       "                         COVID-19-NIH/NIAID - Hospitalized                        39\n",
       "                         COVID-19-NIH/NIAID - Hospitalized - ICU                   7\n",
       "      HIV                HIV - before anti-retroviral therapy                     30\n",
       "      Healthy/Background Healthy/Background - CMV Unknown                         26\n",
       "                         Healthy/Background - CMV+                               340\n",
       "                         Healthy/Background - CMV-                               420\n",
       "                         Healthy/Background - CVID negative                       22\n",
       "                         Healthy/Background - RA negative                         13\n",
       "                         Healthy/Background - T1D negative                        25\n",
       "                         Healthy/Background - TCRBv4-control                      88\n",
       "      RA                 RA - sero-negative                                        6\n",
       "                         RA - sero-positive                                       31\n",
       "      T1D                T1D - new onset                                         143"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual.groupby([\"locus\", \"disease\", \"disease_subtype\"], observed=True)[\n",
    "    \"participant_label\"\n",
    "].nunique().to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>participant_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>study</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"19\" valign=\"top\">TCRB</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">v1</th>\n",
       "      <th>CVID</th>\n",
       "      <th>CVID</th>\n",
       "      <th>ramesh-2015-ci</th>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">Healthy/Background</th>\n",
       "      <th>Healthy/Background - CMV Unknown</th>\n",
       "      <th>emerson-2017-natgen_train</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - CMV+</th>\n",
       "      <th>emerson-2017-natgen_train</th>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - CMV-</th>\n",
       "      <th>emerson-2017-natgen_train</th>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - CVID negative</th>\n",
       "      <th>ramesh-2015-ci</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">v2</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">Healthy/Background</th>\n",
       "      <th>Healthy/Background - CMV Unknown</th>\n",
       "      <th>emerson-2017-natgen_train</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - CMV+</th>\n",
       "      <th>emerson-2017-natgen_train</th>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - CMV-</th>\n",
       "      <th>emerson-2017-natgen_train</th>\n",
       "      <td>305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">RA</th>\n",
       "      <th>RA - sero-negative</th>\n",
       "      <th>mustjoki-2017-natcomms</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA - sero-positive</th>\n",
       "      <th>mustjoki-2017-natcomms</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">v3</th>\n",
       "      <th>HIV</th>\n",
       "      <th>HIV - before anti-retroviral therapy</th>\n",
       "      <th>towlerton-2022-hiv</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Healthy/Background</th>\n",
       "      <th>Healthy/Background - CMV+</th>\n",
       "      <th>emerson-2017-natgen_validation</th>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - CMV-</th>\n",
       "      <th>emerson-2017-natgen_validation</th>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - RA negative</th>\n",
       "      <th>mustjoki-2017-natcomms</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">RA</th>\n",
       "      <th>RA - sero-negative</th>\n",
       "      <th>mustjoki-2017-natcomms</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA - sero-positive</th>\n",
       "      <th>mustjoki-2017-natcomms</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">v4</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Healthy/Background</th>\n",
       "      <th>Healthy/Background - T1D negative</th>\n",
       "      <th>mitchell-2022-jcii</th>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy/Background - TCRBv4-control</th>\n",
       "      <th>TCRBv4-control</th>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D</th>\n",
       "      <th>T1D - new onset</th>\n",
       "      <th>mitchell-2022-jcii</th>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                              participant_label\n",
       "locus counting_method disease            disease_subtype                      study                                            \n",
       "TCRB  v1              CVID               CVID                                 ramesh-2015-ci                                 37\n",
       "                      Healthy/Background Healthy/Background - CMV Unknown     emerson-2017-natgen_train                       2\n",
       "                                         Healthy/Background - CMV+            emerson-2017-natgen_train                      28\n",
       "                                         Healthy/Background - CMV-            emerson-2017-natgen_train                      46\n",
       "                                         Healthy/Background - CVID negative   ramesh-2015-ci                                 22\n",
       "      v2              Healthy/Background Healthy/Background - CMV Unknown     emerson-2017-natgen_train                      24\n",
       "                                         Healthy/Background - CMV+            emerson-2017-natgen_train                     261\n",
       "                                         Healthy/Background - CMV-            emerson-2017-natgen_train                     305\n",
       "                      RA                 RA - sero-negative                   mustjoki-2017-natcomms                          1\n",
       "                                         RA - sero-positive                   mustjoki-2017-natcomms                          1\n",
       "      v3              HIV                HIV - before anti-retroviral therapy towlerton-2022-hiv                             30\n",
       "                      Healthy/Background Healthy/Background - CMV+            emerson-2017-natgen_validation                 51\n",
       "                                         Healthy/Background - CMV-            emerson-2017-natgen_validation                 69\n",
       "                                         Healthy/Background - RA negative     mustjoki-2017-natcomms                         13\n",
       "                      RA                 RA - sero-negative                   mustjoki-2017-natcomms                          5\n",
       "                                         RA - sero-positive                   mustjoki-2017-natcomms                         30\n",
       "      v4              Healthy/Background Healthy/Background - T1D negative    mitchell-2022-jcii                             25\n",
       "                                         Healthy/Background - TCRBv4-control  TCRBv4-control                                 88\n",
       "                      T1D                T1D - new onset                      mitchell-2022-jcii                            143"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual.groupby(\n",
    "    [\"locus\", \"counting_method\", \"disease\", \"disease_subtype\", \"study\"], observed=True\n",
    ")[\"participant_label\"].nunique().to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_tags</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>species</th>\n",
       "      <th>study</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample Timepoint</th>\n",
       "      <th>Race</th>\n",
       "      <th>replicate_label</th>\n",
       "      <th>years_rel_to_art</th>\n",
       "      <th>timePoint</th>\n",
       "      <th>time_group</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>amplification_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIP15860</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>emerson-2017-natgen_train_HIP15860</td>\n",
       "      <td>emerson-2017-natgen_train_HIP15860</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>emerson-2017-natgen_train_HIP15860</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>emerson-2017-natgen_train_HIP15860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP14363</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*02, HLA-A*11, HLA-B*35, Human...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14363</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14363</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14363</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIP14178</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*01, HLA-A*03, HLA-B*08, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14178</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14178</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14178</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIP13944</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>Cohort 01, HLA-A*02, HLA-A*03, HLA-B*18, HLA-B...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13944</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13944</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13944</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP13911</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>22 Years, Caucasian, Cohort 01, Cytomegaloviru...</td>\n",
       "      <td>v2</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>Human</td>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13911</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13911</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13911</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1372</th>\n",
       "      <td>860011117_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>immunecode_860011117_TCRB</td>\n",
       "      <td>immunecode_ImmuneCode-304752</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>immunecode_860011117_TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>immunecode_860011117_TCRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1373</th>\n",
       "      <td>860011106_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>immunecode_860011106_TCRB</td>\n",
       "      <td>immunecode_ImmuneCode-775827</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>immunecode_860011106_TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57.0</td>\n",
       "      <td>F</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>immunecode_860011106_TCRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1374</th>\n",
       "      <td>BS-GIGI_10-replacement_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>immunecode_BS-GIGI_10-replacement_TCRB</td>\n",
       "      <td>immunecode_ImmuneCode-0000446</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>immunecode_BS-GIGI_10-replacement_TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>immunecode_BS-GIGI_10-replacement_TCRB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1375</th>\n",
       "      <td>BS-EQ-25-T1_BS-GIGI-71-replacement_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>immunecode_BS-EQ-25-T1_BS-GIGI-71-replacement_...</td>\n",
       "      <td>immunecode_ImmuneCode-0000160</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>immunecode_BS-EQ-25-T1_BS-GIGI-71-replacement_...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>immunecode_BS-EQ-25-T1_BS-GIGI-71-replacement_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1376</th>\n",
       "      <td>860011240_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>immunecode</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>immunecode_860011240_TCRB</td>\n",
       "      <td>immunecode_ImmuneCode-506345</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>immunecode_860011240_TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>78.0</td>\n",
       "      <td>F</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>immunecode_860011240_TCRB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1377 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  sample_name locus  \\\n",
       "0                                    HIP15860  TCRB   \n",
       "1                                    HIP14363  TCRB   \n",
       "2                                    HIP14178  TCRB   \n",
       "3                                    HIP13944  TCRB   \n",
       "4                                    HIP13911  TCRB   \n",
       "...                                       ...   ...   \n",
       "1372                           860011117_TCRB  TCRB   \n",
       "1373                           860011106_TCRB  TCRB   \n",
       "1374              BS-GIGI_10-replacement_TCRB  TCRB   \n",
       "1375  BS-EQ-25-T1_BS-GIGI-71-replacement_TCRB  TCRB   \n",
       "1376                           860011240_TCRB  TCRB   \n",
       "\n",
       "                                            sample_tags counting_method  \\\n",
       "0     Cohort 01, HLA-A*01, HLA-A*68, HLA-B*14, HLA-B...              v2   \n",
       "1     Cohort 01, HLA-A*02, HLA-A*11, HLA-B*35, Human...              v2   \n",
       "2     Cohort 01, HLA-A*01, HLA-A*03, HLA-B*08, HLA-B...              v2   \n",
       "3     Cohort 01, HLA-A*02, HLA-A*03, HLA-B*18, HLA-B...              v2   \n",
       "4     22 Years, Caucasian, Cohort 01, Cytomegaloviru...              v2   \n",
       "...                                                 ...             ...   \n",
       "1372                                                NaN             NaN   \n",
       "1373                                                NaN             NaN   \n",
       "1374                                                NaN             NaN   \n",
       "1375                                                NaN             NaN   \n",
       "1376                                                NaN             NaN   \n",
       "\n",
       "           primer_set species                      study sequencing_type  \\\n",
       "0     Human-TCRB-PD1x   Human  emerson-2017-natgen_train            gDNA   \n",
       "1     Human-TCRB-PD1x   Human  emerson-2017-natgen_train            gDNA   \n",
       "2     Human-TCRB-PD1x   Human  emerson-2017-natgen_train            gDNA   \n",
       "3     Human-TCRB-PD1x   Human  emerson-2017-natgen_train            gDNA   \n",
       "4     Human-TCRB-PD1x   Human  emerson-2017-natgen_train            gDNA   \n",
       "...               ...     ...                        ...             ...   \n",
       "1372              NaN   Human                 immunecode            gDNA   \n",
       "1373              NaN   Human                 immunecode            gDNA   \n",
       "1374              NaN   Human                 immunecode            gDNA   \n",
       "1375              NaN   Human                 immunecode            gDNA   \n",
       "1376              NaN   Human                 immunecode            gDNA   \n",
       "\n",
       "                                         specimen_label  \\\n",
       "0                    emerson-2017-natgen_train_HIP15860   \n",
       "1                    emerson-2017-natgen_train_HIP14363   \n",
       "2                    emerson-2017-natgen_train_HIP14178   \n",
       "3                    emerson-2017-natgen_train_HIP13944   \n",
       "4                    emerson-2017-natgen_train_HIP13911   \n",
       "...                                                 ...   \n",
       "1372                          immunecode_860011117_TCRB   \n",
       "1373                          immunecode_860011106_TCRB   \n",
       "1374             immunecode_BS-GIGI_10-replacement_TCRB   \n",
       "1375  immunecode_BS-EQ-25-T1_BS-GIGI-71-replacement_...   \n",
       "1376                          immunecode_860011240_TCRB   \n",
       "\n",
       "                       participant_label  ... Sample Timepoint Race  \\\n",
       "0     emerson-2017-natgen_train_HIP15860  ...              NaN  NaN   \n",
       "1     emerson-2017-natgen_train_HIP14363  ...              NaN  NaN   \n",
       "2     emerson-2017-natgen_train_HIP14178  ...              NaN  NaN   \n",
       "3     emerson-2017-natgen_train_HIP13944  ...              NaN  NaN   \n",
       "4     emerson-2017-natgen_train_HIP13911  ...              NaN  NaN   \n",
       "...                                  ...  ...              ...  ...   \n",
       "1372        immunecode_ImmuneCode-304752  ...              NaN  NaN   \n",
       "1373        immunecode_ImmuneCode-775827  ...              NaN  NaN   \n",
       "1374       immunecode_ImmuneCode-0000446  ...              NaN  NaN   \n",
       "1375       immunecode_ImmuneCode-0000160  ...              NaN  NaN   \n",
       "1376        immunecode_ImmuneCode-506345  ...              NaN  NaN   \n",
       "\n",
       "                                        replicate_label years_rel_to_art  \\\n",
       "0                    emerson-2017-natgen_train_HIP15860              NaN   \n",
       "1                    emerson-2017-natgen_train_HIP14363              NaN   \n",
       "2                    emerson-2017-natgen_train_HIP14178              NaN   \n",
       "3                    emerson-2017-natgen_train_HIP13944              NaN   \n",
       "4                    emerson-2017-natgen_train_HIP13911              NaN   \n",
       "...                                                 ...              ...   \n",
       "1372                          immunecode_860011117_TCRB              NaN   \n",
       "1373                          immunecode_860011106_TCRB              NaN   \n",
       "1374             immunecode_BS-GIGI_10-replacement_TCRB              NaN   \n",
       "1375  immunecode_BS-EQ-25-T1_BS-GIGI-71-replacement_...              NaN   \n",
       "1376                          immunecode_860011240_TCRB              NaN   \n",
       "\n",
       "     timePoint time_group   age  sex  ethnicity  \\\n",
       "0          NaN        NaN   NaN  NaN        NaN   \n",
       "1          NaN        NaN   NaN  NaN        NaN   \n",
       "2          NaN        NaN   NaN  NaN        NaN   \n",
       "3          NaN        NaN   NaN  NaN        NaN   \n",
       "4          NaN        NaN   NaN  NaN        NaN   \n",
       "...        ...        ...   ...  ...        ...   \n",
       "1372       NaN        NaN  60.0    M  Caucasian   \n",
       "1373       NaN        NaN  57.0    F  Caucasian   \n",
       "1374       NaN        NaN  75.0    M  Caucasian   \n",
       "1375       NaN        NaN  72.0    M  Caucasian   \n",
       "1376       NaN        NaN  78.0    F  Caucasian   \n",
       "\n",
       "                                    amplification_label  \n",
       "0                    emerson-2017-natgen_train_HIP15860  \n",
       "1                    emerson-2017-natgen_train_HIP14363  \n",
       "2                    emerson-2017-natgen_train_HIP14178  \n",
       "3                    emerson-2017-natgen_train_HIP13944  \n",
       "4                    emerson-2017-natgen_train_HIP13911  \n",
       "...                                                 ...  \n",
       "1372                          immunecode_860011117_TCRB  \n",
       "1373                          immunecode_860011106_TCRB  \n",
       "1374             immunecode_BS-GIGI_10-replacement_TCRB  \n",
       "1375  immunecode_BS-EQ-25-T1_BS-GIGI-71-replacement_...  \n",
       "1376                          immunecode_860011240_TCRB  \n",
       "\n",
       "[1377 rows x 26 columns]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual = dfs_adaptive_manual[\n",
    "    dfs_adaptive_manual[\"sequencing_type\"] == \"gDNA\"\n",
    "]\n",
    "dfs_adaptive_manual = dfs_adaptive_manual[dfs_adaptive_manual[\"locus\"] == \"TCRB\"]\n",
    "dfs_adaptive_manual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "disease\n",
       "HIV                    30\n",
       "CVID                   37\n",
       "RA                     37\n",
       "Covid19                88\n",
       "T1D                   143\n",
       "Healthy/Background    934\n",
       "Name: participant_label, dtype: int64"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual.groupby(\"disease\")[\"participant_label\"].nunique().sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>replicate_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1042</th>\n",
       "      <td>mustjoki-2017-natcomms_RA23</td>\n",
       "      <td>mustjoki-2017-natcomms_RA23_CD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1043</th>\n",
       "      <td>mustjoki-2017-natcomms_RA23</td>\n",
       "      <td>mustjoki-2017-natcomms_RA23_CD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1070</th>\n",
       "      <td>mustjoki-2017-natcomms_RA6</td>\n",
       "      <td>mustjoki-2017-natcomms_RA6_CD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1071</th>\n",
       "      <td>mustjoki-2017-natcomms_RA6</td>\n",
       "      <td>mustjoki-2017-natcomms_RA6_CD8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   specimen_label                  replicate_label\n",
       "1042  mustjoki-2017-natcomms_RA23  mustjoki-2017-natcomms_RA23_CD4\n",
       "1043  mustjoki-2017-natcomms_RA23  mustjoki-2017-natcomms_RA23_CD8\n",
       "1070   mustjoki-2017-natcomms_RA6   mustjoki-2017-natcomms_RA6_CD4\n",
       "1071   mustjoki-2017-natcomms_RA6   mustjoki-2017-natcomms_RA6_CD8"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Review which replicates are getting combined into which specimens\n",
    "# dfs_adaptive_manual[dfs_adaptive_manual['replicate_label'] != dfs_adaptive_manual['specimen_label']].groupby('specimen_label')['replicate_label'].unique().tolist()\n",
    "dfs_adaptive_manual[\n",
    "    dfs_adaptive_manual[\"replicate_label\"] != dfs_adaptive_manual[\"specimen_label\"]\n",
    "][[\"specimen_label\", \"replicate_label\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>mustjoki-2017-natcomms_RA23</th>\n",
       "      <td>mustjoki-2017-natcomms_RA23_CD4</td>\n",
       "      <td>mustjoki-2017-natcomms_RA23_CD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mustjoki-2017-natcomms_RA6</th>\n",
       "      <td>mustjoki-2017-natcomms_RA6_CD4</td>\n",
       "      <td>mustjoki-2017-natcomms_RA6_CD8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           0  \\\n",
       "specimen_label                                                 \n",
       "mustjoki-2017-natcomms_RA23  mustjoki-2017-natcomms_RA23_CD4   \n",
       "mustjoki-2017-natcomms_RA6    mustjoki-2017-natcomms_RA6_CD4   \n",
       "\n",
       "                                                           1  \n",
       "specimen_label                                                \n",
       "mustjoki-2017-natcomms_RA23  mustjoki-2017-natcomms_RA23_CD8  \n",
       "mustjoki-2017-natcomms_RA6    mustjoki-2017-natcomms_RA6_CD8  "
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Review which replicates are getting combined into which specimens\n",
    "replicates_being_merged_into_same_specimen = (\n",
    "    dfs_adaptive_manual[\n",
    "        dfs_adaptive_manual[\"replicate_label\"] != dfs_adaptive_manual[\"specimen_label\"]\n",
    "    ]\n",
    "    .groupby(\"specimen_label\")[\"replicate_label\"]\n",
    "    .unique()\n",
    "    .apply(pd.Series)\n",
    ")\n",
    "# remove rows where single replicate (but just happened to have different label) - no merging happening\n",
    "replicates_being_merged_into_same_specimen = replicates_being_merged_into_same_specimen[\n",
    "    replicates_being_merged_into_same_specimen.notna().sum(axis=1) > 1\n",
    "]\n",
    "replicates_being_merged_into_same_specimen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Also review the cell_type helper column we made for ourselves:\n",
    "# (TODO: bring this back)\n",
    "# dfs_adaptive_manual[\"cell_type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Human    1377\n",
       "Name: species, dtype: int64"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_adaptive_manual[\"species\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['age', 'Age', 'amplification_label', 'cmv', 'counting_method',\n",
       "       'disease', 'disease_subtype', 'ethnicity', 'Ethnicity', 'locus',\n",
       "       'participant_label', 'primer_set', 'Race', 'replicate_label',\n",
       "       'Sample Timepoint', 'sample_name', 'sample_tags', 'sequencing_type',\n",
       "       'Sex', 'sex', 'species', 'specimen_label', 'study', 'time_group',\n",
       "       'timePoint', 'years_rel_to_art'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# all available columns, in case-insensitive sorted order\n",
    "dfs_adaptive_manual.columns.sort_values(key=lambda idx: idx.str.lower())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Symptom metadata columns:\n",
    "# Add more here if we add more studies\n",
    "symptoms_columns = [\n",
    "    \"cmv\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<= 0 month Pre-ART     15\n",
       "<= 1 year Pre-ART      14\n",
       "<= 6 months Pre-ART    12\n",
       "<= 3 months Pre-ART    10\n",
       "<= 2 years Pre-ART      5\n",
       "<= 3 years Pre-ART      3\n",
       "<= 4 years Pre-ART      2\n",
       "Name: specimen_description, dtype: int64"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# if future studies have participant-level description fields, rename those columns into \"participant_description\"\n",
    "\n",
    "# specimen description can come in several fields:\n",
    "specimen_description_fields = [\"time_group\", \"timePoint\"]\n",
    "\n",
    "# They are either all NA or one is set. Never have multiple of these set:\n",
    "assert dfs_adaptive_manual[specimen_description_fields].notna().sum(axis=1).max()\n",
    "\n",
    "# So we can just take first non-null value (if any) per row from these columns (https://stackoverflow.com/a/37938780/130164):\n",
    "dfs_adaptive_manual[\"specimen_description\"] = (\n",
    "    dfs_adaptive_manual[specimen_description_fields]\n",
    "    .fillna(method=\"bfill\", axis=1)\n",
    "    .iloc[:, 0]\n",
    ")\n",
    "dfs_adaptive_manual[\"specimen_description\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the two age columns are never set at the same time, so we can fillna to combine\n",
    "assert (\n",
    "    dfs_adaptive_manual[\"age\"].notna().astype(int)\n",
    "    + dfs_adaptive_manual[\"Age\"].notna().astype(int)\n",
    ").max() == 1\n",
    "dfs_adaptive_manual[\"Age\"].fillna(dfs_adaptive_manual[\"age\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the two sex columns are never set at the same time, so we can fillna to combine\n",
    "assert (\n",
    "    dfs_adaptive_manual[\"sex\"].notna().astype(int)\n",
    "    + dfs_adaptive_manual[\"Sex\"].notna().astype(int)\n",
    ").max() == 1\n",
    "dfs_adaptive_manual[\"Sex\"].fillna(dfs_adaptive_manual[\"sex\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ethnicity and Ethnicity columns are never set at the same time, so we can fillna to combine\n",
    "assert (\n",
    "    dfs_adaptive_manual[\"ethnicity\"].notna().astype(int)\n",
    "    + dfs_adaptive_manual[\"Ethnicity\"].notna().astype(int)\n",
    ").max() == 1\n",
    "dfs_adaptive_manual[\"Ethnicity\"].fillna(dfs_adaptive_manual[\"ethnicity\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "White, Not Hispanic or Latino                                        460\n",
       "Unknown                                                              226\n",
       "White - Non-Hispanic                                                 200\n",
       "Caucasian                                                             74\n",
       "Asian, Not Hispanic or Latino                                         42\n",
       "Unknown, Hispanic or Latino                                           26\n",
       "White - Hispanic                                                      19\n",
       "White - Hispanic                                                      16\n",
       "Black or African American, Not Hispanic or Latino                     11\n",
       "Hispanic/Latino                                                       10\n",
       "American Indian or Alaska Native, Not Hispanic or Latino              10\n",
       "Asian - Non-Hispanic                                                   4\n",
       "Native Hawaiian or other Pacific Islander, Not Hispanic or Latino      3\n",
       "White, Hispanic or Latino                                              3\n",
       "Black - Non-Hispanic                                                   2\n",
       "Other, Not Hispanic or Latino                                          2\n",
       "Asian, Hispanic or Latino                                              1\n",
       "Not reported - Not reported                                            1\n",
       "Other, Hispanic or Latino                                              1\n",
       "Hispanic - Hispanic                                                    1\n",
       "Asian                                                                  1\n",
       "Name: Ethnicity, dtype: int64"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Race and Ethnicity are indeed set together, so we need fancier logic to combine\n",
    "assert (\n",
    "    dfs_adaptive_manual[\"Race\"].notna().astype(int)\n",
    "    + dfs_adaptive_manual[\"Ethnicity\"].notna().astype(int)\n",
    ").max() > 1\n",
    "\n",
    "\n",
    "def _combine_race_ethnicity_cols(row):\n",
    "    non_empty = row.dropna()\n",
    "    if len(non_empty) == 0:\n",
    "        return np.nan\n",
    "    return \" - \".join(non_empty)\n",
    "\n",
    "\n",
    "dfs_adaptive_manual[\"Ethnicity\"] = dfs_adaptive_manual[[\"Race\", \"Ethnicity\"]].apply(\n",
    "    _combine_race_ethnicity_cols, axis=1\n",
    ")\n",
    "dfs_adaptive_manual[\"Ethnicity\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>locus</th>\n",
       "      <th>counting_method</th>\n",
       "      <th>disease</th>\n",
       "      <th>sequencing_type</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>amplification_label</th>\n",
       "      <th>replicate_label</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>specimen_description</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>symptoms_cmv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>HIP15860</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>v2</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background - CMV Unknown</td>\n",
       "      <td>emerson-2017-natgen_train_HIP15860</td>\n",
       "      <td>emerson-2017-natgen_train_HIP15860</td>\n",
       "      <td>emerson-2017-natgen_train_HIP15860</td>\n",
       "      <td>emerson-2017-natgen_train_HIP15860</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>HIP14363</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>v2</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background - CMV Unknown</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14363</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14363</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14363</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14363</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>HIP14178</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>v2</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background - CMV Unknown</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14178</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14178</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14178</td>\n",
       "      <td>emerson-2017-natgen_train_HIP14178</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>HIP13944</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>v2</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background - CMV Unknown</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13944</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13944</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13944</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13944</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>emerson-2017-natgen_train</td>\n",
       "      <td>HIP13911</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>v2</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>Healthy/Background - CMV+</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13911</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13911</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13911</td>\n",
       "      <td>emerson-2017-natgen_train_HIP13911</td>\n",
       "      <td>Male</td>\n",
       "      <td>22.469541</td>\n",
       "      <td>White, Not Hispanic or Latino</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human-TCRB-PD1x</td>\n",
       "      <td>CMV+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1372</th>\n",
       "      <td>immunecode</td>\n",
       "      <td>860011117_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>COVID-19-HUniv12Oct - Hospitalized - ICU</td>\n",
       "      <td>immunecode_ImmuneCode-304752</td>\n",
       "      <td>immunecode_860011117_TCRB</td>\n",
       "      <td>immunecode_860011117_TCRB</td>\n",
       "      <td>immunecode_860011117_TCRB</td>\n",
       "      <td>M</td>\n",
       "      <td>60.0</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1373</th>\n",
       "      <td>immunecode</td>\n",
       "      <td>860011106_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>COVID-19-HUniv12Oct - Hospitalized</td>\n",
       "      <td>immunecode_ImmuneCode-775827</td>\n",
       "      <td>immunecode_860011106_TCRB</td>\n",
       "      <td>immunecode_860011106_TCRB</td>\n",
       "      <td>immunecode_860011106_TCRB</td>\n",
       "      <td>F</td>\n",
       "      <td>57.0</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1374</th>\n",
       "      <td>immunecode</td>\n",
       "      <td>BS-GIGI_10-replacement_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>COVID-19-NIH/NIAID - Hospitalized</td>\n",
       "      <td>immunecode_ImmuneCode-0000446</td>\n",
       "      <td>immunecode_BS-GIGI_10-replacement_TCRB</td>\n",
       "      <td>immunecode_BS-GIGI_10-replacement_TCRB</td>\n",
       "      <td>immunecode_BS-GIGI_10-replacement_TCRB</td>\n",
       "      <td>M</td>\n",
       "      <td>75.0</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1375</th>\n",
       "      <td>immunecode</td>\n",
       "      <td>BS-EQ-25-T1_BS-GIGI-71-replacement_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>COVID-19-NIH/NIAID - Hospitalized - ICU</td>\n",
       "      <td>immunecode_ImmuneCode-0000160</td>\n",
       "      <td>immunecode_BS-EQ-25-T1_BS-GIGI-71-replacement_...</td>\n",
       "      <td>immunecode_BS-EQ-25-T1_BS-GIGI-71-replacement_...</td>\n",
       "      <td>immunecode_BS-EQ-25-T1_BS-GIGI-71-replacement_...</td>\n",
       "      <td>M</td>\n",
       "      <td>72.0</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1376</th>\n",
       "      <td>immunecode</td>\n",
       "      <td>860011240_TCRB</td>\n",
       "      <td>TCRB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>COVID-19-HUniv12Oct - Hospitalized</td>\n",
       "      <td>immunecode_ImmuneCode-506345</td>\n",
       "      <td>immunecode_860011240_TCRB</td>\n",
       "      <td>immunecode_860011240_TCRB</td>\n",
       "      <td>immunecode_860011240_TCRB</td>\n",
       "      <td>F</td>\n",
       "      <td>78.0</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1377 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          study                              sample_name  \\\n",
       "0     emerson-2017-natgen_train                                 HIP15860   \n",
       "1     emerson-2017-natgen_train                                 HIP14363   \n",
       "2     emerson-2017-natgen_train                                 HIP14178   \n",
       "3     emerson-2017-natgen_train                                 HIP13944   \n",
       "4     emerson-2017-natgen_train                                 HIP13911   \n",
       "...                         ...                                      ...   \n",
       "1372                 immunecode                           860011117_TCRB   \n",
       "1373                 immunecode                           860011106_TCRB   \n",
       "1374                 immunecode              BS-GIGI_10-replacement_TCRB   \n",
       "1375                 immunecode  BS-EQ-25-T1_BS-GIGI-71-replacement_TCRB   \n",
       "1376                 immunecode                           860011240_TCRB   \n",
       "\n",
       "     locus counting_method             disease sequencing_type  \\\n",
       "0     TCRB              v2  Healthy/Background            gDNA   \n",
       "1     TCRB              v2  Healthy/Background            gDNA   \n",
       "2     TCRB              v2  Healthy/Background            gDNA   \n",
       "3     TCRB              v2  Healthy/Background            gDNA   \n",
       "4     TCRB              v2  Healthy/Background            gDNA   \n",
       "...    ...             ...                 ...             ...   \n",
       "1372  TCRB             NaN             Covid19            gDNA   \n",
       "1373  TCRB             NaN             Covid19            gDNA   \n",
       "1374  TCRB             NaN             Covid19            gDNA   \n",
       "1375  TCRB             NaN             Covid19            gDNA   \n",
       "1376  TCRB             NaN             Covid19            gDNA   \n",
       "\n",
       "                               disease_subtype  \\\n",
       "0             Healthy/Background - CMV Unknown   \n",
       "1             Healthy/Background - CMV Unknown   \n",
       "2             Healthy/Background - CMV Unknown   \n",
       "3             Healthy/Background - CMV Unknown   \n",
       "4                    Healthy/Background - CMV+   \n",
       "...                                        ...   \n",
       "1372  COVID-19-HUniv12Oct - Hospitalized - ICU   \n",
       "1373        COVID-19-HUniv12Oct - Hospitalized   \n",
       "1374         COVID-19-NIH/NIAID - Hospitalized   \n",
       "1375   COVID-19-NIH/NIAID - Hospitalized - ICU   \n",
       "1376        COVID-19-HUniv12Oct - Hospitalized   \n",
       "\n",
       "                       participant_label  \\\n",
       "0     emerson-2017-natgen_train_HIP15860   \n",
       "1     emerson-2017-natgen_train_HIP14363   \n",
       "2     emerson-2017-natgen_train_HIP14178   \n",
       "3     emerson-2017-natgen_train_HIP13944   \n",
       "4     emerson-2017-natgen_train_HIP13911   \n",
       "...                                  ...   \n",
       "1372        immunecode_ImmuneCode-304752   \n",
       "1373        immunecode_ImmuneCode-775827   \n",
       "1374       immunecode_ImmuneCode-0000446   \n",
       "1375       immunecode_ImmuneCode-0000160   \n",
       "1376        immunecode_ImmuneCode-506345   \n",
       "\n",
       "                                         specimen_label  \\\n",
       "0                    emerson-2017-natgen_train_HIP15860   \n",
       "1                    emerson-2017-natgen_train_HIP14363   \n",
       "2                    emerson-2017-natgen_train_HIP14178   \n",
       "3                    emerson-2017-natgen_train_HIP13944   \n",
       "4                    emerson-2017-natgen_train_HIP13911   \n",
       "...                                                 ...   \n",
       "1372                          immunecode_860011117_TCRB   \n",
       "1373                          immunecode_860011106_TCRB   \n",
       "1374             immunecode_BS-GIGI_10-replacement_TCRB   \n",
       "1375  immunecode_BS-EQ-25-T1_BS-GIGI-71-replacement_...   \n",
       "1376                          immunecode_860011240_TCRB   \n",
       "\n",
       "                                    amplification_label  \\\n",
       "0                    emerson-2017-natgen_train_HIP15860   \n",
       "1                    emerson-2017-natgen_train_HIP14363   \n",
       "2                    emerson-2017-natgen_train_HIP14178   \n",
       "3                    emerson-2017-natgen_train_HIP13944   \n",
       "4                    emerson-2017-natgen_train_HIP13911   \n",
       "...                                                 ...   \n",
       "1372                          immunecode_860011117_TCRB   \n",
       "1373                          immunecode_860011106_TCRB   \n",
       "1374             immunecode_BS-GIGI_10-replacement_TCRB   \n",
       "1375  immunecode_BS-EQ-25-T1_BS-GIGI-71-replacement_...   \n",
       "1376                          immunecode_860011240_TCRB   \n",
       "\n",
       "                                        replicate_label      Sex        Age  \\\n",
       "0                    emerson-2017-natgen_train_HIP15860  Unknown    Unknown   \n",
       "1                    emerson-2017-natgen_train_HIP14363  Unknown    Unknown   \n",
       "2                    emerson-2017-natgen_train_HIP14178  Unknown    Unknown   \n",
       "3                    emerson-2017-natgen_train_HIP13944  Unknown    Unknown   \n",
       "4                    emerson-2017-natgen_train_HIP13911     Male  22.469541   \n",
       "...                                                 ...      ...        ...   \n",
       "1372                          immunecode_860011117_TCRB        M       60.0   \n",
       "1373                          immunecode_860011106_TCRB        F       57.0   \n",
       "1374             immunecode_BS-GIGI_10-replacement_TCRB        M       75.0   \n",
       "1375  immunecode_BS-EQ-25-T1_BS-GIGI-71-replacement_...        M       72.0   \n",
       "1376                          immunecode_860011240_TCRB        F       78.0   \n",
       "\n",
       "                          Ethnicity specimen_description       primer_set  \\\n",
       "0                           Unknown                  NaN  Human-TCRB-PD1x   \n",
       "1                           Unknown                  NaN  Human-TCRB-PD1x   \n",
       "2                           Unknown                  NaN  Human-TCRB-PD1x   \n",
       "3                           Unknown                  NaN  Human-TCRB-PD1x   \n",
       "4     White, Not Hispanic or Latino                  NaN  Human-TCRB-PD1x   \n",
       "...                             ...                  ...              ...   \n",
       "1372                      Caucasian                  NaN              NaN   \n",
       "1373                      Caucasian                  NaN              NaN   \n",
       "1374                      Caucasian                  NaN              NaN   \n",
       "1375                      Caucasian                  NaN              NaN   \n",
       "1376                      Caucasian                  NaN              NaN   \n",
       "\n",
       "     symptoms_cmv  \n",
       "0             NaN  \n",
       "1             NaN  \n",
       "2             NaN  \n",
       "3             NaN  \n",
       "4            CMV+  \n",
       "...           ...  \n",
       "1372          NaN  \n",
       "1373          NaN  \n",
       "1374          NaN  \n",
       "1375          NaN  \n",
       "1376          NaN  \n",
       "\n",
       "[1377 rows x 17 columns]"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset to these surviving columns\n",
    "dfs_adaptive_manual = dfs_adaptive_manual[\n",
    "    [\n",
    "        \"study\",\n",
    "        \"sample_name\",\n",
    "        \"locus\",\n",
    "        \"counting_method\",\n",
    "        \"disease\",\n",
    "        \"sequencing_type\",\n",
    "        \"disease_subtype\",\n",
    "        \"participant_label\",\n",
    "        \"specimen_label\",\n",
    "        \"amplification_label\",\n",
    "        \"replicate_label\",\n",
    "        # \"cell_type\",  # optional\n",
    "        \"Sex\",\n",
    "        \"Age\",\n",
    "        \"Ethnicity\",\n",
    "        \"specimen_description\",\n",
    "        # \"participant_description\",\n",
    "        \"primer_set\",\n",
    "    ]\n",
    "    + symptoms_columns\n",
    "].rename(columns={col: f\"symptoms_{col}\" for col in symptoms_columns})\n",
    "dfs_adaptive_manual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_specimens = dfs_adaptive_manual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make metadata columns consistent with standard Boydlab pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_specimens = all_specimens.rename(\n",
    "    columns={\n",
    "        \"Age\": \"age\",\n",
    "        \"Sex\": \"sex\",\n",
    "        \"Ethnicity\": \"ethnicity\",\n",
    "        \"study\": \"study_name\",\n",
    "    }\n",
    ").assign(has_BCR=False, has_TCR=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Male       533\n",
       "Female     471\n",
       "M          109\n",
       "F           92\n",
       "Unknown     24\n",
       "Name: sex, dtype: int64"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"sex\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Consolidate\n",
    "all_specimens[\"sex\"] = (\n",
    "    all_specimens[\"sex\"].str.upper().str.strip().replace({\"MALE\": \"M\", \"FEMALE\": \"F\"})\n",
    ")\n",
    "all_specimens[\"sex\"].mask(all_specimens[\"sex\"] == \"UNKNOWN\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "M    642\n",
       "F    563\n",
       "Name: sex, dtype: int64"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"sex\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    1113\n",
       "True      264\n",
       "Name: ethnicity, dtype: int64"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"ethnicity\"].isna().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Healthy/Background    124\n",
       "HIV                    61\n",
       "RA                     39\n",
       "CVID                   37\n",
       "Covid19                 3\n",
       "Name: disease, dtype: int64"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here's who is missing ethnicity:\n",
    "all_specimens[all_specimens[\"ethnicity\"].isna()][\"disease\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TCRBv4-control               88\n",
       "towlerton-2022-hiv           61\n",
       "ramesh-2015-ci               59\n",
       "mustjoki-2017-natcomms       52\n",
       "immunecode                    3\n",
       "emerson-2017-natgen_train     1\n",
       "Name: study_name, dtype: int64"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here's who is missing ethnicity:\n",
    "all_specimens[all_specimens[\"ethnicity\"].isna()][\"study_name\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "White, Not Hispanic or Latino                                        460\n",
       "Unknown                                                              226\n",
       "White - Non-Hispanic                                                 200\n",
       "Caucasian                                                             74\n",
       "Asian, Not Hispanic or Latino                                         42\n",
       "Unknown, Hispanic or Latino                                           26\n",
       "White - Hispanic                                                      19\n",
       "White - Hispanic                                                      16\n",
       "Black or African American, Not Hispanic or Latino                     11\n",
       "Hispanic/Latino                                                       10\n",
       "American Indian or Alaska Native, Not Hispanic or Latino              10\n",
       "Asian - Non-Hispanic                                                   4\n",
       "Native Hawaiian or other Pacific Islander, Not Hispanic or Latino      3\n",
       "White, Hispanic or Latino                                              3\n",
       "Black - Non-Hispanic                                                   2\n",
       "Other, Not Hispanic or Latino                                          2\n",
       "Asian, Hispanic or Latino                                              1\n",
       "Not reported - Not reported                                            1\n",
       "Other, Hispanic or Latino                                              1\n",
       "Hispanic - Hispanic                                                    1\n",
       "Asian                                                                  1\n",
       "Name: ethnicity, dtype: int64"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"ethnicity\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Condense rare ethnicity names\n",
    "all_specimens[\"ethnicity_condensed\"] = (\n",
    "    all_specimens[\"ethnicity\"]\n",
    "    .str.strip()\n",
    "    .replace(\n",
    "        {\n",
    "            \"Black - Non-Hispanic\": \"African\",\n",
    "            \"Black or African American, Not Hispanic or Latino\": \"African\",\n",
    "            #\n",
    "            \"White, Hispanic or Latino\": \"Hispanic/Latino\",\n",
    "            \"White - Hispanic\": \"Hispanic/Latino\",\n",
    "            \"Hispanic - Hispanic\": \"Hispanic/Latino\",\n",
    "            \"Unknown, Hispanic or Latino\": \"Hispanic/Latino\",\n",
    "            \"Other, Hispanic or Latino\": \"Hispanic/Latino\",\n",
    "            #\n",
    "            \"Asian - Non-Hispanic\": \"Asian\",\n",
    "            \"Asian, Not Hispanic or Latino\": \"Asian\",\n",
    "            #\n",
    "            \"American Indian or Alaska Native, Not Hispanic or Latino\": \"Native American\",\n",
    "            #\n",
    "            \"Native Hawaiian or other Pacific Islander, Not Hispanic or Latino\": \"Pacific Islander\",\n",
    "            #\n",
    "            \"White - Non-Hispanic\": \"Caucasian\",\n",
    "            \"White, Not Hispanic or Latino\": \"Caucasian\",\n",
    "            #\n",
    "            \"Not reported - Not reported\": np.nan,\n",
    "            \"Unknown\": np.nan,\n",
    "            \"Other, Not Hispanic or Latino\": np.nan,\n",
    "            \"Asian, Hispanic or Latino\": np.nan,\n",
    "        }\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Caucasian           734\n",
       "Hispanic/Latino      76\n",
       "Asian                47\n",
       "African              13\n",
       "Native American      10\n",
       "Pacific Islander      3\n",
       "Name: ethnicity_condensed, dtype: int64"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"ethnicity_condensed\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    883\n",
       "True     494\n",
       "Name: ethnicity_condensed, dtype: int64"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"ethnicity_condensed\"].isna().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Healthy/Background    353\n",
       "HIV                    61\n",
       "RA                     39\n",
       "CVID                   37\n",
       "Covid19                 3\n",
       "T1D                     1\n",
       "Name: disease, dtype: int64"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here's who is missing ethnicity_condensed:\n",
    "all_specimens[all_specimens[\"ethnicity_condensed\"].isna()][\"disease\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Unknown                          226\n",
       "Other, Not Hispanic or Latino      2\n",
       "Asian, Hispanic or Latino          1\n",
       "Not reported - Not reported        1\n",
       "Name: ethnicity, dtype: int64"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here's who is missing ethnicity_condensed:\n",
    "# *Important*: If we see entries here that can be resolved, update the ethnicity_condensed rules above.\n",
    "all_specimens[all_specimens[\"ethnicity_condensed\"].isna()][\"ethnicity\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "emerson-2017-natgen_train         227\n",
       "TCRBv4-control                     88\n",
       "towlerton-2022-hiv                 61\n",
       "ramesh-2015-ci                     59\n",
       "mustjoki-2017-natcomms             52\n",
       "emerson-2017-natgen_validation      3\n",
       "immunecode                          3\n",
       "mitchell-2022-jcii                  1\n",
       "Name: study_name, dtype: int64"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here's who is missing ethnicity_condensed:\n",
    "all_specimens[all_specimens[\"ethnicity_condensed\"].isna()][\"study_name\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Healthy/Background    1009\n",
       "T1D                    143\n",
       "Covid19                 88\n",
       "HIV                     61\n",
       "RA                      39\n",
       "CVID                    37\n",
       "Name: disease, dtype: int64"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Versus total counts\n",
    "all_specimens[\"disease\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ethnicity_condensed  disease           \n",
       "African              Healthy/Background     11\n",
       "                     T1D                     2\n",
       "Asian                Covid19                 1\n",
       "                     Healthy/Background     42\n",
       "                     T1D                     4\n",
       "Caucasian            Covid19                74\n",
       "                     Healthy/Background    544\n",
       "                     T1D                   116\n",
       "Hispanic/Latino      Covid19                10\n",
       "                     Healthy/Background     46\n",
       "                     T1D                    20\n",
       "Native American      Healthy/Background     10\n",
       "Pacific Islander     Healthy/Background      3\n",
       "dtype: int64"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens.groupby([\"ethnicity_condensed\", \"disease\"]).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         Unknown\n",
       "1         Unknown\n",
       "2         Unknown\n",
       "3         Unknown\n",
       "4       22.469541\n",
       "          ...    \n",
       "1372         60.0\n",
       "1373         57.0\n",
       "1374         75.0\n",
       "1375         72.0\n",
       "1376         78.0\n",
       "Name: age, Length: 1229, dtype: object"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"age\"].dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove \"Unknown\"\n",
    "all_specimens[\"age\"].mask(all_specimens[\"age\"] == \"Unknown\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now we can convert to float\n",
    "all_specimens[\"age\"] = all_specimens[\"age\"].astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    1118.000000\n",
       "mean       33.563947\n",
       "std        18.811632\n",
       "min         0.717316\n",
       "25%        18.277259\n",
       "50%        33.567420\n",
       "75%        47.724846\n",
       "max        89.000000\n",
       "Name: age, dtype: float64"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set age_group column as well, just as in assemble_etl_metadata\n",
    "all_specimens[\"age\"].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<20      293\n",
       "30-40    200\n",
       "40-50    200\n",
       "20-30    195\n",
       "50-60    142\n",
       "60-70     53\n",
       "70-80     24\n",
       "80+       11\n",
       "Name: age_group, dtype: int64"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"age_group\"] = pd.cut(\n",
    "    all_specimens[\"age\"],\n",
    "    bins=[0, 20, 30, 40, 50, 60, 70, 80, 100],\n",
    "    labels=[\"<20\", \"20-30\", \"30-40\", \"40-50\", \"50-60\", \"60-70\", \"70-80\", \"80+\"],\n",
    "    right=False,\n",
    ")\n",
    "all_specimens[\"age_group\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['<20', '20-30', '30-40', '40-50', '50-60', '60-70', '70-80', '80+'], dtype='object')"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"age_group\"].cat.categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    1118\n",
       "True      259\n",
       "Name: age, dtype: int64"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"age\"].isna().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    1118\n",
       "True      259\n",
       "Name: age_group, dtype: int64"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"age_group\"].isna().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<20 0.717316 19.843942\n",
      "20-30 20.0 29.987679671457904\n",
      "30-40 30.0 39.99452429842574\n",
      "40-50 40.0 49.91375770020534\n",
      "50-60 50.0 59.97262149212868\n",
      "60-70 60.0 69.6\n",
      "70-80 70.0 79.0\n",
      "80+ 80.0 89.0\n"
     ]
    }
   ],
   "source": [
    "for age_group, grp in all_specimens.groupby(\"age_group\"):\n",
    "    print(age_group, grp[\"age\"].min(), grp[\"age\"].max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just as in assemble_etl_metadata:\n",
    "# Null out \"age_group\" column for extreme ages with small sample size.\n",
    "\n",
    "# Note that we are not getting rid of these specimens altogether,\n",
    "# but marking age_group NaN will disable their use for demographics-controlling models\n",
    "\n",
    "orig_shapes = all_specimens.shape[0], all_specimens[\"age_group\"].isna().sum()\n",
    "mask = all_specimens[\"age_group\"].isin([\"80+\"])\n",
    "all_specimens.loc[mask, \"age_group\"] = np.nan\n",
    "new_shapes = all_specimens.shape[0], all_specimens[\"age_group\"].isna().sum()\n",
    "\n",
    "# sanity checks:\n",
    "# - we did not drop any specimens\n",
    "assert orig_shapes[0] == new_shapes[0]\n",
    "# - but we did null out some age_group entries\n",
    "assert orig_shapes[1] < new_shapes[1]\n",
    "# - we nulled out the right amount\n",
    "assert new_shapes[1] - orig_shapes[1] == mask.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # Fillna for cohorts that are single-locus\n",
    "# if \"specimen_label_by_locus\" not in all_specimens:\n",
    "#     # in case we had no BCR+TCR combined cohorts that set this field already\n",
    "#     all_specimens[\"specimen_label_by_locus\"] = all_specimens[\"specimen_label\"]\n",
    "# else:\n",
    "#     all_specimens[\"specimen_label_by_locus\"].fillna(\n",
    "#         all_specimens[\"specimen_label\"], inplace=True\n",
    "#     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {
    "lines_to_next_cell": 0,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # make sure input fnames exist\n",
    "# assert all_specimens[\"fname\"].apply(os.path.exists).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1377, 21)"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # confirm all specimen labels are unique within each locus (may have one BCR and one TCR line per specimen)\n",
    "# # TODO: in the future, allow for replicates of each specimen\n",
    "# assert not all_specimens[\"specimen_label_by_locus\"].duplicated().any()\n",
    "# for locus, grp in all_specimens.groupby(\"gene_locus\"):\n",
    "#     assert not grp[\"specimen_label\"].duplicated().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Which specimens are in multiple loci?\n",
    "# all_specimens[all_specimens[\"specimen_label\"].duplicated(keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "emerson-2017-natgen_train         666\n",
       "mitchell-2022-jcii                243\n",
       "emerson-2017-natgen_validation    120\n",
       "TCRBv4-control                     88\n",
       "immunecode                         88\n",
       "towlerton-2022-hiv                 61\n",
       "ramesh-2015-ci                     59\n",
       "mustjoki-2017-natcomms             52\n",
       "Name: study_name, dtype: int64"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"study_name\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Healthy/Background    1009\n",
       "T1D                    143\n",
       "Covid19                 88\n",
       "HIV                     61\n",
       "RA                      39\n",
       "CVID                    37\n",
       "Name: disease, dtype: int64"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"disease\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TCRB    1377\n",
       "Name: locus, dtype: int64"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"locus\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Healthy/Background - CMV-                   420\n",
       "Healthy/Background - CMV+                   340\n",
       "T1D - new onset                             143\n",
       "Healthy/Background - T1D negative           100\n",
       "Healthy/Background - TCRBv4-control          88\n",
       "HIV - before anti-retroviral therapy         61\n",
       "COVID-19-NIH/NIAID - Hospitalized            39\n",
       "CVID                                         37\n",
       "RA - sero-positive                           32\n",
       "Healthy/Background - CMV Unknown             26\n",
       "COVID-19-HUniv12Oct - Hospitalized           26\n",
       "Healthy/Background - CVID negative           22\n",
       "Healthy/Background - RA negative             13\n",
       "COVID-19-ISB                                  9\n",
       "RA - sero-negative                            7\n",
       "COVID-19-NIH/NIAID - Hospitalized - ICU       7\n",
       "COVID-19-HUniv12Oct - Hospitalized - ICU      6\n",
       "COVID-19-HUniv12Oct                           1\n",
       "Name: disease_subtype, dtype: int64"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_specimens[\"disease_subtype\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CVID\n",
      "CVID    37\n",
      "Name: disease_subtype, dtype: int64\n",
      "\n",
      "Covid19\n",
      "COVID-19-NIH/NIAID - Hospitalized           39\n",
      "COVID-19-HUniv12Oct - Hospitalized          26\n",
      "COVID-19-ISB                                 9\n",
      "COVID-19-NIH/NIAID - Hospitalized - ICU      7\n",
      "COVID-19-HUniv12Oct - Hospitalized - ICU     6\n",
      "COVID-19-HUniv12Oct                          1\n",
      "Name: disease_subtype, dtype: int64\n",
      "\n",
      "HIV\n",
      "HIV - before anti-retroviral therapy    61\n",
      "Name: disease_subtype, dtype: int64\n",
      "\n",
      "Healthy/Background\n",
      "Healthy/Background - CMV-              420\n",
      "Healthy/Background - CMV+              340\n",
      "Healthy/Background - T1D negative      100\n",
      "Healthy/Background - TCRBv4-control     88\n",
      "Healthy/Background - CMV Unknown        26\n",
      "Healthy/Background - CVID negative      22\n",
      "Healthy/Background - RA negative        13\n",
      "Name: disease_subtype, dtype: int64\n",
      "\n",
      "RA\n",
      "RA - sero-positive    32\n",
      "RA - sero-negative     7\n",
      "Name: disease_subtype, dtype: int64\n",
      "\n",
      "T1D\n",
      "T1D - new onset    143\n",
      "Name: disease_subtype, dtype: int64\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for key, grp in all_specimens.groupby(\"disease\"):\n",
    "    print(key)\n",
    "    print(grp[\"disease_subtype\"].value_counts())\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CVID\n",
      "Series([], Name: specimen_description, dtype: int64)\n",
      "\n",
      "Covid19\n",
      "Series([], Name: specimen_description, dtype: int64)\n",
      "\n",
      "HIV\n",
      "<= 0 month Pre-ART     15\n",
      "<= 1 year Pre-ART      14\n",
      "<= 6 months Pre-ART    12\n",
      "<= 3 months Pre-ART    10\n",
      "<= 2 years Pre-ART      5\n",
      "<= 3 years Pre-ART      3\n",
      "<= 4 years Pre-ART      2\n",
      "Name: specimen_description, dtype: int64\n",
      "\n",
      "Healthy/Background\n",
      "Series([], Name: specimen_description, dtype: int64)\n",
      "\n",
      "RA\n",
      "Series([], Name: specimen_description, dtype: int64)\n",
      "\n",
      "T1D\n",
      "Series([], Name: specimen_description, dtype: int64)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for key, grp in all_specimens.groupby(\"disease\"):\n",
    "    print(key)\n",
    "    print(grp[\"specimen_description\"].value_counts())\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "age\n",
      "27.000000    15\n",
      "25.000000    15\n",
      "23.000000    14\n",
      "24.000000    13\n",
      "22.000000    12\n",
      "             ..\n",
      "14.918549     1\n",
      "54.067077     1\n",
      "64.082136     1\n",
      "44.136893     1\n",
      "72.000000     1\n",
      "Name: age, Length: 830, dtype: int64\n",
      "False    1118\n",
      "True      259\n",
      "Name: age, dtype: int64\n",
      "\n",
      "age_group\n",
      "<20      293\n",
      "30-40    200\n",
      "40-50    200\n",
      "20-30    195\n",
      "50-60    142\n",
      "60-70     53\n",
      "70-80     24\n",
      "80+        0\n",
      "Name: age_group, dtype: int64\n",
      "False    1107\n",
      "True      270\n",
      "Name: age_group, dtype: int64\n",
      "\n",
      "sex\n",
      "M    642\n",
      "F    563\n",
      "Name: sex, dtype: int64\n",
      "False    1205\n",
      "True      172\n",
      "Name: sex, dtype: int64\n",
      "\n",
      "ethnicity_condensed\n",
      "Caucasian           734\n",
      "Hispanic/Latino      76\n",
      "Asian                47\n",
      "African              13\n",
      "Native American      10\n",
      "Pacific Islander      3\n",
      "Name: ethnicity_condensed, dtype: int64\n",
      "False    883\n",
      "True     494\n",
      "Name: ethnicity_condensed, dtype: int64\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for demographics_column in [\"age\", \"age_group\", \"sex\", \"ethnicity_condensed\"]:\n",
    "    print(demographics_column)\n",
    "    print(all_specimens[demographics_column].value_counts())\n",
    "    print(all_specimens[demographics_column].isna().value_counts())\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_specimens.drop(columns=[\"ethnicity\"]).to_csv(\n",
    "    config.paths.metadata_dir / \"adaptive\" / \"generated.adaptive_external_cohorts.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index=None,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "py39-cuda-env",
   "language": "python",
   "name": "py39-cuda-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
