{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e39262b3-62c1-48ab-ae59-c606321e1c80",
   "metadata": {},
   "source": [
    "# Mal-ID Tutorial\n",
    "\n",
    "Mal-ID uses B cell receptor (BCR) and T cell receptor (TCR) sequencing data to classify disease or immune state.\n",
    "\n",
    "In the Mal-ID framework, we train three BCR and three TCR disease classifiers with different ways of extracting feature from immune receptor sequencing data.\n",
    "Then we train an ensemble metamodel, a classifier that uses the predicted disease probabilities from the six base models to make a final prediction of disease status.\n",
    "\n",
    "#### This tutorial covers:\n",
    "\n",
    "1. Mal-ID configuration settings\n",
    "2. Accessing sample metadata\n",
    "3. Loading sequence data\n",
    "4. Loading Models 1, 2, and 3 and making predictions with these base models\n",
    "5. Loading the ensemble metamodel and making predictions\n",
    "\n",
    "We will cover all of the components in this schematic:\n",
    "\n",
    "<div>\n",
    "<img src=\"../schematic.png\" width=\"800\"/>\n",
    "</div>\n",
    "\n",
    "This tutorial assumes you've already followed the \"runbook\": the commands in the readme to load the data and train the models.\n",
    "\n",
    "**([Main repo here](https://github.com/maximz/malid))**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "340665c0-4d1e-4e55-a33a-8ada32aa2fa2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8d187d8f-155e-49d2-801f-eabd74c732a3",
   "metadata": {},
   "source": [
    "## Start with some necessary imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a0e58e05-3304-4352-b08a-305373e03f9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from malid import config, io, helpers\n",
    "from malid.trained_model_wrappers import (\n",
    "    RepertoireClassifier,\n",
    "    ConvergentClusterClassifier,\n",
    "    VJGeneSpecificSequenceModelRollupClassifier,\n",
    "    BlendingMetamodel,\n",
    ")\n",
    "from malid.datamodels import GeneLocus, TargetObsColumnEnum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cb1a69a-006f-4c68-b7cb-b531bc9c850a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "62ced55a-61d2-4c60-a183-8d82dc318e2c",
   "metadata": {},
   "source": [
    "## Review Mal-ID configuration settings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0fed4c6-123e-455d-bfa6-ace23662039e",
   "metadata": {},
   "source": [
    "The inclusion criteria for the dataset — meaning which samples get divided into cross validation folds — are defined as a `CrossValidationSplitStrategy` object in `malid/datamodels.py`.\n",
    "\n",
    "The default strategy is `CrossValidationSplitStrategy.in_house_peak_disease_timepoints`, which includes peak disease timepoints from our primary in-house dataset. Indeed, that's the active cross validation split strategy:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5681bee7-6d8d-4b3b-9cfb-b488acbc4c16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<CrossValidationSplitStrategy.in_house_peak_disease_timepoints: CrossValidationSplitStrategyValue(data_sources_keep=[<DataSource.in_house: 1>], stratify_by='disease', diseases_to_keep_all_subtypes=['Healthy/Background', 'HIV', 'Lupus', 'T1D'], subtypes_keep=['Covid19 - Sero-positive (ICU)', 'Covid19 - Sero-positive (Admit)', 'Covid19 - Acute 2', 'Covid19 - Admit', 'Covid19 - ICU', 'Influenza vaccine 2021 - day 7'], filter_specimens_func_by_study_name={'Covid19-buffycoat': <function acute_disease_choose_most_peak_timepoint at 0x7f1856a5edc0>, 'Covid19-Stanford': <function acute_disease_choose_most_peak_timepoint at 0x7f1856a5edc0>}, gene_loci_supported=<GeneLocus.BCR|TCR: 3>, exclude_study_names=['IBD pre-pandemic Yoni'], include_study_names=None, filter_out_specimens_funcs_global=[], study_names_for_held_out_set=None)>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config.cross_validation_split_strategy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d0035e6-6cfd-470d-818f-0e6a0d37f061",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9e1d1ae9-7c02-4a99-a7ee-b56a55caaa8a",
   "metadata": {},
   "source": [
    "The data is divided into three folds: `fold_id` can be 0, 1, or 2.\n",
    "\n",
    "Each fold has a `train_smaller`, `validation`, and `test` set, referred to as a `fold_label`. (The `train_smaller` set is further subdivided into `train_smaller1` and `train_smaller2`.)\n",
    "\n",
    "Each sample will be in one test set. We also make sure that all samples from the same person have the same `fold_label`.\n",
    "\n",
    "Finally, we define a special `fold_id=-1` \"global\" fold that does not have a `test` set. All the data is instead used in the `train_smaller` and `validation` fold labels. (The `train_smaller` to `validation` proportion is the same as for other fold IDs, but both sets are larger than usual.)\n",
    "\n",
    "<div>\n",
    "<img src=\"../cross_validation.png\" width=\"600\"/>\n",
    "</div>\n",
    "\n",
    "\n",
    "The list of all fold IDs is therefore 0, 1, 2, and -1:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "65b185d4-e189-481f-ab68-269e4c82fdb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 1, 2, -1]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config.all_fold_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6db4fc7-318f-4fc7-bb51-4d5c7e74ed4f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bf844ecb-a212-4769-bd0a-ab0ccb82c7cc",
   "metadata": {},
   "source": [
    "The language model being used is ESM-2, applied to the CDR3 region:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2a9debdd-32be-45f2-89de-5016a4d63518",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'esm2_cdr3'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config.embedder.name  # The name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0271fc1a-8b7a-4bee-a8d5-412139a72686",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<EmbedderSequenceContent.CDR3: 3>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config.embedder.embedder_sequence_content  # The sequence region being embedded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "baad9399-30d3-46f4-9ae4-cffb7867e7b6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "14e15c7e-3399-4b7f-86c4-0c0726e7f43b",
   "metadata": {},
   "source": [
    "Our models are configured to use both BCR and TCR data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a961386b-777b-4407-9b25-181f1e0be48b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<GeneLocus.BCR|TCR: 3>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config.gene_loci_used"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0611780-ee10-4967-be71-1000c60c79c2",
   "metadata": {},
   "source": [
    "Just to clarify, `GeneLocus.BCR|TCR` means the union of BCR and TCR — both are active. This is the same as writing `GeneLocus.BCR | GeneLocus.TCR`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "49520946-c2e4-4a8d-a24d-dc3d9e1c13c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<GeneLocus.BCR|TCR: 3>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GeneLocus.BCR | GeneLocus.TCR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e64775dd-5666-4993-b7c5-6de20bd5971c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e6005a6-03bb-4649-b908-a9910a9a0ced",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3ba686b2-8396-4103-887d-b416030fd119",
   "metadata": {},
   "source": [
    "## Load metadata\n",
    "\n",
    "Here's a Pandas DataFrame with all of the samples in our database.\n",
    "\n",
    "The key fields are:\n",
    "\n",
    "- `participant_label`: the patient ID\n",
    "- `specimen_label`: the sample ID\n",
    "- `disease`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a9f0265d-9961-46bb-81b6-c9f9cd50ef54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant_label</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>specimen_time_point</th>\n",
       "      <th>participant_description</th>\n",
       "      <th>data_source</th>\n",
       "      <th>study_name</th>\n",
       "      <th>available_gene_loci</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>age</th>\n",
       "      <th>...</th>\n",
       "      <th>symptoms_cmv</th>\n",
       "      <th>symptoms_healthy_in_resequencing_experiment</th>\n",
       "      <th>specimen_time_point_days</th>\n",
       "      <th>survived_filters</th>\n",
       "      <th>is_selected_for_cv_strategy</th>\n",
       "      <th>in_training_set</th>\n",
       "      <th>past_exposure</th>\n",
       "      <th>disease.separate_past_exposures</th>\n",
       "      <th>disease.rollup</th>\n",
       "      <th>test_fold_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BFI-0000234</td>\n",
       "      <td>M124-S014</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Location: USA</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>HIV</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>27.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BFI-0000234</td>\n",
       "      <td>M132-S014</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Location: USA</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>HIV</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>27.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BFI-0002850</td>\n",
       "      <td>M124-S042</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Location: USA</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>HIV</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BFI-0002850</td>\n",
       "      <td>M132-S040</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Location: USA</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>HIV</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BFI-0002851</td>\n",
       "      <td>M124-S041</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Location: USA</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>HIV</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>27.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2540</th>\n",
       "      <td>ramesh-2015-ci_378</td>\n",
       "      <td>ramesh-2015-ci_378</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataSource.adaptive</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>(((GeneLocus.TCR)))</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2541</th>\n",
       "      <td>ramesh-2015-ci_386</td>\n",
       "      <td>ramesh-2015-ci_386</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataSource.adaptive</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>(((GeneLocus.TCR)))</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2542</th>\n",
       "      <td>ramesh-2015-ci_400</td>\n",
       "      <td>ramesh-2015-ci_400</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataSource.adaptive</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>(((GeneLocus.TCR)))</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2543</th>\n",
       "      <td>ramesh-2015-ci_441</td>\n",
       "      <td>ramesh-2015-ci_441</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataSource.adaptive</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>(((GeneLocus.TCR)))</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2544</th>\n",
       "      <td>ramesh-2015-ci_462</td>\n",
       "      <td>ramesh-2015-ci_462</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataSource.adaptive</td>\n",
       "      <td>ramesh-2015-ci</td>\n",
       "      <td>(((GeneLocus.TCR)))</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CVID</td>\n",
       "      <td>CVID</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2545 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       participant_label      specimen_label             disease  \\\n",
       "0            BFI-0000234           M124-S014  Healthy/Background   \n",
       "1            BFI-0000234           M132-S014  Healthy/Background   \n",
       "2            BFI-0002850           M124-S042  Healthy/Background   \n",
       "3            BFI-0002850           M132-S040  Healthy/Background   \n",
       "4            BFI-0002851           M124-S041  Healthy/Background   \n",
       "...                  ...                 ...                 ...   \n",
       "2540  ramesh-2015-ci_378  ramesh-2015-ci_378                CVID   \n",
       "2541  ramesh-2015-ci_386  ramesh-2015-ci_386                CVID   \n",
       "2542  ramesh-2015-ci_400  ramesh-2015-ci_400                CVID   \n",
       "2543  ramesh-2015-ci_441  ramesh-2015-ci_441                CVID   \n",
       "2544  ramesh-2015-ci_462  ramesh-2015-ci_462                CVID   \n",
       "\n",
       "     specimen_time_point participant_description          data_source  \\\n",
       "0                    NaN           Location: USA  DataSource.in_house   \n",
       "1                    NaN           Location: USA  DataSource.in_house   \n",
       "2                    NaN           Location: USA  DataSource.in_house   \n",
       "3                    NaN           Location: USA  DataSource.in_house   \n",
       "4                    NaN           Location: USA  DataSource.in_house   \n",
       "...                  ...                     ...                  ...   \n",
       "2540                 NaN                     NaN  DataSource.adaptive   \n",
       "2541                 NaN                     NaN  DataSource.adaptive   \n",
       "2542                 NaN                     NaN  DataSource.adaptive   \n",
       "2543                 NaN                     NaN  DataSource.adaptive   \n",
       "2544                 NaN                     NaN  DataSource.adaptive   \n",
       "\n",
       "          study_name                     available_gene_loci  \\\n",
       "0                HIV  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "1                HIV  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "2                HIV  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "3                HIV  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "4                HIV  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "...              ...                                     ...   \n",
       "2540  ramesh-2015-ci                     (((GeneLocus.TCR)))   \n",
       "2541  ramesh-2015-ci                     (((GeneLocus.TCR)))   \n",
       "2542  ramesh-2015-ci                     (((GeneLocus.TCR)))   \n",
       "2543  ramesh-2015-ci                     (((GeneLocus.TCR)))   \n",
       "2544  ramesh-2015-ci                     (((GeneLocus.TCR)))   \n",
       "\n",
       "                        disease_subtype   age  ... symptoms_cmv  \\\n",
       "0     Healthy/Background - HIV Negative  27.0  ...          NaN   \n",
       "1     Healthy/Background - HIV Negative  27.0  ...          NaN   \n",
       "2     Healthy/Background - HIV Negative  26.0  ...          NaN   \n",
       "3     Healthy/Background - HIV Negative  26.0  ...          NaN   \n",
       "4     Healthy/Background - HIV Negative  27.0  ...          NaN   \n",
       "...                                 ...   ...  ...          ...   \n",
       "2540                               CVID   NaN  ...          NaN   \n",
       "2541                               CVID   NaN  ...          NaN   \n",
       "2542                               CVID   NaN  ...          NaN   \n",
       "2543                               CVID   NaN  ...          NaN   \n",
       "2544                               CVID   NaN  ...          NaN   \n",
       "\n",
       "     symptoms_healthy_in_resequencing_experiment specimen_time_point_days  \\\n",
       "0                                            NaN                      NaN   \n",
       "1                                            NaN                      NaN   \n",
       "2                                            NaN                      NaN   \n",
       "3                                            NaN                      NaN   \n",
       "4                                            NaN                      NaN   \n",
       "...                                          ...                      ...   \n",
       "2540                                         NaN                      NaN   \n",
       "2541                                         NaN                      NaN   \n",
       "2542                                         NaN                      NaN   \n",
       "2543                                         NaN                      NaN   \n",
       "2544                                         NaN                      NaN   \n",
       "\n",
       "     survived_filters is_selected_for_cv_strategy in_training_set  \\\n",
       "0                True                        True            True   \n",
       "1               False                        True           False   \n",
       "2                True                        True            True   \n",
       "3               False                        True           False   \n",
       "4                True                        True            True   \n",
       "...               ...                         ...             ...   \n",
       "2540             True                       False           False   \n",
       "2541             True                       False           False   \n",
       "2542             True                       False           False   \n",
       "2543             True                       False           False   \n",
       "2544             True                       False           False   \n",
       "\n",
       "     past_exposure disease.separate_past_exposures      disease.rollup  \\\n",
       "0            False              Healthy/Background  Healthy/Background   \n",
       "1            False              Healthy/Background  Healthy/Background   \n",
       "2            False              Healthy/Background  Healthy/Background   \n",
       "3            False              Healthy/Background  Healthy/Background   \n",
       "4            False              Healthy/Background  Healthy/Background   \n",
       "...            ...                             ...                 ...   \n",
       "2540         False                            CVID                CVID   \n",
       "2541         False                            CVID                CVID   \n",
       "2542         False                            CVID                CVID   \n",
       "2543         False                            CVID                CVID   \n",
       "2544         False                            CVID                CVID   \n",
       "\n",
       "     test_fold_id  \n",
       "0             1.0  \n",
       "1             NaN  \n",
       "2             0.0  \n",
       "3             NaN  \n",
       "4             2.0  \n",
       "...           ...  \n",
       "2540          NaN  \n",
       "2541          NaN  \n",
       "2542          NaN  \n",
       "2543          NaN  \n",
       "2544          NaN  \n",
       "\n",
       "[2545 rows x 57 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = helpers.get_all_specimen_info()\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8ed69dd-7159-4e6a-abef-bc46f61b6648",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "94513dba-4862-4f35-9f9e-4d794ee4b1af",
   "metadata": {},
   "source": [
    "Each sample is identified by a `specimen_label` and has a boolean column named `in_training_set`, indicating whether a sample met the requirements for inclusion in the cross validation divisions and passed QC requirements (see the readme for more details).\n",
    "\n",
    "Let's look at only the samples that passed those filters:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7a19e941-e34d-4ead-97b5-629e255b2261",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant_label</th>\n",
       "      <th>specimen_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>specimen_time_point</th>\n",
       "      <th>participant_description</th>\n",
       "      <th>data_source</th>\n",
       "      <th>study_name</th>\n",
       "      <th>available_gene_loci</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>age</th>\n",
       "      <th>...</th>\n",
       "      <th>symptoms_cmv</th>\n",
       "      <th>symptoms_healthy_in_resequencing_experiment</th>\n",
       "      <th>specimen_time_point_days</th>\n",
       "      <th>survived_filters</th>\n",
       "      <th>is_selected_for_cv_strategy</th>\n",
       "      <th>in_training_set</th>\n",
       "      <th>past_exposure</th>\n",
       "      <th>disease.separate_past_exposures</th>\n",
       "      <th>disease.rollup</th>\n",
       "      <th>test_fold_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BFI-0000234</td>\n",
       "      <td>M124-S014</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Location: USA</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>HIV</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>27.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BFI-0002850</td>\n",
       "      <td>M124-S042</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Location: USA</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>HIV</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BFI-0002851</td>\n",
       "      <td>M124-S041</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Location: USA</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>HIV</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>27.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BFI-0002852</td>\n",
       "      <td>M124-S012</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Location: USA</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>HIV</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BFI-0002861</td>\n",
       "      <td>M124-S037</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Location: Malawi</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>HIV</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>34.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2320</th>\n",
       "      <td>BFI-0010791</td>\n",
       "      <td>M491-S149</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes 35453-study biobank: pediatric; T1D +...</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>11.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2321</th>\n",
       "      <td>BFI-0010792</td>\n",
       "      <td>M491-S150</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes 35453-study biobank: adult; TID</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>T1D - adult</td>\n",
       "      <td>59.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2322</th>\n",
       "      <td>BFI-0010794</td>\n",
       "      <td>M491-S153</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes 35453-study biobank: pediatric; TID</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2323</th>\n",
       "      <td>BFI-0010800</td>\n",
       "      <td>M491-S160</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes 35453-study biobank: pediatric; T1D</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2327</th>\n",
       "      <td>BFI-0010806</td>\n",
       "      <td>M491-S167</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes 35453-study biobank: pediatric; T1D</td>\n",
       "      <td>DataSource.in_house</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "      <td>(((GeneLocus.BCR)), ((GeneLocus.TCR)))</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>616 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     participant_label specimen_label             disease specimen_time_point  \\\n",
       "0          BFI-0000234      M124-S014  Healthy/Background                 NaN   \n",
       "2          BFI-0002850      M124-S042  Healthy/Background                 NaN   \n",
       "4          BFI-0002851      M124-S041  Healthy/Background                 NaN   \n",
       "6          BFI-0002852      M124-S012  Healthy/Background                 NaN   \n",
       "8          BFI-0002861      M124-S037  Healthy/Background                 NaN   \n",
       "...                ...            ...                 ...                 ...   \n",
       "2320       BFI-0010791      M491-S149                 T1D                 NaN   \n",
       "2321       BFI-0010792      M491-S150                 T1D                 NaN   \n",
       "2322       BFI-0010794      M491-S153                 T1D                 NaN   \n",
       "2323       BFI-0010800      M491-S160                 T1D                 NaN   \n",
       "2327       BFI-0010806      M491-S167                 T1D                 NaN   \n",
       "\n",
       "                                participant_description          data_source  \\\n",
       "0                                         Location: USA  DataSource.in_house   \n",
       "2                                         Location: USA  DataSource.in_house   \n",
       "4                                         Location: USA  DataSource.in_house   \n",
       "6                                         Location: USA  DataSource.in_house   \n",
       "8                                      Location: Malawi  DataSource.in_house   \n",
       "...                                                 ...                  ...   \n",
       "2320  Diabetes 35453-study biobank: pediatric; T1D +...  DataSource.in_house   \n",
       "2321           Diabetes 35453-study biobank: adult; TID  DataSource.in_house   \n",
       "2322       Diabetes 35453-study biobank: pediatric; TID  DataSource.in_house   \n",
       "2323       Diabetes 35453-study biobank: pediatric; T1D  DataSource.in_house   \n",
       "2327       Diabetes 35453-study biobank: pediatric; T1D  DataSource.in_house   \n",
       "\n",
       "            study_name                     available_gene_loci  \\\n",
       "0                  HIV  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "2                  HIV  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "4                  HIV  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "6                  HIV  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "8                  HIV  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "...                ...                                     ...   \n",
       "2320  Diabetes biobank  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "2321  Diabetes biobank  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "2322  Diabetes biobank  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "2323  Diabetes biobank  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "2327  Diabetes biobank  (((GeneLocus.BCR)), ((GeneLocus.TCR)))   \n",
       "\n",
       "                        disease_subtype   age  ... symptoms_cmv  \\\n",
       "0     Healthy/Background - HIV Negative  27.0  ...          NaN   \n",
       "2     Healthy/Background - HIV Negative  26.0  ...          NaN   \n",
       "4     Healthy/Background - HIV Negative  27.0  ...          NaN   \n",
       "6     Healthy/Background - HIV Negative  26.0  ...          NaN   \n",
       "8     Healthy/Background - HIV Negative  34.0  ...          NaN   \n",
       "...                                 ...   ...  ...          ...   \n",
       "2320                    T1D - pediatric  11.0  ...          NaN   \n",
       "2321                        T1D - adult  59.0  ...          NaN   \n",
       "2322                    T1D - pediatric  16.0  ...          NaN   \n",
       "2323                    T1D - pediatric  16.0  ...          NaN   \n",
       "2327                    T1D - pediatric   NaN  ...          NaN   \n",
       "\n",
       "     symptoms_healthy_in_resequencing_experiment specimen_time_point_days  \\\n",
       "0                                            NaN                      NaN   \n",
       "2                                            NaN                      NaN   \n",
       "4                                            NaN                      NaN   \n",
       "6                                            NaN                      NaN   \n",
       "8                                            NaN                      NaN   \n",
       "...                                          ...                      ...   \n",
       "2320                                         NaN                      NaN   \n",
       "2321                                         NaN                      NaN   \n",
       "2322                                         NaN                      NaN   \n",
       "2323                                         NaN                      NaN   \n",
       "2327                                         NaN                      NaN   \n",
       "\n",
       "     survived_filters is_selected_for_cv_strategy in_training_set  \\\n",
       "0                True                        True            True   \n",
       "2                True                        True            True   \n",
       "4                True                        True            True   \n",
       "6                True                        True            True   \n",
       "8                True                        True            True   \n",
       "...               ...                         ...             ...   \n",
       "2320             True                        True            True   \n",
       "2321             True                        True            True   \n",
       "2322             True                        True            True   \n",
       "2323             True                        True            True   \n",
       "2327             True                        True            True   \n",
       "\n",
       "     past_exposure disease.separate_past_exposures      disease.rollup  \\\n",
       "0            False              Healthy/Background  Healthy/Background   \n",
       "2            False              Healthy/Background  Healthy/Background   \n",
       "4            False              Healthy/Background  Healthy/Background   \n",
       "6            False              Healthy/Background  Healthy/Background   \n",
       "8            False              Healthy/Background  Healthy/Background   \n",
       "...            ...                             ...                 ...   \n",
       "2320         False                             T1D                 T1D   \n",
       "2321         False                             T1D                 T1D   \n",
       "2322         False                             T1D                 T1D   \n",
       "2323         False                             T1D                 T1D   \n",
       "2327         False                             T1D                 T1D   \n",
       "\n",
       "     test_fold_id  \n",
       "0             1.0  \n",
       "2             0.0  \n",
       "4             2.0  \n",
       "6             1.0  \n",
       "8             2.0  \n",
       "...           ...  \n",
       "2320          0.0  \n",
       "2321          1.0  \n",
       "2322          0.0  \n",
       "2323          0.0  \n",
       "2327          1.0  \n",
       "\n",
       "[616 rows x 57 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = metadata[metadata[\"in_training_set\"]]\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "244ffa34-bfe9-4b12-8197-10cf6a4f2275",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0905740d-a70c-436d-bfc8-085b28201745",
   "metadata": {},
   "source": [
    "The metadata also includes `disease_subtype`, `study_name`, `age`, `sex`, and `ethnicity_condensed` (ancestry) for a particular sample, as available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19288292-728c-40c3-89bf-aa3c85ac017b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6537718f-e362-4c2f-8f40-a776ee81f219",
   "metadata": {},
   "source": [
    "## Load data\n",
    "\n",
    "Let's choose a specific fold ID we will work with. As described above, our options are 0, 1, and 2 for the cross validation folds, and -1 for the global fold. Let's choose the first fold:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ef618ec7-82c5-423b-ad6d-f461000d7121",
   "metadata": {},
   "outputs": [],
   "source": [
    "fold_id = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1da44657-1b5e-4d78-a07f-c107b35b9978",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d929c2e3-4174-4001-839c-9a1a3ade3e54",
   "metadata": {},
   "source": [
    "Classification targets are defined in an enumeration called `TargetObsColumnEnum`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "54a91ffc-26fc-4c86-8f55-b810d5067923",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['disease',\n",
       " 'disease_all_demographics_present',\n",
       " 'ethnicity_condensed_healthy_only',\n",
       " 'age_group_healthy_only',\n",
       " 'age_group_binary_healthy_only',\n",
       " 'age_group_pediatric_healthy_only',\n",
       " 'sex_healthy_only',\n",
       " 'covid_vs_healthy',\n",
       " 'hiv_vs_healthy',\n",
       " 't1d_vs_healthy',\n",
       " 'lupus_vs_healthy',\n",
       " 'flu_vs_healthy',\n",
       " 'lupus_nephritis']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[t.name for t in TargetObsColumnEnum]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56a316a7-9caf-4f2f-a5f5-a3c1de31167c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6149fd62-a95b-4806-8c2b-53b34e200463",
   "metadata": {},
   "source": [
    "Each classification target is associated with a metadata field. Let's focus on `TargetObsColumnEnum.disease`, which is our main classification goal to predict the `disease` metadata column:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9e2ff612-fd00-4290-a967-dc5aba5e5829",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_obs_column = TargetObsColumnEnum.disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c60360b2-b716-4ef7-958f-b5cb92f0f7b7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c119f382-0fc7-4628-8a05-14c56daba327",
   "metadata": {},
   "source": [
    "This target is associated with the \"disease\" metadata field. You can tell by looking at the `obs_column_name` attribute:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "033bb90e-14aa-40a3-afd6-f4e745203bbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'disease'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_obs_column.value.obs_column_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38faf4e7-83de-437a-8d3c-74c358dda80b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c6bb95cf-3063-4a80-81a0-acca23c81280",
   "metadata": {},
   "source": [
    "Here are the values of that metadata field:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "526ec58d-e7ba-47de-9688-f409278e31c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Healthy/Background    224\n",
       "HIV                    98\n",
       "Lupus                  98\n",
       "T1D                    96\n",
       "Covid19                63\n",
       "Influenza              37\n",
       "Name: disease, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata[\"disease\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01923005-64f4-4904-9b35-4b3069a1f0d0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "82b04aa1-e367-4656-bedb-781294ddb648",
   "metadata": {},
   "source": [
    "At this point, the data have been split into cross-validation folds and the sequences have been transformed into language model embeddings.\n",
    "\n",
    "BCR and TCR data can be loaded with separate calls to `io.load_fold_embeddings()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9e6845f0-fa2d-4190-9e66-32b4dda6a53d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "{\"message\": \"Reading network file from local machine cache: /users/maximz/code/boyd-immune-repertoire-classification/data/data_v_20231027/in_house_peak_disease_timepoints/embedded/esm2_cdr3/anndatas_scaled/BCR/fold.0.test.h5ad -> /srv/scratch/maximz/cache/3d20a5c23fcc35b23cb822073f9cb2d6b0bebaee62c3a334176a736d.0.test.h5ad\", \"time\": \"2024-08-07T13:45:27.002580\"}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Your filename has more than two extensions: ['.0', '.test', '.h5ad'].\n",
      "Only considering the two last: ['.test', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.0', '.test', '.h5ad'].\n",
      "Only considering the two last: ['.test', '.h5ad'].\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5669053 × 640\n",
       "    obs: 'amplification_label', 'v_gene', 'j_gene', 'disease', 'fr1_seq_aa_q_trim', 'cdr1_seq_aa_q_trim', 'fr2_seq_aa_q_trim', 'cdr2_seq_aa_q_trim', 'fr3_seq_aa_q_trim', 'cdr3_seq_aa_q_trim', 'post_seq_aa_q_trim', 'cdr3_aa_sequence_trim_len', 'extracted_isotype', 'isotype_supergroup', 'v_mut', 'num_reads', 'igh_or_tcrb_clone_id', 'total_clone_num_reads', 'num_clone_members', 'specimen_label', 'past_exposure', 'disease.separate_past_exposures', 'disease.rollup', 'v_family', 'sample_weight_isotype_rebalance', 'sample_weight_clone_size', 'study_name', 'participant_label', 'specimen_time_point', 'disease_subtype', 'study_name_condensed', 'age', 'sex', 'ethnicity_condensed', 'age_group', 'age_group_binary', 'age_group_pediatric', 'disease_severity', 'specimen_description', 'symptoms_IBD_Disease duration (yrs)', 'symptoms_IBD_Disease for ≥15 years?', 'symptoms_IBD_Disease location', 'symptoms_IBD_Estraintestinal manifestations?', 'symptoms_IBD_Flare/Remission', 'symptoms_IBD_HBI or Partial Mayo score', 'symptoms_IBD_Onset ≤18 years', 'symptoms_IBD_Stricturing/fistulizing/inflammatory/perianal/ulcerative', 'symptoms_IBD_Tx', 'symptoms_Lupus_ANA', 'symptoms_Lupus_Arthritis', 'symptoms_Lupus_Discoid_Rash', 'symptoms_Lupus_Hematologic_Disorder', 'symptoms_Lupus_Immunologic_Disorder', 'symptoms_Lupus_La +/-', 'symptoms_Lupus_Malar_Rash', 'symptoms_Lupus_Neurologic_Disorder', 'symptoms_Lupus_Number Positive', 'symptoms_Lupus_Oral_Ulcers', 'symptoms_Lupus_Photosensitivity', 'symptoms_Lupus_Renal Disorder', 'symptoms_Lupus_Ribo_P +/-', 'symptoms_Lupus_Ro +/-', 'symptoms_Lupus_Serositis', 'symptoms_Lupus_Total ACR', 'symptoms_Lupus_aPL +/-', 'symptoms_Lupus_dsDNA +/-', 'symptoms_Lupus_sm +/-', 'symptoms_Lupus_sm_nRNP +/-', 'symptoms_cmv', 'symptoms_healthy_in_resequencing_experiment', 'symptoms_healthy_old_vs_new_batch_identifier', 'isotype_proportion:IGHG', 'isotype_proportion:IGHA', 'isotype_proportion:IGHD-M', 'fold_id', 'fold_label'\n",
       "    uns: 'embedded', 'original_uns_keys', 'original_obs_columns'\n",
       "    obsm: 'X_pca'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_bcr = io.load_fold_embeddings(\n",
    "    fold_id=fold_id,\n",
    "    fold_label=\"test\",  # Load the held out data\n",
    "    gene_locus=GeneLocus.BCR,\n",
    "    target_obs_column=target_obs_column,\n",
    ")\n",
    "adata_bcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4ec9dd40-de68-4742-ac1e-77c65bda3a2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "{\"message\": \"Reading network file from local machine cache: /users/maximz/code/boyd-immune-repertoire-classification/data/data_v_20231027/in_house_peak_disease_timepoints/embedded/esm2_cdr3/anndatas_scaled/TCR/fold.0.test.h5ad -> /srv/scratch/maximz/cache/41935c853da5d38d5d39cd74a38bd148951d1e8ded05e20393494c35.0.test.h5ad\", \"time\": \"2024-08-07T13:46:41.518369\"}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Your filename has more than two extensions: ['.0', '.test', '.h5ad'].\n",
      "Only considering the two last: ['.test', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.0', '.test', '.h5ad'].\n",
      "Only considering the two last: ['.test', '.h5ad'].\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 8358079 × 640\n",
       "    obs: 'amplification_label', 'v_gene', 'j_gene', 'disease', 'fr1_seq_aa_q_trim', 'cdr1_seq_aa_q_trim', 'fr2_seq_aa_q_trim', 'cdr2_seq_aa_q_trim', 'fr3_seq_aa_q_trim', 'cdr3_seq_aa_q_trim', 'post_seq_aa_q_trim', 'cdr3_aa_sequence_trim_len', 'extracted_isotype', 'isotype_supergroup', 'v_mut', 'num_reads', 'igh_or_tcrb_clone_id', 'total_clone_num_reads', 'num_clone_members', 'specimen_label', 'past_exposure', 'disease.separate_past_exposures', 'disease.rollup', 'v_family', 'sample_weight_isotype_rebalance', 'sample_weight_clone_size', 'study_name', 'participant_label', 'specimen_time_point', 'disease_subtype', 'study_name_condensed', 'age', 'sex', 'ethnicity_condensed', 'age_group', 'age_group_binary', 'age_group_pediatric', 'disease_severity', 'specimen_description', 'symptoms_IBD_Disease duration (yrs)', 'symptoms_IBD_Disease for ≥15 years?', 'symptoms_IBD_Disease location', 'symptoms_IBD_Estraintestinal manifestations?', 'symptoms_IBD_Flare/Remission', 'symptoms_IBD_HBI or Partial Mayo score', 'symptoms_IBD_Onset ≤18 years', 'symptoms_IBD_Stricturing/fistulizing/inflammatory/perianal/ulcerative', 'symptoms_IBD_Tx', 'symptoms_Lupus_ANA', 'symptoms_Lupus_Arthritis', 'symptoms_Lupus_Discoid_Rash', 'symptoms_Lupus_Hematologic_Disorder', 'symptoms_Lupus_Immunologic_Disorder', 'symptoms_Lupus_La +/-', 'symptoms_Lupus_Malar_Rash', 'symptoms_Lupus_Neurologic_Disorder', 'symptoms_Lupus_Number Positive', 'symptoms_Lupus_Oral_Ulcers', 'symptoms_Lupus_Photosensitivity', 'symptoms_Lupus_Renal Disorder', 'symptoms_Lupus_Ribo_P +/-', 'symptoms_Lupus_Ro +/-', 'symptoms_Lupus_Serositis', 'symptoms_Lupus_Total ACR', 'symptoms_Lupus_aPL +/-', 'symptoms_Lupus_dsDNA +/-', 'symptoms_Lupus_sm +/-', 'symptoms_Lupus_sm_nRNP +/-', 'symptoms_cmv', 'symptoms_healthy_in_resequencing_experiment', 'symptoms_healthy_old_vs_new_batch_identifier', 'fold_id', 'fold_label'\n",
       "    uns: 'embedded', 'original_uns_keys', 'original_obs_columns'\n",
       "    obsm: 'X_pca'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_tcr = io.load_fold_embeddings(\n",
    "    fold_id=fold_id,\n",
    "    fold_label=\"test\",\n",
    "    gene_locus=GeneLocus.TCR,  # Same as previous code block, except now we are loading TCR data\n",
    "    target_obs_column=TargetObsColumnEnum.disease,\n",
    ")\n",
    "adata_tcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1895738e-082b-4e72-9d8d-1078fec2099d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d408dd18-0c2d-4994-ab8f-a721a533d566",
   "metadata": {},
   "source": [
    "In Mal-ID, we store data in [AnnData containers](https://anndata.readthedocs.io/en/latest/). We often use the variable name `adata` for these objects.\n",
    "\n",
    "AnnData containers have a `.X` property with the language model embedding vector for each sequence, and a `.obs` property with the metadata for each sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cc8a9300-8969-4a39-882f-8a1389261cd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-0.614  ,  0.1725 , -0.325  , ...,  0.512  , -0.7983 , -0.5854 ],\n",
       "       [ 0.1841 ,  0.9688 ,  1.127  , ...,  0.6943 ,  0.0789 , -0.02266],\n",
       "       [ 0.732  , -1.323  , -1.272  , ...,  0.0448 ,  0.6924 , -0.761  ],\n",
       "       ...,\n",
       "       [-1.221  ,  1.755  ,  0.1779 , ...,  0.4229 ,  1.312  , -0.4229 ],\n",
       "       [ 0.9478 , -0.7085 , -1.136  , ..., -1.077  ,  0.1272 ,  0.1335 ],\n",
       "       [-0.8794 ,  0.4502 ,  1.741  , ...,  0.906  , -1.046  , -0.3447 ]],\n",
       "      dtype=float16)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_bcr.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e4485e75-c719-4936-8fbd-1aa6c0d932ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amplification_label</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>disease</th>\n",
       "      <th>fr1_seq_aa_q_trim</th>\n",
       "      <th>cdr1_seq_aa_q_trim</th>\n",
       "      <th>fr2_seq_aa_q_trim</th>\n",
       "      <th>cdr2_seq_aa_q_trim</th>\n",
       "      <th>fr3_seq_aa_q_trim</th>\n",
       "      <th>cdr3_seq_aa_q_trim</th>\n",
       "      <th>...</th>\n",
       "      <th>symptoms_Lupus_sm +/-</th>\n",
       "      <th>symptoms_Lupus_sm_nRNP +/-</th>\n",
       "      <th>symptoms_cmv</th>\n",
       "      <th>symptoms_healthy_in_resequencing_experiment</th>\n",
       "      <th>symptoms_healthy_old_vs_new_batch_identifier</th>\n",
       "      <th>isotype_proportion:IGHG</th>\n",
       "      <th>isotype_proportion:IGHA</th>\n",
       "      <th>isotype_proportion:IGHD-M</th>\n",
       "      <th>fold_id</th>\n",
       "      <th>fold_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>__null_dask_index__</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>M111-S037_cDNA_PCR</td>\n",
       "      <td>IGHV3-20</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>HIV</td>\n",
       "      <td>AAS</td>\n",
       "      <td>GFTFDDYG</td>\n",
       "      <td>MSWVRQAPGKGLEWVSG</td>\n",
       "      <td>LNWNGGST</td>\n",
       "      <td>GYADSVKGRFTISRDNAKNSLYLQMNSLRAEDTALYYC</td>\n",
       "      <td>ARDLLPGSYPTYYFEY</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.177207</td>\n",
       "      <td>0.140070</td>\n",
       "      <td>0.682723</td>\n",
       "      <td>0</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M111-S037_cDNA_PCR</td>\n",
       "      <td>IGHV1-69</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>HIV</td>\n",
       "      <td>AS</td>\n",
       "      <td>GGTFSSYA</td>\n",
       "      <td>ISWVRQAPGQGLEWMGR</td>\n",
       "      <td>IIPILGIA</td>\n",
       "      <td>NYAQKFQGRVTMTTDTSTSTAYMELRSLRSDDTAVYYC</td>\n",
       "      <td>AREDSFGAAVEY</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.177207</td>\n",
       "      <td>0.140070</td>\n",
       "      <td>0.682723</td>\n",
       "      <td>0</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15397</th>\n",
       "      <td>M111-S037_cDNA_PCR</td>\n",
       "      <td>IGHV6-1</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>HIV</td>\n",
       "      <td>IS</td>\n",
       "      <td>GDNVSTNSAA</td>\n",
       "      <td>WNWIRQSPSRGLEWLGR</td>\n",
       "      <td>THFQSRWLY</td>\n",
       "      <td>DYAESVRGRITINPDTSKNQITLQLKSMTPDDTGIYYC</td>\n",
       "      <td>ARDQRYPKYYFDY</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.177207</td>\n",
       "      <td>0.140070</td>\n",
       "      <td>0.682723</td>\n",
       "      <td>0</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M111-S037_cDNA_PCR</td>\n",
       "      <td>IGHV3-7</td>\n",
       "      <td>IGHJ5</td>\n",
       "      <td>HIV</td>\n",
       "      <td>AAS</td>\n",
       "      <td>GFSFRGYW</td>\n",
       "      <td>MTWVRQAPGKGLEWVAN</td>\n",
       "      <td>IKQDGSET</td>\n",
       "      <td>YYVDSVNGRFTISRDNAKNSLYLQMNSLRAEDTAVYFC</td>\n",
       "      <td>ARSGTYMGFWDP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.177207</td>\n",
       "      <td>0.140070</td>\n",
       "      <td>0.682723</td>\n",
       "      <td>0</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9787</th>\n",
       "      <td>M111-S037_cDNA_PCR</td>\n",
       "      <td>IGHV3-23</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>HIV</td>\n",
       "      <td>ATS</td>\n",
       "      <td>GFTFSSCA</td>\n",
       "      <td>MSWVRQAPGKGLEWVSA</td>\n",
       "      <td>ISAGSSAT</td>\n",
       "      <td>YYADSVKGRFTISRDNSKNTLFLQMNSLRAEDTAVYYC</td>\n",
       "      <td>AKGGADCTTAACRGSYYYMDV</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.177207</td>\n",
       "      <td>0.140070</td>\n",
       "      <td>0.682723</td>\n",
       "      <td>0</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21465-75</th>\n",
       "      <td>M491-S160_cDNA_PCR</td>\n",
       "      <td>IGHV3-74</td>\n",
       "      <td>IGHJ5</td>\n",
       "      <td>T1D</td>\n",
       "      <td>AAS</td>\n",
       "      <td>GFTFSNYW</td>\n",
       "      <td>MHWVRQAPGKGRVWGSR</td>\n",
       "      <td>LNSDERST</td>\n",
       "      <td>SYADSVKGRFTISRDNAKNTLYLQMNSLRVEDTAVYYC</td>\n",
       "      <td>TKAPVGSCNDASCYPLDL</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.051986</td>\n",
       "      <td>0.130453</td>\n",
       "      <td>0.817561</td>\n",
       "      <td>0</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21486-72</th>\n",
       "      <td>M491-S160_cDNA_PCR</td>\n",
       "      <td>IGHV3-23</td>\n",
       "      <td>IGHJ5</td>\n",
       "      <td>T1D</td>\n",
       "      <td>AAS</td>\n",
       "      <td>GFTFSSYV</td>\n",
       "      <td>MRWVRQAPGKGLEWVSA</td>\n",
       "      <td>ITGSGGST</td>\n",
       "      <td>YYADSVKGRFTISRDNSKKTVSLQMNSLRPEDTALYYC</td>\n",
       "      <td>VRFDSASSFDS</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.051986</td>\n",
       "      <td>0.130453</td>\n",
       "      <td>0.817561</td>\n",
       "      <td>0</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21534-71</th>\n",
       "      <td>M491-S160_cDNA_PCR</td>\n",
       "      <td>IGHV3-7</td>\n",
       "      <td>IGHJ5</td>\n",
       "      <td>T1D</td>\n",
       "      <td>AAS</td>\n",
       "      <td>GFTFSTYW</td>\n",
       "      <td>MSWVRQAPGKGLEWVAN</td>\n",
       "      <td>IKQDGSEK</td>\n",
       "      <td>YYVDSVKGRFTISRDNAKNSLYLQMNNLRPEDTAVYYC</td>\n",
       "      <td>VKVERFCGTSSCAPFDP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.051986</td>\n",
       "      <td>0.130453</td>\n",
       "      <td>0.817561</td>\n",
       "      <td>0</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21548-68</th>\n",
       "      <td>M491-S160_cDNA_PCR</td>\n",
       "      <td>IGHV3-7</td>\n",
       "      <td>IGHJ5</td>\n",
       "      <td>T1D</td>\n",
       "      <td>AAS</td>\n",
       "      <td>GFTFSRYW</td>\n",
       "      <td>MNWVRQAPGKGLEWVAD</td>\n",
       "      <td>IKQDGREK</td>\n",
       "      <td>YYVDSVKGRFTISRDNAKNSLYLQMNSLRAEETAVYYC</td>\n",
       "      <td>ASQSDYYDSSGYYAWFAP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.051986</td>\n",
       "      <td>0.130453</td>\n",
       "      <td>0.817561</td>\n",
       "      <td>0</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21568-79</th>\n",
       "      <td>M491-S160_cDNA_PCR</td>\n",
       "      <td>IGHV3-30</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>T1D</td>\n",
       "      <td>AAS</td>\n",
       "      <td>GFTFSSYG</td>\n",
       "      <td>MHWVRQAPGKGLEWVAF</td>\n",
       "      <td>IRYDGSNK</td>\n",
       "      <td>YYQDSVKGRFTISRDNSKNTLYLQMNSLRAEDTAVYYC</td>\n",
       "      <td>ARETGTTLYNGMDV</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.051986</td>\n",
       "      <td>0.130453</td>\n",
       "      <td>0.817561</td>\n",
       "      <td>0</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5669053 rows × 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    amplification_label    v_gene j_gene disease  \\\n",
       "__null_dask_index__                                                \n",
       "487                  M111-S037_cDNA_PCR  IGHV3-20  IGHJ4     HIV   \n",
       "1                    M111-S037_cDNA_PCR  IGHV1-69  IGHJ4     HIV   \n",
       "15397                M111-S037_cDNA_PCR   IGHV6-1  IGHJ4     HIV   \n",
       "3                    M111-S037_cDNA_PCR   IGHV3-7  IGHJ5     HIV   \n",
       "9787                 M111-S037_cDNA_PCR  IGHV3-23  IGHJ6     HIV   \n",
       "...                                 ...       ...    ...     ...   \n",
       "21465-75             M491-S160_cDNA_PCR  IGHV3-74  IGHJ5     T1D   \n",
       "21486-72             M491-S160_cDNA_PCR  IGHV3-23  IGHJ5     T1D   \n",
       "21534-71             M491-S160_cDNA_PCR   IGHV3-7  IGHJ5     T1D   \n",
       "21548-68             M491-S160_cDNA_PCR   IGHV3-7  IGHJ5     T1D   \n",
       "21568-79             M491-S160_cDNA_PCR  IGHV3-30  IGHJ6     T1D   \n",
       "\n",
       "                    fr1_seq_aa_q_trim cdr1_seq_aa_q_trim  fr2_seq_aa_q_trim  \\\n",
       "__null_dask_index__                                                           \n",
       "487                               AAS           GFTFDDYG  MSWVRQAPGKGLEWVSG   \n",
       "1                                  AS           GGTFSSYA  ISWVRQAPGQGLEWMGR   \n",
       "15397                              IS         GDNVSTNSAA  WNWIRQSPSRGLEWLGR   \n",
       "3                                 AAS           GFSFRGYW  MTWVRQAPGKGLEWVAN   \n",
       "9787                              ATS           GFTFSSCA  MSWVRQAPGKGLEWVSA   \n",
       "...                               ...                ...                ...   \n",
       "21465-75                          AAS           GFTFSNYW  MHWVRQAPGKGRVWGSR   \n",
       "21486-72                          AAS           GFTFSSYV  MRWVRQAPGKGLEWVSA   \n",
       "21534-71                          AAS           GFTFSTYW  MSWVRQAPGKGLEWVAN   \n",
       "21548-68                          AAS           GFTFSRYW  MNWVRQAPGKGLEWVAD   \n",
       "21568-79                          AAS           GFTFSSYG  MHWVRQAPGKGLEWVAF   \n",
       "\n",
       "                    cdr2_seq_aa_q_trim  \\\n",
       "__null_dask_index__                      \n",
       "487                           LNWNGGST   \n",
       "1                             IIPILGIA   \n",
       "15397                        THFQSRWLY   \n",
       "3                             IKQDGSET   \n",
       "9787                          ISAGSSAT   \n",
       "...                                ...   \n",
       "21465-75                      LNSDERST   \n",
       "21486-72                      ITGSGGST   \n",
       "21534-71                      IKQDGSEK   \n",
       "21548-68                      IKQDGREK   \n",
       "21568-79                      IRYDGSNK   \n",
       "\n",
       "                                          fr3_seq_aa_q_trim  \\\n",
       "__null_dask_index__                                           \n",
       "487                  GYADSVKGRFTISRDNAKNSLYLQMNSLRAEDTALYYC   \n",
       "1                    NYAQKFQGRVTMTTDTSTSTAYMELRSLRSDDTAVYYC   \n",
       "15397                DYAESVRGRITINPDTSKNQITLQLKSMTPDDTGIYYC   \n",
       "3                    YYVDSVNGRFTISRDNAKNSLYLQMNSLRAEDTAVYFC   \n",
       "9787                 YYADSVKGRFTISRDNSKNTLFLQMNSLRAEDTAVYYC   \n",
       "...                                                     ...   \n",
       "21465-75             SYADSVKGRFTISRDNAKNTLYLQMNSLRVEDTAVYYC   \n",
       "21486-72             YYADSVKGRFTISRDNSKKTVSLQMNSLRPEDTALYYC   \n",
       "21534-71             YYVDSVKGRFTISRDNAKNSLYLQMNNLRPEDTAVYYC   \n",
       "21548-68             YYVDSVKGRFTISRDNAKNSLYLQMNSLRAEETAVYYC   \n",
       "21568-79             YYQDSVKGRFTISRDNSKNTLYLQMNSLRAEDTAVYYC   \n",
       "\n",
       "                        cdr3_seq_aa_q_trim  ... symptoms_Lupus_sm +/-  \\\n",
       "__null_dask_index__                         ...                         \n",
       "487                       ARDLLPGSYPTYYFEY  ...                   NaN   \n",
       "1                             AREDSFGAAVEY  ...                   NaN   \n",
       "15397                        ARDQRYPKYYFDY  ...                   NaN   \n",
       "3                             ARSGTYMGFWDP  ...                   NaN   \n",
       "9787                 AKGGADCTTAACRGSYYYMDV  ...                   NaN   \n",
       "...                                    ...  ...                   ...   \n",
       "21465-75                TKAPVGSCNDASCYPLDL  ...                   NaN   \n",
       "21486-72                       VRFDSASSFDS  ...                   NaN   \n",
       "21534-71                 VKVERFCGTSSCAPFDP  ...                   NaN   \n",
       "21548-68                ASQSDYYDSSGYYAWFAP  ...                   NaN   \n",
       "21568-79                    ARETGTTLYNGMDV  ...                   NaN   \n",
       "\n",
       "                     symptoms_Lupus_sm_nRNP +/- symptoms_cmv  \\\n",
       "__null_dask_index__                                            \n",
       "487                                         NaN          NaN   \n",
       "1                                           NaN          NaN   \n",
       "15397                                       NaN          NaN   \n",
       "3                                           NaN          NaN   \n",
       "9787                                        NaN          NaN   \n",
       "...                                         ...          ...   \n",
       "21465-75                                    NaN          NaN   \n",
       "21486-72                                    NaN          NaN   \n",
       "21534-71                                    NaN          NaN   \n",
       "21548-68                                    NaN          NaN   \n",
       "21568-79                                    NaN          NaN   \n",
       "\n",
       "                    symptoms_healthy_in_resequencing_experiment  \\\n",
       "__null_dask_index__                                               \n",
       "487                                                         NaN   \n",
       "1                                                           NaN   \n",
       "15397                                                       NaN   \n",
       "3                                                           NaN   \n",
       "9787                                                        NaN   \n",
       "...                                                         ...   \n",
       "21465-75                                                    NaN   \n",
       "21486-72                                                    NaN   \n",
       "21534-71                                                    NaN   \n",
       "21548-68                                                    NaN   \n",
       "21568-79                                                    NaN   \n",
       "\n",
       "                     symptoms_healthy_old_vs_new_batch_identifier  \\\n",
       "__null_dask_index__                                                 \n",
       "487                                                           NaN   \n",
       "1                                                             NaN   \n",
       "15397                                                         NaN   \n",
       "3                                                             NaN   \n",
       "9787                                                          NaN   \n",
       "...                                                           ...   \n",
       "21465-75                                                      NaN   \n",
       "21486-72                                                      NaN   \n",
       "21534-71                                                      NaN   \n",
       "21548-68                                                      NaN   \n",
       "21568-79                                                      NaN   \n",
       "\n",
       "                     isotype_proportion:IGHG  isotype_proportion:IGHA  \\\n",
       "__null_dask_index__                                                     \n",
       "487                                 0.177207                 0.140070   \n",
       "1                                   0.177207                 0.140070   \n",
       "15397                               0.177207                 0.140070   \n",
       "3                                   0.177207                 0.140070   \n",
       "9787                                0.177207                 0.140070   \n",
       "...                                      ...                      ...   \n",
       "21465-75                            0.051986                 0.130453   \n",
       "21486-72                            0.051986                 0.130453   \n",
       "21534-71                            0.051986                 0.130453   \n",
       "21548-68                            0.051986                 0.130453   \n",
       "21568-79                            0.051986                 0.130453   \n",
       "\n",
       "                     isotype_proportion:IGHD-M  fold_id fold_label  \n",
       "__null_dask_index__                                                 \n",
       "487                                   0.682723        0       test  \n",
       "1                                     0.682723        0       test  \n",
       "15397                                 0.682723        0       test  \n",
       "3                                     0.682723        0       test  \n",
       "9787                                  0.682723        0       test  \n",
       "...                                        ...      ...        ...  \n",
       "21465-75                              0.817561        0       test  \n",
       "21486-72                              0.817561        0       test  \n",
       "21534-71                              0.817561        0       test  \n",
       "21548-68                              0.817561        0       test  \n",
       "21568-79                              0.817561        0       test  \n",
       "\n",
       "[5669053 rows x 76 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_bcr.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ede4b48b-0bbb-4054-b8c9-7301ecf1c2b2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a341dec9-9988-4b98-83ee-b1a544c81fa5",
   "metadata": {},
   "source": [
    "`adata_bcr.X` has one row per sequence and one column per language model embedding dimension. `adata_bcr.obs` has one row per sequence and one column per metadata field. The same is true for `adata_tcr`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3d4d02f1-5222-49b1-a06a-add60229c6a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5669053, 640)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_bcr.X.shape  # samples x embedding dimensions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7d509f43-5cc3-4723-a8a8-8ac944542417",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5669053, 76)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_bcr.obs.shape  # samples x metadata fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ba01a72-cbaf-43d8-b708-86765eb71736",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "973e0c2c-36d0-4350-b5c3-30514b3e7f5c",
   "metadata": {},
   "source": [
    "Sequence-level metadata fields include:\n",
    "\n",
    "- The `participant_label` (patient ID) and `specimen_label` (sample ID) from which the sequence originated\n",
    "- `v_gene`: IGHV or TRBV gene segment name\n",
    "- `j_gene`: IGHJ or TRBJ gene segment name\n",
    "- `cdr3_seq_aa_q_trim`: CDR3 amino acid sequence\n",
    "- `v_mut`: Somatic hypermutation rate, specifically the fraction of V region nucleotides that are mutated (BCR only)\n",
    "- `isotype_supergroup`: either `IGHG`, `IGHA`, `IGHD-M` (combining IgD and IgM), or `TCRB`. The name \"supergroup\" refers to the fact that we are overloading the definition of isotypes by combining IgD and IgM, and by labeling all TCR data as `isotype_supergroup=\"TCRB\"` for convenience, even though TCRs don't have isotypes by definition. (Note that IgE is filtered out in the subsampling step, as are unmutated IgD and IgM sequences which represent naive B cells.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f2c3063-3980-4d70-8b47-fa26e1c69f8a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b81b337e-8ab8-4865-b493-69221aa6e44b",
   "metadata": {},
   "source": [
    "The choice of `target_obs_column` argument in the data loading `io.load_fold_embeddings()` call matters. The AnnData object is filtered down to samples that are \"in scope\" for a particular classification target. For example, if you choose `target_obs_column=TargetObsColumnEnum.sex_healthy_only`, the AnnDatas will only have samples from healthy individuals for whom the sex is known. (See `malid/datamodels.py` for the exact definition of each `TargetObsColumnEnum` option.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f88306d7-d6d7-4a1c-8ce4-d4950a1edf3b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "24db5eda-0e91-4487-9d2f-d1604ed16010",
   "metadata": {},
   "source": [
    "> *Aside*\n",
    ">\n",
    "> When we ran `io.load_fold_embeddings()` above, you may have noticed some log messages about caching the data. When the large AnnData object is first loaded, it's copied from its original file path — usually on a network-mounted file system — to scratch storage on the machine where you're running this notebook. This speeds up further data loading.\n",
    ">\n",
    "> Another type of caching happens silently: the imported data is cached in memory, so that repeated calls to `io.load_fold_embeddings` with the same arguments don't result in slow loads of the same data over and over. The data is cached before we run the filtering for a particular classification target, so we can still leverage the cached version even if we switch the `target_obs_column` argument. This helps in our training loop, so we can train the model for many classification targets seamlessly. The cache is capped at four entries, and older data is automatically removed from the cache to make room for what's being used now. (You can also manually clear the cache with `io.clear_cached_fold_embeddings()`, or disable the cache altogether by setting the environment variable `MALID_DISABLE_IN_MEMORY_CACHE=true` before loading any Mal-ID Python code.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a48e547-8fd8-46e3-821c-05a096ae5790",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "72448a45-bbbe-4c70-942f-70db1775b159",
   "metadata": {},
   "source": [
    "## Model 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8980864-6e73-4d81-a539-6be0837441ab",
   "metadata": {},
   "source": [
    "Model 1 uses overall summary statistics of the BCR or TCR repertoire to predict disease status.\n",
    "\n",
    "This way of generating features can be tied into many classification algorithms, e.g. logistic regression or random forests. We try a bunch of classification algorithms and choose the one with highest performance on the validation set, which is not seen during training. The same is true for Models 2 and 3.\n",
    "\n",
    "We've recorded our choices of the chosen \"model name\" for Model 1, Model 2, and Model 3 in `config.metamodel_base_model_names`. The ensemble metamodel will use these versions of the base Mal-ID components:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "677ea04e-fa13-424a-9ee6-3f0d322939d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "namespace(model_name_overall_repertoire_composition={<GeneLocus.BCR: 1>: 'elasticnet_cv0.25',\n",
       "                                                     <GeneLocus.TCR: 2>: 'lasso_cv'},\n",
       "          model_name_convergent_clustering={<GeneLocus.BCR: 1>: 'ridge_cv',\n",
       "                                            <GeneLocus.TCR: 2>: 'lasso_cv'},\n",
       "          base_sequence_model_name={<GeneLocus.BCR: 1>: 'rf_multiclass',\n",
       "                                    <GeneLocus.TCR: 2>: 'ridge_cv_ovr'},\n",
       "          base_sequence_model_subset_strategy=<SequenceSubsetStrategy.split_Vgene_and_isotype: 3>,\n",
       "          aggregation_sequence_model_name={<GeneLocus.BCR: 1>: 'rf_multiclass_mean_aggregated_as_binary_ovr_reweighed_by_subset_frequencies',\n",
       "                                           <GeneLocus.TCR: 2>: 'rf_multiclass_entropy_twenty_percent_cutoff_aggregated_as_binary_ovr_reweighed_by_subset_frequencies'})"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config.metamodel_base_model_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24d97ac0-451d-4057-b838-edf4020cf86e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a186a9f0-816a-4ccc-a321-1be9738682a1",
   "metadata": {},
   "source": [
    "Let's look specifically at the version of Model 1 chosen for BCR data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "14bf03cb-b800-4460-8a13-cc3b59976465",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'elasticnet_cv0.25'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config.metamodel_base_model_names.model_name_overall_repertoire_composition[\n",
    "    GeneLocus.BCR\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2d8795a-d2fb-42c6-bdd0-1f256deb82c7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b7831555-c7cd-47ef-8998-5937bfe9e0c6",
   "metadata": {},
   "source": [
    "The model name is `elasticnet_cv0.25`, which is elastic net regularized logistic regression with an L1-L2 ratio of 0.25. (The exact definition is in `malid/train/model_definitions.py`.)\n",
    "\n",
    "> _Aside_:\n",
    "> >\n",
    "> Models with `_cv` in the name use internal (nested) cross validation to tune their hyperparameters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96fe3118-8a22-4564-b01c-20023fe095c1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9f9d9160-9c92-4bdf-9077-9722e4114dd6",
   "metadata": {},
   "source": [
    "Let's load this version of Model 1 for BCR data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d2615a38-7b16-4ded-ac86-98cae273e6fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>RepertoireClassifier: Pipeline(steps=[('columntransformer',\n",
       "                 ColumnTransformer(remainder='passthrough',\n",
       "                                   transformers=[('log1p-scale-PCA_IGHG',\n",
       "                                                  Pipeline(steps=[('log1p',\n",
       "                                                                   FunctionTransformer(feature_names_out='one-to-one',\n",
       "                                                                                       func=<ufunc 'log1p'>,\n",
       "                                                                                       validate=True)),\n",
       "                                                                  ('scale',\n",
       "                                                                   StandardScalerThatPreservesInputType()),\n",
       "                                                                  ('pca',\n",
       "                                                                   PCA(n_components=15,\n",
       "                                                                       random_state=0))]),\n",
       "                                                  <sklearn.compose._column_t...\n",
       "                 GlmnetLogitNetWrapper(alpha=0.25, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=1, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=<function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0>, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])</pre>"
      ],
      "text/plain": [
       "RepertoireClassifier: Pipeline(steps=[('columntransformer',\n",
       "                 ColumnTransformer(remainder='passthrough',\n",
       "                                   transformers=[('log1p-scale-PCA_IGHG',\n",
       "                                                  Pipeline(steps=[('log1p',\n",
       "                                                                   FunctionTransformer(feature_names_out='one-to-one',\n",
       "                                                                                       func=<ufunc 'log1p'>,\n",
       "                                                                                       validate=True)),\n",
       "                                                                  ('scale',\n",
       "                                                                   StandardScalerThatPreservesInputType()),\n",
       "                                                                  ('pca',\n",
       "                                                                   PCA(n_components=15,\n",
       "                                                                       random_state=0))]),\n",
       "                                                  <sklearn.compose._column_t...\n",
       "                 GlmnetLogitNetWrapper(alpha=0.25, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=1, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=<function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0>, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf1 = RepertoireClassifier(\n",
    "    fold_id=fold_id,\n",
    "    # Load \"elasticnet_cv0.25\"\n",
    "    model_name=config.metamodel_base_model_names.model_name_overall_repertoire_composition[\n",
    "        GeneLocus.BCR\n",
    "    ],\n",
    "    fold_label_train=\"train_smaller\",  # Indicates which part of the data the model was trained on\n",
    "    gene_locus=GeneLocus.BCR,  # A different model is trained for each sequencing locus\n",
    "    target_obs_column=target_obs_column,  # A different model is trained for each classification target\n",
    ")\n",
    "clf1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8939d114-5ff3-447d-bee4-0a6c464d34aa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f11d9150-e2ea-46f0-9f09-94ada5ef1aa5",
   "metadata": {},
   "source": [
    "We've loaded a `RepertoireClassifier`, which is a wrapper around a scikit-learn model stored in `_inner`.\n",
    "\n",
    "In this case, it's a wrapper around a scikit-learn Pipeline:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "63fec475-3b7c-40c3-bf2d-a5a1f965f7ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sklearn.pipeline.Pipeline"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(clf1._inner)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6a6f0ced-8b60-4580-ad20-cf01eb2878f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {color: black;background-color: white;}#sk-container-id-1 pre{padding: 0;}#sk-container-id-1 div.sk-toggleable {background-color: white;}#sk-container-id-1 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-1 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-1 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-1 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-1 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-1 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-1 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-1 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-1 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-1 div.sk-item {position: relative;z-index: 1;}#sk-container-id-1 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-1 div.sk-item::before, #sk-container-id-1 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-1 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-1 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-1 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-1 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-1 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-1 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-1 div.sk-label-container {text-align: center;}#sk-container-id-1 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-1 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>Pipeline(steps=[(&#x27;columntransformer&#x27;,\n",
       "                 ColumnTransformer(remainder=&#x27;passthrough&#x27;,\n",
       "                                   transformers=[(&#x27;log1p-scale-PCA_IGHG&#x27;,\n",
       "                                                  Pipeline(steps=[(&#x27;log1p&#x27;,\n",
       "                                                                   FunctionTransformer(feature_names_out=&#x27;one-to-one&#x27;,\n",
       "                                                                                       func=&lt;ufunc &#x27;log1p&#x27;&gt;,\n",
       "                                                                                       validate=True)),\n",
       "                                                                  (&#x27;scale&#x27;,\n",
       "                                                                   StandardScalerThatPreservesInputType()),\n",
       "                                                                  (&#x27;pca&#x27;,\n",
       "                                                                   PCA(n_components=15,\n",
       "                                                                       random_state=0))]),\n",
       "                                                  &lt;sklearn.compose._column_t...\n",
       "                 GlmnetLogitNetWrapper(alpha=0.25, class_weight=&#x27;balanced&#x27;, cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=1, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=&lt;function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0&gt;, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" ><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">Pipeline</label><div class=\"sk-toggleable__content\"><pre>Pipeline(steps=[(&#x27;columntransformer&#x27;,\n",
       "                 ColumnTransformer(remainder=&#x27;passthrough&#x27;,\n",
       "                                   transformers=[(&#x27;log1p-scale-PCA_IGHG&#x27;,\n",
       "                                                  Pipeline(steps=[(&#x27;log1p&#x27;,\n",
       "                                                                   FunctionTransformer(feature_names_out=&#x27;one-to-one&#x27;,\n",
       "                                                                                       func=&lt;ufunc &#x27;log1p&#x27;&gt;,\n",
       "                                                                                       validate=True)),\n",
       "                                                                  (&#x27;scale&#x27;,\n",
       "                                                                   StandardScalerThatPreservesInputType()),\n",
       "                                                                  (&#x27;pca&#x27;,\n",
       "                                                                   PCA(n_components=15,\n",
       "                                                                       random_state=0))]),\n",
       "                                                  &lt;sklearn.compose._column_t...\n",
       "                 GlmnetLogitNetWrapper(alpha=0.25, class_weight=&#x27;balanced&#x27;, cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=1, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=&lt;function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0&gt;, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-2\" type=\"checkbox\" ><label for=\"sk-estimator-id-2\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">columntransformer: ColumnTransformer</label><div class=\"sk-toggleable__content\"><pre>ColumnTransformer(remainder=&#x27;passthrough&#x27;,\n",
       "                  transformers=[(&#x27;log1p-scale-PCA_IGHG&#x27;,\n",
       "                                 Pipeline(steps=[(&#x27;log1p&#x27;,\n",
       "                                                  FunctionTransformer(feature_names_out=&#x27;one-to-one&#x27;,\n",
       "                                                                      func=&lt;ufunc &#x27;log1p&#x27;&gt;,\n",
       "                                                                      validate=True)),\n",
       "                                                 (&#x27;scale&#x27;,\n",
       "                                                  StandardScalerThatPreservesInputType()),\n",
       "                                                 (&#x27;pca&#x27;,\n",
       "                                                  PCA(n_components=15,\n",
       "                                                      random_state=0))]),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object...\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x7f1451340b20&gt;),\n",
       "                                (&#x27;log1p-scale-PCA_IGHD-M&#x27;,\n",
       "                                 Pipeline(steps=[(&#x27;log1p&#x27;,\n",
       "                                                  FunctionTransformer(feature_names_out=&#x27;one-to-one&#x27;,\n",
       "                                                                      func=&lt;ufunc &#x27;log1p&#x27;&gt;,\n",
       "                                                                      validate=True)),\n",
       "                                                 (&#x27;scale&#x27;,\n",
       "                                                  StandardScalerThatPreservesInputType()),\n",
       "                                                 (&#x27;pca&#x27;,\n",
       "                                                  PCA(n_components=15,\n",
       "                                                      random_state=0))]),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x7f1451340730&gt;)])</pre></div></div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-3\" type=\"checkbox\" ><label for=\"sk-estimator-id-3\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">log1p-scale-PCA_IGHG</label><div class=\"sk-toggleable__content\"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x7f1451340400&gt;</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-4\" type=\"checkbox\" ><label for=\"sk-estimator-id-4\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">FunctionTransformer</label><div class=\"sk-toggleable__content\"><pre>FunctionTransformer(feature_names_out=&#x27;one-to-one&#x27;, func=&lt;ufunc &#x27;log1p&#x27;&gt;,\n",
       "                    validate=True)</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-5\" type=\"checkbox\" ><label for=\"sk-estimator-id-5\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">StandardScalerThatPreservesInputType</label><div class=\"sk-toggleable__content\"><pre>StandardScalerThatPreservesInputType()</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-6\" type=\"checkbox\" ><label for=\"sk-estimator-id-6\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">PCA</label><div class=\"sk-toggleable__content\"><pre>PCA(n_components=15, random_state=0)</pre></div></div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-7\" type=\"checkbox\" ><label for=\"sk-estimator-id-7\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">log1p-scale-PCA_IGHA</label><div class=\"sk-toggleable__content\"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x7f1451340b20&gt;</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-8\" type=\"checkbox\" ><label for=\"sk-estimator-id-8\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">FunctionTransformer</label><div class=\"sk-toggleable__content\"><pre>FunctionTransformer(feature_names_out=&#x27;one-to-one&#x27;, func=&lt;ufunc &#x27;log1p&#x27;&gt;,\n",
       "                    validate=True)</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-9\" type=\"checkbox\" ><label for=\"sk-estimator-id-9\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">StandardScalerThatPreservesInputType</label><div class=\"sk-toggleable__content\"><pre>StandardScalerThatPreservesInputType()</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-10\" type=\"checkbox\" ><label for=\"sk-estimator-id-10\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">PCA</label><div class=\"sk-toggleable__content\"><pre>PCA(n_components=15, random_state=0)</pre></div></div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-11\" type=\"checkbox\" ><label for=\"sk-estimator-id-11\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">log1p-scale-PCA_IGHD-M</label><div class=\"sk-toggleable__content\"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x7f1451340730&gt;</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-12\" type=\"checkbox\" ><label for=\"sk-estimator-id-12\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">FunctionTransformer</label><div class=\"sk-toggleable__content\"><pre>FunctionTransformer(feature_names_out=&#x27;one-to-one&#x27;, func=&lt;ufunc &#x27;log1p&#x27;&gt;,\n",
       "                    validate=True)</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-13\" type=\"checkbox\" ><label for=\"sk-estimator-id-13\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">StandardScalerThatPreservesInputType</label><div class=\"sk-toggleable__content\"><pre>StandardScalerThatPreservesInputType()</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-14\" type=\"checkbox\" ><label for=\"sk-estimator-id-14\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">PCA</label><div class=\"sk-toggleable__content\"><pre>PCA(n_components=15, random_state=0)</pre></div></div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-15\" type=\"checkbox\" ><label for=\"sk-estimator-id-15\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">remainder</label><div class=\"sk-toggleable__content\"><pre>[&#x27;v_mut_median_per_specimen:IGHG&#x27;, &#x27;v_sequence_is_mutated:IGHG&#x27;, &#x27;v_mut_median_per_specimen:IGHA&#x27;, &#x27;v_sequence_is_mutated:IGHA&#x27;, &#x27;v_mut_median_per_specimen:IGHD-M&#x27;, &#x27;v_sequence_is_mutated:IGHD-M&#x27;]</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-16\" type=\"checkbox\" ><label for=\"sk-estimator-id-16\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">passthrough</label><div class=\"sk-toggleable__content\"><pre>passthrough</pre></div></div></div></div></div></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-17\" type=\"checkbox\" ><label for=\"sk-estimator-id-17\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">StandardScalerThatPreservesInputType</label><div class=\"sk-toggleable__content\"><pre>StandardScalerThatPreservesInputType()</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-18\" type=\"checkbox\" ><label for=\"sk-estimator-id-18\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">GlmnetLogitNetWrapper</label><div class=\"sk-toggleable__content\"><pre>GlmnetLogitNetWrapper: LogitNet(alpha=0.25, n_splits=5, random_state=0,\n",
       "         scoring=&lt;function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0&gt;,\n",
       "         standardize=False, verbose=True)</pre></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "Pipeline(steps=[('columntransformer',\n",
       "                 ColumnTransformer(remainder='passthrough',\n",
       "                                   transformers=[('log1p-scale-PCA_IGHG',\n",
       "                                                  Pipeline(steps=[('log1p',\n",
       "                                                                   FunctionTransformer(feature_names_out='one-to-one',\n",
       "                                                                                       func=<ufunc 'log1p'>,\n",
       "                                                                                       validate=True)),\n",
       "                                                                  ('scale',\n",
       "                                                                   StandardScalerThatPreservesInputType()),\n",
       "                                                                  ('pca',\n",
       "                                                                   PCA(n_components=15,\n",
       "                                                                       random_state=0))]),\n",
       "                                                  <sklearn.compose._column_t...\n",
       "                 GlmnetLogitNetWrapper(alpha=0.25, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=1, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=<function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0>, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf1._inner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "203d4354-cdfc-466d-9ab8-dcdeaafaa295",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b6f7eb36-a695-4209-9bcb-26f6d856c8fc",
   "metadata": {},
   "source": [
    "The scikit-learn Pipeline passes V-J gene use counts through `log1p`, scaling, and PCA transformations. This happens for IgG, IgA, and IgM separately, and is coordinated by a `ColumnTransformer` step.\n",
    "\n",
    "The resulting PCs — along with the somatic hypermutation features marked \"remainder\" and \"passthrough\" — then go through standardization (`StandardScalerThatPreservesInputType`) and logistic regression (`GlmnetLogitNetWrapper`).\n",
    "\n",
    "\n",
    "Here are the original input feature names. Notice how there's a feature for each V gene, J gene, and isotype combination here, which then gets reduced into a smaller set of features by PCA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c12fde0e-9e68-46ca-bbea-c95154dbf664",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['v_mut_median_per_specimen:IGHG', 'v_sequence_is_mutated:IGHG',\n",
       "       'v_mut_median_per_specimen:IGHA', 'v_sequence_is_mutated:IGHA',\n",
       "       'v_mut_median_per_specimen:IGHD-M', 'v_sequence_is_mutated:IGHD-M',\n",
       "       'IGHG:pca_IGHV1-18|IGHJ1:IGHG', 'IGHG:pca_IGHV1-18|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV1-18|IGHJ3:IGHG', 'IGHG:pca_IGHV1-18|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV1-18|IGHJ5:IGHG', 'IGHG:pca_IGHV1-18|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV1-24|IGHJ1:IGHG', 'IGHG:pca_IGHV1-24|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV1-24|IGHJ3:IGHG', 'IGHG:pca_IGHV1-24|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV1-24|IGHJ5:IGHG', 'IGHG:pca_IGHV1-24|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV1-2|IGHJ1:IGHG', 'IGHG:pca_IGHV1-2|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV1-2|IGHJ3:IGHG', 'IGHG:pca_IGHV1-2|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV1-2|IGHJ5:IGHG', 'IGHG:pca_IGHV1-2|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV1-3|IGHJ1:IGHG', 'IGHG:pca_IGHV1-3|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV1-3|IGHJ3:IGHG', 'IGHG:pca_IGHV1-3|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV1-3|IGHJ5:IGHG', 'IGHG:pca_IGHV1-3|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV1-46|IGHJ1:IGHG', 'IGHG:pca_IGHV1-46|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV1-46|IGHJ3:IGHG', 'IGHG:pca_IGHV1-46|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV1-46|IGHJ5:IGHG', 'IGHG:pca_IGHV1-46|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV1-69|IGHJ1:IGHG', 'IGHG:pca_IGHV1-69|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV1-69|IGHJ3:IGHG', 'IGHG:pca_IGHV1-69|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV1-69|IGHJ5:IGHG', 'IGHG:pca_IGHV1-69|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV1-8|IGHJ1:IGHG', 'IGHG:pca_IGHV1-8|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV1-8|IGHJ3:IGHG', 'IGHG:pca_IGHV1-8|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV1-8|IGHJ5:IGHG', 'IGHG:pca_IGHV1-8|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV2-5|IGHJ1:IGHG', 'IGHG:pca_IGHV2-5|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV2-5|IGHJ3:IGHG', 'IGHG:pca_IGHV2-5|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV2-5|IGHJ5:IGHG', 'IGHG:pca_IGHV2-5|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-11|IGHJ1:IGHG', 'IGHG:pca_IGHV3-11|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-11|IGHJ3:IGHG', 'IGHG:pca_IGHV3-11|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-11|IGHJ5:IGHG', 'IGHG:pca_IGHV3-11|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-13|IGHJ1:IGHG', 'IGHG:pca_IGHV3-13|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-13|IGHJ3:IGHG', 'IGHG:pca_IGHV3-13|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-13|IGHJ5:IGHG', 'IGHG:pca_IGHV3-13|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-15|IGHJ1:IGHG', 'IGHG:pca_IGHV3-15|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-15|IGHJ3:IGHG', 'IGHG:pca_IGHV3-15|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-15|IGHJ5:IGHG', 'IGHG:pca_IGHV3-15|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-20|IGHJ1:IGHG', 'IGHG:pca_IGHV3-20|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-20|IGHJ3:IGHG', 'IGHG:pca_IGHV3-20|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-20|IGHJ5:IGHG', 'IGHG:pca_IGHV3-20|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-21|IGHJ1:IGHG', 'IGHG:pca_IGHV3-21|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-21|IGHJ3:IGHG', 'IGHG:pca_IGHV3-21|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-21|IGHJ5:IGHG', 'IGHG:pca_IGHV3-21|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-23|IGHJ1:IGHG', 'IGHG:pca_IGHV3-23|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-23|IGHJ3:IGHG', 'IGHG:pca_IGHV3-23|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-23|IGHJ5:IGHG', 'IGHG:pca_IGHV3-23|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-30-3|IGHJ1:IGHG', 'IGHG:pca_IGHV3-30-3|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-30-3|IGHJ3:IGHG', 'IGHG:pca_IGHV3-30-3|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-30-3|IGHJ5:IGHG', 'IGHG:pca_IGHV3-30-3|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-30|IGHJ1:IGHG', 'IGHG:pca_IGHV3-30|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-30|IGHJ3:IGHG', 'IGHG:pca_IGHV3-30|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-30|IGHJ5:IGHG', 'IGHG:pca_IGHV3-30|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-33|IGHJ1:IGHG', 'IGHG:pca_IGHV3-33|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-33|IGHJ3:IGHG', 'IGHG:pca_IGHV3-33|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-33|IGHJ5:IGHG', 'IGHG:pca_IGHV3-33|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-43|IGHJ1:IGHG', 'IGHG:pca_IGHV3-43|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-43|IGHJ3:IGHG', 'IGHG:pca_IGHV3-43|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-43|IGHJ5:IGHG', 'IGHG:pca_IGHV3-43|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-48|IGHJ1:IGHG', 'IGHG:pca_IGHV3-48|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-48|IGHJ3:IGHG', 'IGHG:pca_IGHV3-48|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-48|IGHJ5:IGHG', 'IGHG:pca_IGHV3-48|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-49|IGHJ1:IGHG', 'IGHG:pca_IGHV3-49|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-49|IGHJ3:IGHG', 'IGHG:pca_IGHV3-49|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-49|IGHJ5:IGHG', 'IGHG:pca_IGHV3-49|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-53|IGHJ1:IGHG', 'IGHG:pca_IGHV3-53|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-53|IGHJ3:IGHG', 'IGHG:pca_IGHV3-53|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-53|IGHJ5:IGHG', 'IGHG:pca_IGHV3-53|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-64|IGHJ1:IGHG', 'IGHG:pca_IGHV3-64|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-64|IGHJ3:IGHG', 'IGHG:pca_IGHV3-64|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-64|IGHJ5:IGHG', 'IGHG:pca_IGHV3-64|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-66|IGHJ1:IGHG', 'IGHG:pca_IGHV3-66|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-66|IGHJ3:IGHG', 'IGHG:pca_IGHV3-66|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-66|IGHJ5:IGHG', 'IGHG:pca_IGHV3-66|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-73|IGHJ1:IGHG', 'IGHG:pca_IGHV3-73|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-73|IGHJ3:IGHG', 'IGHG:pca_IGHV3-73|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-73|IGHJ5:IGHG', 'IGHG:pca_IGHV3-73|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-74|IGHJ1:IGHG', 'IGHG:pca_IGHV3-74|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-74|IGHJ3:IGHG', 'IGHG:pca_IGHV3-74|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-74|IGHJ5:IGHG', 'IGHG:pca_IGHV3-74|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-7|IGHJ1:IGHG', 'IGHG:pca_IGHV3-7|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-7|IGHJ3:IGHG', 'IGHG:pca_IGHV3-7|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-7|IGHJ5:IGHG', 'IGHG:pca_IGHV3-7|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV3-9|IGHJ1:IGHG', 'IGHG:pca_IGHV3-9|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV3-9|IGHJ3:IGHG', 'IGHG:pca_IGHV3-9|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV3-9|IGHJ5:IGHG', 'IGHG:pca_IGHV3-9|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV4-30-2|IGHJ1:IGHG', 'IGHG:pca_IGHV4-30-2|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV4-30-2|IGHJ3:IGHG', 'IGHG:pca_IGHV4-30-2|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV4-30-2|IGHJ5:IGHG', 'IGHG:pca_IGHV4-30-2|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV4-30-4|IGHJ1:IGHG', 'IGHG:pca_IGHV4-30-4|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV4-30-4|IGHJ3:IGHG', 'IGHG:pca_IGHV4-30-4|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV4-30-4|IGHJ5:IGHG', 'IGHG:pca_IGHV4-30-4|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV4-31|IGHJ1:IGHG', 'IGHG:pca_IGHV4-31|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV4-31|IGHJ3:IGHG', 'IGHG:pca_IGHV4-31|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV4-31|IGHJ5:IGHG', 'IGHG:pca_IGHV4-31|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV4-34|IGHJ1:IGHG', 'IGHG:pca_IGHV4-34|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV4-34|IGHJ3:IGHG', 'IGHG:pca_IGHV4-34|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV4-34|IGHJ5:IGHG', 'IGHG:pca_IGHV4-34|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV4-39|IGHJ1:IGHG', 'IGHG:pca_IGHV4-39|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV4-39|IGHJ3:IGHG', 'IGHG:pca_IGHV4-39|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV4-39|IGHJ5:IGHG', 'IGHG:pca_IGHV4-39|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV4-4|IGHJ1:IGHG', 'IGHG:pca_IGHV4-4|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV4-4|IGHJ3:IGHG', 'IGHG:pca_IGHV4-4|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV4-4|IGHJ5:IGHG', 'IGHG:pca_IGHV4-4|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV4-59|IGHJ1:IGHG', 'IGHG:pca_IGHV4-59|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV4-59|IGHJ3:IGHG', 'IGHG:pca_IGHV4-59|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV4-59|IGHJ5:IGHG', 'IGHG:pca_IGHV4-59|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV4-61|IGHJ1:IGHG', 'IGHG:pca_IGHV4-61|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV4-61|IGHJ3:IGHG', 'IGHG:pca_IGHV4-61|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV4-61|IGHJ5:IGHG', 'IGHG:pca_IGHV4-61|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV4-b|IGHJ1:IGHG', 'IGHG:pca_IGHV4-b|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV4-b|IGHJ3:IGHG', 'IGHG:pca_IGHV4-b|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV4-b|IGHJ5:IGHG', 'IGHG:pca_IGHV4-b|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV5-51|IGHJ1:IGHG', 'IGHG:pca_IGHV5-51|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV5-51|IGHJ3:IGHG', 'IGHG:pca_IGHV5-51|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV5-51|IGHJ5:IGHG', 'IGHG:pca_IGHV5-51|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV5-a|IGHJ1:IGHG', 'IGHG:pca_IGHV5-a|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV5-a|IGHJ3:IGHG', 'IGHG:pca_IGHV5-a|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV5-a|IGHJ5:IGHG', 'IGHG:pca_IGHV5-a|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV6-1|IGHJ1:IGHG', 'IGHG:pca_IGHV6-1|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV6-1|IGHJ3:IGHG', 'IGHG:pca_IGHV6-1|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV6-1|IGHJ5:IGHG', 'IGHG:pca_IGHV6-1|IGHJ6:IGHG',\n",
       "       'IGHG:pca_IGHV7-4-1|IGHJ1:IGHG', 'IGHG:pca_IGHV7-4-1|IGHJ2:IGHG',\n",
       "       'IGHG:pca_IGHV7-4-1|IGHJ3:IGHG', 'IGHG:pca_IGHV7-4-1|IGHJ4:IGHG',\n",
       "       'IGHG:pca_IGHV7-4-1|IGHJ5:IGHG', 'IGHG:pca_IGHV7-4-1|IGHJ6:IGHG',\n",
       "       'IGHA:pca_IGHV1-18|IGHJ1:IGHA', 'IGHA:pca_IGHV1-18|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV1-18|IGHJ3:IGHA', 'IGHA:pca_IGHV1-18|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV1-18|IGHJ5:IGHA', 'IGHA:pca_IGHV1-18|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV1-24|IGHJ1:IGHA', 'IGHA:pca_IGHV1-24|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV1-24|IGHJ3:IGHA', 'IGHA:pca_IGHV1-24|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV1-24|IGHJ5:IGHA', 'IGHA:pca_IGHV1-24|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV1-2|IGHJ1:IGHA', 'IGHA:pca_IGHV1-2|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV1-2|IGHJ3:IGHA', 'IGHA:pca_IGHV1-2|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV1-2|IGHJ5:IGHA', 'IGHA:pca_IGHV1-2|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV1-3|IGHJ1:IGHA', 'IGHA:pca_IGHV1-3|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV1-3|IGHJ3:IGHA', 'IGHA:pca_IGHV1-3|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV1-3|IGHJ5:IGHA', 'IGHA:pca_IGHV1-3|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV1-46|IGHJ1:IGHA', 'IGHA:pca_IGHV1-46|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV1-46|IGHJ3:IGHA', 'IGHA:pca_IGHV1-46|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV1-46|IGHJ5:IGHA', 'IGHA:pca_IGHV1-46|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV1-69|IGHJ1:IGHA', 'IGHA:pca_IGHV1-69|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV1-69|IGHJ3:IGHA', 'IGHA:pca_IGHV1-69|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV1-69|IGHJ5:IGHA', 'IGHA:pca_IGHV1-69|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV1-8|IGHJ1:IGHA', 'IGHA:pca_IGHV1-8|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV1-8|IGHJ3:IGHA', 'IGHA:pca_IGHV1-8|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV1-8|IGHJ5:IGHA', 'IGHA:pca_IGHV1-8|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV2-5|IGHJ1:IGHA', 'IGHA:pca_IGHV2-5|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV2-5|IGHJ3:IGHA', 'IGHA:pca_IGHV2-5|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV2-5|IGHJ5:IGHA', 'IGHA:pca_IGHV2-5|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-11|IGHJ1:IGHA', 'IGHA:pca_IGHV3-11|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-11|IGHJ3:IGHA', 'IGHA:pca_IGHV3-11|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-11|IGHJ5:IGHA', 'IGHA:pca_IGHV3-11|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-13|IGHJ1:IGHA', 'IGHA:pca_IGHV3-13|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-13|IGHJ3:IGHA', 'IGHA:pca_IGHV3-13|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-13|IGHJ5:IGHA', 'IGHA:pca_IGHV3-13|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-15|IGHJ1:IGHA', 'IGHA:pca_IGHV3-15|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-15|IGHJ3:IGHA', 'IGHA:pca_IGHV3-15|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-15|IGHJ5:IGHA', 'IGHA:pca_IGHV3-15|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-20|IGHJ1:IGHA', 'IGHA:pca_IGHV3-20|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-20|IGHJ3:IGHA', 'IGHA:pca_IGHV3-20|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-20|IGHJ5:IGHA', 'IGHA:pca_IGHV3-20|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-21|IGHJ1:IGHA', 'IGHA:pca_IGHV3-21|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-21|IGHJ3:IGHA', 'IGHA:pca_IGHV3-21|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-21|IGHJ5:IGHA', 'IGHA:pca_IGHV3-21|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-23|IGHJ1:IGHA', 'IGHA:pca_IGHV3-23|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-23|IGHJ3:IGHA', 'IGHA:pca_IGHV3-23|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-23|IGHJ5:IGHA', 'IGHA:pca_IGHV3-23|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-30-3|IGHJ1:IGHA', 'IGHA:pca_IGHV3-30-3|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-30-3|IGHJ3:IGHA', 'IGHA:pca_IGHV3-30-3|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-30-3|IGHJ5:IGHA', 'IGHA:pca_IGHV3-30-3|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-30|IGHJ1:IGHA', 'IGHA:pca_IGHV3-30|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-30|IGHJ3:IGHA', 'IGHA:pca_IGHV3-30|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-30|IGHJ5:IGHA', 'IGHA:pca_IGHV3-30|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-33|IGHJ1:IGHA', 'IGHA:pca_IGHV3-33|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-33|IGHJ3:IGHA', 'IGHA:pca_IGHV3-33|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-33|IGHJ5:IGHA', 'IGHA:pca_IGHV3-33|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-43|IGHJ1:IGHA', 'IGHA:pca_IGHV3-43|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-43|IGHJ3:IGHA', 'IGHA:pca_IGHV3-43|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-43|IGHJ5:IGHA', 'IGHA:pca_IGHV3-43|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-48|IGHJ1:IGHA', 'IGHA:pca_IGHV3-48|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-48|IGHJ3:IGHA', 'IGHA:pca_IGHV3-48|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-48|IGHJ5:IGHA', 'IGHA:pca_IGHV3-48|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-49|IGHJ1:IGHA', 'IGHA:pca_IGHV3-49|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-49|IGHJ3:IGHA', 'IGHA:pca_IGHV3-49|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-49|IGHJ5:IGHA', 'IGHA:pca_IGHV3-49|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-53|IGHJ1:IGHA', 'IGHA:pca_IGHV3-53|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-53|IGHJ3:IGHA', 'IGHA:pca_IGHV3-53|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-53|IGHJ5:IGHA', 'IGHA:pca_IGHV3-53|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-64|IGHJ1:IGHA', 'IGHA:pca_IGHV3-64|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-64|IGHJ3:IGHA', 'IGHA:pca_IGHV3-64|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-64|IGHJ5:IGHA', 'IGHA:pca_IGHV3-64|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-66|IGHJ1:IGHA', 'IGHA:pca_IGHV3-66|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-66|IGHJ3:IGHA', 'IGHA:pca_IGHV3-66|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-66|IGHJ5:IGHA', 'IGHA:pca_IGHV3-66|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-73|IGHJ1:IGHA', 'IGHA:pca_IGHV3-73|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-73|IGHJ3:IGHA', 'IGHA:pca_IGHV3-73|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-73|IGHJ5:IGHA', 'IGHA:pca_IGHV3-73|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-74|IGHJ1:IGHA', 'IGHA:pca_IGHV3-74|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-74|IGHJ3:IGHA', 'IGHA:pca_IGHV3-74|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-74|IGHJ5:IGHA', 'IGHA:pca_IGHV3-74|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-7|IGHJ1:IGHA', 'IGHA:pca_IGHV3-7|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-7|IGHJ3:IGHA', 'IGHA:pca_IGHV3-7|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-7|IGHJ5:IGHA', 'IGHA:pca_IGHV3-7|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV3-9|IGHJ1:IGHA', 'IGHA:pca_IGHV3-9|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV3-9|IGHJ3:IGHA', 'IGHA:pca_IGHV3-9|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV3-9|IGHJ5:IGHA', 'IGHA:pca_IGHV3-9|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV4-30-2|IGHJ1:IGHA', 'IGHA:pca_IGHV4-30-2|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV4-30-2|IGHJ3:IGHA', 'IGHA:pca_IGHV4-30-2|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV4-30-2|IGHJ5:IGHA', 'IGHA:pca_IGHV4-30-2|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV4-30-4|IGHJ1:IGHA', 'IGHA:pca_IGHV4-30-4|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV4-30-4|IGHJ3:IGHA', 'IGHA:pca_IGHV4-30-4|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV4-30-4|IGHJ5:IGHA', 'IGHA:pca_IGHV4-30-4|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV4-31|IGHJ1:IGHA', 'IGHA:pca_IGHV4-31|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV4-31|IGHJ3:IGHA', 'IGHA:pca_IGHV4-31|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV4-31|IGHJ5:IGHA', 'IGHA:pca_IGHV4-31|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV4-34|IGHJ1:IGHA', 'IGHA:pca_IGHV4-34|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV4-34|IGHJ3:IGHA', 'IGHA:pca_IGHV4-34|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV4-34|IGHJ5:IGHA', 'IGHA:pca_IGHV4-34|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV4-39|IGHJ1:IGHA', 'IGHA:pca_IGHV4-39|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV4-39|IGHJ3:IGHA', 'IGHA:pca_IGHV4-39|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV4-39|IGHJ5:IGHA', 'IGHA:pca_IGHV4-39|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV4-4|IGHJ1:IGHA', 'IGHA:pca_IGHV4-4|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV4-4|IGHJ3:IGHA', 'IGHA:pca_IGHV4-4|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV4-4|IGHJ5:IGHA', 'IGHA:pca_IGHV4-4|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV4-59|IGHJ1:IGHA', 'IGHA:pca_IGHV4-59|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV4-59|IGHJ3:IGHA', 'IGHA:pca_IGHV4-59|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV4-59|IGHJ5:IGHA', 'IGHA:pca_IGHV4-59|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV4-61|IGHJ1:IGHA', 'IGHA:pca_IGHV4-61|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV4-61|IGHJ3:IGHA', 'IGHA:pca_IGHV4-61|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV4-61|IGHJ5:IGHA', 'IGHA:pca_IGHV4-61|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV4-b|IGHJ1:IGHA', 'IGHA:pca_IGHV4-b|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV4-b|IGHJ3:IGHA', 'IGHA:pca_IGHV4-b|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV4-b|IGHJ5:IGHA', 'IGHA:pca_IGHV4-b|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV5-51|IGHJ1:IGHA', 'IGHA:pca_IGHV5-51|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV5-51|IGHJ3:IGHA', 'IGHA:pca_IGHV5-51|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV5-51|IGHJ5:IGHA', 'IGHA:pca_IGHV5-51|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV5-a|IGHJ1:IGHA', 'IGHA:pca_IGHV5-a|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV5-a|IGHJ3:IGHA', 'IGHA:pca_IGHV5-a|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV5-a|IGHJ5:IGHA', 'IGHA:pca_IGHV5-a|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV6-1|IGHJ1:IGHA', 'IGHA:pca_IGHV6-1|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV6-1|IGHJ3:IGHA', 'IGHA:pca_IGHV6-1|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV6-1|IGHJ5:IGHA', 'IGHA:pca_IGHV6-1|IGHJ6:IGHA',\n",
       "       'IGHA:pca_IGHV7-4-1|IGHJ1:IGHA', 'IGHA:pca_IGHV7-4-1|IGHJ2:IGHA',\n",
       "       'IGHA:pca_IGHV7-4-1|IGHJ3:IGHA', 'IGHA:pca_IGHV7-4-1|IGHJ4:IGHA',\n",
       "       'IGHA:pca_IGHV7-4-1|IGHJ5:IGHA', 'IGHA:pca_IGHV7-4-1|IGHJ6:IGHA',\n",
       "       'IGHD-M:pca_IGHV1-18|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-18|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-18|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-18|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-18|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-18|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-24|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-24|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-24|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-24|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-24|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-24|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-2|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-2|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-2|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-2|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-2|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-2|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-3|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-3|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-3|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-3|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-3|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-3|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-46|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-46|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-46|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-46|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-46|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-46|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-69|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-69|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-69|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-69|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-69|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-69|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-8|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-8|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-8|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-8|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-8|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV1-8|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV2-5|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV2-5|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV2-5|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV2-5|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV2-5|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV2-5|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-11|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-11|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-11|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-11|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-11|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-11|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-13|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-13|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-13|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-13|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-13|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-13|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-15|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-15|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-15|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-15|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-15|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-15|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-20|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-20|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-20|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-20|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-20|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-20|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-21|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-21|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-21|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-21|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-21|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-21|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-23|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-23|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-23|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-23|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-23|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-23|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-30-3|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-30-3|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-30-3|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-30-3|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-30-3|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-30-3|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-30|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-30|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-30|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-30|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-30|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-30|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-33|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-33|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-33|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-33|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-33|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-33|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-43|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-43|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-43|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-43|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-43|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-43|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-48|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-48|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-48|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-48|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-48|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-48|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-49|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-49|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-49|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-49|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-49|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-49|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-53|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-53|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-53|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-53|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-53|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-53|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-64|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-64|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-64|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-64|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-64|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-64|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-66|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-66|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-66|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-66|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-66|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-66|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-73|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-73|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-73|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-73|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-73|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-73|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-74|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-74|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-74|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-74|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-74|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-74|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-7|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-7|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-7|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-7|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-7|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-7|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-9|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-9|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-9|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-9|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-9|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV3-9|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-30-2|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-30-2|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-30-2|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-30-2|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-30-2|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-30-2|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-30-4|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-30-4|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-30-4|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-30-4|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-30-4|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-30-4|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-31|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-31|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-31|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-31|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-31|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-31|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-34|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-34|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-34|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-34|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-34|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-34|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-39|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-39|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-39|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-39|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-39|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-39|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-4|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-4|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-4|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-4|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-4|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-4|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-59|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-59|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-59|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-59|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-59|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-59|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-61|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-61|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-61|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-61|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-61|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-61|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-b|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-b|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-b|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-b|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-b|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV4-b|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV5-51|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV5-51|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV5-51|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV5-51|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV5-51|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV5-51|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV5-a|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV5-a|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV5-a|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV5-a|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV5-a|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV5-a|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV6-1|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV6-1|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV6-1|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV6-1|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV6-1|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV6-1|IGHJ6:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV7-4-1|IGHJ1:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV7-4-1|IGHJ2:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV7-4-1|IGHJ3:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV7-4-1|IGHJ4:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV7-4-1|IGHJ5:IGHD-M',\n",
       "       'IGHD-M:pca_IGHV7-4-1|IGHJ6:IGHD-M'], dtype=object)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf1.feature_names_in_  # Attribute access is passed through to the _inner scikit-learn Pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c107d7c-751c-4774-9021-98873faad3fe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c3835633-800f-4b0e-8e7b-e8818c6712b6",
   "metadata": {},
   "source": [
    "And now here is the reduced set of features coming out of the `ColumnTransformer` step. Notice how the V gene/J gene count features have turned into 15 PCs per isotype:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "59446f52-2367-4774-9c73-cadf259c90cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['log1p-scale-PCA_IGHG__pca0', 'log1p-scale-PCA_IGHG__pca1',\n",
       "       'log1p-scale-PCA_IGHG__pca2', 'log1p-scale-PCA_IGHG__pca3',\n",
       "       'log1p-scale-PCA_IGHG__pca4', 'log1p-scale-PCA_IGHG__pca5',\n",
       "       'log1p-scale-PCA_IGHG__pca6', 'log1p-scale-PCA_IGHG__pca7',\n",
       "       'log1p-scale-PCA_IGHG__pca8', 'log1p-scale-PCA_IGHG__pca9',\n",
       "       'log1p-scale-PCA_IGHG__pca10', 'log1p-scale-PCA_IGHG__pca11',\n",
       "       'log1p-scale-PCA_IGHG__pca12', 'log1p-scale-PCA_IGHG__pca13',\n",
       "       'log1p-scale-PCA_IGHG__pca14', 'log1p-scale-PCA_IGHA__pca0',\n",
       "       'log1p-scale-PCA_IGHA__pca1', 'log1p-scale-PCA_IGHA__pca2',\n",
       "       'log1p-scale-PCA_IGHA__pca3', 'log1p-scale-PCA_IGHA__pca4',\n",
       "       'log1p-scale-PCA_IGHA__pca5', 'log1p-scale-PCA_IGHA__pca6',\n",
       "       'log1p-scale-PCA_IGHA__pca7', 'log1p-scale-PCA_IGHA__pca8',\n",
       "       'log1p-scale-PCA_IGHA__pca9', 'log1p-scale-PCA_IGHA__pca10',\n",
       "       'log1p-scale-PCA_IGHA__pca11', 'log1p-scale-PCA_IGHA__pca12',\n",
       "       'log1p-scale-PCA_IGHA__pca13', 'log1p-scale-PCA_IGHA__pca14',\n",
       "       'log1p-scale-PCA_IGHD-M__pca0', 'log1p-scale-PCA_IGHD-M__pca1',\n",
       "       'log1p-scale-PCA_IGHD-M__pca2', 'log1p-scale-PCA_IGHD-M__pca3',\n",
       "       'log1p-scale-PCA_IGHD-M__pca4', 'log1p-scale-PCA_IGHD-M__pca5',\n",
       "       'log1p-scale-PCA_IGHD-M__pca6', 'log1p-scale-PCA_IGHD-M__pca7',\n",
       "       'log1p-scale-PCA_IGHD-M__pca8', 'log1p-scale-PCA_IGHD-M__pca9',\n",
       "       'log1p-scale-PCA_IGHD-M__pca10', 'log1p-scale-PCA_IGHD-M__pca11',\n",
       "       'log1p-scale-PCA_IGHD-M__pca12', 'log1p-scale-PCA_IGHD-M__pca13',\n",
       "       'log1p-scale-PCA_IGHD-M__pca14',\n",
       "       'remainder__v_mut_median_per_specimen:IGHG',\n",
       "       'remainder__v_sequence_is_mutated:IGHG',\n",
       "       'remainder__v_mut_median_per_specimen:IGHA',\n",
       "       'remainder__v_sequence_is_mutated:IGHA',\n",
       "       'remainder__v_mut_median_per_specimen:IGHD-M',\n",
       "       'remainder__v_sequence_is_mutated:IGHD-M'], dtype=object)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf1.named_steps[\"columntransformer\"].get_feature_names_out()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e100c158-69fc-48cb-901e-00e89b3838f7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fb6c32bf-3ba4-4117-b8fa-b6277a5b9c1a",
   "metadata": {},
   "source": [
    "Here is the final step of the pipeline. As expected based on the `elasticnet_cv0.25` model name we specified when loading the trained model from disk, it's a elasticnet logistic regression with an 0.25 L1-L2 ratio:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5d040c6a-27ae-4b15-ab4d-9723e527266c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('glmnetlogitnetwrapper',\n",
       " GlmnetLogitNetWrapper: LogitNet(alpha=0.25, n_splits=5, random_state=0,\n",
       "          scoring=<function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0>,\n",
       "          standardize=False, verbose=True))"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf1.steps[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e29475ba-e457-4ea8-b36d-700da42629b7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e2ca44a5-1cca-4c82-b377-493ea70c6564",
   "metadata": {},
   "source": [
    "**Mal-ID models have the following API:**\n",
    "\n",
    "- **`clf.featurize(adata)`**: this function accepts an AnnData object and generates features specific for the model. The features and metadata are returned in a `FeaturizedData` container, which we'll explore below. (The features themselves are in the `.X` attribute of the `FeaturizedData` container.)\n",
    "- **`clf.predict_proba(features)`**: this function accepts features and returns predicted class probabilities by running the model.\n",
    "- **`clf.predict(features)`**: this function accepts features and returns predicted class labels by running the model.\n",
    "\n",
    "**_Common pattern:_** `predicted_class_probabilities = clf.predict_proba(clf.featurize(adata).X)`.\n",
    "\n",
    "Let's walk through this with Model 1. First, let's generate features from the held-out BCR data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6d0165ce-9ee4-47f3-9724-0839c987fd82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "crosseval.featurized_data.FeaturizedData"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "featurized_model1_data = clf1.featurize(adata_bcr)\n",
    "type(featurized_model1_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "007c589d-5193-4f07-9fe2-b78bd2276a60",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "512a5a52-a5ef-4562-ab25-8817863c1f67",
   "metadata": {},
   "source": [
    "We now have a `FeaturizedData` container. Let's unpack its contents:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0b15500b-5031-4cfc-a13c-a458c526bef0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>v_mut_median_per_specimen:IGHG</th>\n",
       "      <th>v_sequence_is_mutated:IGHG</th>\n",
       "      <th>v_mut_median_per_specimen:IGHA</th>\n",
       "      <th>v_sequence_is_mutated:IGHA</th>\n",
       "      <th>v_mut_median_per_specimen:IGHD-M</th>\n",
       "      <th>v_sequence_is_mutated:IGHD-M</th>\n",
       "      <th>IGHG:pca_IGHV1-18|IGHJ1:IGHG</th>\n",
       "      <th>IGHG:pca_IGHV1-18|IGHJ2:IGHG</th>\n",
       "      <th>IGHG:pca_IGHV1-18|IGHJ3:IGHG</th>\n",
       "      <th>IGHG:pca_IGHV1-18|IGHJ4:IGHG</th>\n",
       "      <th>...</th>\n",
       "      <th>IGHD-M:pca_IGHV6-1|IGHJ3:IGHD-M</th>\n",
       "      <th>IGHD-M:pca_IGHV6-1|IGHJ4:IGHD-M</th>\n",
       "      <th>IGHD-M:pca_IGHV6-1|IGHJ5:IGHD-M</th>\n",
       "      <th>IGHD-M:pca_IGHV6-1|IGHJ6:IGHD-M</th>\n",
       "      <th>IGHD-M:pca_IGHV7-4-1|IGHJ1:IGHD-M</th>\n",
       "      <th>IGHD-M:pca_IGHV7-4-1|IGHJ2:IGHD-M</th>\n",
       "      <th>IGHD-M:pca_IGHV7-4-1|IGHJ3:IGHD-M</th>\n",
       "      <th>IGHD-M:pca_IGHV7-4-1|IGHJ4:IGHD-M</th>\n",
       "      <th>IGHD-M:pca_IGHV7-4-1|IGHJ5:IGHD-M</th>\n",
       "      <th>IGHD-M:pca_IGHV7-4-1|IGHJ6:IGHD-M</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M111-S037</th>\n",
       "      <td>0.057778</td>\n",
       "      <td>0.911347</td>\n",
       "      <td>0.088496</td>\n",
       "      <td>0.979737</td>\n",
       "      <td>0.026549</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001832</td>\n",
       "      <td>0.001603</td>\n",
       "      <td>0.005955</td>\n",
       "      <td>0.023362</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000177</td>\n",
       "      <td>0.000473</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000827</td>\n",
       "      <td>0.003251</td>\n",
       "      <td>0.000827</td>\n",
       "      <td>0.000709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S042</th>\n",
       "      <td>0.079295</td>\n",
       "      <td>0.959693</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.990588</td>\n",
       "      <td>0.030973</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000850</td>\n",
       "      <td>0.000567</td>\n",
       "      <td>0.003967</td>\n",
       "      <td>0.020686</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000527</td>\n",
       "      <td>0.002636</td>\n",
       "      <td>0.000527</td>\n",
       "      <td>0.000527</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S016</th>\n",
       "      <td>0.074890</td>\n",
       "      <td>0.951521</td>\n",
       "      <td>0.076233</td>\n",
       "      <td>0.982593</td>\n",
       "      <td>0.026549</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000305</td>\n",
       "      <td>0.007324</td>\n",
       "      <td>0.024413</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001624</td>\n",
       "      <td>0.001662</td>\n",
       "      <td>0.001082</td>\n",
       "      <td>0.000889</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.000850</td>\n",
       "      <td>0.001855</td>\n",
       "      <td>0.000618</td>\n",
       "      <td>0.000541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S043</th>\n",
       "      <td>0.047414</td>\n",
       "      <td>0.874591</td>\n",
       "      <td>0.066079</td>\n",
       "      <td>0.939325</td>\n",
       "      <td>0.029787</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001112</td>\n",
       "      <td>0.000953</td>\n",
       "      <td>0.006989</td>\n",
       "      <td>0.015248</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001325</td>\n",
       "      <td>0.002261</td>\n",
       "      <td>0.000780</td>\n",
       "      <td>0.000546</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S036</th>\n",
       "      <td>0.077586</td>\n",
       "      <td>0.966753</td>\n",
       "      <td>0.074236</td>\n",
       "      <td>0.966128</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.002915</td>\n",
       "      <td>0.002624</td>\n",
       "      <td>0.006122</td>\n",
       "      <td>0.027697</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S146</th>\n",
       "      <td>0.052863</td>\n",
       "      <td>0.837728</td>\n",
       "      <td>0.034632</td>\n",
       "      <td>0.691589</td>\n",
       "      <td>0.026667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002088</td>\n",
       "      <td>0.014614</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.002330</td>\n",
       "      <td>0.000736</td>\n",
       "      <td>0.001103</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000981</td>\n",
       "      <td>0.006253</td>\n",
       "      <td>0.003065</td>\n",
       "      <td>0.003188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S148</th>\n",
       "      <td>0.057269</td>\n",
       "      <td>0.940083</td>\n",
       "      <td>0.070485</td>\n",
       "      <td>0.969115</td>\n",
       "      <td>0.026549</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001055</td>\n",
       "      <td>0.005274</td>\n",
       "      <td>0.021097</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000660</td>\n",
       "      <td>0.001885</td>\n",
       "      <td>0.000377</td>\n",
       "      <td>0.000566</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>0.001225</td>\n",
       "      <td>0.006127</td>\n",
       "      <td>0.002168</td>\n",
       "      <td>0.002451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S149</th>\n",
       "      <td>0.069565</td>\n",
       "      <td>0.977945</td>\n",
       "      <td>0.058296</td>\n",
       "      <td>0.929610</td>\n",
       "      <td>0.030568</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000962</td>\n",
       "      <td>0.005775</td>\n",
       "      <td>0.020693</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000670</td>\n",
       "      <td>0.004019</td>\n",
       "      <td>0.002010</td>\n",
       "      <td>0.002345</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S153</th>\n",
       "      <td>0.065502</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.070485</td>\n",
       "      <td>0.923849</td>\n",
       "      <td>0.026432</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001368</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009576</td>\n",
       "      <td>0.019152</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000553</td>\n",
       "      <td>0.001660</td>\n",
       "      <td>0.000830</td>\n",
       "      <td>0.000277</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S160</th>\n",
       "      <td>0.074236</td>\n",
       "      <td>0.919571</td>\n",
       "      <td>0.065217</td>\n",
       "      <td>0.909188</td>\n",
       "      <td>0.026087</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.002747</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008242</td>\n",
       "      <td>0.019231</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000348</td>\n",
       "      <td>0.001915</td>\n",
       "      <td>0.000522</td>\n",
       "      <td>0.001045</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>206 rows × 726 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                v_mut_median_per_specimen:IGHG  v_sequence_is_mutated:IGHG  \\\n",
       "specimen_label                                                               \n",
       "M111-S037                             0.057778                    0.911347   \n",
       "M124-S042                             0.079295                    0.959693   \n",
       "M111-S016                             0.074890                    0.951521   \n",
       "M111-S043                             0.047414                    0.874591   \n",
       "M124-S036                             0.077586                    0.966753   \n",
       "...                                        ...                         ...   \n",
       "M491-S146                             0.052863                    0.837728   \n",
       "M491-S148                             0.057269                    0.940083   \n",
       "M491-S149                             0.069565                    0.977945   \n",
       "M491-S153                             0.065502                    0.916667   \n",
       "M491-S160                             0.074236                    0.919571   \n",
       "\n",
       "                v_mut_median_per_specimen:IGHA  v_sequence_is_mutated:IGHA  \\\n",
       "specimen_label                                                               \n",
       "M111-S037                             0.088496                    0.979737   \n",
       "M124-S042                             0.083333                    0.990588   \n",
       "M111-S016                             0.076233                    0.982593   \n",
       "M111-S043                             0.066079                    0.939325   \n",
       "M124-S036                             0.074236                    0.966128   \n",
       "...                                        ...                         ...   \n",
       "M491-S146                             0.034632                    0.691589   \n",
       "M491-S148                             0.070485                    0.969115   \n",
       "M491-S149                             0.058296                    0.929610   \n",
       "M491-S153                             0.070485                    0.923849   \n",
       "M491-S160                             0.065217                    0.909188   \n",
       "\n",
       "                v_mut_median_per_specimen:IGHD-M  \\\n",
       "specimen_label                                     \n",
       "M111-S037                               0.026549   \n",
       "M124-S042                               0.030973   \n",
       "M111-S016                               0.026549   \n",
       "M111-S043                               0.029787   \n",
       "M124-S036                               0.030303   \n",
       "...                                          ...   \n",
       "M491-S146                               0.026667   \n",
       "M491-S148                               0.026549   \n",
       "M491-S149                               0.030568   \n",
       "M491-S153                               0.026432   \n",
       "M491-S160                               0.026087   \n",
       "\n",
       "                v_sequence_is_mutated:IGHD-M  IGHG:pca_IGHV1-18|IGHJ1:IGHG  \\\n",
       "specimen_label                                                               \n",
       "M111-S037                                1.0                      0.001832   \n",
       "M124-S042                                1.0                      0.000850   \n",
       "M111-S016                                1.0                      0.000000   \n",
       "M111-S043                                1.0                      0.001112   \n",
       "M124-S036                                1.0                      0.002915   \n",
       "...                                      ...                           ...   \n",
       "M491-S146                                1.0                      0.000000   \n",
       "M491-S148                                1.0                      0.000000   \n",
       "M491-S149                                1.0                      0.000000   \n",
       "M491-S153                                1.0                      0.001368   \n",
       "M491-S160                                1.0                      0.002747   \n",
       "\n",
       "                IGHG:pca_IGHV1-18|IGHJ2:IGHG  IGHG:pca_IGHV1-18|IGHJ3:IGHG  \\\n",
       "specimen_label                                                               \n",
       "M111-S037                           0.001603                      0.005955   \n",
       "M124-S042                           0.000567                      0.003967   \n",
       "M111-S016                           0.000305                      0.007324   \n",
       "M111-S043                           0.000953                      0.006989   \n",
       "M124-S036                           0.002624                      0.006122   \n",
       "...                                      ...                           ...   \n",
       "M491-S146                           0.000000                      0.002088   \n",
       "M491-S148                           0.001055                      0.005274   \n",
       "M491-S149                           0.000962                      0.005775   \n",
       "M491-S153                           0.000000                      0.009576   \n",
       "M491-S160                           0.000000                      0.008242   \n",
       "\n",
       "                IGHG:pca_IGHV1-18|IGHJ4:IGHG  ...  \\\n",
       "specimen_label                                ...   \n",
       "M111-S037                           0.023362  ...   \n",
       "M124-S042                           0.020686  ...   \n",
       "M111-S016                           0.024413  ...   \n",
       "M111-S043                           0.015248  ...   \n",
       "M124-S036                           0.027697  ...   \n",
       "...                                      ...  ...   \n",
       "M491-S146                           0.014614  ...   \n",
       "M491-S148                           0.021097  ...   \n",
       "M491-S149                           0.020693  ...   \n",
       "M491-S153                           0.019152  ...   \n",
       "M491-S160                           0.019231  ...   \n",
       "\n",
       "                IGHD-M:pca_IGHV6-1|IGHJ3:IGHD-M  \\\n",
       "specimen_label                                    \n",
       "M111-S037                              0.000177   \n",
       "M124-S042                              0.000527   \n",
       "M111-S016                              0.001624   \n",
       "M111-S043                              0.001325   \n",
       "M124-S036                              0.000000   \n",
       "...                                         ...   \n",
       "M491-S146                              0.000123   \n",
       "M491-S148                              0.000660   \n",
       "M491-S149                              0.000670   \n",
       "M491-S153                              0.000553   \n",
       "M491-S160                              0.000348   \n",
       "\n",
       "                IGHD-M:pca_IGHV6-1|IGHJ4:IGHD-M  \\\n",
       "specimen_label                                    \n",
       "M111-S037                              0.000473   \n",
       "M124-S042                              0.002636   \n",
       "M111-S016                              0.001662   \n",
       "M111-S043                              0.002261   \n",
       "M124-S036                              0.000085   \n",
       "...                                         ...   \n",
       "M491-S146                              0.002330   \n",
       "M491-S148                              0.001885   \n",
       "M491-S149                              0.004019   \n",
       "M491-S153                              0.001660   \n",
       "M491-S160                              0.001915   \n",
       "\n",
       "                IGHD-M:pca_IGHV6-1|IGHJ5:IGHD-M  \\\n",
       "specimen_label                                    \n",
       "M111-S037                              0.000059   \n",
       "M124-S042                              0.000527   \n",
       "M111-S016                              0.001082   \n",
       "M111-S043                              0.000780   \n",
       "M124-S036                              0.000085   \n",
       "...                                         ...   \n",
       "M491-S146                              0.000736   \n",
       "M491-S148                              0.000377   \n",
       "M491-S149                              0.002010   \n",
       "M491-S153                              0.000830   \n",
       "M491-S160                              0.000522   \n",
       "\n",
       "                IGHD-M:pca_IGHV6-1|IGHJ6:IGHD-M  \\\n",
       "specimen_label                                    \n",
       "M111-S037                              0.000059   \n",
       "M124-S042                              0.000527   \n",
       "M111-S016                              0.000889   \n",
       "M111-S043                              0.000546   \n",
       "M124-S036                              0.000085   \n",
       "...                                         ...   \n",
       "M491-S146                              0.001103   \n",
       "M491-S148                              0.000566   \n",
       "M491-S149                              0.002345   \n",
       "M491-S153                              0.000277   \n",
       "M491-S160                              0.001045   \n",
       "\n",
       "                IGHD-M:pca_IGHV7-4-1|IGHJ1:IGHD-M  \\\n",
       "specimen_label                                      \n",
       "M111-S037                                0.000118   \n",
       "M124-S042                                0.000000   \n",
       "M111-S016                                0.000077   \n",
       "M111-S043                                0.000000   \n",
       "M124-S036                                0.000000   \n",
       "...                                           ...   \n",
       "M491-S146                                0.000000   \n",
       "M491-S148                                0.000094   \n",
       "M491-S149                                0.000000   \n",
       "M491-S153                                0.000000   \n",
       "M491-S160                                0.000000   \n",
       "\n",
       "                IGHD-M:pca_IGHV7-4-1|IGHJ2:IGHD-M  \\\n",
       "specimen_label                                      \n",
       "M111-S037                                0.000118   \n",
       "M124-S042                                0.000000   \n",
       "M111-S016                                0.000077   \n",
       "M111-S043                                0.000000   \n",
       "M124-S036                                0.000000   \n",
       "...                                           ...   \n",
       "M491-S146                                0.000000   \n",
       "M491-S148                                0.000189   \n",
       "M491-S149                                0.000000   \n",
       "M491-S153                                0.000000   \n",
       "M491-S160                                0.000000   \n",
       "\n",
       "                IGHD-M:pca_IGHV7-4-1|IGHJ3:IGHD-M  \\\n",
       "specimen_label                                      \n",
       "M111-S037                                0.000827   \n",
       "M124-S042                                0.000000   \n",
       "M111-S016                                0.000850   \n",
       "M111-S043                                0.000000   \n",
       "M124-S036                                0.000000   \n",
       "...                                           ...   \n",
       "M491-S146                                0.000981   \n",
       "M491-S148                                0.001225   \n",
       "M491-S149                                0.000167   \n",
       "M491-S153                                0.000000   \n",
       "M491-S160                                0.000000   \n",
       "\n",
       "                IGHD-M:pca_IGHV7-4-1|IGHJ4:IGHD-M  \\\n",
       "specimen_label                                      \n",
       "M111-S037                                0.003251   \n",
       "M124-S042                                0.000105   \n",
       "M111-S016                                0.001855   \n",
       "M111-S043                                0.000000   \n",
       "M124-S036                                0.000000   \n",
       "...                                           ...   \n",
       "M491-S146                                0.006253   \n",
       "M491-S148                                0.006127   \n",
       "M491-S149                                0.000000   \n",
       "M491-S153                                0.000000   \n",
       "M491-S160                                0.000000   \n",
       "\n",
       "                IGHD-M:pca_IGHV7-4-1|IGHJ5:IGHD-M  \\\n",
       "specimen_label                                      \n",
       "M111-S037                                0.000827   \n",
       "M124-S042                                0.000000   \n",
       "M111-S016                                0.000618   \n",
       "M111-S043                                0.000000   \n",
       "M124-S036                                0.000000   \n",
       "...                                           ...   \n",
       "M491-S146                                0.003065   \n",
       "M491-S148                                0.002168   \n",
       "M491-S149                                0.000167   \n",
       "M491-S153                                0.000000   \n",
       "M491-S160                                0.000000   \n",
       "\n",
       "                IGHD-M:pca_IGHV7-4-1|IGHJ6:IGHD-M  \n",
       "specimen_label                                     \n",
       "M111-S037                                0.000709  \n",
       "M124-S042                                0.000000  \n",
       "M111-S016                                0.000541  \n",
       "M111-S043                                0.000000  \n",
       "M124-S036                                0.000000  \n",
       "...                                           ...  \n",
       "M491-S146                                0.003188  \n",
       "M491-S148                                0.002451  \n",
       "M491-S149                                0.000000  \n",
       "M491-S153                                0.000000  \n",
       "M491-S160                                0.000000  \n",
       "\n",
       "[206 rows x 726 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Features\n",
    "featurized_model1_data.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c6d808d0-576a-4401-816a-c7d144695d73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "specimen_label\n",
       "M111-S037                   HIV\n",
       "M124-S042    Healthy/Background\n",
       "M111-S016                   HIV\n",
       "M111-S043                   HIV\n",
       "M124-S036    Healthy/Background\n",
       "                    ...        \n",
       "M491-S146                   T1D\n",
       "M491-S148                   T1D\n",
       "M491-S149                   T1D\n",
       "M491-S153                   T1D\n",
       "M491-S160                   T1D\n",
       "Name: disease, Length: 206, dtype: category\n",
       "Categories (6, object): ['HIV', 'Healthy/Background', 'Lupus', 'Covid19', 'Influenza', 'T1D']"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ground truth\n",
    "featurized_model1_data.y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "02a0e975-7b17-4812-9c3f-fb684ab6d46a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['M111-S037', 'M124-S042', 'M111-S016', 'M111-S043', 'M124-S036',\n",
       "       'M124-S040', 'M111-S035', 'M111-S025', 'M111-S011', 'M111-S036',\n",
       "       ...\n",
       "       'M491-S111', 'M491-S117', 'M491-S124', 'M491-S126', 'M491-S136',\n",
       "       'M491-S146', 'M491-S148', 'M491-S149', 'M491-S153', 'M491-S160'],\n",
       "      dtype='object', name='specimen_label', length=206)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample names\n",
    "featurized_model1_data.sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c2d5a5a9-84fa-4e58-b6c2-82614b31bdd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease.rollup</th>\n",
       "      <th>disease.separate_past_exposures</th>\n",
       "      <th>disease_severity</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>ethnicity_condensed</th>\n",
       "      <th>isotype_proportion:IGHA</th>\n",
       "      <th>isotype_proportion:IGHD-M</th>\n",
       "      <th>isotype_proportion:IGHG</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>past_exposure</th>\n",
       "      <th>sex</th>\n",
       "      <th>study_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M111-S037</th>\n",
       "      <td>48.0</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV Broad Neutralizing</td>\n",
       "      <td>African</td>\n",
       "      <td>0.140070</td>\n",
       "      <td>0.682723</td>\n",
       "      <td>0.177207</td>\n",
       "      <td>BFI-0000254</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S042</th>\n",
       "      <td>26.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.246533</td>\n",
       "      <td>0.547060</td>\n",
       "      <td>0.206407</td>\n",
       "      <td>BFI-0002850</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S016</th>\n",
       "      <td>22.0</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV Broad Neutralizing</td>\n",
       "      <td>African</td>\n",
       "      <td>0.101753</td>\n",
       "      <td>0.796830</td>\n",
       "      <td>0.101417</td>\n",
       "      <td>BFI-0002855</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S043</th>\n",
       "      <td>23.0</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV Non Neutralizing</td>\n",
       "      <td>African</td>\n",
       "      <td>0.228808</td>\n",
       "      <td>0.516170</td>\n",
       "      <td>0.255022</td>\n",
       "      <td>BFI-0002856</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S036</th>\n",
       "      <td>29.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>African</td>\n",
       "      <td>0.144986</td>\n",
       "      <td>0.660305</td>\n",
       "      <td>0.194710</td>\n",
       "      <td>BFI-0002862</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S146</th>\n",
       "      <td>15.0</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.098597</td>\n",
       "      <td>0.850927</td>\n",
       "      <td>0.050476</td>\n",
       "      <td>BFI-0010788</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S148</th>\n",
       "      <td>14.0</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.110664</td>\n",
       "      <td>0.815813</td>\n",
       "      <td>0.073523</td>\n",
       "      <td>BFI-0010790</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S149</th>\n",
       "      <td>11.0</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.136773</td>\n",
       "      <td>0.640226</td>\n",
       "      <td>0.223001</td>\n",
       "      <td>BFI-0010791</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S153</th>\n",
       "      <td>16.0</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.160569</td>\n",
       "      <td>0.697251</td>\n",
       "      <td>0.142180</td>\n",
       "      <td>BFI-0010794</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S160</th>\n",
       "      <td>16.0</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.130453</td>\n",
       "      <td>0.817561</td>\n",
       "      <td>0.051986</td>\n",
       "      <td>BFI-0010800</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>206 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 age             disease      disease.rollup  \\\n",
       "specimen_label                                                 \n",
       "M111-S037       48.0                 HIV                 HIV   \n",
       "M124-S042       26.0  Healthy/Background  Healthy/Background   \n",
       "M111-S016       22.0                 HIV                 HIV   \n",
       "M111-S043       23.0                 HIV                 HIV   \n",
       "M124-S036       29.0  Healthy/Background  Healthy/Background   \n",
       "...              ...                 ...                 ...   \n",
       "M491-S146       15.0                 T1D                 T1D   \n",
       "M491-S148       14.0                 T1D                 T1D   \n",
       "M491-S149       11.0                 T1D                 T1D   \n",
       "M491-S153       16.0                 T1D                 T1D   \n",
       "M491-S160       16.0                 T1D                 T1D   \n",
       "\n",
       "               disease.separate_past_exposures disease_severity  \\\n",
       "specimen_label                                                    \n",
       "M111-S037                                  HIV              NaN   \n",
       "M124-S042                   Healthy/Background              NaN   \n",
       "M111-S016                                  HIV              NaN   \n",
       "M111-S043                                  HIV              NaN   \n",
       "M124-S036                   Healthy/Background              NaN   \n",
       "...                                        ...              ...   \n",
       "M491-S146                                  T1D              NaN   \n",
       "M491-S148                                  T1D              NaN   \n",
       "M491-S149                                  T1D              NaN   \n",
       "M491-S153                                  T1D              NaN   \n",
       "M491-S160                                  T1D              NaN   \n",
       "\n",
       "                                  disease_subtype ethnicity_condensed  \\\n",
       "specimen_label                                                          \n",
       "M111-S037                  HIV Broad Neutralizing             African   \n",
       "M124-S042       Healthy/Background - HIV Negative                 NaN   \n",
       "M111-S016                  HIV Broad Neutralizing             African   \n",
       "M111-S043                    HIV Non Neutralizing             African   \n",
       "M124-S036       Healthy/Background - HIV Negative             African   \n",
       "...                                           ...                 ...   \n",
       "M491-S146                         T1D - pediatric                 NaN   \n",
       "M491-S148                         T1D - pediatric                 NaN   \n",
       "M491-S149                         T1D - pediatric                 NaN   \n",
       "M491-S153                         T1D - pediatric                 NaN   \n",
       "M491-S160                         T1D - pediatric                 NaN   \n",
       "\n",
       "                isotype_proportion:IGHA  isotype_proportion:IGHD-M  \\\n",
       "specimen_label                                                       \n",
       "M111-S037                      0.140070                   0.682723   \n",
       "M124-S042                      0.246533                   0.547060   \n",
       "M111-S016                      0.101753                   0.796830   \n",
       "M111-S043                      0.228808                   0.516170   \n",
       "M124-S036                      0.144986                   0.660305   \n",
       "...                                 ...                        ...   \n",
       "M491-S146                      0.098597                   0.850927   \n",
       "M491-S148                      0.110664                   0.815813   \n",
       "M491-S149                      0.136773                   0.640226   \n",
       "M491-S153                      0.160569                   0.697251   \n",
       "M491-S160                      0.130453                   0.817561   \n",
       "\n",
       "                isotype_proportion:IGHG participant_label  past_exposure  sex  \\\n",
       "specimen_label                                                                  \n",
       "M111-S037                      0.177207       BFI-0000254          False    F   \n",
       "M124-S042                      0.206407       BFI-0002850          False    F   \n",
       "M111-S016                      0.101417       BFI-0002855          False    F   \n",
       "M111-S043                      0.255022       BFI-0002856          False    F   \n",
       "M124-S036                      0.194710       BFI-0002862          False    F   \n",
       "...                                 ...               ...            ...  ...   \n",
       "M491-S146                      0.050476       BFI-0010788          False  NaN   \n",
       "M491-S148                      0.073523       BFI-0010790          False  NaN   \n",
       "M491-S149                      0.223001       BFI-0010791          False  NaN   \n",
       "M491-S153                      0.142180       BFI-0010794          False  NaN   \n",
       "M491-S160                      0.051986       BFI-0010800          False  NaN   \n",
       "\n",
       "                      study_name  \n",
       "specimen_label                    \n",
       "M111-S037                    HIV  \n",
       "M124-S042                    HIV  \n",
       "M111-S016                    HIV  \n",
       "M111-S043                    HIV  \n",
       "M124-S036                    HIV  \n",
       "...                          ...  \n",
       "M491-S146       Diabetes biobank  \n",
       "M491-S148       Diabetes biobank  \n",
       "M491-S149       Diabetes biobank  \n",
       "M491-S153       Diabetes biobank  \n",
       "M491-S160       Diabetes biobank  \n",
       "\n",
       "[206 rows x 14 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample metadata\n",
    "featurized_model1_data.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dda81c30-e991-41e9-bac5-50f9ba703e38",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "45e19414-d3cf-4047-857c-32696625b802",
   "metadata": {},
   "source": [
    "Now run the model to predict the per-class probabilities:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e8323642-89c3-4188-a753-8a14df7ec9d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.20373269e-05, 9.93021939e-01, 6.68838158e-03, 4.56024978e-05,\n",
       "        2.21748447e-04, 1.02907163e-05],\n",
       "       [9.59384622e-04, 5.99845787e-03, 9.50646515e-01, 1.68303975e-03,\n",
       "        3.85957328e-02, 2.11686958e-03],\n",
       "       [8.50644976e-03, 7.52307114e-01, 1.39096021e-01, 4.49604811e-02,\n",
       "        4.66933810e-02, 8.43655264e-03],\n",
       "       ...,\n",
       "       [1.17617906e-02, 1.45660594e-03, 1.84622816e-01, 1.39845976e-02,\n",
       "        1.82523485e-01, 6.05650705e-01],\n",
       "       [2.04910658e-02, 2.52044988e-03, 3.15463558e-01, 2.78819744e-03,\n",
       "        6.38096684e-01, 2.06400456e-02],\n",
       "       [1.04655981e-02, 1.80573281e-03, 3.49655098e-01, 1.45640859e-02,\n",
       "        4.69149578e-04, 6.23040335e-01]])"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf1.predict_proba(featurized_model1_data.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cfdaf4f-16ba-43f4-96a4-5815ca44e18d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "54564f6c-5f97-476b-8ec5-3a16f3e0b87b",
   "metadata": {},
   "source": [
    "To make this easier to read, let's bring in row and column names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "b2f3dc8f-d051-4722-8be9-e701a3b2aece",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['M111-S037', 'M124-S042', 'M111-S016', 'M111-S043', 'M124-S036',\n",
       "       'M124-S040', 'M111-S035', 'M111-S025', 'M111-S011', 'M111-S036',\n",
       "       ...\n",
       "       'M491-S111', 'M491-S117', 'M491-S124', 'M491-S126', 'M491-S136',\n",
       "       'M491-S146', 'M491-S148', 'M491-S149', 'M491-S153', 'M491-S160'],\n",
       "      dtype='object', name='specimen_label', length=206)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Row names\n",
    "featurized_model1_data.sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "395762e5-f729-4d7c-b68e-de4d5fa34587",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Covid19', 'HIV', 'Healthy/Background', 'Influenza', 'Lupus',\n",
       "       'T1D'], dtype=object)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Column names\n",
    "clf1.classes_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "5c293e0b-dba0-45cc-953d-e7338c468dcb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Covid19</th>\n",
       "      <th>HIV</th>\n",
       "      <th>Healthy/Background</th>\n",
       "      <th>Influenza</th>\n",
       "      <th>Lupus</th>\n",
       "      <th>T1D</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M111-S037</th>\n",
       "      <td>0.000012</td>\n",
       "      <td>9.930219e-01</td>\n",
       "      <td>0.006688</td>\n",
       "      <td>4.560250e-05</td>\n",
       "      <td>0.000222</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S042</th>\n",
       "      <td>0.000959</td>\n",
       "      <td>5.998458e-03</td>\n",
       "      <td>0.950647</td>\n",
       "      <td>1.683040e-03</td>\n",
       "      <td>0.038596</td>\n",
       "      <td>0.002117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S016</th>\n",
       "      <td>0.008506</td>\n",
       "      <td>7.523071e-01</td>\n",
       "      <td>0.139096</td>\n",
       "      <td>4.496048e-02</td>\n",
       "      <td>0.046693</td>\n",
       "      <td>0.008437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S043</th>\n",
       "      <td>0.000105</td>\n",
       "      <td>9.961163e-01</td>\n",
       "      <td>0.003342</td>\n",
       "      <td>4.690368e-05</td>\n",
       "      <td>0.000294</td>\n",
       "      <td>0.000096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S036</th>\n",
       "      <td>0.000226</td>\n",
       "      <td>2.429462e-01</td>\n",
       "      <td>0.704671</td>\n",
       "      <td>5.227891e-03</td>\n",
       "      <td>0.041741</td>\n",
       "      <td>0.005189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S146</th>\n",
       "      <td>0.000003</td>\n",
       "      <td>1.615749e-07</td>\n",
       "      <td>0.000527</td>\n",
       "      <td>1.550379e-07</td>\n",
       "      <td>0.000238</td>\n",
       "      <td>0.999232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S148</th>\n",
       "      <td>0.099963</td>\n",
       "      <td>4.009954e-02</td>\n",
       "      <td>0.477593</td>\n",
       "      <td>8.308356e-02</td>\n",
       "      <td>0.177256</td>\n",
       "      <td>0.122004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S149</th>\n",
       "      <td>0.011762</td>\n",
       "      <td>1.456606e-03</td>\n",
       "      <td>0.184623</td>\n",
       "      <td>1.398460e-02</td>\n",
       "      <td>0.182523</td>\n",
       "      <td>0.605651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S153</th>\n",
       "      <td>0.020491</td>\n",
       "      <td>2.520450e-03</td>\n",
       "      <td>0.315464</td>\n",
       "      <td>2.788197e-03</td>\n",
       "      <td>0.638097</td>\n",
       "      <td>0.020640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S160</th>\n",
       "      <td>0.010466</td>\n",
       "      <td>1.805733e-03</td>\n",
       "      <td>0.349655</td>\n",
       "      <td>1.456409e-02</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>0.623040</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>206 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Covid19           HIV  Healthy/Background     Influenza  \\\n",
       "specimen_label                                                             \n",
       "M111-S037       0.000012  9.930219e-01            0.006688  4.560250e-05   \n",
       "M124-S042       0.000959  5.998458e-03            0.950647  1.683040e-03   \n",
       "M111-S016       0.008506  7.523071e-01            0.139096  4.496048e-02   \n",
       "M111-S043       0.000105  9.961163e-01            0.003342  4.690368e-05   \n",
       "M124-S036       0.000226  2.429462e-01            0.704671  5.227891e-03   \n",
       "...                  ...           ...                 ...           ...   \n",
       "M491-S146       0.000003  1.615749e-07            0.000527  1.550379e-07   \n",
       "M491-S148       0.099963  4.009954e-02            0.477593  8.308356e-02   \n",
       "M491-S149       0.011762  1.456606e-03            0.184623  1.398460e-02   \n",
       "M491-S153       0.020491  2.520450e-03            0.315464  2.788197e-03   \n",
       "M491-S160       0.010466  1.805733e-03            0.349655  1.456409e-02   \n",
       "\n",
       "                   Lupus       T1D  \n",
       "specimen_label                      \n",
       "M111-S037       0.000222  0.000010  \n",
       "M124-S042       0.038596  0.002117  \n",
       "M111-S016       0.046693  0.008437  \n",
       "M111-S043       0.000294  0.000096  \n",
       "M124-S036       0.041741  0.005189  \n",
       "...                  ...       ...  \n",
       "M491-S146       0.000238  0.999232  \n",
       "M491-S148       0.177256  0.122004  \n",
       "M491-S149       0.182523  0.605651  \n",
       "M491-S153       0.638097  0.020640  \n",
       "M491-S160       0.000469  0.623040  \n",
       "\n",
       "[206 rows x 6 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Table of predicted class probabilities\n",
    "pd.DataFrame(\n",
    "    clf1.predict_proba(featurized_model1_data.X),\n",
    "    index=featurized_model1_data.sample_names,\n",
    "    columns=clf1.classes_,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6196293e-f797-4c70-9b10-91bae66f731a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "203e26e9-3116-4304-91a9-0e8f666c7485",
   "metadata": {},
   "source": [
    "**We just generated a table of predicted class probabilities for Model 1, using `featurize()` and `predict_proba()`.**\n",
    "\n",
    "Alternatively, we can use the model to generate a single predicted label for each sample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "2e75ce7d-f455-42fc-8c50-538d6ce040e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['HIV', 'Healthy/Background', 'HIV', 'HIV', 'Healthy/Background',\n",
       "       'Healthy/Background', 'HIV', 'HIV', 'HIV', 'HIV',\n",
       "       'Healthy/Background', 'T1D', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'T1D', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Lupus', 'Healthy/Background', 'Healthy/Background', 'Covid19',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Covid19', 'HIV',\n",
       "       'HIV', 'HIV', 'HIV', 'HIV', 'HIV', 'HIV', 'HIV', 'HIV', 'HIV',\n",
       "       'HIV', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'HIV', 'Healthy/Background', 'T1D', 'Healthy/Background',\n",
       "       'Healthy/Background', 'HIV', 'HIV', 'HIV', 'HIV', 'HIV', 'HIV',\n",
       "       'Healthy/Background', 'HIV', 'HIV', 'HIV', 'Healthy/Background',\n",
       "       'Healthy/Background', 'HIV', 'HIV', 'HIV', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Covid19', 'Lupus', 'Healthy/Background',\n",
       "       'T1D', 'Healthy/Background', 'Lupus', 'Lupus',\n",
       "       'Healthy/Background', 'T1D', 'T1D', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'T1D', 'Covid19', 'Covid19', 'Covid19',\n",
       "       'Covid19', 'Covid19', 'Covid19', 'Lupus', 'Covid19', 'T1D',\n",
       "       'Covid19', 'Covid19', 'Covid19', 'Healthy/Background', 'Lupus',\n",
       "       'Covid19', 'Covid19', 'Covid19', 'Covid19', 'Covid19', 'Covid19',\n",
       "       'T1D', 'Lupus', 'Lupus', 'Lupus', 'Lupus', 'Lupus', 'Lupus',\n",
       "       'Lupus', 'Lupus', 'Lupus', 'Lupus', 'Covid19', 'Lupus', 'Lupus',\n",
       "       'Lupus', 'Lupus', 'Lupus', 'Influenza', 'T1D', 'Influenza',\n",
       "       'Influenza', 'Influenza', 'Healthy/Background', 'Influenza',\n",
       "       'Influenza', 'Healthy/Background', 'Healthy/Background', 'Lupus',\n",
       "       'Healthy/Background', 'Lupus', 'Lupus', 'Lupus', 'Lupus', 'T1D',\n",
       "       'Covid19', 'Healthy/Background', 'Lupus', 'Lupus',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Lupus', 'T1D', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'T1D', 'T1D', 'T1D', 'T1D', 'T1D', 'T1D',\n",
       "       'T1D', 'T1D', 'T1D', 'T1D', 'Healthy/Background',\n",
       "       'Healthy/Background', 'T1D', 'T1D', 'T1D', 'Lupus', 'T1D', 'Lupus',\n",
       "       'Healthy/Background', 'Covid19', 'T1D', 'T1D', 'Lupus', 'Lupus',\n",
       "       'Healthy/Background', 'T1D', 'T1D', 'Healthy/Background', 'T1D',\n",
       "       'Healthy/Background', 'T1D', 'Lupus', 'T1D'], dtype=object)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf1.predict(featurized_model1_data.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a99e7d07-7e78-4246-a563-a6e76322583c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "08ccadb2-f182-482e-971b-c0f6dd303347",
   "metadata": {},
   "source": [
    "## Model 2\n",
    "\n",
    "Model 2 uses clustering to identify shared groups of sequences across individuals with the same disease. Then we predict disease using the number of disease-associated cluster hits per sample.\n",
    "\n",
    "As before, let's first check which version is used in the ensemble metamodel. It's the ridge logistic regression version of Model 2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "9537b3b6-8765-4ea3-93b0-f14836c73656",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "namespace(model_name_overall_repertoire_composition={<GeneLocus.BCR: 1>: 'elasticnet_cv0.25',\n",
       "                                                     <GeneLocus.TCR: 2>: 'lasso_cv'},\n",
       "          model_name_convergent_clustering={<GeneLocus.BCR: 1>: 'ridge_cv',\n",
       "                                            <GeneLocus.TCR: 2>: 'lasso_cv'},\n",
       "          base_sequence_model_name={<GeneLocus.BCR: 1>: 'rf_multiclass',\n",
       "                                    <GeneLocus.TCR: 2>: 'ridge_cv_ovr'},\n",
       "          base_sequence_model_subset_strategy=<SequenceSubsetStrategy.split_Vgene_and_isotype: 3>,\n",
       "          aggregation_sequence_model_name={<GeneLocus.BCR: 1>: 'rf_multiclass_mean_aggregated_as_binary_ovr_reweighed_by_subset_frequencies',\n",
       "                                           <GeneLocus.TCR: 2>: 'rf_multiclass_entropy_twenty_percent_cutoff_aggregated_as_binary_ovr_reweighed_by_subset_frequencies'})"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Chosen model names for Models 1, 2, and 3\n",
    "config.metamodel_base_model_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "902dbb25-4bc2-4b5d-8db1-249513b59889",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ridge_cv'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The model name chosen for Model 2 - BCR\n",
    "config.metamodel_base_model_names.model_name_convergent_clustering[GeneLocus.BCR]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca7fc6cd-a42f-4c76-8444-6b8c31c44c87",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5205b108-2dff-4ba8-a7df-2af22afc8cb9",
   "metadata": {},
   "source": [
    "Now let's load that version:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "75568d99-fb7a-48d4-b790-a9190cc26af0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>ConvergentClusterClassifier: Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                ('standardscalerthatpreservesinputtype',\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                ('glmnetlogitnetwrapper',\n",
       "                 GlmnetLogitNetWrapper(alpha=0, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=8, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=<function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0>, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])</pre>"
      ],
      "text/plain": [
       "ConvergentClusterClassifier: Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                ('standardscalerthatpreservesinputtype',\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                ('glmnetlogitnetwrapper',\n",
       "                 GlmnetLogitNetWrapper(alpha=0, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=8, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=<function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0>, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf2 = ConvergentClusterClassifier(\n",
    "    fold_id=fold_id,\n",
    "    model_name=config.metamodel_base_model_names.model_name_convergent_clustering[\n",
    "        GeneLocus.BCR\n",
    "    ],\n",
    "    fold_label_train=\"train_smaller1\",  # The model was trained on train_smaller1, with hyperparameter tuning on train_smaller2.\n",
    "    gene_locus=GeneLocus.BCR,\n",
    "    target_obs_column=target_obs_column,\n",
    ")\n",
    "clf2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "158d3b08-44d1-4ff3-a35e-3ba2dbd619e1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f63cc976-9072-4795-b7d6-df8e7f07fbcd",
   "metadata": {},
   "source": [
    "We get a `ConvergentClusterClassifier` object.\n",
    "\n",
    "Just as with `RepertoireClassifier` for Model 1, `ConvergentClusterClassifier` is a wrapper around a scikit-learn Pipeline:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "109d6474-98a9-40ba-b5d9-de8727808b83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sklearn.pipeline.Pipeline"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(clf2._inner)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "887855e1-5bfc-4c63-a5c1-6562aa4d2a61",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "24a0f635-582f-4371-ab22-8ab2e87373e2",
   "metadata": {},
   "source": [
    "But this time, the Pipeline is simpler:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "356aa0e1-f96a-4fd3-91dc-c193ea920740",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-2 {color: black;background-color: white;}#sk-container-id-2 pre{padding: 0;}#sk-container-id-2 div.sk-toggleable {background-color: white;}#sk-container-id-2 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-2 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-2 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-2 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-2 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-2 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-2 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-2 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-2 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-2 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-2 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-2 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-2 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-2 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-2 div.sk-item {position: relative;z-index: 1;}#sk-container-id-2 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-2 div.sk-item::before, #sk-container-id-2 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-2 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-2 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-2 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-2 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-2 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-2 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-2 div.sk-label-container {text-align: center;}#sk-container-id-2 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-2 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-2\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>Pipeline(steps=[(&#x27;matchvariables&#x27;, MatchVariables()),\n",
       "                (&#x27;standardscalerthatpreservesinputtype&#x27;,\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                (&#x27;glmnetlogitnetwrapper&#x27;,\n",
       "                 GlmnetLogitNetWrapper(alpha=0, class_weight=&#x27;balanced&#x27;, cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=8, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=&lt;function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0&gt;, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-19\" type=\"checkbox\" ><label for=\"sk-estimator-id-19\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">Pipeline</label><div class=\"sk-toggleable__content\"><pre>Pipeline(steps=[(&#x27;matchvariables&#x27;, MatchVariables()),\n",
       "                (&#x27;standardscalerthatpreservesinputtype&#x27;,\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                (&#x27;glmnetlogitnetwrapper&#x27;,\n",
       "                 GlmnetLogitNetWrapper(alpha=0, class_weight=&#x27;balanced&#x27;, cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=8, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=&lt;function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0&gt;, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-20\" type=\"checkbox\" ><label for=\"sk-estimator-id-20\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">MatchVariables</label><div class=\"sk-toggleable__content\"><pre>MatchVariables()</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-21\" type=\"checkbox\" ><label for=\"sk-estimator-id-21\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">StandardScalerThatPreservesInputType</label><div class=\"sk-toggleable__content\"><pre>StandardScalerThatPreservesInputType()</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-22\" type=\"checkbox\" ><label for=\"sk-estimator-id-22\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">GlmnetLogitNetWrapper</label><div class=\"sk-toggleable__content\"><pre>GlmnetLogitNetWrapper: LogitNet(alpha=0, n_jobs=8, n_splits=5, random_state=0,\n",
       "         scoring=&lt;function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0&gt;,\n",
       "         standardize=False, verbose=True)</pre></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                ('standardscalerthatpreservesinputtype',\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                ('glmnetlogitnetwrapper',\n",
       "                 GlmnetLogitNetWrapper(alpha=0, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=8, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=<function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0>, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf2._inner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6678de91-08b7-4b0d-aecf-0717c058502c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "943781dc-10c3-48f1-ac29-b3b9b0aae915",
   "metadata": {},
   "source": [
    "What's happening here: The pipeline confirms that the expected feature names are present, it standardizes the features, and then it runs logistic regression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "defea6c5-96e0-4b5b-93f5-81add6fd7fa1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f81a0971-0187-416d-8012-288747dd7434",
   "metadata": {},
   "source": [
    "To run the model, let's start by featurizing the dataset, just as we did with Model 1:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "f0d6e9bf-f8ed-4d2c-a1be-24768d5e92df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "crosseval.featurized_data.FeaturizedData"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "featurized_model2_data = clf2.featurize(adata_bcr)\n",
    "type(featurized_model2_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b66ca754-2e5e-46f5-976a-d03e4e3db88d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "85a2489f-cea7-4612-bfe8-bbefffc29808",
   "metadata": {},
   "source": [
    "The `FeaturizedData` container can be unpacked the same way:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "0612eb38-998c-453d-adaa-00a2793b3002",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>cluster_dominant_label</th>\n",
       "      <th>Covid19</th>\n",
       "      <th>HIV</th>\n",
       "      <th>Healthy/Background</th>\n",
       "      <th>Influenza</th>\n",
       "      <th>Lupus</th>\n",
       "      <th>T1D</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M111-S037</th>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S042</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S016</th>\n",
       "      <td>1.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S043</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S036</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S146</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S148</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S149</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S153</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S160</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "cluster_dominant_label  Covid19   HIV  Healthy/Background  Influenza  Lupus  \\\n",
       "specimen_label                                                                \n",
       "M111-S037                   0.0  13.0                 1.0        1.0   11.0   \n",
       "M124-S042                   0.0   0.0                 2.0        4.0    5.0   \n",
       "M111-S016                   1.0  19.0                 1.0        2.0    8.0   \n",
       "M111-S043                   0.0  24.0                 1.0        0.0    4.0   \n",
       "M124-S036                   0.0   5.0                 1.0        2.0    3.0   \n",
       "...                         ...   ...                 ...        ...    ...   \n",
       "M491-S146                   0.0   3.0                 1.0        0.0    3.0   \n",
       "M491-S148                   0.0   0.0                 2.0        1.0    2.0   \n",
       "M491-S149                   0.0   1.0                 4.0        5.0    1.0   \n",
       "M491-S153                   0.0   1.0                 1.0        0.0    5.0   \n",
       "M491-S160                   0.0   0.0                 2.0        1.0    1.0   \n",
       "\n",
       "cluster_dominant_label  T1D  \n",
       "specimen_label               \n",
       "M111-S037               0.0  \n",
       "M124-S042               0.0  \n",
       "M111-S016               0.0  \n",
       "M111-S043               0.0  \n",
       "M124-S036               0.0  \n",
       "...                     ...  \n",
       "M491-S146               0.0  \n",
       "M491-S148               0.0  \n",
       "M491-S149               1.0  \n",
       "M491-S153               1.0  \n",
       "M491-S160               0.0  \n",
       "\n",
       "[205 rows x 6 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Features (disease-associated cluster hit counts)\n",
    "featurized_model2_data.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "89ad67a3-8956-4e0a-a52d-3f50422366d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "specimen_label\n",
       "M111-S037                   HIV\n",
       "M124-S042    Healthy/Background\n",
       "M111-S016                   HIV\n",
       "M111-S043                   HIV\n",
       "M124-S036    Healthy/Background\n",
       "                    ...        \n",
       "M491-S146                   T1D\n",
       "M491-S148                   T1D\n",
       "M491-S149                   T1D\n",
       "M491-S153                   T1D\n",
       "M491-S160                   T1D\n",
       "Name: disease, Length: 205, dtype: category\n",
       "Categories (6, object): ['HIV', 'Healthy/Background', 'Lupus', 'Covid19', 'Influenza', 'T1D']"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ground truth\n",
    "featurized_model2_data.y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "ea0cd80f-5592-4018-a802-f97a4b087ca3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['M111-S037', 'M124-S042', 'M111-S016', 'M111-S043', 'M124-S036',\n",
       "       'M124-S040', 'M111-S035', 'M111-S025', 'M111-S011', 'M111-S036',\n",
       "       ...\n",
       "       'M491-S111', 'M491-S117', 'M491-S124', 'M491-S126', 'M491-S136',\n",
       "       'M491-S146', 'M491-S148', 'M491-S149', 'M491-S153', 'M491-S160'],\n",
       "      dtype='object', name='specimen_label', length=205)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample names\n",
    "featurized_model2_data.sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9f1a160a-5453-45d0-8afc-3cdd63edd699",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease.rollup</th>\n",
       "      <th>disease.separate_past_exposures</th>\n",
       "      <th>disease_severity</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>ethnicity_condensed</th>\n",
       "      <th>isotype_proportion:IGHA</th>\n",
       "      <th>isotype_proportion:IGHD-M</th>\n",
       "      <th>isotype_proportion:IGHG</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>past_exposure</th>\n",
       "      <th>sex</th>\n",
       "      <th>study_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M111-S037</th>\n",
       "      <td>48.0</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV Broad Neutralizing</td>\n",
       "      <td>African</td>\n",
       "      <td>0.140070</td>\n",
       "      <td>0.682723</td>\n",
       "      <td>0.177207</td>\n",
       "      <td>BFI-0000254</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S042</th>\n",
       "      <td>26.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.246533</td>\n",
       "      <td>0.547060</td>\n",
       "      <td>0.206407</td>\n",
       "      <td>BFI-0002850</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S016</th>\n",
       "      <td>22.0</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV Broad Neutralizing</td>\n",
       "      <td>African</td>\n",
       "      <td>0.101753</td>\n",
       "      <td>0.796830</td>\n",
       "      <td>0.101417</td>\n",
       "      <td>BFI-0002855</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S043</th>\n",
       "      <td>23.0</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV Non Neutralizing</td>\n",
       "      <td>African</td>\n",
       "      <td>0.228808</td>\n",
       "      <td>0.516170</td>\n",
       "      <td>0.255022</td>\n",
       "      <td>BFI-0002856</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S036</th>\n",
       "      <td>29.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>African</td>\n",
       "      <td>0.144986</td>\n",
       "      <td>0.660305</td>\n",
       "      <td>0.194710</td>\n",
       "      <td>BFI-0002862</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S146</th>\n",
       "      <td>15.0</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.098597</td>\n",
       "      <td>0.850927</td>\n",
       "      <td>0.050476</td>\n",
       "      <td>BFI-0010788</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S148</th>\n",
       "      <td>14.0</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.110664</td>\n",
       "      <td>0.815813</td>\n",
       "      <td>0.073523</td>\n",
       "      <td>BFI-0010790</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S149</th>\n",
       "      <td>11.0</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.136773</td>\n",
       "      <td>0.640226</td>\n",
       "      <td>0.223001</td>\n",
       "      <td>BFI-0010791</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S153</th>\n",
       "      <td>16.0</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.160569</td>\n",
       "      <td>0.697251</td>\n",
       "      <td>0.142180</td>\n",
       "      <td>BFI-0010794</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S160</th>\n",
       "      <td>16.0</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.130453</td>\n",
       "      <td>0.817561</td>\n",
       "      <td>0.051986</td>\n",
       "      <td>BFI-0010800</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 age             disease      disease.rollup  \\\n",
       "specimen_label                                                 \n",
       "M111-S037       48.0                 HIV                 HIV   \n",
       "M124-S042       26.0  Healthy/Background  Healthy/Background   \n",
       "M111-S016       22.0                 HIV                 HIV   \n",
       "M111-S043       23.0                 HIV                 HIV   \n",
       "M124-S036       29.0  Healthy/Background  Healthy/Background   \n",
       "...              ...                 ...                 ...   \n",
       "M491-S146       15.0                 T1D                 T1D   \n",
       "M491-S148       14.0                 T1D                 T1D   \n",
       "M491-S149       11.0                 T1D                 T1D   \n",
       "M491-S153       16.0                 T1D                 T1D   \n",
       "M491-S160       16.0                 T1D                 T1D   \n",
       "\n",
       "               disease.separate_past_exposures disease_severity  \\\n",
       "specimen_label                                                    \n",
       "M111-S037                                  HIV              NaN   \n",
       "M124-S042                   Healthy/Background              NaN   \n",
       "M111-S016                                  HIV              NaN   \n",
       "M111-S043                                  HIV              NaN   \n",
       "M124-S036                   Healthy/Background              NaN   \n",
       "...                                        ...              ...   \n",
       "M491-S146                                  T1D              NaN   \n",
       "M491-S148                                  T1D              NaN   \n",
       "M491-S149                                  T1D              NaN   \n",
       "M491-S153                                  T1D              NaN   \n",
       "M491-S160                                  T1D              NaN   \n",
       "\n",
       "                                  disease_subtype ethnicity_condensed  \\\n",
       "specimen_label                                                          \n",
       "M111-S037                  HIV Broad Neutralizing             African   \n",
       "M124-S042       Healthy/Background - HIV Negative                 NaN   \n",
       "M111-S016                  HIV Broad Neutralizing             African   \n",
       "M111-S043                    HIV Non Neutralizing             African   \n",
       "M124-S036       Healthy/Background - HIV Negative             African   \n",
       "...                                           ...                 ...   \n",
       "M491-S146                         T1D - pediatric                 NaN   \n",
       "M491-S148                         T1D - pediatric                 NaN   \n",
       "M491-S149                         T1D - pediatric                 NaN   \n",
       "M491-S153                         T1D - pediatric                 NaN   \n",
       "M491-S160                         T1D - pediatric                 NaN   \n",
       "\n",
       "                isotype_proportion:IGHA  isotype_proportion:IGHD-M  \\\n",
       "specimen_label                                                       \n",
       "M111-S037                      0.140070                   0.682723   \n",
       "M124-S042                      0.246533                   0.547060   \n",
       "M111-S016                      0.101753                   0.796830   \n",
       "M111-S043                      0.228808                   0.516170   \n",
       "M124-S036                      0.144986                   0.660305   \n",
       "...                                 ...                        ...   \n",
       "M491-S146                      0.098597                   0.850927   \n",
       "M491-S148                      0.110664                   0.815813   \n",
       "M491-S149                      0.136773                   0.640226   \n",
       "M491-S153                      0.160569                   0.697251   \n",
       "M491-S160                      0.130453                   0.817561   \n",
       "\n",
       "                isotype_proportion:IGHG participant_label  past_exposure  sex  \\\n",
       "specimen_label                                                                  \n",
       "M111-S037                      0.177207       BFI-0000254          False    F   \n",
       "M124-S042                      0.206407       BFI-0002850          False    F   \n",
       "M111-S016                      0.101417       BFI-0002855          False    F   \n",
       "M111-S043                      0.255022       BFI-0002856          False    F   \n",
       "M124-S036                      0.194710       BFI-0002862          False    F   \n",
       "...                                 ...               ...            ...  ...   \n",
       "M491-S146                      0.050476       BFI-0010788          False  NaN   \n",
       "M491-S148                      0.073523       BFI-0010790          False  NaN   \n",
       "M491-S149                      0.223001       BFI-0010791          False  NaN   \n",
       "M491-S153                      0.142180       BFI-0010794          False  NaN   \n",
       "M491-S160                      0.051986       BFI-0010800          False  NaN   \n",
       "\n",
       "                      study_name  \n",
       "specimen_label                    \n",
       "M111-S037                    HIV  \n",
       "M124-S042                    HIV  \n",
       "M111-S016                    HIV  \n",
       "M111-S043                    HIV  \n",
       "M124-S036                    HIV  \n",
       "...                          ...  \n",
       "M491-S146       Diabetes biobank  \n",
       "M491-S148       Diabetes biobank  \n",
       "M491-S149       Diabetes biobank  \n",
       "M491-S153       Diabetes biobank  \n",
       "M491-S160       Diabetes biobank  \n",
       "\n",
       "[205 rows x 14 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample metadata\n",
    "featurized_model2_data.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09a7f218-4d0f-408b-abac-7e18d3ad6c76",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "61f11231-6735-446d-b5b2-ac4578ef2614",
   "metadata": {},
   "source": [
    "But this time, the `FeaturizedData` container also has some extra fields. Unlike the other components of Mal-ID, Model 2 abstains from prediction if none of the sequences in a sample match any disease-associated clusters. These abstentions are also stored in the `FeaturizedData` container too:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "b696383b-8602-4635-b06e-b7054fdefffa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['M418-S118'], dtype='object', name='specimen_label')"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# These samples had no features generated — they are abstentions:\n",
    "featurized_model2_data.abstained_sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "05882d8b-743a-4052-b8ce-5e716dde5bbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "specimen_label\n",
       "M418-S118    Covid19\n",
       "Name: disease, dtype: category\n",
       "Categories (6, object): ['HIV', 'Healthy/Background', 'Lupus', 'Covid19', 'Influenza', 'T1D']"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ground truth for abstained samples\n",
    "featurized_model2_data.abstained_sample_y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "6715d1cb-bd8c-45f7-adef-00638d301e21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease.rollup</th>\n",
       "      <th>disease.separate_past_exposures</th>\n",
       "      <th>disease_severity</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>ethnicity_condensed</th>\n",
       "      <th>isotype_proportion:IGHA</th>\n",
       "      <th>isotype_proportion:IGHD-M</th>\n",
       "      <th>isotype_proportion:IGHG</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>past_exposure</th>\n",
       "      <th>sex</th>\n",
       "      <th>study_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M418-S118</th>\n",
       "      <td>68.0</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>Covid19</td>\n",
       "      <td>ICU</td>\n",
       "      <td>Covid19 - ICU</td>\n",
       "      <td>Hispanic/Latino</td>\n",
       "      <td>0.497761</td>\n",
       "      <td>0.294776</td>\n",
       "      <td>0.207463</td>\n",
       "      <td>BFI-0009059</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>Covid19-Stanford</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 age  disease disease.rollup disease.separate_past_exposures  \\\n",
       "specimen_label                                                                 \n",
       "M418-S118       68.0  Covid19        Covid19                         Covid19   \n",
       "\n",
       "               disease_severity disease_subtype ethnicity_condensed  \\\n",
       "specimen_label                                                        \n",
       "M418-S118                   ICU   Covid19 - ICU     Hispanic/Latino   \n",
       "\n",
       "                isotype_proportion:IGHA  isotype_proportion:IGHD-M  \\\n",
       "specimen_label                                                       \n",
       "M418-S118                      0.497761                   0.294776   \n",
       "\n",
       "                isotype_proportion:IGHG participant_label  past_exposure sex  \\\n",
       "specimen_label                                                                 \n",
       "M418-S118                      0.207463       BFI-0009059          False   M   \n",
       "\n",
       "                      study_name  \n",
       "specimen_label                    \n",
       "M418-S118       Covid19-Stanford  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Metadata for abstained samples\n",
    "featurized_model2_data.abstained_sample_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cddbbc0d-c6c8-4583-9544-d6b9b80cb3bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "24e64760-27f6-4199-bc2d-cf8fd7bb9fb9",
   "metadata": {},
   "source": [
    "Let's run the model to predict the per-class probabilities for the samples that are not abstentions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "648b39c4-bd61-4e61-90eb-40eb20e7aca9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Covid19</th>\n",
       "      <th>HIV</th>\n",
       "      <th>Healthy/Background</th>\n",
       "      <th>Influenza</th>\n",
       "      <th>Lupus</th>\n",
       "      <th>T1D</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M111-S037</th>\n",
       "      <td>0.093203</td>\n",
       "      <td>0.231276</td>\n",
       "      <td>0.203522</td>\n",
       "      <td>0.038482</td>\n",
       "      <td>0.261447</td>\n",
       "      <td>0.172070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S042</th>\n",
       "      <td>0.110002</td>\n",
       "      <td>0.085210</td>\n",
       "      <td>0.311180</td>\n",
       "      <td>0.051944</td>\n",
       "      <td>0.202014</td>\n",
       "      <td>0.239650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S016</th>\n",
       "      <td>0.094130</td>\n",
       "      <td>0.355081</td>\n",
       "      <td>0.177844</td>\n",
       "      <td>0.039048</td>\n",
       "      <td>0.189403</td>\n",
       "      <td>0.144494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S043</th>\n",
       "      <td>0.072264</td>\n",
       "      <td>0.485880</td>\n",
       "      <td>0.155931</td>\n",
       "      <td>0.028994</td>\n",
       "      <td>0.128728</td>\n",
       "      <td>0.128202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S036</th>\n",
       "      <td>0.116146</td>\n",
       "      <td>0.138989</td>\n",
       "      <td>0.255802</td>\n",
       "      <td>0.046437</td>\n",
       "      <td>0.190750</td>\n",
       "      <td>0.251876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S146</th>\n",
       "      <td>0.118626</td>\n",
       "      <td>0.117655</td>\n",
       "      <td>0.262798</td>\n",
       "      <td>0.039239</td>\n",
       "      <td>0.195331</td>\n",
       "      <td>0.266350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S148</th>\n",
       "      <td>0.113949</td>\n",
       "      <td>0.087438</td>\n",
       "      <td>0.323998</td>\n",
       "      <td>0.040867</td>\n",
       "      <td>0.172615</td>\n",
       "      <td>0.261133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S149</th>\n",
       "      <td>0.090366</td>\n",
       "      <td>0.056158</td>\n",
       "      <td>0.267491</td>\n",
       "      <td>0.045139</td>\n",
       "      <td>0.087322</td>\n",
       "      <td>0.453525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S153</th>\n",
       "      <td>0.101486</td>\n",
       "      <td>0.059672</td>\n",
       "      <td>0.144173</td>\n",
       "      <td>0.031095</td>\n",
       "      <td>0.134344</td>\n",
       "      <td>0.529229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S160</th>\n",
       "      <td>0.114940</td>\n",
       "      <td>0.088163</td>\n",
       "      <td>0.326824</td>\n",
       "      <td>0.040983</td>\n",
       "      <td>0.162981</td>\n",
       "      <td>0.266109</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Covid19       HIV  Healthy/Background  Influenza     Lupus  \\\n",
       "specimen_label                                                                \n",
       "M111-S037       0.093203  0.231276            0.203522   0.038482  0.261447   \n",
       "M124-S042       0.110002  0.085210            0.311180   0.051944  0.202014   \n",
       "M111-S016       0.094130  0.355081            0.177844   0.039048  0.189403   \n",
       "M111-S043       0.072264  0.485880            0.155931   0.028994  0.128728   \n",
       "M124-S036       0.116146  0.138989            0.255802   0.046437  0.190750   \n",
       "...                  ...       ...                 ...        ...       ...   \n",
       "M491-S146       0.118626  0.117655            0.262798   0.039239  0.195331   \n",
       "M491-S148       0.113949  0.087438            0.323998   0.040867  0.172615   \n",
       "M491-S149       0.090366  0.056158            0.267491   0.045139  0.087322   \n",
       "M491-S153       0.101486  0.059672            0.144173   0.031095  0.134344   \n",
       "M491-S160       0.114940  0.088163            0.326824   0.040983  0.162981   \n",
       "\n",
       "                     T1D  \n",
       "specimen_label            \n",
       "M111-S037       0.172070  \n",
       "M124-S042       0.239650  \n",
       "M111-S016       0.144494  \n",
       "M111-S043       0.128202  \n",
       "M124-S036       0.251876  \n",
       "...                  ...  \n",
       "M491-S146       0.266350  \n",
       "M491-S148       0.261133  \n",
       "M491-S149       0.453525  \n",
       "M491-S153       0.529229  \n",
       "M491-S160       0.266109  \n",
       "\n",
       "[205 rows x 6 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Table of predicted class probabilities\n",
    "pd.DataFrame(\n",
    "    clf2.predict_proba(featurized_model2_data.X),\n",
    "    index=featurized_model2_data.sample_names,\n",
    "    columns=clf2.classes_,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2872086-e2d9-44e5-b8a0-cea9ba88bbfd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c8d53290-b034-4680-82a0-bc167ce9ef6b",
   "metadata": {},
   "source": [
    "The `clf2.predict_proba(clf2.featurize(adata_bcr).X)` pattern is the same as what we saw above for Model 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5ca878c-5731-4d6a-84c9-a9fea7dd6eeb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a9a5bd95-427e-47f7-ba58-4e752fef02a8",
   "metadata": {},
   "source": [
    "## Model 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b441090-3d72-4536-8dd8-cae194a519c9",
   "metadata": {},
   "source": [
    "Model 3 uses language model embeddings of BCR and TCR sequences to predict disease state, in two stages:\n",
    "\n",
    "* **Sequence stage**: Predict which type of patient a sequence comes from, based on the language model embedding for that sequence. _This is a sequence-level model._ It's actually trained separately for each V gene and isotype.\n",
    "* **Aggregation stage**: Predict the disease status for an entire sample, based on the per-sequence predictions from the sequence stage. _This is a sample-level or person-level model._\n",
    "\n",
    "Each stage is trained separately. The sequence stage is trained on the `train_smaller1` fold, and the aggregation stage is trained on the `train_smaller2` fold.\n",
    "\n",
    "The selected model names are recorded in `config.metamodel_base_model_names` as `base_sequence_model_name` and `aggregation_sequence_model_name` for the two stages, respectively:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "2213e35c-e77b-4111-90e8-2ee781731751",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "namespace(model_name_overall_repertoire_composition={<GeneLocus.BCR: 1>: 'elasticnet_cv0.25',\n",
       "                                                     <GeneLocus.TCR: 2>: 'lasso_cv'},\n",
       "          model_name_convergent_clustering={<GeneLocus.BCR: 1>: 'ridge_cv',\n",
       "                                            <GeneLocus.TCR: 2>: 'lasso_cv'},\n",
       "          base_sequence_model_name={<GeneLocus.BCR: 1>: 'rf_multiclass',\n",
       "                                    <GeneLocus.TCR: 2>: 'ridge_cv_ovr'},\n",
       "          base_sequence_model_subset_strategy=<SequenceSubsetStrategy.split_Vgene_and_isotype: 3>,\n",
       "          aggregation_sequence_model_name={<GeneLocus.BCR: 1>: 'rf_multiclass_mean_aggregated_as_binary_ovr_reweighed_by_subset_frequencies',\n",
       "                                           <GeneLocus.TCR: 2>: 'rf_multiclass_entropy_twenty_percent_cutoff_aggregated_as_binary_ovr_reweighed_by_subset_frequencies'})"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config.metamodel_base_model_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fe93310-ce2b-4223-8411-3c9b585f25a6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b2780b6e-addb-46b0-83db-3f91515daf93",
   "metadata": {},
   "source": [
    "Let's load the selected versions of both stages.\n",
    "\n",
    "Both stages can be loaded together through the `VJGeneSpecificSequenceModelRollupClassifier` class, which represents the aggregation stage:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "c5105288-1c6f-42f0-8397-8fe0d713471f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>VJGeneSpecificSequenceModelRollupClassifier: Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                ('standardscalerthatpreservesinputtype',\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                ('binaryovrclassifierwithfeaturesubsettingbyclass',\n",
       "                 BinaryOvRClassifierWithFeatureSubsettingByClass(allow_some_classes_to_fail_to_train=True,\n",
       "                                                                 estimator=RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                                                                                  n_jobs=1,\n",
       "                                                                                                  random_state=0),\n",
       "                                                                 feature_name_class_delimeter='_',\n",
       "                                                                 n_jobs=1))])</pre>"
      ],
      "text/plain": [
       "VJGeneSpecificSequenceModelRollupClassifier: Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                ('standardscalerthatpreservesinputtype',\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                ('binaryovrclassifierwithfeaturesubsettingbyclass',\n",
       "                 BinaryOvRClassifierWithFeatureSubsettingByClass(allow_some_classes_to_fail_to_train=True,\n",
       "                                                                 estimator=RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                                                                                  n_jobs=1,\n",
       "                                                                                                  random_state=0),\n",
       "                                                                 feature_name_class_delimeter='_',\n",
       "                                                                 n_jobs=1))])"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf3 = VJGeneSpecificSequenceModelRollupClassifier(\n",
    "    # First, all the usual parameters, like fold ID, sequencing locus, and classification target:\n",
    "    fold_id=fold_id,\n",
    "    gene_locus=GeneLocus.BCR,\n",
    "    target_obs_column=target_obs_column,\n",
    "    #\n",
    "    # Model 3 includes a seqeunce stage and an aggregation stage.\n",
    "    # The aggregation stage is trained on top of the sequence stage, so to speak.\n",
    "    # First, provide the sequence stage model name:\n",
    "    base_sequence_model_name=config.metamodel_base_model_names.base_sequence_model_name[\n",
    "        GeneLocus.BCR\n",
    "    ],\n",
    "    # The sequence stage was trained on train_smaller1:\n",
    "    base_model_train_fold_label=\"train_smaller1\",\n",
    "    #\n",
    "    # Next, provide the aggregation stage model name here:\n",
    "    rollup_model_name=config.metamodel_base_model_names.aggregation_sequence_model_name[\n",
    "        GeneLocus.BCR\n",
    "    ],\n",
    "    # The aggregation stage was trained on train_smaller2:\n",
    "    fold_label_train=\"train_smaller2\",\n",
    ")\n",
    "clf3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13c7dee3-4504-44ad-8878-cd3e5220a3bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4038ef1b-e8f3-43d0-9133-c7a805a7e52d",
   "metadata": {},
   "source": [
    "The sequence stage is automatically loaded and stored inside of the `VJGeneSpecificSequenceModelRollupClassifier`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "faa4e4af-6058-4936-bf94-562ca20230fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<malid.trained_model_wrappers.vj_gene_specific_sequence_classifier.VGeneIsotypeSpecificSequenceClassifier at 0x7f09ec36cdf0>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf3.sequence_classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8a08643-c543-4d6c-8100-a6489a470b44",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "80de1a8f-b46b-482c-86cf-6b1d5d318d7f",
   "metadata": {},
   "source": [
    "The sequence stage, a `VGeneIsotypeSpecificSequenceClassifier` object, is actually a collection of models trained separately for each V gene and isotype:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "9fd30972-bc6f-45f0-a3a0-ad80fede070b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('IGHV1-2',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-21',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-34',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-34',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-39',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-b',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV5-51',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-39',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-18',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-31',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-39',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-59',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV2-5',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-48',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-74',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-7',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-23',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-b',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV5-51',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-7',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-4',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-69',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-20',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-18',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-59',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-61',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-23',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-11',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-48',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-33',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-2',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-30',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-61',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-66',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-69',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-30-2',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-15',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-48',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-74',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV5-a',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-23',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-21',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV7-4-1',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-30',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-59',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-9',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-4',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-3',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV5-51',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-30',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-49',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-30-2',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-66',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-30-3',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-53',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-69',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-18',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-8',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-30-4',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-34',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-30-3',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV5-a',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-8',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-3',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-20',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-74',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-11',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-7',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-61',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV7-4-1',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-15',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-4',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-46',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-b',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-46',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-31',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-2',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-33',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV2-5',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-33',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-20',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-46',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-21',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-53',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-11',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-53',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-30-3',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-8',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-15',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-30-4',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-66',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-30-2',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-3',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV2-5',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-24',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-30-4',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV2-70',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-49',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV6-1',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-13',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-24',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-9',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-49',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV1-24',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV5-a',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-9',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV7-4-1',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-13',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV3-13',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV2-70',\n",
       "  'IGHD-M'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV4-31',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV6-1',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV2-70',\n",
       "  'IGHG'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))]),\n",
       " ('IGHV6-1',\n",
       "  'IGHA'): Pipeline(steps=[('standardscalerthatpreservesinputtype',\n",
       "                  StandardScalerThatPreservesInputType()),\n",
       "                 ('randomforestclassifier',\n",
       "                  RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                         n_jobs=2, random_state=0))])}"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf3.sequence_classifier.models_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7edb8829-af15-4807-9874-59631ba59aad",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b55f8bd6-9bf0-45a0-9124-bc6bb5c2a48a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6958976c-750f-40e8-8d29-3a0b7e57c787",
   "metadata": {},
   "source": [
    "Returning to the aggregation stage, it accepts the following features:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ac8a7c3b-df68-4a74-83f2-fc4f84b2c188",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Influenza_IGHV3-23_IGHA',\n",
       " 'Influenza_IGHV3-23_IGHD-M',\n",
       " 'Influenza_IGHV3-23_IGHG',\n",
       " 'Influenza_IGHV4-b_IGHA',\n",
       " 'Influenza_IGHV4-b_IGHD-M',\n",
       " 'Influenza_IGHV4-b_IGHG',\n",
       " 'Influenza_IGHV3-7_IGHA',\n",
       " 'Influenza_IGHV3-7_IGHD-M',\n",
       " 'Influenza_IGHV3-7_IGHG',\n",
       " 'Influenza_IGHV1-18_IGHA',\n",
       " 'Influenza_IGHV1-18_IGHD-M',\n",
       " 'Influenza_IGHV1-18_IGHG',\n",
       " 'Influenza_IGHV4-59_IGHA',\n",
       " 'Influenza_IGHV4-59_IGHD-M',\n",
       " 'Influenza_IGHV4-59_IGHG',\n",
       " 'Influenza_IGHV4-61_IGHA',\n",
       " 'Influenza_IGHV4-61_IGHD-M',\n",
       " 'Influenza_IGHV4-61_IGHG',\n",
       " 'Influenza_IGHV1-69_IGHA',\n",
       " 'Influenza_IGHV1-69_IGHD-M',\n",
       " 'Influenza_IGHV1-69_IGHG',\n",
       " 'Influenza_IGHV3-21_IGHA',\n",
       " 'Influenza_IGHV3-21_IGHD-M',\n",
       " 'Influenza_IGHV3-21_IGHG',\n",
       " 'Influenza_IGHV3-48_IGHA',\n",
       " 'Influenza_IGHV3-48_IGHD-M',\n",
       " 'Influenza_IGHV3-48_IGHG',\n",
       " 'Influenza_IGHV1-46_IGHA',\n",
       " 'Influenza_IGHV1-46_IGHD-M',\n",
       " 'Influenza_IGHV1-46_IGHG',\n",
       " 'Influenza_IGHV4-39_IGHA',\n",
       " 'Influenza_IGHV4-39_IGHD-M',\n",
       " 'Influenza_IGHV4-39_IGHG',\n",
       " 'Influenza_IGHV5-51_IGHA',\n",
       " 'Influenza_IGHV5-51_IGHD-M',\n",
       " 'Influenza_IGHV5-51_IGHG',\n",
       " 'Influenza_IGHV6-1_IGHA',\n",
       " 'Influenza_IGHV6-1_IGHD-M',\n",
       " 'Influenza_IGHV6-1_IGHG',\n",
       " 'Influenza_IGHV1-2_IGHA',\n",
       " 'Influenza_IGHV1-2_IGHD-M',\n",
       " 'Influenza_IGHV1-2_IGHG',\n",
       " 'Influenza_IGHV3-11_IGHA',\n",
       " 'Influenza_IGHV3-11_IGHD-M',\n",
       " 'Influenza_IGHV3-11_IGHG',\n",
       " 'Influenza_IGHV4-34_IGHA',\n",
       " 'Influenza_IGHV4-34_IGHD-M',\n",
       " 'Influenza_IGHV4-34_IGHG',\n",
       " 'Influenza_IGHV3-20_IGHA',\n",
       " 'Influenza_IGHV3-20_IGHD-M',\n",
       " 'Influenza_IGHV3-20_IGHG',\n",
       " 'Influenza_IGHV4-31_IGHA',\n",
       " 'Influenza_IGHV4-31_IGHD-M',\n",
       " 'Influenza_IGHV4-31_IGHG',\n",
       " 'Influenza_IGHV3-30_IGHA',\n",
       " 'Influenza_IGHV3-30_IGHD-M',\n",
       " 'Influenza_IGHV3-30_IGHG',\n",
       " 'Influenza_IGHV4-4_IGHA',\n",
       " 'Influenza_IGHV4-4_IGHD-M',\n",
       " 'Influenza_IGHV4-4_IGHG',\n",
       " 'Influenza_IGHV3-33_IGHA',\n",
       " 'Influenza_IGHV3-33_IGHD-M',\n",
       " 'Influenza_IGHV3-33_IGHG',\n",
       " 'Influenza_IGHV1-8_IGHA',\n",
       " 'Influenza_IGHV1-8_IGHD-M',\n",
       " 'Influenza_IGHV1-8_IGHG',\n",
       " 'Influenza_IGHV3-9_IGHA',\n",
       " 'Influenza_IGHV3-9_IGHD-M',\n",
       " 'Influenza_IGHV3-9_IGHG',\n",
       " 'Influenza_IGHV3-53_IGHA',\n",
       " 'Influenza_IGHV3-53_IGHD-M',\n",
       " 'Influenza_IGHV3-53_IGHG',\n",
       " 'Influenza_IGHV3-74_IGHA',\n",
       " 'Influenza_IGHV3-74_IGHD-M',\n",
       " 'Influenza_IGHV3-74_IGHG',\n",
       " 'Influenza_IGHV3-15_IGHA',\n",
       " 'Influenza_IGHV3-15_IGHD-M',\n",
       " 'Influenza_IGHV3-15_IGHG',\n",
       " 'Influenza_IGHV2-5_IGHA',\n",
       " 'Influenza_IGHV2-5_IGHD-M',\n",
       " 'Influenza_IGHV2-5_IGHG',\n",
       " 'Influenza_IGHV4-30-2_IGHA',\n",
       " 'Influenza_IGHV4-30-2_IGHD-M',\n",
       " 'Influenza_IGHV4-30-2_IGHG',\n",
       " 'Influenza_IGHV3-13_IGHA',\n",
       " 'Influenza_IGHV3-13_IGHD-M',\n",
       " 'Influenza_IGHV3-13_IGHG',\n",
       " 'Influenza_IGHV3-49_IGHA',\n",
       " 'Influenza_IGHV3-49_IGHD-M',\n",
       " 'Influenza_IGHV3-49_IGHG',\n",
       " 'Influenza_IGHV1-24_IGHA',\n",
       " 'Influenza_IGHV1-24_IGHD-M',\n",
       " 'Influenza_IGHV1-24_IGHG',\n",
       " 'Influenza_IGHV3-30-3_IGHA',\n",
       " 'Influenza_IGHV3-30-3_IGHD-M',\n",
       " 'Influenza_IGHV3-30-3_IGHG',\n",
       " 'Influenza_IGHV2-70_IGHA',\n",
       " 'Influenza_IGHV2-70_IGHD-M',\n",
       " 'Influenza_IGHV2-70_IGHG',\n",
       " 'Influenza_IGHV3-66_IGHA',\n",
       " 'Influenza_IGHV3-66_IGHD-M',\n",
       " 'Influenza_IGHV3-66_IGHG',\n",
       " 'Influenza_IGHV4-30-4_IGHA',\n",
       " 'Influenza_IGHV4-30-4_IGHD-M',\n",
       " 'Influenza_IGHV4-30-4_IGHG',\n",
       " 'Influenza_IGHV1-3_IGHA',\n",
       " 'Influenza_IGHV1-3_IGHD-M',\n",
       " 'Influenza_IGHV1-3_IGHG',\n",
       " 'Influenza_IGHV5-a_IGHA',\n",
       " 'Influenza_IGHV5-a_IGHD-M',\n",
       " 'Influenza_IGHV5-a_IGHG',\n",
       " 'Influenza_IGHV7-4-1_IGHA',\n",
       " 'Influenza_IGHV7-4-1_IGHD-M',\n",
       " 'Influenza_IGHV7-4-1_IGHG',\n",
       " 'Lupus_IGHV3-23_IGHA',\n",
       " 'Lupus_IGHV3-23_IGHD-M',\n",
       " 'Lupus_IGHV3-23_IGHG',\n",
       " 'Lupus_IGHV4-b_IGHA',\n",
       " 'Lupus_IGHV4-b_IGHD-M',\n",
       " 'Lupus_IGHV4-b_IGHG',\n",
       " 'Lupus_IGHV3-7_IGHA',\n",
       " 'Lupus_IGHV3-7_IGHD-M',\n",
       " 'Lupus_IGHV3-7_IGHG',\n",
       " 'Lupus_IGHV1-18_IGHA',\n",
       " 'Lupus_IGHV1-18_IGHD-M',\n",
       " 'Lupus_IGHV1-18_IGHG',\n",
       " 'Lupus_IGHV4-59_IGHA',\n",
       " 'Lupus_IGHV4-59_IGHD-M',\n",
       " 'Lupus_IGHV4-59_IGHG',\n",
       " 'Lupus_IGHV4-61_IGHA',\n",
       " 'Lupus_IGHV4-61_IGHD-M',\n",
       " 'Lupus_IGHV4-61_IGHG',\n",
       " 'Lupus_IGHV1-69_IGHA',\n",
       " 'Lupus_IGHV1-69_IGHD-M',\n",
       " 'Lupus_IGHV1-69_IGHG',\n",
       " 'Lupus_IGHV3-21_IGHA',\n",
       " 'Lupus_IGHV3-21_IGHD-M',\n",
       " 'Lupus_IGHV3-21_IGHG',\n",
       " 'Lupus_IGHV3-48_IGHA',\n",
       " 'Lupus_IGHV3-48_IGHD-M',\n",
       " 'Lupus_IGHV3-48_IGHG',\n",
       " 'Lupus_IGHV1-46_IGHA',\n",
       " 'Lupus_IGHV1-46_IGHD-M',\n",
       " 'Lupus_IGHV1-46_IGHG',\n",
       " 'Lupus_IGHV4-39_IGHA',\n",
       " 'Lupus_IGHV4-39_IGHD-M',\n",
       " 'Lupus_IGHV4-39_IGHG',\n",
       " 'Lupus_IGHV5-51_IGHA',\n",
       " 'Lupus_IGHV5-51_IGHD-M',\n",
       " 'Lupus_IGHV5-51_IGHG',\n",
       " 'Lupus_IGHV6-1_IGHA',\n",
       " 'Lupus_IGHV6-1_IGHD-M',\n",
       " 'Lupus_IGHV6-1_IGHG',\n",
       " 'Lupus_IGHV1-2_IGHA',\n",
       " 'Lupus_IGHV1-2_IGHD-M',\n",
       " 'Lupus_IGHV1-2_IGHG',\n",
       " 'Lupus_IGHV3-11_IGHA',\n",
       " 'Lupus_IGHV3-11_IGHD-M',\n",
       " 'Lupus_IGHV3-11_IGHG',\n",
       " 'Lupus_IGHV4-34_IGHA',\n",
       " 'Lupus_IGHV4-34_IGHD-M',\n",
       " 'Lupus_IGHV4-34_IGHG',\n",
       " 'Lupus_IGHV3-20_IGHA',\n",
       " 'Lupus_IGHV3-20_IGHD-M',\n",
       " 'Lupus_IGHV3-20_IGHG',\n",
       " 'Lupus_IGHV4-31_IGHA',\n",
       " 'Lupus_IGHV4-31_IGHD-M',\n",
       " 'Lupus_IGHV4-31_IGHG',\n",
       " 'Lupus_IGHV3-30_IGHA',\n",
       " 'Lupus_IGHV3-30_IGHD-M',\n",
       " 'Lupus_IGHV3-30_IGHG',\n",
       " 'Lupus_IGHV4-4_IGHA',\n",
       " 'Lupus_IGHV4-4_IGHD-M',\n",
       " 'Lupus_IGHV4-4_IGHG',\n",
       " 'Lupus_IGHV3-33_IGHA',\n",
       " 'Lupus_IGHV3-33_IGHD-M',\n",
       " 'Lupus_IGHV3-33_IGHG',\n",
       " 'Lupus_IGHV1-8_IGHA',\n",
       " 'Lupus_IGHV1-8_IGHD-M',\n",
       " 'Lupus_IGHV1-8_IGHG',\n",
       " 'Lupus_IGHV3-9_IGHA',\n",
       " 'Lupus_IGHV3-9_IGHD-M',\n",
       " 'Lupus_IGHV3-9_IGHG',\n",
       " 'Lupus_IGHV3-53_IGHA',\n",
       " 'Lupus_IGHV3-53_IGHD-M',\n",
       " 'Lupus_IGHV3-53_IGHG',\n",
       " 'Lupus_IGHV3-74_IGHA',\n",
       " 'Lupus_IGHV3-74_IGHD-M',\n",
       " 'Lupus_IGHV3-74_IGHG',\n",
       " 'Lupus_IGHV3-15_IGHA',\n",
       " 'Lupus_IGHV3-15_IGHD-M',\n",
       " 'Lupus_IGHV3-15_IGHG',\n",
       " 'Lupus_IGHV2-5_IGHA',\n",
       " 'Lupus_IGHV2-5_IGHD-M',\n",
       " 'Lupus_IGHV2-5_IGHG',\n",
       " 'Lupus_IGHV4-30-2_IGHA',\n",
       " 'Lupus_IGHV4-30-2_IGHD-M',\n",
       " 'Lupus_IGHV4-30-2_IGHG',\n",
       " 'Lupus_IGHV3-13_IGHA',\n",
       " 'Lupus_IGHV3-13_IGHD-M',\n",
       " 'Lupus_IGHV3-13_IGHG',\n",
       " 'Lupus_IGHV3-49_IGHA',\n",
       " 'Lupus_IGHV3-49_IGHD-M',\n",
       " 'Lupus_IGHV3-49_IGHG',\n",
       " 'Lupus_IGHV1-24_IGHA',\n",
       " 'Lupus_IGHV1-24_IGHD-M',\n",
       " 'Lupus_IGHV1-24_IGHG',\n",
       " 'Lupus_IGHV3-30-3_IGHA',\n",
       " 'Lupus_IGHV3-30-3_IGHD-M',\n",
       " 'Lupus_IGHV3-30-3_IGHG',\n",
       " 'Lupus_IGHV2-70_IGHA',\n",
       " 'Lupus_IGHV2-70_IGHD-M',\n",
       " 'Lupus_IGHV2-70_IGHG',\n",
       " 'Lupus_IGHV3-66_IGHA',\n",
       " 'Lupus_IGHV3-66_IGHD-M',\n",
       " 'Lupus_IGHV3-66_IGHG',\n",
       " 'Lupus_IGHV4-30-4_IGHA',\n",
       " 'Lupus_IGHV4-30-4_IGHD-M',\n",
       " 'Lupus_IGHV4-30-4_IGHG',\n",
       " 'Lupus_IGHV1-3_IGHA',\n",
       " 'Lupus_IGHV1-3_IGHD-M',\n",
       " 'Lupus_IGHV1-3_IGHG',\n",
       " 'Lupus_IGHV5-a_IGHA',\n",
       " 'Lupus_IGHV5-a_IGHD-M',\n",
       " 'Lupus_IGHV5-a_IGHG',\n",
       " 'Lupus_IGHV7-4-1_IGHA',\n",
       " 'Lupus_IGHV7-4-1_IGHD-M',\n",
       " 'Lupus_IGHV7-4-1_IGHG',\n",
       " 'HIV_IGHV3-23_IGHA',\n",
       " 'HIV_IGHV3-23_IGHD-M',\n",
       " 'HIV_IGHV3-23_IGHG',\n",
       " 'HIV_IGHV4-b_IGHA',\n",
       " 'HIV_IGHV4-b_IGHD-M',\n",
       " 'HIV_IGHV4-b_IGHG',\n",
       " 'HIV_IGHV3-7_IGHA',\n",
       " 'HIV_IGHV3-7_IGHD-M',\n",
       " 'HIV_IGHV3-7_IGHG',\n",
       " 'HIV_IGHV1-18_IGHA',\n",
       " 'HIV_IGHV1-18_IGHD-M',\n",
       " 'HIV_IGHV1-18_IGHG',\n",
       " 'HIV_IGHV4-59_IGHA',\n",
       " 'HIV_IGHV4-59_IGHD-M',\n",
       " 'HIV_IGHV4-59_IGHG',\n",
       " 'HIV_IGHV4-61_IGHA',\n",
       " 'HIV_IGHV4-61_IGHD-M',\n",
       " 'HIV_IGHV4-61_IGHG',\n",
       " 'HIV_IGHV1-69_IGHA',\n",
       " 'HIV_IGHV1-69_IGHD-M',\n",
       " 'HIV_IGHV1-69_IGHG',\n",
       " 'HIV_IGHV3-21_IGHA',\n",
       " 'HIV_IGHV3-21_IGHD-M',\n",
       " 'HIV_IGHV3-21_IGHG',\n",
       " 'HIV_IGHV3-48_IGHA',\n",
       " 'HIV_IGHV3-48_IGHD-M',\n",
       " 'HIV_IGHV3-48_IGHG',\n",
       " 'HIV_IGHV1-46_IGHA',\n",
       " 'HIV_IGHV1-46_IGHD-M',\n",
       " 'HIV_IGHV1-46_IGHG',\n",
       " 'HIV_IGHV4-39_IGHA',\n",
       " 'HIV_IGHV4-39_IGHD-M',\n",
       " 'HIV_IGHV4-39_IGHG',\n",
       " 'HIV_IGHV5-51_IGHA',\n",
       " 'HIV_IGHV5-51_IGHD-M',\n",
       " 'HIV_IGHV5-51_IGHG',\n",
       " 'HIV_IGHV6-1_IGHA',\n",
       " 'HIV_IGHV6-1_IGHD-M',\n",
       " 'HIV_IGHV6-1_IGHG',\n",
       " 'HIV_IGHV1-2_IGHA',\n",
       " 'HIV_IGHV1-2_IGHD-M',\n",
       " 'HIV_IGHV1-2_IGHG',\n",
       " 'HIV_IGHV3-11_IGHA',\n",
       " 'HIV_IGHV3-11_IGHD-M',\n",
       " 'HIV_IGHV3-11_IGHG',\n",
       " 'HIV_IGHV4-34_IGHA',\n",
       " 'HIV_IGHV4-34_IGHD-M',\n",
       " 'HIV_IGHV4-34_IGHG',\n",
       " 'HIV_IGHV3-20_IGHA',\n",
       " 'HIV_IGHV3-20_IGHD-M',\n",
       " 'HIV_IGHV3-20_IGHG',\n",
       " 'HIV_IGHV4-31_IGHA',\n",
       " 'HIV_IGHV4-31_IGHD-M',\n",
       " 'HIV_IGHV4-31_IGHG',\n",
       " 'HIV_IGHV3-30_IGHA',\n",
       " 'HIV_IGHV3-30_IGHD-M',\n",
       " 'HIV_IGHV3-30_IGHG',\n",
       " 'HIV_IGHV4-4_IGHA',\n",
       " 'HIV_IGHV4-4_IGHD-M',\n",
       " 'HIV_IGHV4-4_IGHG',\n",
       " 'HIV_IGHV3-33_IGHA',\n",
       " 'HIV_IGHV3-33_IGHD-M',\n",
       " 'HIV_IGHV3-33_IGHG',\n",
       " 'HIV_IGHV1-8_IGHA',\n",
       " 'HIV_IGHV1-8_IGHD-M',\n",
       " 'HIV_IGHV1-8_IGHG',\n",
       " 'HIV_IGHV3-9_IGHA',\n",
       " 'HIV_IGHV3-9_IGHD-M',\n",
       " 'HIV_IGHV3-9_IGHG',\n",
       " 'HIV_IGHV3-53_IGHA',\n",
       " 'HIV_IGHV3-53_IGHD-M',\n",
       " 'HIV_IGHV3-53_IGHG',\n",
       " 'HIV_IGHV3-74_IGHA',\n",
       " 'HIV_IGHV3-74_IGHD-M',\n",
       " 'HIV_IGHV3-74_IGHG',\n",
       " 'HIV_IGHV3-15_IGHA',\n",
       " 'HIV_IGHV3-15_IGHD-M',\n",
       " 'HIV_IGHV3-15_IGHG',\n",
       " 'HIV_IGHV2-5_IGHA',\n",
       " 'HIV_IGHV2-5_IGHD-M',\n",
       " 'HIV_IGHV2-5_IGHG',\n",
       " 'HIV_IGHV4-30-2_IGHA',\n",
       " 'HIV_IGHV4-30-2_IGHD-M',\n",
       " 'HIV_IGHV4-30-2_IGHG',\n",
       " 'HIV_IGHV3-13_IGHA',\n",
       " 'HIV_IGHV3-13_IGHD-M',\n",
       " 'HIV_IGHV3-13_IGHG',\n",
       " 'HIV_IGHV3-49_IGHA',\n",
       " 'HIV_IGHV3-49_IGHD-M',\n",
       " 'HIV_IGHV3-49_IGHG',\n",
       " 'HIV_IGHV1-24_IGHA',\n",
       " 'HIV_IGHV1-24_IGHD-M',\n",
       " 'HIV_IGHV1-24_IGHG',\n",
       " 'HIV_IGHV3-30-3_IGHA',\n",
       " 'HIV_IGHV3-30-3_IGHD-M',\n",
       " 'HIV_IGHV3-30-3_IGHG',\n",
       " 'HIV_IGHV2-70_IGHA',\n",
       " 'HIV_IGHV2-70_IGHD-M',\n",
       " 'HIV_IGHV2-70_IGHG',\n",
       " 'HIV_IGHV3-66_IGHA',\n",
       " 'HIV_IGHV3-66_IGHD-M',\n",
       " 'HIV_IGHV3-66_IGHG',\n",
       " 'HIV_IGHV4-30-4_IGHA',\n",
       " 'HIV_IGHV4-30-4_IGHD-M',\n",
       " 'HIV_IGHV4-30-4_IGHG',\n",
       " 'HIV_IGHV1-3_IGHA',\n",
       " 'HIV_IGHV1-3_IGHD-M',\n",
       " 'HIV_IGHV1-3_IGHG',\n",
       " 'HIV_IGHV5-a_IGHA',\n",
       " 'HIV_IGHV5-a_IGHD-M',\n",
       " 'HIV_IGHV5-a_IGHG',\n",
       " 'HIV_IGHV7-4-1_IGHA',\n",
       " 'HIV_IGHV7-4-1_IGHD-M',\n",
       " 'HIV_IGHV7-4-1_IGHG',\n",
       " 'Healthy/Background_IGHV3-23_IGHA',\n",
       " 'Healthy/Background_IGHV3-23_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-23_IGHG',\n",
       " 'Healthy/Background_IGHV4-b_IGHA',\n",
       " 'Healthy/Background_IGHV4-b_IGHD-M',\n",
       " 'Healthy/Background_IGHV4-b_IGHG',\n",
       " 'Healthy/Background_IGHV3-7_IGHA',\n",
       " 'Healthy/Background_IGHV3-7_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-7_IGHG',\n",
       " 'Healthy/Background_IGHV1-18_IGHA',\n",
       " 'Healthy/Background_IGHV1-18_IGHD-M',\n",
       " 'Healthy/Background_IGHV1-18_IGHG',\n",
       " 'Healthy/Background_IGHV4-59_IGHA',\n",
       " 'Healthy/Background_IGHV4-59_IGHD-M',\n",
       " 'Healthy/Background_IGHV4-59_IGHG',\n",
       " 'Healthy/Background_IGHV4-61_IGHA',\n",
       " 'Healthy/Background_IGHV4-61_IGHD-M',\n",
       " 'Healthy/Background_IGHV4-61_IGHG',\n",
       " 'Healthy/Background_IGHV1-69_IGHA',\n",
       " 'Healthy/Background_IGHV1-69_IGHD-M',\n",
       " 'Healthy/Background_IGHV1-69_IGHG',\n",
       " 'Healthy/Background_IGHV3-21_IGHA',\n",
       " 'Healthy/Background_IGHV3-21_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-21_IGHG',\n",
       " 'Healthy/Background_IGHV3-48_IGHA',\n",
       " 'Healthy/Background_IGHV3-48_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-48_IGHG',\n",
       " 'Healthy/Background_IGHV1-46_IGHA',\n",
       " 'Healthy/Background_IGHV1-46_IGHD-M',\n",
       " 'Healthy/Background_IGHV1-46_IGHG',\n",
       " 'Healthy/Background_IGHV4-39_IGHA',\n",
       " 'Healthy/Background_IGHV4-39_IGHD-M',\n",
       " 'Healthy/Background_IGHV4-39_IGHG',\n",
       " 'Healthy/Background_IGHV5-51_IGHA',\n",
       " 'Healthy/Background_IGHV5-51_IGHD-M',\n",
       " 'Healthy/Background_IGHV5-51_IGHG',\n",
       " 'Healthy/Background_IGHV6-1_IGHA',\n",
       " 'Healthy/Background_IGHV6-1_IGHD-M',\n",
       " 'Healthy/Background_IGHV6-1_IGHG',\n",
       " 'Healthy/Background_IGHV1-2_IGHA',\n",
       " 'Healthy/Background_IGHV1-2_IGHD-M',\n",
       " 'Healthy/Background_IGHV1-2_IGHG',\n",
       " 'Healthy/Background_IGHV3-11_IGHA',\n",
       " 'Healthy/Background_IGHV3-11_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-11_IGHG',\n",
       " 'Healthy/Background_IGHV4-34_IGHA',\n",
       " 'Healthy/Background_IGHV4-34_IGHD-M',\n",
       " 'Healthy/Background_IGHV4-34_IGHG',\n",
       " 'Healthy/Background_IGHV3-20_IGHA',\n",
       " 'Healthy/Background_IGHV3-20_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-20_IGHG',\n",
       " 'Healthy/Background_IGHV4-31_IGHA',\n",
       " 'Healthy/Background_IGHV4-31_IGHD-M',\n",
       " 'Healthy/Background_IGHV4-31_IGHG',\n",
       " 'Healthy/Background_IGHV3-30_IGHA',\n",
       " 'Healthy/Background_IGHV3-30_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-30_IGHG',\n",
       " 'Healthy/Background_IGHV4-4_IGHA',\n",
       " 'Healthy/Background_IGHV4-4_IGHD-M',\n",
       " 'Healthy/Background_IGHV4-4_IGHG',\n",
       " 'Healthy/Background_IGHV3-33_IGHA',\n",
       " 'Healthy/Background_IGHV3-33_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-33_IGHG',\n",
       " 'Healthy/Background_IGHV1-8_IGHA',\n",
       " 'Healthy/Background_IGHV1-8_IGHD-M',\n",
       " 'Healthy/Background_IGHV1-8_IGHG',\n",
       " 'Healthy/Background_IGHV3-9_IGHA',\n",
       " 'Healthy/Background_IGHV3-9_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-9_IGHG',\n",
       " 'Healthy/Background_IGHV3-53_IGHA',\n",
       " 'Healthy/Background_IGHV3-53_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-53_IGHG',\n",
       " 'Healthy/Background_IGHV3-74_IGHA',\n",
       " 'Healthy/Background_IGHV3-74_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-74_IGHG',\n",
       " 'Healthy/Background_IGHV3-15_IGHA',\n",
       " 'Healthy/Background_IGHV3-15_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-15_IGHG',\n",
       " 'Healthy/Background_IGHV2-5_IGHA',\n",
       " 'Healthy/Background_IGHV2-5_IGHD-M',\n",
       " 'Healthy/Background_IGHV2-5_IGHG',\n",
       " 'Healthy/Background_IGHV4-30-2_IGHA',\n",
       " 'Healthy/Background_IGHV4-30-2_IGHD-M',\n",
       " 'Healthy/Background_IGHV4-30-2_IGHG',\n",
       " 'Healthy/Background_IGHV3-13_IGHA',\n",
       " 'Healthy/Background_IGHV3-13_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-13_IGHG',\n",
       " 'Healthy/Background_IGHV3-49_IGHA',\n",
       " 'Healthy/Background_IGHV3-49_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-49_IGHG',\n",
       " 'Healthy/Background_IGHV1-24_IGHA',\n",
       " 'Healthy/Background_IGHV1-24_IGHD-M',\n",
       " 'Healthy/Background_IGHV1-24_IGHG',\n",
       " 'Healthy/Background_IGHV3-30-3_IGHA',\n",
       " 'Healthy/Background_IGHV3-30-3_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-30-3_IGHG',\n",
       " 'Healthy/Background_IGHV2-70_IGHA',\n",
       " 'Healthy/Background_IGHV2-70_IGHD-M',\n",
       " 'Healthy/Background_IGHV2-70_IGHG',\n",
       " 'Healthy/Background_IGHV3-66_IGHA',\n",
       " 'Healthy/Background_IGHV3-66_IGHD-M',\n",
       " 'Healthy/Background_IGHV3-66_IGHG',\n",
       " 'Healthy/Background_IGHV4-30-4_IGHA',\n",
       " 'Healthy/Background_IGHV4-30-4_IGHD-M',\n",
       " 'Healthy/Background_IGHV4-30-4_IGHG',\n",
       " 'Healthy/Background_IGHV1-3_IGHA',\n",
       " 'Healthy/Background_IGHV1-3_IGHD-M',\n",
       " 'Healthy/Background_IGHV1-3_IGHG',\n",
       " 'Healthy/Background_IGHV5-a_IGHA',\n",
       " 'Healthy/Background_IGHV5-a_IGHD-M',\n",
       " 'Healthy/Background_IGHV5-a_IGHG',\n",
       " 'Healthy/Background_IGHV7-4-1_IGHA',\n",
       " 'Healthy/Background_IGHV7-4-1_IGHD-M',\n",
       " 'Healthy/Background_IGHV7-4-1_IGHG',\n",
       " 'T1D_IGHV3-23_IGHA',\n",
       " 'T1D_IGHV3-23_IGHD-M',\n",
       " 'T1D_IGHV3-23_IGHG',\n",
       " 'T1D_IGHV4-b_IGHA',\n",
       " 'T1D_IGHV4-b_IGHD-M',\n",
       " 'T1D_IGHV4-b_IGHG',\n",
       " 'T1D_IGHV3-7_IGHA',\n",
       " 'T1D_IGHV3-7_IGHD-M',\n",
       " 'T1D_IGHV3-7_IGHG',\n",
       " 'T1D_IGHV1-18_IGHA',\n",
       " 'T1D_IGHV1-18_IGHD-M',\n",
       " 'T1D_IGHV1-18_IGHG',\n",
       " 'T1D_IGHV4-59_IGHA',\n",
       " 'T1D_IGHV4-59_IGHD-M',\n",
       " 'T1D_IGHV4-59_IGHG',\n",
       " 'T1D_IGHV4-61_IGHA',\n",
       " 'T1D_IGHV4-61_IGHD-M',\n",
       " 'T1D_IGHV4-61_IGHG',\n",
       " 'T1D_IGHV1-69_IGHA',\n",
       " 'T1D_IGHV1-69_IGHD-M',\n",
       " 'T1D_IGHV1-69_IGHG',\n",
       " 'T1D_IGHV3-21_IGHA',\n",
       " 'T1D_IGHV3-21_IGHD-M',\n",
       " 'T1D_IGHV3-21_IGHG',\n",
       " 'T1D_IGHV3-48_IGHA',\n",
       " 'T1D_IGHV3-48_IGHD-M',\n",
       " 'T1D_IGHV3-48_IGHG',\n",
       " 'T1D_IGHV1-46_IGHA',\n",
       " 'T1D_IGHV1-46_IGHD-M',\n",
       " 'T1D_IGHV1-46_IGHG',\n",
       " 'T1D_IGHV4-39_IGHA',\n",
       " 'T1D_IGHV4-39_IGHD-M',\n",
       " 'T1D_IGHV4-39_IGHG',\n",
       " 'T1D_IGHV5-51_IGHA',\n",
       " 'T1D_IGHV5-51_IGHD-M',\n",
       " 'T1D_IGHV5-51_IGHG',\n",
       " 'T1D_IGHV6-1_IGHA',\n",
       " 'T1D_IGHV6-1_IGHD-M',\n",
       " 'T1D_IGHV6-1_IGHG',\n",
       " 'T1D_IGHV1-2_IGHA',\n",
       " 'T1D_IGHV1-2_IGHD-M',\n",
       " 'T1D_IGHV1-2_IGHG',\n",
       " 'T1D_IGHV3-11_IGHA',\n",
       " 'T1D_IGHV3-11_IGHD-M',\n",
       " 'T1D_IGHV3-11_IGHG',\n",
       " 'T1D_IGHV4-34_IGHA',\n",
       " 'T1D_IGHV4-34_IGHD-M',\n",
       " 'T1D_IGHV4-34_IGHG',\n",
       " 'T1D_IGHV3-20_IGHA',\n",
       " 'T1D_IGHV3-20_IGHD-M',\n",
       " 'T1D_IGHV3-20_IGHG',\n",
       " 'T1D_IGHV4-31_IGHA',\n",
       " 'T1D_IGHV4-31_IGHD-M',\n",
       " 'T1D_IGHV4-31_IGHG',\n",
       " 'T1D_IGHV3-30_IGHA',\n",
       " 'T1D_IGHV3-30_IGHD-M',\n",
       " 'T1D_IGHV3-30_IGHG',\n",
       " 'T1D_IGHV4-4_IGHA',\n",
       " 'T1D_IGHV4-4_IGHD-M',\n",
       " 'T1D_IGHV4-4_IGHG',\n",
       " 'T1D_IGHV3-33_IGHA',\n",
       " 'T1D_IGHV3-33_IGHD-M',\n",
       " 'T1D_IGHV3-33_IGHG',\n",
       " 'T1D_IGHV1-8_IGHA',\n",
       " 'T1D_IGHV1-8_IGHD-M',\n",
       " 'T1D_IGHV1-8_IGHG',\n",
       " 'T1D_IGHV3-9_IGHA',\n",
       " 'T1D_IGHV3-9_IGHD-M',\n",
       " 'T1D_IGHV3-9_IGHG',\n",
       " 'T1D_IGHV3-53_IGHA',\n",
       " 'T1D_IGHV3-53_IGHD-M',\n",
       " 'T1D_IGHV3-53_IGHG',\n",
       " 'T1D_IGHV3-74_IGHA',\n",
       " 'T1D_IGHV3-74_IGHD-M',\n",
       " 'T1D_IGHV3-74_IGHG',\n",
       " 'T1D_IGHV3-15_IGHA',\n",
       " 'T1D_IGHV3-15_IGHD-M',\n",
       " 'T1D_IGHV3-15_IGHG',\n",
       " 'T1D_IGHV2-5_IGHA',\n",
       " 'T1D_IGHV2-5_IGHD-M',\n",
       " 'T1D_IGHV2-5_IGHG',\n",
       " 'T1D_IGHV4-30-2_IGHA',\n",
       " 'T1D_IGHV4-30-2_IGHD-M',\n",
       " 'T1D_IGHV4-30-2_IGHG',\n",
       " 'T1D_IGHV3-13_IGHA',\n",
       " 'T1D_IGHV3-13_IGHD-M',\n",
       " 'T1D_IGHV3-13_IGHG',\n",
       " 'T1D_IGHV3-49_IGHA',\n",
       " 'T1D_IGHV3-49_IGHD-M',\n",
       " 'T1D_IGHV3-49_IGHG',\n",
       " 'T1D_IGHV1-24_IGHA',\n",
       " 'T1D_IGHV1-24_IGHD-M',\n",
       " 'T1D_IGHV1-24_IGHG',\n",
       " 'T1D_IGHV3-30-3_IGHA',\n",
       " 'T1D_IGHV3-30-3_IGHD-M',\n",
       " 'T1D_IGHV3-30-3_IGHG',\n",
       " 'T1D_IGHV2-70_IGHA',\n",
       " 'T1D_IGHV2-70_IGHD-M',\n",
       " 'T1D_IGHV2-70_IGHG',\n",
       " 'T1D_IGHV3-66_IGHA',\n",
       " 'T1D_IGHV3-66_IGHD-M',\n",
       " 'T1D_IGHV3-66_IGHG',\n",
       " 'T1D_IGHV4-30-4_IGHA',\n",
       " 'T1D_IGHV4-30-4_IGHD-M',\n",
       " 'T1D_IGHV4-30-4_IGHG',\n",
       " 'T1D_IGHV1-3_IGHA',\n",
       " 'T1D_IGHV1-3_IGHD-M',\n",
       " 'T1D_IGHV1-3_IGHG',\n",
       " 'T1D_IGHV5-a_IGHA',\n",
       " 'T1D_IGHV5-a_IGHD-M',\n",
       " 'T1D_IGHV5-a_IGHG',\n",
       " 'T1D_IGHV7-4-1_IGHA',\n",
       " 'T1D_IGHV7-4-1_IGHD-M',\n",
       " 'T1D_IGHV7-4-1_IGHG',\n",
       " 'Covid19_IGHV3-23_IGHA',\n",
       " 'Covid19_IGHV3-23_IGHD-M',\n",
       " 'Covid19_IGHV3-23_IGHG',\n",
       " 'Covid19_IGHV4-b_IGHA',\n",
       " 'Covid19_IGHV4-b_IGHD-M',\n",
       " 'Covid19_IGHV4-b_IGHG',\n",
       " 'Covid19_IGHV3-7_IGHA',\n",
       " 'Covid19_IGHV3-7_IGHD-M',\n",
       " 'Covid19_IGHV3-7_IGHG',\n",
       " 'Covid19_IGHV1-18_IGHA',\n",
       " 'Covid19_IGHV1-18_IGHD-M',\n",
       " 'Covid19_IGHV1-18_IGHG',\n",
       " 'Covid19_IGHV4-59_IGHA',\n",
       " 'Covid19_IGHV4-59_IGHD-M',\n",
       " 'Covid19_IGHV4-59_IGHG',\n",
       " 'Covid19_IGHV4-61_IGHA',\n",
       " 'Covid19_IGHV4-61_IGHD-M',\n",
       " 'Covid19_IGHV4-61_IGHG',\n",
       " 'Covid19_IGHV1-69_IGHA',\n",
       " 'Covid19_IGHV1-69_IGHD-M',\n",
       " 'Covid19_IGHV1-69_IGHG',\n",
       " 'Covid19_IGHV3-21_IGHA',\n",
       " 'Covid19_IGHV3-21_IGHD-M',\n",
       " 'Covid19_IGHV3-21_IGHG',\n",
       " 'Covid19_IGHV3-48_IGHA',\n",
       " 'Covid19_IGHV3-48_IGHD-M',\n",
       " 'Covid19_IGHV3-48_IGHG',\n",
       " 'Covid19_IGHV1-46_IGHA',\n",
       " 'Covid19_IGHV1-46_IGHD-M',\n",
       " 'Covid19_IGHV1-46_IGHG',\n",
       " 'Covid19_IGHV4-39_IGHA',\n",
       " 'Covid19_IGHV4-39_IGHD-M',\n",
       " 'Covid19_IGHV4-39_IGHG',\n",
       " 'Covid19_IGHV5-51_IGHA',\n",
       " 'Covid19_IGHV5-51_IGHD-M',\n",
       " 'Covid19_IGHV5-51_IGHG',\n",
       " 'Covid19_IGHV6-1_IGHA',\n",
       " 'Covid19_IGHV6-1_IGHD-M',\n",
       " 'Covid19_IGHV6-1_IGHG',\n",
       " 'Covid19_IGHV1-2_IGHA',\n",
       " 'Covid19_IGHV1-2_IGHD-M',\n",
       " 'Covid19_IGHV1-2_IGHG',\n",
       " 'Covid19_IGHV3-11_IGHA',\n",
       " 'Covid19_IGHV3-11_IGHD-M',\n",
       " 'Covid19_IGHV3-11_IGHG',\n",
       " 'Covid19_IGHV4-34_IGHA',\n",
       " 'Covid19_IGHV4-34_IGHD-M',\n",
       " 'Covid19_IGHV4-34_IGHG',\n",
       " 'Covid19_IGHV3-20_IGHA',\n",
       " 'Covid19_IGHV3-20_IGHD-M',\n",
       " 'Covid19_IGHV3-20_IGHG',\n",
       " 'Covid19_IGHV4-31_IGHA',\n",
       " 'Covid19_IGHV4-31_IGHD-M',\n",
       " 'Covid19_IGHV4-31_IGHG',\n",
       " 'Covid19_IGHV3-30_IGHA',\n",
       " 'Covid19_IGHV3-30_IGHD-M',\n",
       " 'Covid19_IGHV3-30_IGHG',\n",
       " 'Covid19_IGHV4-4_IGHA',\n",
       " 'Covid19_IGHV4-4_IGHD-M',\n",
       " 'Covid19_IGHV4-4_IGHG',\n",
       " 'Covid19_IGHV3-33_IGHA',\n",
       " 'Covid19_IGHV3-33_IGHD-M',\n",
       " 'Covid19_IGHV3-33_IGHG',\n",
       " 'Covid19_IGHV1-8_IGHA',\n",
       " 'Covid19_IGHV1-8_IGHD-M',\n",
       " 'Covid19_IGHV1-8_IGHG',\n",
       " 'Covid19_IGHV3-9_IGHA',\n",
       " 'Covid19_IGHV3-9_IGHD-M',\n",
       " 'Covid19_IGHV3-9_IGHG',\n",
       " 'Covid19_IGHV3-53_IGHA',\n",
       " 'Covid19_IGHV3-53_IGHD-M',\n",
       " 'Covid19_IGHV3-53_IGHG',\n",
       " 'Covid19_IGHV3-74_IGHA',\n",
       " 'Covid19_IGHV3-74_IGHD-M',\n",
       " 'Covid19_IGHV3-74_IGHG',\n",
       " 'Covid19_IGHV3-15_IGHA',\n",
       " 'Covid19_IGHV3-15_IGHD-M',\n",
       " 'Covid19_IGHV3-15_IGHG',\n",
       " 'Covid19_IGHV2-5_IGHA',\n",
       " 'Covid19_IGHV2-5_IGHD-M',\n",
       " 'Covid19_IGHV2-5_IGHG',\n",
       " 'Covid19_IGHV4-30-2_IGHA',\n",
       " 'Covid19_IGHV4-30-2_IGHD-M',\n",
       " 'Covid19_IGHV4-30-2_IGHG',\n",
       " 'Covid19_IGHV3-13_IGHA',\n",
       " 'Covid19_IGHV3-13_IGHD-M',\n",
       " 'Covid19_IGHV3-13_IGHG',\n",
       " 'Covid19_IGHV3-49_IGHA',\n",
       " 'Covid19_IGHV3-49_IGHD-M',\n",
       " 'Covid19_IGHV3-49_IGHG',\n",
       " 'Covid19_IGHV1-24_IGHA',\n",
       " 'Covid19_IGHV1-24_IGHD-M',\n",
       " 'Covid19_IGHV1-24_IGHG',\n",
       " 'Covid19_IGHV3-30-3_IGHA',\n",
       " 'Covid19_IGHV3-30-3_IGHD-M',\n",
       " 'Covid19_IGHV3-30-3_IGHG',\n",
       " 'Covid19_IGHV2-70_IGHA',\n",
       " 'Covid19_IGHV2-70_IGHD-M',\n",
       " 'Covid19_IGHV2-70_IGHG',\n",
       " 'Covid19_IGHV3-66_IGHA',\n",
       " 'Covid19_IGHV3-66_IGHD-M',\n",
       " 'Covid19_IGHV3-66_IGHG',\n",
       " 'Covid19_IGHV4-30-4_IGHA',\n",
       " 'Covid19_IGHV4-30-4_IGHD-M',\n",
       " 'Covid19_IGHV4-30-4_IGHG',\n",
       " 'Covid19_IGHV1-3_IGHA',\n",
       " 'Covid19_IGHV1-3_IGHD-M',\n",
       " 'Covid19_IGHV1-3_IGHG',\n",
       " 'Covid19_IGHV5-a_IGHA',\n",
       " 'Covid19_IGHV5-a_IGHD-M',\n",
       " 'Covid19_IGHV5-a_IGHG',\n",
       " 'Covid19_IGHV7-4-1_IGHA',\n",
       " 'Covid19_IGHV7-4-1_IGHD-M',\n",
       " 'Covid19_IGHV7-4-1_IGHG']"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf3.feature_names_in_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f5dbe35-85d3-46e0-a91e-4dc244adb90e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6290069d-c00d-4f0f-bdbc-625986b86f70",
   "metadata": {},
   "source": [
    "**How do we get features like `Influenza_IGHV3-23_IGHA`, which represents the average probability of the Influenza class across the IGHV3-23, IgA sequences in each sample?**\n",
    "\n",
    "First, a vector of per-class predicted probabilities is generated for each sequence, using the model associated with the V gene and isotype the sequence belongs to.\n",
    "\n",
    "Then the probabilities are aggregated across sequences from the same sample. Probabilities are only comparable between sequences scored by the same model, so the aggregation happens separately for each V gene and isotype group. For BCR, the aggregation strategy used is just an average:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "afe8c786-3871-4d07-b314-20c9650cca28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AggregationStrategy.mean: 1>"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf3.aggregation_strategy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81975c76-3c32-42ff-a7c4-f78dba5314ca",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "09cb9adc-0991-45a0-86c2-d82c8c2b9e1c",
   "metadata": {},
   "source": [
    "Once we have those features, the \"aggregation stage\" involves running this scikit-learn Pipeline:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "3f1e638b-b506-4b1d-96cc-7806966c0f63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-3 {color: black;background-color: white;}#sk-container-id-3 pre{padding: 0;}#sk-container-id-3 div.sk-toggleable {background-color: white;}#sk-container-id-3 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-3 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-3 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-3 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-3 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-3 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-3 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-3 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-3 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-3 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-3 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-3 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-3 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-3 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-3 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-3 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-3 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-3 div.sk-item {position: relative;z-index: 1;}#sk-container-id-3 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-3 div.sk-item::before, #sk-container-id-3 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-3 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-3 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-3 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-3 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-3 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-3 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-3 div.sk-label-container {text-align: center;}#sk-container-id-3 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-3 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-3\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>Pipeline(steps=[(&#x27;matchvariables&#x27;, MatchVariables()),\n",
       "                (&#x27;standardscalerthatpreservesinputtype&#x27;,\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                (&#x27;binaryovrclassifierwithfeaturesubsettingbyclass&#x27;,\n",
       "                 BinaryOvRClassifierWithFeatureSubsettingByClass(allow_some_classes_to_fail_to_train=True,\n",
       "                                                                 estimator=RandomForestClassifier(class_weight=&#x27;balanced_subsample&#x27;,\n",
       "                                                                                                  n_jobs=1,\n",
       "                                                                                                  random_state=0),\n",
       "                                                                 feature_name_class_delimeter=&#x27;_&#x27;,\n",
       "                                                                 n_jobs=1))])</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-23\" type=\"checkbox\" ><label for=\"sk-estimator-id-23\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">Pipeline</label><div class=\"sk-toggleable__content\"><pre>Pipeline(steps=[(&#x27;matchvariables&#x27;, MatchVariables()),\n",
       "                (&#x27;standardscalerthatpreservesinputtype&#x27;,\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                (&#x27;binaryovrclassifierwithfeaturesubsettingbyclass&#x27;,\n",
       "                 BinaryOvRClassifierWithFeatureSubsettingByClass(allow_some_classes_to_fail_to_train=True,\n",
       "                                                                 estimator=RandomForestClassifier(class_weight=&#x27;balanced_subsample&#x27;,\n",
       "                                                                                                  n_jobs=1,\n",
       "                                                                                                  random_state=0),\n",
       "                                                                 feature_name_class_delimeter=&#x27;_&#x27;,\n",
       "                                                                 n_jobs=1))])</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-24\" type=\"checkbox\" ><label for=\"sk-estimator-id-24\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">MatchVariables</label><div class=\"sk-toggleable__content\"><pre>MatchVariables()</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-25\" type=\"checkbox\" ><label for=\"sk-estimator-id-25\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">StandardScalerThatPreservesInputType</label><div class=\"sk-toggleable__content\"><pre>StandardScalerThatPreservesInputType()</pre></div></div></div><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-26\" type=\"checkbox\" ><label for=\"sk-estimator-id-26\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">binaryovrclassifierwithfeaturesubsettingbyclass: BinaryOvRClassifierWithFeatureSubsettingByClass</label><div class=\"sk-toggleable__content\"><pre>BinaryOvRClassifierWithFeatureSubsettingByClass(allow_some_classes_to_fail_to_train=True,\n",
       "                                                estimator=RandomForestClassifier(class_weight=&#x27;balanced_subsample&#x27;,\n",
       "                                                                                 n_jobs=1,\n",
       "                                                                                 random_state=0),\n",
       "                                                feature_name_class_delimeter=&#x27;_&#x27;,\n",
       "                                                n_jobs=1)</pre></div></div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-27\" type=\"checkbox\" ><label for=\"sk-estimator-id-27\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">estimator: RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier(class_weight=&#x27;balanced_subsample&#x27;, n_jobs=1,\n",
       "                       random_state=0)</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-28\" type=\"checkbox\" ><label for=\"sk-estimator-id-28\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier(class_weight=&#x27;balanced_subsample&#x27;, n_jobs=1,\n",
       "                       random_state=0)</pre></div></div></div></div></div></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                ('standardscalerthatpreservesinputtype',\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                ('binaryovrclassifierwithfeaturesubsettingbyclass',\n",
       "                 BinaryOvRClassifierWithFeatureSubsettingByClass(allow_some_classes_to_fail_to_train=True,\n",
       "                                                                 estimator=RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                                                                                  n_jobs=1,\n",
       "                                                                                                  random_state=0),\n",
       "                                                                 feature_name_class_delimeter='_',\n",
       "                                                                 n_jobs=1))])"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf3._inner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7da929a0-3040-49f8-88a4-94f3a425efcf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d8b2df0a-d91e-4d63-b84b-6b8daa0dd9c8",
   "metadata": {},
   "source": [
    "You may recognize the first two steps of the Pipeline from before:\n",
    "\n",
    "* `MatchVariables` confirms the expected feature names are present.\n",
    "* `StandardScalerThatPreservesInputType` standardizes the features.\n",
    "\n",
    "But the third step, `BinaryOvRClassifierWithFeatureSubsettingByClass`, is new.\n",
    "\n",
    "What's happening here is that the aggregation stage model is fitted in a one-versus-rest fashion, with one model for each class (e.g. Covid-19 vs rest):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "9098769e-13cf-4e91-ac51-5307f0c32546",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[InnerEstimator(clf=RandomForestClassifier(class_weight='balanced_subsample', n_jobs=1,\n",
       "                        random_state=0), negative_class='not Covid19', positive_class='Covid19'),\n",
       " InnerEstimator(clf=RandomForestClassifier(class_weight='balanced_subsample', n_jobs=1,\n",
       "                        random_state=0), negative_class='not HIV', positive_class='HIV'),\n",
       " InnerEstimator(clf=RandomForestClassifier(class_weight='balanced_subsample', n_jobs=1,\n",
       "                        random_state=0), negative_class='not Healthy/Background', positive_class='Healthy/Background'),\n",
       " InnerEstimator(clf=RandomForestClassifier(class_weight='balanced_subsample', n_jobs=1,\n",
       "                        random_state=0), negative_class='not Influenza', positive_class='Influenza'),\n",
       " InnerEstimator(clf=RandomForestClassifier(class_weight='balanced_subsample', n_jobs=1,\n",
       "                        random_state=0), negative_class='not Lupus', positive_class='Lupus'),\n",
       " InnerEstimator(clf=RandomForestClassifier(class_weight='balanced_subsample', n_jobs=1,\n",
       "                        random_state=0), negative_class='not T1D', positive_class='T1D')]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf3.named_steps[\"binaryovrclassifierwithfeaturesubsettingbyclass\"].estimators_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d2ca015-c23a-43e0-95df-824b1035f387",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "dfeb5a4e-0399-49ec-990b-c3845c192004",
   "metadata": {},
   "source": [
    "Additionally, the submodel for each class is trained _only with features corresponding to that class_.\n",
    "\n",
    "For example, the sequence-level model generates predicted class probabilities `P(Covid-19)`, `P(HIV)`, `P(Lupus)`, and so forth for every sequence. But when making predictions of Covid-19 using the \"Covid-19 vs rest\" model, we should only look at the `P(Covid-19)` values. Sequence-level probabilities for the other classes, like `P(HIV)`, should have no bearing.\n",
    "\n",
    "Let's confirm that only Covid-19 specific features enter the \"Covid-19 vs rest\" model. All of these are `P(Covid-19)` features:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "601ef24f-2911-4b1d-9c6f-d7f101b1c823",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Covid19_IGHV3-23_IGHA', 'Covid19_IGHV3-23_IGHD-M',\n",
       "       'Covid19_IGHV3-23_IGHG', 'Covid19_IGHV4-b_IGHA',\n",
       "       'Covid19_IGHV4-b_IGHD-M', 'Covid19_IGHV4-b_IGHG',\n",
       "       'Covid19_IGHV3-7_IGHA', 'Covid19_IGHV3-7_IGHD-M',\n",
       "       'Covid19_IGHV3-7_IGHG', 'Covid19_IGHV1-18_IGHA',\n",
       "       'Covid19_IGHV1-18_IGHD-M', 'Covid19_IGHV1-18_IGHG',\n",
       "       'Covid19_IGHV4-59_IGHA', 'Covid19_IGHV4-59_IGHD-M',\n",
       "       'Covid19_IGHV4-59_IGHG', 'Covid19_IGHV4-61_IGHA',\n",
       "       'Covid19_IGHV4-61_IGHD-M', 'Covid19_IGHV4-61_IGHG',\n",
       "       'Covid19_IGHV1-69_IGHA', 'Covid19_IGHV1-69_IGHD-M',\n",
       "       'Covid19_IGHV1-69_IGHG', 'Covid19_IGHV3-21_IGHA',\n",
       "       'Covid19_IGHV3-21_IGHD-M', 'Covid19_IGHV3-21_IGHG',\n",
       "       'Covid19_IGHV3-48_IGHA', 'Covid19_IGHV3-48_IGHD-M',\n",
       "       'Covid19_IGHV3-48_IGHG', 'Covid19_IGHV1-46_IGHA',\n",
       "       'Covid19_IGHV1-46_IGHD-M', 'Covid19_IGHV1-46_IGHG',\n",
       "       'Covid19_IGHV4-39_IGHA', 'Covid19_IGHV4-39_IGHD-M',\n",
       "       'Covid19_IGHV4-39_IGHG', 'Covid19_IGHV5-51_IGHA',\n",
       "       'Covid19_IGHV5-51_IGHD-M', 'Covid19_IGHV5-51_IGHG',\n",
       "       'Covid19_IGHV6-1_IGHA', 'Covid19_IGHV6-1_IGHD-M',\n",
       "       'Covid19_IGHV6-1_IGHG', 'Covid19_IGHV1-2_IGHA',\n",
       "       'Covid19_IGHV1-2_IGHD-M', 'Covid19_IGHV1-2_IGHG',\n",
       "       'Covid19_IGHV3-11_IGHA', 'Covid19_IGHV3-11_IGHD-M',\n",
       "       'Covid19_IGHV3-11_IGHG', 'Covid19_IGHV4-34_IGHA',\n",
       "       'Covid19_IGHV4-34_IGHD-M', 'Covid19_IGHV4-34_IGHG',\n",
       "       'Covid19_IGHV3-20_IGHA', 'Covid19_IGHV3-20_IGHD-M',\n",
       "       'Covid19_IGHV3-20_IGHG', 'Covid19_IGHV4-31_IGHA',\n",
       "       'Covid19_IGHV4-31_IGHD-M', 'Covid19_IGHV4-31_IGHG',\n",
       "       'Covid19_IGHV3-30_IGHA', 'Covid19_IGHV3-30_IGHD-M',\n",
       "       'Covid19_IGHV3-30_IGHG', 'Covid19_IGHV4-4_IGHA',\n",
       "       'Covid19_IGHV4-4_IGHD-M', 'Covid19_IGHV4-4_IGHG',\n",
       "       'Covid19_IGHV3-33_IGHA', 'Covid19_IGHV3-33_IGHD-M',\n",
       "       'Covid19_IGHV3-33_IGHG', 'Covid19_IGHV1-8_IGHA',\n",
       "       'Covid19_IGHV1-8_IGHD-M', 'Covid19_IGHV1-8_IGHG',\n",
       "       'Covid19_IGHV3-9_IGHA', 'Covid19_IGHV3-9_IGHD-M',\n",
       "       'Covid19_IGHV3-9_IGHG', 'Covid19_IGHV3-53_IGHA',\n",
       "       'Covid19_IGHV3-53_IGHD-M', 'Covid19_IGHV3-53_IGHG',\n",
       "       'Covid19_IGHV3-74_IGHA', 'Covid19_IGHV3-74_IGHD-M',\n",
       "       'Covid19_IGHV3-74_IGHG', 'Covid19_IGHV3-15_IGHA',\n",
       "       'Covid19_IGHV3-15_IGHD-M', 'Covid19_IGHV3-15_IGHG',\n",
       "       'Covid19_IGHV2-5_IGHA', 'Covid19_IGHV2-5_IGHD-M',\n",
       "       'Covid19_IGHV2-5_IGHG', 'Covid19_IGHV4-30-2_IGHA',\n",
       "       'Covid19_IGHV4-30-2_IGHD-M', 'Covid19_IGHV4-30-2_IGHG',\n",
       "       'Covid19_IGHV3-13_IGHA', 'Covid19_IGHV3-13_IGHD-M',\n",
       "       'Covid19_IGHV3-13_IGHG', 'Covid19_IGHV3-49_IGHA',\n",
       "       'Covid19_IGHV3-49_IGHD-M', 'Covid19_IGHV3-49_IGHG',\n",
       "       'Covid19_IGHV1-24_IGHA', 'Covid19_IGHV1-24_IGHD-M',\n",
       "       'Covid19_IGHV1-24_IGHG', 'Covid19_IGHV3-30-3_IGHA',\n",
       "       'Covid19_IGHV3-30-3_IGHD-M', 'Covid19_IGHV3-30-3_IGHG',\n",
       "       'Covid19_IGHV2-70_IGHA', 'Covid19_IGHV2-70_IGHD-M',\n",
       "       'Covid19_IGHV2-70_IGHG', 'Covid19_IGHV3-66_IGHA',\n",
       "       'Covid19_IGHV3-66_IGHD-M', 'Covid19_IGHV3-66_IGHG',\n",
       "       'Covid19_IGHV4-30-4_IGHA', 'Covid19_IGHV4-30-4_IGHD-M',\n",
       "       'Covid19_IGHV4-30-4_IGHG', 'Covid19_IGHV1-3_IGHA',\n",
       "       'Covid19_IGHV1-3_IGHD-M', 'Covid19_IGHV1-3_IGHG',\n",
       "       'Covid19_IGHV5-a_IGHA', 'Covid19_IGHV5-a_IGHD-M',\n",
       "       'Covid19_IGHV5-a_IGHG', 'Covid19_IGHV7-4-1_IGHA',\n",
       "       'Covid19_IGHV7-4-1_IGHD-M', 'Covid19_IGHV7-4-1_IGHG'], dtype=object)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf3.named_steps[\"binaryovrclassifierwithfeaturesubsettingbyclass\"].estimators_[\n",
    "    0\n",
    "].clf.feature_names_in_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cddbb951-097b-47d3-b9da-797ddf713e0f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7e18bd06-dd58-4f0e-9e65-43892e5de0dc",
   "metadata": {},
   "source": [
    "Now that we have Model 3 loaded, let's featurize the dataset to be able to use the model.\n",
    "\n",
    "This means every sequence is run through its associated sequence-level classifier, then the sequence probabilities are aggregated into sample-level features:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "fdd4a7fa-fa77-41a0-b7ce-f97eab429969",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "{\"message\": \"Number of VJGeneSpecificSequenceModelRollupClassifier featurization matrix N/As due to specimens not having any sequences with particular V/J gene pairs: 8538 / 140904 = 6.06%\", \"time\": \"2024-08-07T13:54:29.165469\"}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "malid.trained_model_wrappers.vj_gene_specific_sequence_model_rollup_classifier.SubsetRollupClassifierFeaturizedData"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "featurized_model3_data = clf3.featurize(adata_bcr)\n",
    "type(featurized_model3_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0da12948-a468-4fd1-b9ac-fea23b58b1a4",
   "metadata": {},
   "source": [
    "> _Aside:_\n",
    ">\n",
    "> The log message about \"VJGeneSpecificSequenceModelRollupClassifier featurization matrix N/As due to specimens not having any sequences with particular V/J gene pairs\" is a bit misleading.\n",
    ">\n",
    "> What it's actually referring to is missing values due to some samples having no sequences to score and aggregate for a particular V gene and isotype. For these samples, the `P(Covid-19)`, `P(HIV)`, `P(Lupus)`, and so on are set to the value `1 / n_classes` for the V gene and isotype combination that are missing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33a32fd2-d1ea-433f-85ab-5ec7ba6e4670",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ffe399dd-e73c-4246-81b7-b6f2b0dc5e58",
   "metadata": {},
   "source": [
    "Let's review the features that were generated:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "5412981a-1318-4d89-838f-48282e41fc1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Influenza_IGHV3-23_IGHA</th>\n",
       "      <th>Influenza_IGHV3-23_IGHD-M</th>\n",
       "      <th>Influenza_IGHV3-23_IGHG</th>\n",
       "      <th>Influenza_IGHV4-b_IGHA</th>\n",
       "      <th>Influenza_IGHV4-b_IGHD-M</th>\n",
       "      <th>Influenza_IGHV4-b_IGHG</th>\n",
       "      <th>Influenza_IGHV3-7_IGHA</th>\n",
       "      <th>Influenza_IGHV3-7_IGHD-M</th>\n",
       "      <th>Influenza_IGHV3-7_IGHG</th>\n",
       "      <th>Influenza_IGHV1-18_IGHA</th>\n",
       "      <th>...</th>\n",
       "      <th>Covid19_IGHV4-30-4_IGHG</th>\n",
       "      <th>Covid19_IGHV1-3_IGHA</th>\n",
       "      <th>Covid19_IGHV1-3_IGHD-M</th>\n",
       "      <th>Covid19_IGHV1-3_IGHG</th>\n",
       "      <th>Covid19_IGHV5-a_IGHA</th>\n",
       "      <th>Covid19_IGHV5-a_IGHD-M</th>\n",
       "      <th>Covid19_IGHV5-a_IGHG</th>\n",
       "      <th>Covid19_IGHV7-4-1_IGHA</th>\n",
       "      <th>Covid19_IGHV7-4-1_IGHD-M</th>\n",
       "      <th>Covid19_IGHV7-4-1_IGHG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M111-S003</th>\n",
       "      <td>0.010139</td>\n",
       "      <td>-0.128048</td>\n",
       "      <td>-0.010928</td>\n",
       "      <td>-0.018978</td>\n",
       "      <td>-0.020769</td>\n",
       "      <td>-0.029884</td>\n",
       "      <td>-0.026154</td>\n",
       "      <td>-0.038170</td>\n",
       "      <td>-0.014322</td>\n",
       "      <td>-0.005550</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000056</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.003164</td>\n",
       "      <td>-0.002320</td>\n",
       "      <td>-0.002893</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S007</th>\n",
       "      <td>0.032499</td>\n",
       "      <td>-0.014721</td>\n",
       "      <td>0.057095</td>\n",
       "      <td>-0.008114</td>\n",
       "      <td>-0.010291</td>\n",
       "      <td>-0.009460</td>\n",
       "      <td>0.037804</td>\n",
       "      <td>-0.012315</td>\n",
       "      <td>0.035763</td>\n",
       "      <td>0.013249</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000180</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000212</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S009</th>\n",
       "      <td>-0.030703</td>\n",
       "      <td>-0.054979</td>\n",
       "      <td>-0.021037</td>\n",
       "      <td>-0.000240</td>\n",
       "      <td>-0.000024</td>\n",
       "      <td>-0.000127</td>\n",
       "      <td>0.005277</td>\n",
       "      <td>-0.030897</td>\n",
       "      <td>-0.015195</td>\n",
       "      <td>-0.003173</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000830</td>\n",
       "      <td>-0.004305</td>\n",
       "      <td>-0.001096</td>\n",
       "      <td>-0.001610</td>\n",
       "      <td>-0.006311</td>\n",
       "      <td>-0.001547</td>\n",
       "      <td>-0.001782</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000266</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S011</th>\n",
       "      <td>-0.083829</td>\n",
       "      <td>-0.054190</td>\n",
       "      <td>-0.017302</td>\n",
       "      <td>-0.035306</td>\n",
       "      <td>-0.037384</td>\n",
       "      <td>-0.019191</td>\n",
       "      <td>-0.079265</td>\n",
       "      <td>-0.045720</td>\n",
       "      <td>-0.011754</td>\n",
       "      <td>-0.080297</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.006675</td>\n",
       "      <td>-0.002341</td>\n",
       "      <td>-0.002151</td>\n",
       "      <td>-0.006111</td>\n",
       "      <td>-0.002172</td>\n",
       "      <td>-0.005744</td>\n",
       "      <td>-0.006852</td>\n",
       "      <td>-0.004383</td>\n",
       "      <td>-0.002286</td>\n",
       "      <td>-0.013172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S016</th>\n",
       "      <td>0.125323</td>\n",
       "      <td>0.033698</td>\n",
       "      <td>0.045521</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000200</td>\n",
       "      <td>0.091379</td>\n",
       "      <td>-0.008926</td>\n",
       "      <td>0.018937</td>\n",
       "      <td>-0.044331</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000388</td>\n",
       "      <td>-0.002148</td>\n",
       "      <td>-0.002241</td>\n",
       "      <td>-0.000831</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.005406</td>\n",
       "      <td>-0.002961</td>\n",
       "      <td>-0.007080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-107</th>\n",
       "      <td>-0.034607</td>\n",
       "      <td>-0.149861</td>\n",
       "      <td>0.124894</td>\n",
       "      <td>-0.006630</td>\n",
       "      <td>-0.012994</td>\n",
       "      <td>-0.009094</td>\n",
       "      <td>0.018204</td>\n",
       "      <td>-0.034933</td>\n",
       "      <td>-0.062698</td>\n",
       "      <td>0.014481</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.003967</td>\n",
       "      <td>-0.001134</td>\n",
       "      <td>-0.004544</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000073</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000117</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-109</th>\n",
       "      <td>-0.003658</td>\n",
       "      <td>-0.124797</td>\n",
       "      <td>0.051024</td>\n",
       "      <td>-0.010079</td>\n",
       "      <td>-0.013158</td>\n",
       "      <td>-0.005915</td>\n",
       "      <td>-0.005914</td>\n",
       "      <td>-0.025667</td>\n",
       "      <td>0.056407</td>\n",
       "      <td>-0.077331</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.028257</td>\n",
       "      <td>-0.002327</td>\n",
       "      <td>-0.003638</td>\n",
       "      <td>-0.016851</td>\n",
       "      <td>-0.015336</td>\n",
       "      <td>-0.006743</td>\n",
       "      <td>-0.011526</td>\n",
       "      <td>-0.000539</td>\n",
       "      <td>-0.000074</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-111</th>\n",
       "      <td>0.128735</td>\n",
       "      <td>-0.082885</td>\n",
       "      <td>0.007388</td>\n",
       "      <td>-0.008783</td>\n",
       "      <td>-0.007355</td>\n",
       "      <td>-0.000910</td>\n",
       "      <td>0.075859</td>\n",
       "      <td>-0.047132</td>\n",
       "      <td>-0.014003</td>\n",
       "      <td>0.074678</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004160</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>0.002157</td>\n",
       "      <td>-0.016897</td>\n",
       "      <td>-0.008180</td>\n",
       "      <td>-0.004634</td>\n",
       "      <td>-0.025562</td>\n",
       "      <td>-0.001787</td>\n",
       "      <td>-0.001867</td>\n",
       "      <td>-0.005862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-112</th>\n",
       "      <td>-0.233664</td>\n",
       "      <td>-0.009533</td>\n",
       "      <td>0.016623</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000212</td>\n",
       "      <td>-0.001248</td>\n",
       "      <td>-0.065664</td>\n",
       "      <td>-0.018313</td>\n",
       "      <td>0.002118</td>\n",
       "      <td>-0.057121</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000634</td>\n",
       "      <td>-0.005376</td>\n",
       "      <td>-0.001812</td>\n",
       "      <td>-0.011419</td>\n",
       "      <td>-0.008069</td>\n",
       "      <td>-0.002266</td>\n",
       "      <td>-0.010332</td>\n",
       "      <td>-0.009078</td>\n",
       "      <td>-0.002326</td>\n",
       "      <td>-0.019100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-114</th>\n",
       "      <td>0.084389</td>\n",
       "      <td>-0.150883</td>\n",
       "      <td>-0.000692</td>\n",
       "      <td>-0.000131</td>\n",
       "      <td>-0.000026</td>\n",
       "      <td>-0.000024</td>\n",
       "      <td>-0.005460</td>\n",
       "      <td>-0.014968</td>\n",
       "      <td>0.014817</td>\n",
       "      <td>0.019185</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000521</td>\n",
       "      <td>-0.000213</td>\n",
       "      <td>-0.000100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.010839</td>\n",
       "      <td>-0.003951</td>\n",
       "      <td>-0.010037</td>\n",
       "      <td>-0.000570</td>\n",
       "      <td>-0.000014</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>206 rows × 684 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Influenza_IGHV3-23_IGHA  Influenza_IGHV3-23_IGHD-M  \\\n",
       "specimen_label                                                       \n",
       "M111-S003                      0.010139                  -0.128048   \n",
       "M111-S007                      0.032499                  -0.014721   \n",
       "M111-S009                     -0.030703                  -0.054979   \n",
       "M111-S011                     -0.083829                  -0.054190   \n",
       "M111-S016                      0.125323                   0.033698   \n",
       "...                                 ...                        ...   \n",
       "M64-107                       -0.034607                  -0.149861   \n",
       "M64-109                       -0.003658                  -0.124797   \n",
       "M64-111                        0.128735                  -0.082885   \n",
       "M64-112                       -0.233664                  -0.009533   \n",
       "M64-114                        0.084389                  -0.150883   \n",
       "\n",
       "                Influenza_IGHV3-23_IGHG  Influenza_IGHV4-b_IGHA  \\\n",
       "specimen_label                                                    \n",
       "M111-S003                     -0.010928               -0.018978   \n",
       "M111-S007                      0.057095               -0.008114   \n",
       "M111-S009                     -0.021037               -0.000240   \n",
       "M111-S011                     -0.017302               -0.035306   \n",
       "M111-S016                      0.045521                0.000000   \n",
       "...                                 ...                     ...   \n",
       "M64-107                        0.124894               -0.006630   \n",
       "M64-109                        0.051024               -0.010079   \n",
       "M64-111                        0.007388               -0.008783   \n",
       "M64-112                        0.016623                0.000000   \n",
       "M64-114                       -0.000692               -0.000131   \n",
       "\n",
       "                Influenza_IGHV4-b_IGHD-M  Influenza_IGHV4-b_IGHG  \\\n",
       "specimen_label                                                     \n",
       "M111-S003                      -0.020769               -0.029884   \n",
       "M111-S007                      -0.010291               -0.009460   \n",
       "M111-S009                      -0.000024               -0.000127   \n",
       "M111-S011                      -0.037384               -0.019191   \n",
       "M111-S016                       0.000000               -0.000200   \n",
       "...                                  ...                     ...   \n",
       "M64-107                        -0.012994               -0.009094   \n",
       "M64-109                        -0.013158               -0.005915   \n",
       "M64-111                        -0.007355               -0.000910   \n",
       "M64-112                        -0.000212               -0.001248   \n",
       "M64-114                        -0.000026               -0.000024   \n",
       "\n",
       "                Influenza_IGHV3-7_IGHA  Influenza_IGHV3-7_IGHD-M  \\\n",
       "specimen_label                                                     \n",
       "M111-S003                    -0.026154                 -0.038170   \n",
       "M111-S007                     0.037804                 -0.012315   \n",
       "M111-S009                     0.005277                 -0.030897   \n",
       "M111-S011                    -0.079265                 -0.045720   \n",
       "M111-S016                     0.091379                 -0.008926   \n",
       "...                                ...                       ...   \n",
       "M64-107                       0.018204                 -0.034933   \n",
       "M64-109                      -0.005914                 -0.025667   \n",
       "M64-111                       0.075859                 -0.047132   \n",
       "M64-112                      -0.065664                 -0.018313   \n",
       "M64-114                      -0.005460                 -0.014968   \n",
       "\n",
       "                Influenza_IGHV3-7_IGHG  Influenza_IGHV1-18_IGHA  ...  \\\n",
       "specimen_label                                                   ...   \n",
       "M111-S003                    -0.014322                -0.005550  ...   \n",
       "M111-S007                     0.035763                 0.013249  ...   \n",
       "M111-S009                    -0.015195                -0.003173  ...   \n",
       "M111-S011                    -0.011754                -0.080297  ...   \n",
       "M111-S016                     0.018937                -0.044331  ...   \n",
       "...                                ...                      ...  ...   \n",
       "M64-107                      -0.062698                 0.014481  ...   \n",
       "M64-109                       0.056407                -0.077331  ...   \n",
       "M64-111                      -0.014003                 0.074678  ...   \n",
       "M64-112                       0.002118                -0.057121  ...   \n",
       "M64-114                       0.014817                 0.019185  ...   \n",
       "\n",
       "                Covid19_IGHV4-30-4_IGHG  Covid19_IGHV1-3_IGHA  \\\n",
       "specimen_label                                                  \n",
       "M111-S003                     -0.000056              0.000000   \n",
       "M111-S007                     -0.000180              0.000000   \n",
       "M111-S009                     -0.000830             -0.004305   \n",
       "M111-S011                     -0.006675             -0.002341   \n",
       "M111-S016                     -0.000388             -0.002148   \n",
       "...                                 ...                   ...   \n",
       "M64-107                        0.000000             -0.003967   \n",
       "M64-109                       -0.028257             -0.002327   \n",
       "M64-111                       -0.004160              0.000058   \n",
       "M64-112                       -0.000634             -0.005376   \n",
       "M64-114                       -0.000521             -0.000213   \n",
       "\n",
       "                Covid19_IGHV1-3_IGHD-M  Covid19_IGHV1-3_IGHG  \\\n",
       "specimen_label                                                 \n",
       "M111-S003                     0.000068              0.000000   \n",
       "M111-S007                    -0.000212              0.000000   \n",
       "M111-S009                    -0.001096             -0.001610   \n",
       "M111-S011                    -0.002151             -0.006111   \n",
       "M111-S016                    -0.002241             -0.000831   \n",
       "...                                ...                   ...   \n",
       "M64-107                      -0.001134             -0.004544   \n",
       "M64-109                      -0.003638             -0.016851   \n",
       "M64-111                       0.002157             -0.016897   \n",
       "M64-112                      -0.001812             -0.011419   \n",
       "M64-114                      -0.000100              0.000000   \n",
       "\n",
       "                Covid19_IGHV5-a_IGHA  Covid19_IGHV5-a_IGHD-M  \\\n",
       "specimen_label                                                 \n",
       "M111-S003                  -0.003164               -0.002320   \n",
       "M111-S007                   0.000000                0.000000   \n",
       "M111-S009                  -0.006311               -0.001547   \n",
       "M111-S011                  -0.002172               -0.005744   \n",
       "M111-S016                   0.000000                0.000000   \n",
       "...                              ...                     ...   \n",
       "M64-107                     0.000000               -0.000073   \n",
       "M64-109                    -0.015336               -0.006743   \n",
       "M64-111                    -0.008180               -0.004634   \n",
       "M64-112                    -0.008069               -0.002266   \n",
       "M64-114                    -0.010839               -0.003951   \n",
       "\n",
       "                Covid19_IGHV5-a_IGHG  Covid19_IGHV7-4-1_IGHA  \\\n",
       "specimen_label                                                 \n",
       "M111-S003                  -0.002893                0.000000   \n",
       "M111-S007                   0.000000                0.000000   \n",
       "M111-S009                  -0.001782                0.000000   \n",
       "M111-S011                  -0.006852               -0.004383   \n",
       "M111-S016                   0.000000               -0.005406   \n",
       "...                              ...                     ...   \n",
       "M64-107                     0.000000                0.000000   \n",
       "M64-109                    -0.011526               -0.000539   \n",
       "M64-111                    -0.025562               -0.001787   \n",
       "M64-112                    -0.010332               -0.009078   \n",
       "M64-114                    -0.010037               -0.000570   \n",
       "\n",
       "                Covid19_IGHV7-4-1_IGHD-M  Covid19_IGHV7-4-1_IGHG  \n",
       "specimen_label                                                    \n",
       "M111-S003                       0.000000                0.000000  \n",
       "M111-S007                       0.000000                0.000000  \n",
       "M111-S009                      -0.000266                0.000000  \n",
       "M111-S011                      -0.002286               -0.013172  \n",
       "M111-S016                      -0.002961               -0.007080  \n",
       "...                                  ...                     ...  \n",
       "M64-107                        -0.000117                0.000000  \n",
       "M64-109                        -0.000074                0.000000  \n",
       "M64-111                        -0.001867               -0.005862  \n",
       "M64-112                        -0.002326               -0.019100  \n",
       "M64-114                        -0.000014                0.000000  \n",
       "\n",
       "[206 rows x 684 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Features (in this case, they've already been standardized)\n",
    "featurized_model3_data.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "29b8eb7e-a9e2-428b-b5df-01068f161847",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "specimen_label\n",
       "M111-S003                   HIV\n",
       "M111-S007                   HIV\n",
       "M111-S009                   HIV\n",
       "M111-S011                   HIV\n",
       "M111-S016                   HIV\n",
       "                    ...        \n",
       "M64-107      Healthy/Background\n",
       "M64-109      Healthy/Background\n",
       "M64-111      Healthy/Background\n",
       "M64-112      Healthy/Background\n",
       "M64-114      Healthy/Background\n",
       "Name: disease, Length: 206, dtype: category\n",
       "Categories (6, object): ['HIV', 'Healthy/Background', 'Lupus', 'Covid19', 'Influenza', 'T1D']"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ground truth\n",
    "featurized_model3_data.y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "60c79a9b-c496-4f65-b1d9-f5f3eb87ae1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['M111-S003', 'M111-S007', 'M111-S009', 'M111-S011', 'M111-S016',\n",
       "       'M111-S020', 'M111-S021', 'M111-S023', 'M111-S025', 'M111-S028',\n",
       "       ...\n",
       "       'M64-086', 'M64-089', 'M64-090', 'M64-101', 'M64-103', 'M64-107',\n",
       "       'M64-109', 'M64-111', 'M64-112', 'M64-114'],\n",
       "      dtype='object', name='specimen_label', length=206)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample names\n",
    "featurized_model3_data.sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "b46d38f9-626e-424b-8616-3714c8397751",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease.rollup</th>\n",
       "      <th>disease.separate_past_exposures</th>\n",
       "      <th>disease_severity</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>ethnicity_condensed</th>\n",
       "      <th>isotype_proportion:IGHA</th>\n",
       "      <th>isotype_proportion:IGHD-M</th>\n",
       "      <th>isotype_proportion:IGHG</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>past_exposure</th>\n",
       "      <th>sex</th>\n",
       "      <th>study_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M111-S003</th>\n",
       "      <td>49.0</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV Broad Neutralizing</td>\n",
       "      <td>African</td>\n",
       "      <td>0.120568</td>\n",
       "      <td>0.741955</td>\n",
       "      <td>0.137476</td>\n",
       "      <td>BFI-0003466</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S007</th>\n",
       "      <td>48.0</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV Broad Neutralizing</td>\n",
       "      <td>African</td>\n",
       "      <td>0.114607</td>\n",
       "      <td>0.709431</td>\n",
       "      <td>0.175962</td>\n",
       "      <td>BFI-0003469</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S009</th>\n",
       "      <td>47.0</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV Broad Neutralizing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.174015</td>\n",
       "      <td>0.644778</td>\n",
       "      <td>0.181207</td>\n",
       "      <td>BFI-0003453</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S011</th>\n",
       "      <td>24.0</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV Broad Neutralizing</td>\n",
       "      <td>African</td>\n",
       "      <td>0.056556</td>\n",
       "      <td>0.721739</td>\n",
       "      <td>0.221704</td>\n",
       "      <td>BFI-0002875</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S016</th>\n",
       "      <td>22.0</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV Broad Neutralizing</td>\n",
       "      <td>African</td>\n",
       "      <td>0.101753</td>\n",
       "      <td>0.796830</td>\n",
       "      <td>0.101417</td>\n",
       "      <td>BFI-0002855</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-107</th>\n",
       "      <td>39.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - CMV+</td>\n",
       "      <td>Hispanic/Latino</td>\n",
       "      <td>0.086075</td>\n",
       "      <td>0.877653</td>\n",
       "      <td>0.036272</td>\n",
       "      <td>BFI-0003156</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>Healthy-StanfordBloodCenter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-109</th>\n",
       "      <td>19.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - CMV-</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>0.067132</td>\n",
       "      <td>0.876106</td>\n",
       "      <td>0.056762</td>\n",
       "      <td>BFI-0003158</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>Healthy-StanfordBloodCenter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-111</th>\n",
       "      <td>57.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - CMV+</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>0.084105</td>\n",
       "      <td>0.877510</td>\n",
       "      <td>0.038385</td>\n",
       "      <td>BFI-0003160</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>Healthy-StanfordBloodCenter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-112</th>\n",
       "      <td>18.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - CMV-</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>0.070577</td>\n",
       "      <td>0.897266</td>\n",
       "      <td>0.032157</td>\n",
       "      <td>BFI-0003161</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>Healthy-StanfordBloodCenter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-114</th>\n",
       "      <td>27.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - CMV-</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>0.048393</td>\n",
       "      <td>0.908835</td>\n",
       "      <td>0.042772</td>\n",
       "      <td>BFI-0003163</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>Healthy-StanfordBloodCenter</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>206 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 age             disease      disease.rollup  \\\n",
       "specimen_label                                                 \n",
       "M111-S003       49.0                 HIV                 HIV   \n",
       "M111-S007       48.0                 HIV                 HIV   \n",
       "M111-S009       47.0                 HIV                 HIV   \n",
       "M111-S011       24.0                 HIV                 HIV   \n",
       "M111-S016       22.0                 HIV                 HIV   \n",
       "...              ...                 ...                 ...   \n",
       "M64-107         39.0  Healthy/Background  Healthy/Background   \n",
       "M64-109         19.0  Healthy/Background  Healthy/Background   \n",
       "M64-111         57.0  Healthy/Background  Healthy/Background   \n",
       "M64-112         18.0  Healthy/Background  Healthy/Background   \n",
       "M64-114         27.0  Healthy/Background  Healthy/Background   \n",
       "\n",
       "               disease.separate_past_exposures disease_severity  \\\n",
       "specimen_label                                                    \n",
       "M111-S003                                  HIV              NaN   \n",
       "M111-S007                                  HIV              NaN   \n",
       "M111-S009                                  HIV              NaN   \n",
       "M111-S011                                  HIV              NaN   \n",
       "M111-S016                                  HIV              NaN   \n",
       "...                                        ...              ...   \n",
       "M64-107                     Healthy/Background              NaN   \n",
       "M64-109                     Healthy/Background              NaN   \n",
       "M64-111                     Healthy/Background              NaN   \n",
       "M64-112                     Healthy/Background              NaN   \n",
       "M64-114                     Healthy/Background              NaN   \n",
       "\n",
       "                          disease_subtype ethnicity_condensed  \\\n",
       "specimen_label                                                  \n",
       "M111-S003          HIV Broad Neutralizing             African   \n",
       "M111-S007          HIV Broad Neutralizing             African   \n",
       "M111-S009          HIV Broad Neutralizing                 NaN   \n",
       "M111-S011          HIV Broad Neutralizing             African   \n",
       "M111-S016          HIV Broad Neutralizing             African   \n",
       "...                                   ...                 ...   \n",
       "M64-107         Healthy/Background - CMV+     Hispanic/Latino   \n",
       "M64-109         Healthy/Background - CMV-           Caucasian   \n",
       "M64-111         Healthy/Background - CMV+           Caucasian   \n",
       "M64-112         Healthy/Background - CMV-           Caucasian   \n",
       "M64-114         Healthy/Background - CMV-           Caucasian   \n",
       "\n",
       "                isotype_proportion:IGHA  isotype_proportion:IGHD-M  \\\n",
       "specimen_label                                                       \n",
       "M111-S003                      0.120568                   0.741955   \n",
       "M111-S007                      0.114607                   0.709431   \n",
       "M111-S009                      0.174015                   0.644778   \n",
       "M111-S011                      0.056556                   0.721739   \n",
       "M111-S016                      0.101753                   0.796830   \n",
       "...                                 ...                        ...   \n",
       "M64-107                        0.086075                   0.877653   \n",
       "M64-109                        0.067132                   0.876106   \n",
       "M64-111                        0.084105                   0.877510   \n",
       "M64-112                        0.070577                   0.897266   \n",
       "M64-114                        0.048393                   0.908835   \n",
       "\n",
       "                isotype_proportion:IGHG participant_label  past_exposure sex  \\\n",
       "specimen_label                                                                 \n",
       "M111-S003                      0.137476       BFI-0003466          False   M   \n",
       "M111-S007                      0.175962       BFI-0003469          False   M   \n",
       "M111-S009                      0.181207       BFI-0003453          False   M   \n",
       "M111-S011                      0.221704       BFI-0002875          False   F   \n",
       "M111-S016                      0.101417       BFI-0002855          False   F   \n",
       "...                                 ...               ...            ...  ..   \n",
       "M64-107                        0.036272       BFI-0003156          False   F   \n",
       "M64-109                        0.056762       BFI-0003158          False   M   \n",
       "M64-111                        0.038385       BFI-0003160          False   M   \n",
       "M64-112                        0.032157       BFI-0003161          False   M   \n",
       "M64-114                        0.042772       BFI-0003163          False   M   \n",
       "\n",
       "                                 study_name  \n",
       "specimen_label                               \n",
       "M111-S003                               HIV  \n",
       "M111-S007                               HIV  \n",
       "M111-S009                               HIV  \n",
       "M111-S011                               HIV  \n",
       "M111-S016                               HIV  \n",
       "...                                     ...  \n",
       "M64-107         Healthy-StanfordBloodCenter  \n",
       "M64-109         Healthy-StanfordBloodCenter  \n",
       "M64-111         Healthy-StanfordBloodCenter  \n",
       "M64-112         Healthy-StanfordBloodCenter  \n",
       "M64-114         Healthy-StanfordBloodCenter  \n",
       "\n",
       "[206 rows x 14 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample metadata\n",
    "featurized_model3_data.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa79387b-5085-456c-9f3b-c7062b17549d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a73817ee-9b09-47b8-aa11-98b2e05ec027",
   "metadata": {},
   "source": [
    "Finally, let's run the model to predict the per-class probabilities at a _sample level_:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "466fcaa4-6073-4b9d-921e-11fcfc9f2aaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Covid19</th>\n",
       "      <th>HIV</th>\n",
       "      <th>Healthy/Background</th>\n",
       "      <th>Influenza</th>\n",
       "      <th>Lupus</th>\n",
       "      <th>T1D</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M111-S003</th>\n",
       "      <td>0.01</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S007</th>\n",
       "      <td>0.01</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S009</th>\n",
       "      <td>0.61</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S011</th>\n",
       "      <td>0.01</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M111-S016</th>\n",
       "      <td>0.01</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-107</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.67</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-109</th>\n",
       "      <td>0.05</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-111</th>\n",
       "      <td>0.09</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-112</th>\n",
       "      <td>0.12</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-114</th>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>206 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Covid19   HIV  Healthy/Background  Influenza  Lupus   T1D\n",
       "specimen_label                                                           \n",
       "M111-S003          0.01  0.31                0.14       0.00   0.07  0.04\n",
       "M111-S007          0.01  0.23                0.44       0.03   0.03  0.07\n",
       "M111-S009          0.61  0.00                0.01       0.00   0.23  0.02\n",
       "M111-S011          0.01  0.36                0.11       0.00   0.21  0.15\n",
       "M111-S016          0.01  0.27                0.40       0.05   0.02  0.14\n",
       "...                 ...   ...                 ...        ...    ...   ...\n",
       "M64-107            0.00  0.01                0.67       0.09   0.05  0.20\n",
       "M64-109            0.05  0.04                0.84       0.00   0.04  0.09\n",
       "M64-111            0.09  0.03                0.59       0.05   0.02  0.19\n",
       "M64-112            0.12  0.01                0.54       0.07   0.08  0.16\n",
       "M64-114            0.02  0.01                0.73       0.00   0.00  0.02\n",
       "\n",
       "[206 rows x 6 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Table of predicted class probabilities\n",
    "pd.DataFrame(\n",
    "    clf3.predict_proba(featurized_model3_data.X),\n",
    "    index=featurized_model3_data.sample_names,\n",
    "    columns=clf3.classes_,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1aa44fed-aaae-49bf-bc06-1e6aadffa421",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a982c7e0-aec2-425d-b1c3-5133171e0e81",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "854ce22e-754d-4620-b513-82d392f577e9",
   "metadata": {},
   "source": [
    "## Ensemble metamodel combining Models 1, 2, and 3 and combining BCR and TCR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b6e3ddf-b2b6-4abc-adb7-3b0c9e607bce",
   "metadata": {},
   "source": [
    "Finally, let's load the ensemble metamodel, which brings all the other model components together for a final prediction of disease status.\n",
    "\n",
    "We'll specify `metamodel_name=\"ridge_cv\"` to load the ridge logistic regression version which we highlight in the paper. (The featurization uses the base models chosen in `config.metamodel_base_model_names` as we reviewed above, so the features will be the same regardless of `metamodel_name`. This parameter just controls which classification algorithm is used in the metamodel itself.)\n",
    "\n",
    "We'll also specify `metamodel_flavor=\"default\"`, which refers to combining Models 1, 2, and 3. Other `metamodel_flavor` options include:\n",
    "\n",
    "- `subset_of_submodels_repertoire_stats` (Model 1 only)\n",
    "- `subset_of_submodels_convergent_cluster_model` (Model 2 only)\n",
    "- `subset_of_submodels_sequence_model` (Model 3 only)\n",
    "- `subset_of_submodels_repertoire_stats_convergent_cluster_model` (Models 1 and 2 only)\n",
    "- and so on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "9e1535c1-26b3-49c4-b133-21e3403be165",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>BlendingMetamodel: Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                ('standardscalerthatpreservesinputtype',\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                ('glmnetlogitnetwrapper',\n",
       "                 GlmnetLogitNetWrapper(alpha=0, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=1, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=make_scorer(matthews_corrcoef), standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])</pre>"
      ],
      "text/plain": [
       "BlendingMetamodel: Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                ('standardscalerthatpreservesinputtype',\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                ('glmnetlogitnetwrapper',\n",
       "                 GlmnetLogitNetWrapper(alpha=0, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=1, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=make_scorer(matthews_corrcoef), standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf_metamodel = BlendingMetamodel.from_disk(\n",
    "    fold_id=fold_id,\n",
    "    target_obs_column=target_obs_column,\n",
    "    metamodel_name=\"ridge_cv\",  # Which metamodel version to use\n",
    "    base_model_train_fold_name=\"train_smaller\",  # The base components are fitted on the train_smaller set\n",
    "    metamodel_fold_label_train=\"validation\",  # The metamodel is fitted on the validation set\n",
    "    gene_locus=GeneLocus.BCR | GeneLocus.TCR,  # Use BCR and TCR components together\n",
    "    metamodel_flavor=\"default\",  # Use Models 1 + 2 + 3\n",
    ")\n",
    "\n",
    "clf_metamodel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44217c8a-9432-4991-b8ab-6e48ab0f90de",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b3172550-d962-4208-b28c-f29406e56405",
   "metadata": {},
   "source": [
    "Like we've seen with the other models, `BlendingMetamodel` is a wrapper around a scikit-learn Pipeline that confirms the expected features are present, standardizes them, and then runs ridge logistic regression:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "e5ca2faa-b42c-44dd-bbcc-a790076ba4a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-4 {color: black;background-color: white;}#sk-container-id-4 pre{padding: 0;}#sk-container-id-4 div.sk-toggleable {background-color: white;}#sk-container-id-4 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-4 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-4 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-4 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-4 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-4 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-4 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-4 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-4 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-4 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-4 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-4 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-4 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-4 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-4 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-4 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-4 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-4 div.sk-item {position: relative;z-index: 1;}#sk-container-id-4 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-4 div.sk-item::before, #sk-container-id-4 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-4 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-4 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-4 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-4 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-4 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-4 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-4 div.sk-label-container {text-align: center;}#sk-container-id-4 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-4 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-4\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>Pipeline(steps=[(&#x27;matchvariables&#x27;, MatchVariables()),\n",
       "                (&#x27;standardscalerthatpreservesinputtype&#x27;,\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                (&#x27;glmnetlogitnetwrapper&#x27;,\n",
       "                 GlmnetLogitNetWrapper(alpha=0, class_weight=&#x27;balanced&#x27;, cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=1, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=make_scorer(matthews_corrcoef), standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-29\" type=\"checkbox\" ><label for=\"sk-estimator-id-29\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">Pipeline</label><div class=\"sk-toggleable__content\"><pre>Pipeline(steps=[(&#x27;matchvariables&#x27;, MatchVariables()),\n",
       "                (&#x27;standardscalerthatpreservesinputtype&#x27;,\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                (&#x27;glmnetlogitnetwrapper&#x27;,\n",
       "                 GlmnetLogitNetWrapper(alpha=0, class_weight=&#x27;balanced&#x27;, cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=1, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=make_scorer(matthews_corrcoef), standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-30\" type=\"checkbox\" ><label for=\"sk-estimator-id-30\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">MatchVariables</label><div class=\"sk-toggleable__content\"><pre>MatchVariables()</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-31\" type=\"checkbox\" ><label for=\"sk-estimator-id-31\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">StandardScalerThatPreservesInputType</label><div class=\"sk-toggleable__content\"><pre>StandardScalerThatPreservesInputType()</pre></div></div></div><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-32\" type=\"checkbox\" ><label for=\"sk-estimator-id-32\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">GlmnetLogitNetWrapper</label><div class=\"sk-toggleable__content\"><pre>GlmnetLogitNetWrapper: LogitNet(alpha=0, n_splits=5, random_state=0,\n",
       "         scoring=make_scorer(matthews_corrcoef), standardize=False,\n",
       "         verbose=True)</pre></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                ('standardscalerthatpreservesinputtype',\n",
       "                 StandardScalerThatPreservesInputType()),\n",
       "                ('glmnetlogitnetwrapper',\n",
       "                 GlmnetLogitNetWrapper(alpha=0, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=1, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=make_scorer(matthews_corrcoef), standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))])"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf_metamodel._inner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85bde2f0-0201-4523-8032-62317af48844",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "17ebf412-a8af-4518-9610-ea5e4687ca2d",
   "metadata": {},
   "source": [
    "The features are per-class predicted probabilities from each submodel. For example, `BCR:repertoire_stats:Covid19` is `P(Covid-19)` according to the BCR version of Model 1, the repertoire summary statistics model. Here's the full list of features:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "5c2e1288-ba63-4ca4-af48-18e445e611e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BCR:repertoire_stats:Covid19',\n",
       " 'BCR:repertoire_stats:HIV',\n",
       " 'BCR:repertoire_stats:Healthy/Background',\n",
       " 'BCR:repertoire_stats:Influenza',\n",
       " 'BCR:repertoire_stats:Lupus',\n",
       " 'BCR:repertoire_stats:T1D',\n",
       " 'BCR:convergent_cluster_model:Covid19',\n",
       " 'BCR:convergent_cluster_model:HIV',\n",
       " 'BCR:convergent_cluster_model:Healthy/Background',\n",
       " 'BCR:convergent_cluster_model:Influenza',\n",
       " 'BCR:convergent_cluster_model:Lupus',\n",
       " 'BCR:convergent_cluster_model:T1D',\n",
       " 'BCR:sequence_model:Covid19',\n",
       " 'BCR:sequence_model:HIV',\n",
       " 'BCR:sequence_model:Healthy/Background',\n",
       " 'BCR:sequence_model:Influenza',\n",
       " 'BCR:sequence_model:Lupus',\n",
       " 'BCR:sequence_model:T1D',\n",
       " 'TCR:repertoire_stats:Covid19',\n",
       " 'TCR:repertoire_stats:HIV',\n",
       " 'TCR:repertoire_stats:Healthy/Background',\n",
       " 'TCR:repertoire_stats:Influenza',\n",
       " 'TCR:repertoire_stats:Lupus',\n",
       " 'TCR:repertoire_stats:T1D',\n",
       " 'TCR:convergent_cluster_model:Covid19',\n",
       " 'TCR:convergent_cluster_model:HIV',\n",
       " 'TCR:convergent_cluster_model:Healthy/Background',\n",
       " 'TCR:convergent_cluster_model:Influenza',\n",
       " 'TCR:convergent_cluster_model:Lupus',\n",
       " 'TCR:convergent_cluster_model:T1D',\n",
       " 'TCR:sequence_model:Covid19',\n",
       " 'TCR:sequence_model:HIV',\n",
       " 'TCR:sequence_model:Healthy/Background',\n",
       " 'TCR:sequence_model:Influenza',\n",
       " 'TCR:sequence_model:Lupus',\n",
       " 'TCR:sequence_model:T1D']"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf_metamodel.feature_names_in_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59f8215d-5ad9-4139-8720-942ec2b3cf5f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "feb06cb2-00f9-49e0-8785-c41683d29c3e",
   "metadata": {},
   "source": [
    "As expected, there are three BCR and three TCR submodels:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "923e4e20-6024-4332-ac18-51b301867213",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{<GeneLocus.BCR: 1>: {'repertoire_stats': RepertoireClassifier: Pipeline(steps=[('columntransformer',\n",
       "                   ColumnTransformer(remainder='passthrough',\n",
       "                                     transformers=[('log1p-scale-PCA_IGHG',\n",
       "                                                    Pipeline(steps=[('log1p',\n",
       "                                                                     FunctionTransformer(feature_names_out='one-to-one',\n",
       "                                                                                         func=<ufunc 'log1p'>,\n",
       "                                                                                         validate=True)),\n",
       "                                                                    ('scale',\n",
       "                                                                     StandardScalerThatPreservesInputType()),\n",
       "                                                                    ('pca',\n",
       "                                                                     PCA(n_components=15,\n",
       "                                                                         random_state=0))]),\n",
       "                                                    <sklearn.compose._column_t...\n",
       "                   GlmnetLogitNetWrapper(alpha=0.25, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=1, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=<function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0>, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))]),\n",
       "  'convergent_cluster_model': ConvergentClusterClassifier: Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                  ('standardscalerthatpreservesinputtype',\n",
       "                   StandardScalerThatPreservesInputType()),\n",
       "                  ('glmnetlogitnetwrapper',\n",
       "                   GlmnetLogitNetWrapper(alpha=0, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=8, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=<function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0>, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))]),\n",
       "  'sequence_model': VJGeneSpecificSequenceModelRollupClassifier: Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                  ('standardscalerthatpreservesinputtype',\n",
       "                   StandardScalerThatPreservesInputType()),\n",
       "                  ('binaryovrclassifierwithfeaturesubsettingbyclass',\n",
       "                   BinaryOvRClassifierWithFeatureSubsettingByClass(allow_some_classes_to_fail_to_train=True,\n",
       "                                                                   estimator=RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                                                                                    n_jobs=1,\n",
       "                                                                                                    random_state=0),\n",
       "                                                                   feature_name_class_delimeter='_',\n",
       "                                                                   n_jobs=1))])},\n",
       " <GeneLocus.TCR: 2>: {'repertoire_stats': RepertoireClassifier: Pipeline(steps=[('columntransformer',\n",
       "                   ColumnTransformer(remainder='passthrough',\n",
       "                                     transformers=[('log1p-scale-PCA_TCRB',\n",
       "                                                    Pipeline(steps=[('log1p',\n",
       "                                                                     FunctionTransformer(feature_names_out='one-to-one',\n",
       "                                                                                         func=<ufunc 'log1p'>,\n",
       "                                                                                         validate=True)),\n",
       "                                                                    ('scale',\n",
       "                                                                     StandardScalerThatPreservesInputType()),\n",
       "                                                                    ('pca',\n",
       "                                                                     PCA(n_components=15,\n",
       "                                                                         random_state=0))]),\n",
       "                                                    <sklearn.compose._column_t...\n",
       "                   GlmnetLogitNetWrapper(alpha=1, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=1, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=<function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0>, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))]),\n",
       "  'convergent_cluster_model': ConvergentClusterClassifier: Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                  ('standardscalerthatpreservesinputtype',\n",
       "                   StandardScalerThatPreservesInputType()),\n",
       "                  ('glmnetlogitnetwrapper',\n",
       "                   GlmnetLogitNetWrapper(alpha=1, class_weight='balanced', cut_point=1.0, fit_intercept=True, internal_cv=StratifiedGroupKFoldRequiresGroups(n_splits=5, random_state=0, shuffle=True), lambda_path=None, lower_limits=-inf, max_features=None, max_iter=100000, min_lambda_ratio=0.0001, n_jobs=8, n_lambda=100, n_splits=5, random_state=0, require_cv_group_labels=True, scoring=<function GlmnetLogitNetWrapper.deviance_scorer at 0x7f1af8cb08b0>, standardize=False, tol=1e-07, upper_limits=inf, use_lambda_1se=False, verbose=True))]),\n",
       "  'sequence_model': VJGeneSpecificSequenceModelRollupClassifier: Pipeline(steps=[('matchvariables', MatchVariables()),\n",
       "                  ('standardscalerthatpreservesinputtype',\n",
       "                   StandardScalerThatPreservesInputType()),\n",
       "                  ('binaryovrclassifierwithfeaturesubsettingbyclass',\n",
       "                   BinaryOvRClassifierWithFeatureSubsettingByClass(allow_some_classes_to_fail_to_train=True,\n",
       "                                                                   estimator=RandomForestClassifier(class_weight='balanced_subsample',\n",
       "                                                                                                    n_jobs=1,\n",
       "                                                                                                    random_state=0),\n",
       "                                                                   feature_name_class_delimeter='_',\n",
       "                                                                   n_jobs=1))])}}"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf_metamodel.metamodel_config.submodels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22230116-2cbc-49bf-85e3-9ecef08b1a6c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "43114a08-464d-419a-8aed-bc4d0802a1bf",
   "metadata": {},
   "source": [
    "Let's featurize our input data to use with the metamodel. This time, the call to `featurize()` requires wrapping the input AnnData as a `dict[GeneLocus, AnnData]` — meaning a dictionary that maps from a sequencing locus to an AnnData object. This is because the BCR+TCR metamodel requires both AnnDatas at the same time to generate the features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "ff7fee5b-b25f-4ec4-b711-d7f73a9c9e99",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "{\"message\": \"Metamodel featurization with data keys dict_keys([<GeneLocus.BCR: 1>, <GeneLocus.TCR: 2>]) and gene_locus GeneLocus.BCR|TCR: dropping specimens from GeneLocus.BCR anndata: {'M281redo-S061', 'M281redo-S018', 'M281redo-S058', 'M404-S014', 'M281redo-S054', 'M281redo-S056', 'M281redo-S026', 'M281redo-S045', 'M281redo-S041', 'M281redo-S048', 'M281redo-S019', 'M281redo-S040', 'M281redo-S014', 'M281redo-S017', 'M281redo-S007', 'M281redo-S051', 'M281redo-S008', 'M281redo-S029', 'M281redo-S023', 'M281redo-S020', 'M281redo-S034', 'M281redo-S057'}\", \"time\": \"2024-08-07T13:57:33.424991\"}\n",
      "{\"message\": \"/users/maximz/code/boyd-immune-repertoire-classification/malid/trained_model_wrappers/convergent_cluster_classifier.py:538: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\\n  df[\\\"fold_id\\\"] = pd.to_numeric(\\n\", \"time\": \"2024-08-07T13:57:53.853454\"}\n",
      "{\"message\": \"/users/maximz/anaconda3/envs/cuda-env-py39/lib/python3.9/contextlib.py:126: FutureWarning: X.dtype being converted to np.float32 from float16. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.\\n  next(self.gen)\\n\", \"time\": \"2024-08-07T13:58:05.109361\"}\n",
      "{\"message\": \"Number of VJGeneSpecificSequenceModelRollupClassifier featurization matrix N/As due to specimens not having any sequences with particular V/J gene pairs: 7470 / 125856 = 5.94%\", \"time\": \"2024-08-07T14:03:02.023751\"}\n",
      "{\"message\": \"Number of VJGeneSpecificSequenceModelRollupClassifier featurization matrix N/As due to specimens not having any sequences with particular V/J gene pairs: 0 / 30912 = 0.00%\", \"time\": \"2024-08-07T14:12:51.993458\"}\n",
      "{\"message\": \"Abstained specimens: {'M371-S028', 'M418-S102', 'M418-S118'}\", \"time\": \"2024-08-07T14:12:58.068975\"}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "crosseval.featurized_data.FeaturizedData"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "featurized_metamodel_data = clf_metamodel.featurize(\n",
    "    {GeneLocus.BCR: adata_bcr, GeneLocus.TCR: adata_tcr}\n",
    ")\n",
    "type(featurized_metamodel_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a018723-2dad-4405-8b80-954528676ebf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "48f2ac20-46f5-4d53-93ed-60dd55b37938",
   "metadata": {},
   "source": [
    "> _Aside:_\n",
    ">\n",
    "> There are two new log messages here worth highlighting:\n",
    ">\n",
    "> 1. The first line, with \"dropping specimens from GeneLocus.BCR anndata\": these samples are removed because they have BCR data only — no TCR data. The BCR+TCR metamodel only runs on samples that have both BCR and TCR data available.\n",
    "> 2. The last line, with \"Abstained specimens\": Model 2 abstained from prediction for these samples, because none of the sequences in these samples matched any of the disease-associated clusters. The abstention propogates up into the metamodel."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "654ef907-0637-4ac1-bc07-119487b72443",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "50e0c346-8cce-4776-89bf-da165d6fbcf3",
   "metadata": {},
   "source": [
    "Let's unwrap the `FeaturizedData` container as usual. This time, the features are the predicted class probabilities from the base models:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "300d507f-da74-486d-866b-6a01b7f36845",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BCR:repertoire_stats:Covid19</th>\n",
       "      <th>BCR:repertoire_stats:HIV</th>\n",
       "      <th>BCR:repertoire_stats:Healthy/Background</th>\n",
       "      <th>BCR:repertoire_stats:Influenza</th>\n",
       "      <th>BCR:repertoire_stats:Lupus</th>\n",
       "      <th>BCR:repertoire_stats:T1D</th>\n",
       "      <th>BCR:convergent_cluster_model:Covid19</th>\n",
       "      <th>BCR:convergent_cluster_model:HIV</th>\n",
       "      <th>BCR:convergent_cluster_model:Healthy/Background</th>\n",
       "      <th>BCR:convergent_cluster_model:Influenza</th>\n",
       "      <th>...</th>\n",
       "      <th>TCR:convergent_cluster_model:Healthy/Background</th>\n",
       "      <th>TCR:convergent_cluster_model:Influenza</th>\n",
       "      <th>TCR:convergent_cluster_model:Lupus</th>\n",
       "      <th>TCR:convergent_cluster_model:T1D</th>\n",
       "      <th>TCR:sequence_model:Covid19</th>\n",
       "      <th>TCR:sequence_model:HIV</th>\n",
       "      <th>TCR:sequence_model:Healthy/Background</th>\n",
       "      <th>TCR:sequence_model:Influenza</th>\n",
       "      <th>TCR:sequence_model:Lupus</th>\n",
       "      <th>TCR:sequence_model:T1D</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M124-S066</th>\n",
       "      <td>0.004994</td>\n",
       "      <td>0.376223</td>\n",
       "      <td>0.576059</td>\n",
       "      <td>0.006669</td>\n",
       "      <td>0.002035</td>\n",
       "      <td>0.034018</td>\n",
       "      <td>0.112480</td>\n",
       "      <td>0.229213</td>\n",
       "      <td>0.191971</td>\n",
       "      <td>0.046151</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>1.947732e-07</td>\n",
       "      <td>1.300925e-10</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S071</th>\n",
       "      <td>0.000844</td>\n",
       "      <td>0.000456</td>\n",
       "      <td>0.011374</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.462071</td>\n",
       "      <td>0.525192</td>\n",
       "      <td>0.103348</td>\n",
       "      <td>0.115077</td>\n",
       "      <td>0.291361</td>\n",
       "      <td>0.047232</td>\n",
       "      <td>...</td>\n",
       "      <td>0.613858</td>\n",
       "      <td>1.544074e-04</td>\n",
       "      <td>4.305656e-03</td>\n",
       "      <td>0.307040</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S017</th>\n",
       "      <td>0.206832</td>\n",
       "      <td>0.071715</td>\n",
       "      <td>0.206088</td>\n",
       "      <td>0.024512</td>\n",
       "      <td>0.050140</td>\n",
       "      <td>0.440714</td>\n",
       "      <td>0.114141</td>\n",
       "      <td>0.096168</td>\n",
       "      <td>0.323603</td>\n",
       "      <td>0.044744</td>\n",
       "      <td>...</td>\n",
       "      <td>0.184054</td>\n",
       "      <td>1.429102e-02</td>\n",
       "      <td>1.826828e-02</td>\n",
       "      <td>0.537486</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-029</th>\n",
       "      <td>0.003075</td>\n",
       "      <td>0.000924</td>\n",
       "      <td>0.953790</td>\n",
       "      <td>0.023456</td>\n",
       "      <td>0.003283</td>\n",
       "      <td>0.015472</td>\n",
       "      <td>0.060075</td>\n",
       "      <td>0.058137</td>\n",
       "      <td>0.655783</td>\n",
       "      <td>0.023916</td>\n",
       "      <td>...</td>\n",
       "      <td>0.993845</td>\n",
       "      <td>9.257881e-06</td>\n",
       "      <td>1.073829e-07</td>\n",
       "      <td>0.003435</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-026</th>\n",
       "      <td>0.005413</td>\n",
       "      <td>0.003711</td>\n",
       "      <td>0.941543</td>\n",
       "      <td>0.006129</td>\n",
       "      <td>0.038927</td>\n",
       "      <td>0.004276</td>\n",
       "      <td>0.093520</td>\n",
       "      <td>0.096803</td>\n",
       "      <td>0.435293</td>\n",
       "      <td>0.033410</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999362</td>\n",
       "      <td>7.960142e-07</td>\n",
       "      <td>2.164846e-06</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S027</th>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.827938</td>\n",
       "      <td>0.113491</td>\n",
       "      <td>0.001046</td>\n",
       "      <td>0.056235</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.128990</td>\n",
       "      <td>0.126311</td>\n",
       "      <td>0.222758</td>\n",
       "      <td>0.044442</td>\n",
       "      <td>...</td>\n",
       "      <td>0.217314</td>\n",
       "      <td>6.392670e-04</td>\n",
       "      <td>1.447242e-02</td>\n",
       "      <td>0.319662</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-068</th>\n",
       "      <td>0.007300</td>\n",
       "      <td>0.002270</td>\n",
       "      <td>0.849172</td>\n",
       "      <td>0.004033</td>\n",
       "      <td>0.028994</td>\n",
       "      <td>0.108232</td>\n",
       "      <td>0.033721</td>\n",
       "      <td>0.045151</td>\n",
       "      <td>0.767739</td>\n",
       "      <td>0.018526</td>\n",
       "      <td>...</td>\n",
       "      <td>0.893119</td>\n",
       "      <td>2.317722e-09</td>\n",
       "      <td>4.543618e-07</td>\n",
       "      <td>0.106880</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-085</th>\n",
       "      <td>0.132419</td>\n",
       "      <td>0.207171</td>\n",
       "      <td>0.583211</td>\n",
       "      <td>0.000300</td>\n",
       "      <td>0.073702</td>\n",
       "      <td>0.003197</td>\n",
       "      <td>0.052490</td>\n",
       "      <td>0.114812</td>\n",
       "      <td>0.441408</td>\n",
       "      <td>0.026659</td>\n",
       "      <td>...</td>\n",
       "      <td>0.423783</td>\n",
       "      <td>2.419594e-04</td>\n",
       "      <td>4.430096e-03</td>\n",
       "      <td>0.420812</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M433-S059</th>\n",
       "      <td>0.000164</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.000842</td>\n",
       "      <td>0.988949</td>\n",
       "      <td>0.000588</td>\n",
       "      <td>0.009348</td>\n",
       "      <td>0.103840</td>\n",
       "      <td>0.090363</td>\n",
       "      <td>0.288975</td>\n",
       "      <td>0.107768</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>9.998619e-01</td>\n",
       "      <td>3.968216e-05</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S048</th>\n",
       "      <td>0.019593</td>\n",
       "      <td>0.001947</td>\n",
       "      <td>0.298463</td>\n",
       "      <td>0.114423</td>\n",
       "      <td>0.002473</td>\n",
       "      <td>0.563101</td>\n",
       "      <td>0.121997</td>\n",
       "      <td>0.120857</td>\n",
       "      <td>0.270288</td>\n",
       "      <td>0.039654</td>\n",
       "      <td>...</td>\n",
       "      <td>0.120523</td>\n",
       "      <td>8.047564e-04</td>\n",
       "      <td>1.189822e-03</td>\n",
       "      <td>0.327735</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>181 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                BCR:repertoire_stats:Covid19  BCR:repertoire_stats:HIV  \\\n",
       "specimen_label                                                           \n",
       "M124-S066                           0.004994                  0.376223   \n",
       "M491-S071                           0.000844                  0.000456   \n",
       "M491-S017                           0.206832                  0.071715   \n",
       "M64-029                             0.003075                  0.000924   \n",
       "M64-026                             0.005413                  0.003711   \n",
       "...                                      ...                       ...   \n",
       "M124-S027                           0.001228                  0.827938   \n",
       "M64-068                             0.007300                  0.002270   \n",
       "M64-085                             0.132419                  0.207171   \n",
       "M433-S059                           0.000164                  0.000109   \n",
       "M491-S048                           0.019593                  0.001947   \n",
       "\n",
       "                BCR:repertoire_stats:Healthy/Background  \\\n",
       "specimen_label                                            \n",
       "M124-S066                                      0.576059   \n",
       "M491-S071                                      0.011374   \n",
       "M491-S017                                      0.206088   \n",
       "M64-029                                        0.953790   \n",
       "M64-026                                        0.941543   \n",
       "...                                                 ...   \n",
       "M124-S027                                      0.113491   \n",
       "M64-068                                        0.849172   \n",
       "M64-085                                        0.583211   \n",
       "M433-S059                                      0.000842   \n",
       "M491-S048                                      0.298463   \n",
       "\n",
       "                BCR:repertoire_stats:Influenza  BCR:repertoire_stats:Lupus  \\\n",
       "specimen_label                                                               \n",
       "M124-S066                             0.006669                    0.002035   \n",
       "M491-S071                             0.000062                    0.462071   \n",
       "M491-S017                             0.024512                    0.050140   \n",
       "M64-029                               0.023456                    0.003283   \n",
       "M64-026                               0.006129                    0.038927   \n",
       "...                                        ...                         ...   \n",
       "M124-S027                             0.001046                    0.056235   \n",
       "M64-068                               0.004033                    0.028994   \n",
       "M64-085                               0.000300                    0.073702   \n",
       "M433-S059                             0.988949                    0.000588   \n",
       "M491-S048                             0.114423                    0.002473   \n",
       "\n",
       "                BCR:repertoire_stats:T1D  \\\n",
       "specimen_label                             \n",
       "M124-S066                       0.034018   \n",
       "M491-S071                       0.525192   \n",
       "M491-S017                       0.440714   \n",
       "M64-029                         0.015472   \n",
       "M64-026                         0.004276   \n",
       "...                                  ...   \n",
       "M124-S027                       0.000061   \n",
       "M64-068                         0.108232   \n",
       "M64-085                         0.003197   \n",
       "M433-S059                       0.009348   \n",
       "M491-S048                       0.563101   \n",
       "\n",
       "                BCR:convergent_cluster_model:Covid19  \\\n",
       "specimen_label                                         \n",
       "M124-S066                                   0.112480   \n",
       "M491-S071                                   0.103348   \n",
       "M491-S017                                   0.114141   \n",
       "M64-029                                     0.060075   \n",
       "M64-026                                     0.093520   \n",
       "...                                              ...   \n",
       "M124-S027                                   0.128990   \n",
       "M64-068                                     0.033721   \n",
       "M64-085                                     0.052490   \n",
       "M433-S059                                   0.103840   \n",
       "M491-S048                                   0.121997   \n",
       "\n",
       "                BCR:convergent_cluster_model:HIV  \\\n",
       "specimen_label                                     \n",
       "M124-S066                               0.229213   \n",
       "M491-S071                               0.115077   \n",
       "M491-S017                               0.096168   \n",
       "M64-029                                 0.058137   \n",
       "M64-026                                 0.096803   \n",
       "...                                          ...   \n",
       "M124-S027                               0.126311   \n",
       "M64-068                                 0.045151   \n",
       "M64-085                                 0.114812   \n",
       "M433-S059                               0.090363   \n",
       "M491-S048                               0.120857   \n",
       "\n",
       "                BCR:convergent_cluster_model:Healthy/Background  \\\n",
       "specimen_label                                                    \n",
       "M124-S066                                              0.191971   \n",
       "M491-S071                                              0.291361   \n",
       "M491-S017                                              0.323603   \n",
       "M64-029                                                0.655783   \n",
       "M64-026                                                0.435293   \n",
       "...                                                         ...   \n",
       "M124-S027                                              0.222758   \n",
       "M64-068                                                0.767739   \n",
       "M64-085                                                0.441408   \n",
       "M433-S059                                              0.288975   \n",
       "M491-S048                                              0.270288   \n",
       "\n",
       "                BCR:convergent_cluster_model:Influenza  ...  \\\n",
       "specimen_label                                          ...   \n",
       "M124-S066                                     0.046151  ...   \n",
       "M491-S071                                     0.047232  ...   \n",
       "M491-S017                                     0.044744  ...   \n",
       "M64-029                                       0.023916  ...   \n",
       "M64-026                                       0.033410  ...   \n",
       "...                                                ...  ...   \n",
       "M124-S027                                     0.044442  ...   \n",
       "M64-068                                       0.018526  ...   \n",
       "M64-085                                       0.026659  ...   \n",
       "M433-S059                                     0.107768  ...   \n",
       "M491-S048                                     0.039654  ...   \n",
       "\n",
       "                TCR:convergent_cluster_model:Healthy/Background  \\\n",
       "specimen_label                                                    \n",
       "M124-S066                                              0.000029   \n",
       "M491-S071                                              0.613858   \n",
       "M491-S017                                              0.184054   \n",
       "M64-029                                                0.993845   \n",
       "M64-026                                                0.999362   \n",
       "...                                                         ...   \n",
       "M124-S027                                              0.217314   \n",
       "M64-068                                                0.893119   \n",
       "M64-085                                                0.423783   \n",
       "M433-S059                                              0.000005   \n",
       "M491-S048                                              0.120523   \n",
       "\n",
       "                TCR:convergent_cluster_model:Influenza  \\\n",
       "specimen_label                                           \n",
       "M124-S066                                 1.947732e-07   \n",
       "M491-S071                                 1.544074e-04   \n",
       "M491-S017                                 1.429102e-02   \n",
       "M64-029                                   9.257881e-06   \n",
       "M64-026                                   7.960142e-07   \n",
       "...                                                ...   \n",
       "M124-S027                                 6.392670e-04   \n",
       "M64-068                                   2.317722e-09   \n",
       "M64-085                                   2.419594e-04   \n",
       "M433-S059                                 9.998619e-01   \n",
       "M491-S048                                 8.047564e-04   \n",
       "\n",
       "                TCR:convergent_cluster_model:Lupus  \\\n",
       "specimen_label                                       \n",
       "M124-S066                             1.300925e-10   \n",
       "M491-S071                             4.305656e-03   \n",
       "M491-S017                             1.826828e-02   \n",
       "M64-029                               1.073829e-07   \n",
       "M64-026                               2.164846e-06   \n",
       "...                                            ...   \n",
       "M124-S027                             1.447242e-02   \n",
       "M64-068                               4.543618e-07   \n",
       "M64-085                               4.430096e-03   \n",
       "M433-S059                             3.968216e-05   \n",
       "M491-S048                             1.189822e-03   \n",
       "\n",
       "                TCR:convergent_cluster_model:T1D  TCR:sequence_model:Covid19  \\\n",
       "specimen_label                                                                 \n",
       "M124-S066                               0.000037                        0.00   \n",
       "M491-S071                               0.307040                        0.21   \n",
       "M491-S017                               0.537486                        0.20   \n",
       "M64-029                                 0.003435                        0.03   \n",
       "M64-026                                 0.000635                        0.05   \n",
       "...                                          ...                         ...   \n",
       "M124-S027                               0.319662                        0.02   \n",
       "M64-068                                 0.106880                        0.13   \n",
       "M64-085                                 0.420812                        0.06   \n",
       "M433-S059                               0.000082                        0.02   \n",
       "M491-S048                               0.327735                        0.00   \n",
       "\n",
       "                TCR:sequence_model:HIV  TCR:sequence_model:Healthy/Background  \\\n",
       "specimen_label                                                                  \n",
       "M124-S066                         0.80                                   0.22   \n",
       "M491-S071                         0.08                                   0.12   \n",
       "M491-S017                         0.05                                   0.08   \n",
       "M64-029                           0.21                                   0.59   \n",
       "M64-026                           0.03                                   0.62   \n",
       "...                                ...                                    ...   \n",
       "M124-S027                         0.61                                   0.34   \n",
       "M64-068                           0.01                                   0.65   \n",
       "M64-085                           0.04                                   0.56   \n",
       "M433-S059                         0.36                                   0.00   \n",
       "M491-S048                         0.32                                   0.19   \n",
       "\n",
       "                TCR:sequence_model:Influenza  TCR:sequence_model:Lupus  \\\n",
       "specimen_label                                                           \n",
       "M124-S066                               0.06                      0.01   \n",
       "M491-S071                               0.05                      0.08   \n",
       "M491-S017                               0.02                      0.15   \n",
       "M64-029                                 0.00                      0.01   \n",
       "M64-026                                 0.01                      0.05   \n",
       "...                                      ...                       ...   \n",
       "M124-S027                               0.03                      0.00   \n",
       "M64-068                                 0.00                      0.01   \n",
       "M64-085                                 0.01                      0.04   \n",
       "M433-S059                               0.27                      0.00   \n",
       "M491-S048                               0.04                      0.02   \n",
       "\n",
       "                TCR:sequence_model:T1D  \n",
       "specimen_label                          \n",
       "M124-S066                         0.04  \n",
       "M491-S071                         0.32  \n",
       "M491-S017                         0.21  \n",
       "M64-029                           0.03  \n",
       "M64-026                           0.06  \n",
       "...                                ...  \n",
       "M124-S027                         0.03  \n",
       "M64-068                           0.03  \n",
       "M64-085                           0.09  \n",
       "M433-S059                         0.25  \n",
       "M491-S048                         0.12  \n",
       "\n",
       "[181 rows x 36 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Features\n",
    "featurized_metamodel_data.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "73984959-c5b9-41a0-a84f-f873f630bab4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['HIV', 'T1D', 'T1D', 'Healthy/Background', 'Healthy/Background',\n",
       "       'T1D', 'Influenza', 'Healthy/Background', 'HIV', 'Covid19',\n",
       "       'Healthy/Background', 'Covid19', 'HIV', 'HIV', 'HIV',\n",
       "       'Healthy/Background', 'Healthy/Background', 'T1D', 'Lupus', 'HIV',\n",
       "       'T1D', 'Healthy/Background', 'Covid19', 'Healthy/Background',\n",
       "       'HIV', 'Healthy/Background', 'T1D', 'Influenza', 'Covid19',\n",
       "       'Lupus', 'T1D', 'Healthy/Background', 'Healthy/Background', 'HIV',\n",
       "       'Covid19', 'Lupus', 'HIV', 'Healthy/Background', 'Lupus',\n",
       "       'Healthy/Background', 'HIV', 'Lupus', 'Lupus', 'Lupus',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'HIV', 'HIV', 'Lupus', 'Healthy/Background', 'Healthy/Background',\n",
       "       'HIV', 'Lupus', 'Covid19', 'T1D', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Influenza', 'Influenza', 'T1D', 'HIV',\n",
       "       'Influenza', 'HIV', 'Healthy/Background', 'T1D',\n",
       "       'Healthy/Background', 'T1D', 'Healthy/Background', 'HIV', 'HIV',\n",
       "       'Healthy/Background', 'Lupus', 'Lupus', 'Covid19', 'T1D',\n",
       "       'Healthy/Background', 'T1D', 'HIV', 'Healthy/Background',\n",
       "       'Healthy/Background', 'HIV', 'HIV', 'Healthy/Background',\n",
       "       'Covid19', 'Healthy/Background', 'HIV', 'Healthy/Background',\n",
       "       'HIV', 'Lupus', 'Covid19', 'Healthy/Background', 'T1D',\n",
       "       'Influenza', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Influenza', 'HIV', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Covid19', 'T1D', 'T1D',\n",
       "       'Healthy/Background', 'T1D', 'Influenza', 'Healthy/Background',\n",
       "       'HIV', 'Lupus', 'HIV', 'Healthy/Background', 'Lupus', 'Lupus',\n",
       "       'Covid19', 'Healthy/Background', 'Lupus', 'HIV',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'T1D', 'T1D', 'T1D',\n",
       "       'Influenza', 'Healthy/Background', 'Healthy/Background', 'HIV',\n",
       "       'Lupus', 'Healthy/Background', 'Covid19', 'Healthy/Background',\n",
       "       'T1D', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Lupus', 'Healthy/Background', 'Lupus', 'Covid19', 'HIV',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Influenza', 'T1D',\n",
       "       'T1D', 'T1D', 'T1D', 'HIV', 'HIV', 'HIV', 'HIV', 'T1D', 'Lupus',\n",
       "       'Covid19', 'Influenza', 'Healthy/Background', 'T1D',\n",
       "       'Healthy/Background', 'Healthy/Background', 'Covid19', 'T1D',\n",
       "       'Healthy/Background', 'Lupus', 'Covid19', 'Lupus',\n",
       "       'Healthy/Background', 'T1D', 'T1D', 'Healthy/Background', 'T1D',\n",
       "       'HIV', 'Healthy/Background', 'Healthy/Background',\n",
       "       'Healthy/Background', 'Influenza', 'T1D'], dtype=object)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ground truth\n",
    "featurized_metamodel_data.y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "fd609af7-9902-47c4-91d1-50ebb2b52937",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['M124-S066', 'M491-S071', 'M491-S017', 'M64-029', 'M64-026',\n",
       "       'M491-S107', 'M433-S051', 'M64-078', 'M124-S072', 'M418-S010',\n",
       "       ...\n",
       "       'M491-S064', 'M491-S106', 'M64-077', 'M491-S057', 'M111-S048',\n",
       "       'M124-S027', 'M64-068', 'M64-085', 'M433-S059', 'M491-S048'],\n",
       "      dtype='object', name='specimen_label', length=181)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample names\n",
    "featurized_metamodel_data.sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "65ca919d-d9e6-46bd-8566-1d2d8d8f7bd4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease.rollup</th>\n",
       "      <th>disease.separate_past_exposures</th>\n",
       "      <th>disease_severity</th>\n",
       "      <th>disease_subtype</th>\n",
       "      <th>ethnicity_condensed</th>\n",
       "      <th>isotype_proportion:IGHA</th>\n",
       "      <th>isotype_proportion:IGHD-M</th>\n",
       "      <th>isotype_proportion:IGHG</th>\n",
       "      <th>participant_label</th>\n",
       "      <th>past_exposure</th>\n",
       "      <th>sex</th>\n",
       "      <th>study_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M124-S066</th>\n",
       "      <td>27.0</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>HIV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIV Non Neutralizing</td>\n",
       "      <td>African</td>\n",
       "      <td>0.230346</td>\n",
       "      <td>0.551237</td>\n",
       "      <td>0.218417</td>\n",
       "      <td>BFI-0003755</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S071</th>\n",
       "      <td>18.0</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.178105</td>\n",
       "      <td>0.596202</td>\n",
       "      <td>0.225693</td>\n",
       "      <td>BFI-0010718</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S017</th>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.244323</td>\n",
       "      <td>0.574685</td>\n",
       "      <td>0.180992</td>\n",
       "      <td>BFI-0010665</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-029</th>\n",
       "      <td>66.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - CMV-</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>0.209289</td>\n",
       "      <td>0.541179</td>\n",
       "      <td>0.249532</td>\n",
       "      <td>BFI-0003078</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>Healthy-StanfordBloodCenter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-026</th>\n",
       "      <td>68.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - CMV+</td>\n",
       "      <td>Asian</td>\n",
       "      <td>0.254202</td>\n",
       "      <td>0.569415</td>\n",
       "      <td>0.176384</td>\n",
       "      <td>BFI-0003075</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>Healthy-StanfordBloodCenter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S027</th>\n",
       "      <td>28.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - HIV Negative</td>\n",
       "      <td>African</td>\n",
       "      <td>0.368556</td>\n",
       "      <td>0.410265</td>\n",
       "      <td>0.221179</td>\n",
       "      <td>BFI-0003724</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>HIV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-068</th>\n",
       "      <td>54.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - CMV+</td>\n",
       "      <td>Asian</td>\n",
       "      <td>0.119946</td>\n",
       "      <td>0.779775</td>\n",
       "      <td>0.100280</td>\n",
       "      <td>BFI-0003117</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>Healthy-StanfordBloodCenter_included-in-resequ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-085</th>\n",
       "      <td>55.0</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>Healthy/Background</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy/Background - CMV+</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>0.146258</td>\n",
       "      <td>0.743095</td>\n",
       "      <td>0.110647</td>\n",
       "      <td>BFI-0003134</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>Healthy-StanfordBloodCenter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M433-S059</th>\n",
       "      <td>23.0</td>\n",
       "      <td>Influenza</td>\n",
       "      <td>Influenza</td>\n",
       "      <td>Influenza</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Influenza vaccine 2021 - day 7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.353308</td>\n",
       "      <td>0.433732</td>\n",
       "      <td>0.212960</td>\n",
       "      <td>BFI-0009975</td>\n",
       "      <td>False</td>\n",
       "      <td>F</td>\n",
       "      <td>Flu vaccine UPenn 2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S048</th>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>T1D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T1D - pediatric</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.205114</td>\n",
       "      <td>0.672336</td>\n",
       "      <td>0.122550</td>\n",
       "      <td>BFI-0010695</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diabetes biobank</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>181 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 age             disease      disease.rollup  \\\n",
       "specimen_label                                                 \n",
       "M124-S066       27.0                 HIV                 HIV   \n",
       "M491-S071       18.0                 T1D                 T1D   \n",
       "M491-S017        NaN                 T1D                 T1D   \n",
       "M64-029         66.0  Healthy/Background  Healthy/Background   \n",
       "M64-026         68.0  Healthy/Background  Healthy/Background   \n",
       "...              ...                 ...                 ...   \n",
       "M124-S027       28.0  Healthy/Background  Healthy/Background   \n",
       "M64-068         54.0  Healthy/Background  Healthy/Background   \n",
       "M64-085         55.0  Healthy/Background  Healthy/Background   \n",
       "M433-S059       23.0           Influenza           Influenza   \n",
       "M491-S048        NaN                 T1D                 T1D   \n",
       "\n",
       "               disease.separate_past_exposures disease_severity  \\\n",
       "specimen_label                                                    \n",
       "M124-S066                                  HIV              NaN   \n",
       "M491-S071                                  T1D              NaN   \n",
       "M491-S017                                  T1D              NaN   \n",
       "M64-029                     Healthy/Background              NaN   \n",
       "M64-026                     Healthy/Background              NaN   \n",
       "...                                        ...              ...   \n",
       "M124-S027                   Healthy/Background              NaN   \n",
       "M64-068                     Healthy/Background              NaN   \n",
       "M64-085                     Healthy/Background              NaN   \n",
       "M433-S059                            Influenza              NaN   \n",
       "M491-S048                                  T1D              NaN   \n",
       "\n",
       "                                  disease_subtype ethnicity_condensed  \\\n",
       "specimen_label                                                          \n",
       "M124-S066                    HIV Non Neutralizing             African   \n",
       "M491-S071                         T1D - pediatric                 NaN   \n",
       "M491-S017                             T1D - adult                 NaN   \n",
       "M64-029                 Healthy/Background - CMV-           Caucasian   \n",
       "M64-026                 Healthy/Background - CMV+               Asian   \n",
       "...                                           ...                 ...   \n",
       "M124-S027       Healthy/Background - HIV Negative             African   \n",
       "M64-068                 Healthy/Background - CMV+               Asian   \n",
       "M64-085                 Healthy/Background - CMV+           Caucasian   \n",
       "M433-S059          Influenza vaccine 2021 - day 7                 NaN   \n",
       "M491-S048                         T1D - pediatric                 NaN   \n",
       "\n",
       "                isotype_proportion:IGHA  isotype_proportion:IGHD-M  \\\n",
       "specimen_label                                                       \n",
       "M124-S066                      0.230346                   0.551237   \n",
       "M491-S071                      0.178105                   0.596202   \n",
       "M491-S017                      0.244323                   0.574685   \n",
       "M64-029                        0.209289                   0.541179   \n",
       "M64-026                        0.254202                   0.569415   \n",
       "...                                 ...                        ...   \n",
       "M124-S027                      0.368556                   0.410265   \n",
       "M64-068                        0.119946                   0.779775   \n",
       "M64-085                        0.146258                   0.743095   \n",
       "M433-S059                      0.353308                   0.433732   \n",
       "M491-S048                      0.205114                   0.672336   \n",
       "\n",
       "                isotype_proportion:IGHG participant_label  past_exposure  sex  \\\n",
       "specimen_label                                                                  \n",
       "M124-S066                      0.218417       BFI-0003755          False    F   \n",
       "M491-S071                      0.225693       BFI-0010718          False  NaN   \n",
       "M491-S017                      0.180992       BFI-0010665          False  NaN   \n",
       "M64-029                        0.249532       BFI-0003078          False    M   \n",
       "M64-026                        0.176384       BFI-0003075          False    M   \n",
       "...                                 ...               ...            ...  ...   \n",
       "M124-S027                      0.221179       BFI-0003724          False    M   \n",
       "M64-068                        0.100280       BFI-0003117          False    M   \n",
       "M64-085                        0.110647       BFI-0003134          False    M   \n",
       "M433-S059                      0.212960       BFI-0009975          False    F   \n",
       "M491-S048                      0.122550       BFI-0010695          False  NaN   \n",
       "\n",
       "                                                       study_name  \n",
       "specimen_label                                                     \n",
       "M124-S066                                                     HIV  \n",
       "M491-S071                                        Diabetes biobank  \n",
       "M491-S017                                        Diabetes biobank  \n",
       "M64-029                               Healthy-StanfordBloodCenter  \n",
       "M64-026                               Healthy-StanfordBloodCenter  \n",
       "...                                                           ...  \n",
       "M124-S027                                                     HIV  \n",
       "M64-068         Healthy-StanfordBloodCenter_included-in-resequ...  \n",
       "M64-085                               Healthy-StanfordBloodCenter  \n",
       "M433-S059                                  Flu vaccine UPenn 2021  \n",
       "M491-S048                                        Diabetes biobank  \n",
       "\n",
       "[181 rows x 14 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample metadata\n",
    "featurized_metamodel_data.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdaafff9-69ce-48ae-bab4-e80bbe8129d7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ac3f4177-0389-44e3-9ef0-89742426e693",
   "metadata": {},
   "source": [
    "And we'll run our final prediction, like with the other models:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "dd471399-4402-43fc-b18f-00b26a3c0630",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Covid19</th>\n",
       "      <th>HIV</th>\n",
       "      <th>Healthy/Background</th>\n",
       "      <th>Influenza</th>\n",
       "      <th>Lupus</th>\n",
       "      <th>T1D</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimen_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M124-S066</th>\n",
       "      <td>0.069313</td>\n",
       "      <td>0.475842</td>\n",
       "      <td>0.164825</td>\n",
       "      <td>0.097305</td>\n",
       "      <td>0.089503</td>\n",
       "      <td>0.103212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S071</th>\n",
       "      <td>0.109962</td>\n",
       "      <td>0.101471</td>\n",
       "      <td>0.114104</td>\n",
       "      <td>0.084598</td>\n",
       "      <td>0.206145</td>\n",
       "      <td>0.383721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S017</th>\n",
       "      <td>0.154769</td>\n",
       "      <td>0.124511</td>\n",
       "      <td>0.157118</td>\n",
       "      <td>0.096315</td>\n",
       "      <td>0.151228</td>\n",
       "      <td>0.316059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-029</th>\n",
       "      <td>0.081137</td>\n",
       "      <td>0.121118</td>\n",
       "      <td>0.497514</td>\n",
       "      <td>0.099606</td>\n",
       "      <td>0.112362</td>\n",
       "      <td>0.088264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-026</th>\n",
       "      <td>0.088994</td>\n",
       "      <td>0.118889</td>\n",
       "      <td>0.435706</td>\n",
       "      <td>0.117980</td>\n",
       "      <td>0.131329</td>\n",
       "      <td>0.107103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M124-S027</th>\n",
       "      <td>0.088240</td>\n",
       "      <td>0.416703</td>\n",
       "      <td>0.152640</td>\n",
       "      <td>0.144919</td>\n",
       "      <td>0.095025</td>\n",
       "      <td>0.102473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-068</th>\n",
       "      <td>0.121266</td>\n",
       "      <td>0.114451</td>\n",
       "      <td>0.412180</td>\n",
       "      <td>0.095271</td>\n",
       "      <td>0.132779</td>\n",
       "      <td>0.124052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M64-085</th>\n",
       "      <td>0.096077</td>\n",
       "      <td>0.148324</td>\n",
       "      <td>0.327083</td>\n",
       "      <td>0.170974</td>\n",
       "      <td>0.145739</td>\n",
       "      <td>0.111802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M433-S059</th>\n",
       "      <td>0.021166</td>\n",
       "      <td>0.036373</td>\n",
       "      <td>0.033421</td>\n",
       "      <td>0.801682</td>\n",
       "      <td>0.033510</td>\n",
       "      <td>0.073848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M491-S048</th>\n",
       "      <td>0.119346</td>\n",
       "      <td>0.137936</td>\n",
       "      <td>0.214863</td>\n",
       "      <td>0.145823</td>\n",
       "      <td>0.113642</td>\n",
       "      <td>0.268390</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>181 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Covid19       HIV  Healthy/Background  Influenza     Lupus  \\\n",
       "specimen_label                                                                \n",
       "M124-S066       0.069313  0.475842            0.164825   0.097305  0.089503   \n",
       "M491-S071       0.109962  0.101471            0.114104   0.084598  0.206145   \n",
       "M491-S017       0.154769  0.124511            0.157118   0.096315  0.151228   \n",
       "M64-029         0.081137  0.121118            0.497514   0.099606  0.112362   \n",
       "M64-026         0.088994  0.118889            0.435706   0.117980  0.131329   \n",
       "...                  ...       ...                 ...        ...       ...   \n",
       "M124-S027       0.088240  0.416703            0.152640   0.144919  0.095025   \n",
       "M64-068         0.121266  0.114451            0.412180   0.095271  0.132779   \n",
       "M64-085         0.096077  0.148324            0.327083   0.170974  0.145739   \n",
       "M433-S059       0.021166  0.036373            0.033421   0.801682  0.033510   \n",
       "M491-S048       0.119346  0.137936            0.214863   0.145823  0.113642   \n",
       "\n",
       "                     T1D  \n",
       "specimen_label            \n",
       "M124-S066       0.103212  \n",
       "M491-S071       0.383721  \n",
       "M491-S017       0.316059  \n",
       "M64-029         0.088264  \n",
       "M64-026         0.107103  \n",
       "...                  ...  \n",
       "M124-S027       0.102473  \n",
       "M64-068         0.124052  \n",
       "M64-085         0.111802  \n",
       "M433-S059       0.073848  \n",
       "M491-S048       0.268390  \n",
       "\n",
       "[181 rows x 6 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Table of predicted class probabilities\n",
    "pd.DataFrame(\n",
    "    clf_metamodel.predict_proba(featurized_metamodel_data.X),\n",
    "    index=featurized_metamodel_data.sample_names,\n",
    "    columns=clf_metamodel.classes_,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "203fcca3-271a-4010-8ff6-6831e1d9657e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "02c85113-bb8c-4be3-af53-a55aaf3c9aea",
   "metadata": {},
   "source": [
    "We'll end by calculating the AUROC for fold 0:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "121f298b-6e5a-427a-9081-a76937f77dde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9878390256152686"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sklearn.metrics\n",
    "\n",
    "sklearn.metrics.roc_auc_score(\n",
    "    y_true=featurized_metamodel_data.y,\n",
    "    y_score=clf_metamodel.predict_proba(featurized_metamodel_data.X),\n",
    "    multi_class=\"ovo\",  # Multiclass AUC calculated in one-versus-one fashion\n",
    "    average=\"weighted\",  # Take class size-weighted average of the binary AUROC calculated for each pair of classes\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67bc8fb1-6048-473b-8e4e-a790fcef0eb1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e11144df-d0f9-44ca-af4e-47afa02ec78e",
   "metadata": {},
   "source": [
    "See the readme for instructions about customizing Mal-ID for additional datasets and classification targets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67802899-1e53-4158-9d6c-8961ea194337",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31532c93-2416-4ad3-9c7b-4adc663c2363",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "py39-cuda-env",
   "language": "python",
   "name": "py39-cuda-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
